--- a/drivers/mtd/nand/hinfc_spl_ids.c
+++ b/drivers/mtd/nand/hinfc_spl_ids.c
@@ -493,6 +493,25 @@
 		.flags = 0,
 	},
 	{       /* SLC 8bit/512 */
+		.name      = "TC58BVG0S3HTA00",
+		.id        = {0x98, 0xF1, 0x80, 0x15, 0xF2, 0x16, 0x08, 0x00},
+		.length    = 7,
+		.chipsize  = SZ_128M,
+		.probe     = NULL,
+		.pagesize  = SZ_2K,
+		.erasesize = SZ_128K,
+		.oobsize   = 64,
+		.options   = 0,
+		.read_retry_type = NAND_RR_NONE,
+		/*
+		 * Datasheet: read one column of any page in each block. If the
+		 * data of the column is 00 (Hex), define the block as a bad
+		 * block.
+		 */
+		.badblock_pos    = BBP_FIRST_PAGE,
+		.flags = 0,
+	},
+	{       /* SLC 8bit/512 */
 		.name      = "TC58NVG1S3HTA00",
 		.id        = {0x98, 0xDA, 0x90, 0x15, 0x76, 0x16, 0x08, 0x00},
 		.length    = 7,
