--- a/Makefile
+++ b/Makefile
@@ -552,11 +552,17 @@
 include/config/auto.conf: ;
 endif # $(dot-config)
 
+export PROJECT_NAME = $(shell grep -e '^CONFIG_MACH_FH.*' .config|head -1|sed 's/CONFIG_MACH_\(.*\)=y/\1/'|awk '{print tolower($$0)}')
+
+PHONY += boardconfig
+boardconfig:
+	cp arch/arm/mach-fh/include/mach/board_config.$(PROJECT_NAME).appboard arch/arm/mach-fh/include/mach/board_config.h
+
 # The all: target is the default when no target is given on the
 # command line.
 # This allow a user to issue only 'make' to build a kernel including modules
 # Defaults to vmlinux, but the arch makefile usually adds further targets
-all: vmlinux
+all: boardconfig vmlinux
 
 ifdef CONFIG_CC_OPTIMIZE_FOR_SIZE
 KBUILD_CFLAGS	+= -Os
