diff --git a/scripts/ms_bin_option_update_int.py b/scripts/ms_bin_option_update_int.py
index a515e5e872af..a538d1903802 100755
--- a/scripts/ms_bin_option_update_int.py
+++ b/scripts/ms_bin_option_update_int.py
@@ -6,13 +6,13 @@ if __name__ == '__main__':
 
     name=sys.argv[2]
     if sys.argv[3].upper().startswith( '0X' ):
-        value=long(sys.argv[3],16)
+        value=int(sys.argv[3],16)
     else:
-        value=long(sys.argv[3])
+        value=int(sys.argv[3])
 
     fmap=mmap.mmap(os.open(sys.argv[1],os.O_RDWR),0)
 
-    offset=fmap.find(name)
+    offset=fmap.find(name.encode())
 #    print ('%s:%d\n' % (name,offset))
     if offset < 0:
         print ('error finding ms_bin_option:%s in %s\n' % (name,sys.argv[1]))
diff --git a/scripts/ms_builtin_dtb_update.py b/scripts/ms_builtin_dtb_update.py
index 787ab98e4747..9214d9a79ad0 100755
--- a/scripts/ms_builtin_dtb_update.py
+++ b/scripts/ms_builtin_dtb_update.py
@@ -6,7 +6,7 @@ if __name__ == '__main__':
     
     name='#MS_DTB#'
  
-    dtb_file=open(sys.argv[2])
+    dtb_file=open(sys.argv[2], 'rb')
     dtb_file.seek(0,os.SEEK_END)
     size=dtb_file.tell()
     dtb_file.seek(0,os.SEEK_SET)
@@ -19,7 +19,7 @@ if __name__ == '__main__':
         
     fmap=mmap.mmap(os.open(sys.argv[1],os.O_RDWR),0)
 
-    offset=fmap.find(name)    
+    offset=fmap.find(name.encode())
     if offset >=0:
         print ('offset:0x%08X' % offset)
         print ('  size:0x%08X' % size )
