--- a/arch/mips/xburst/soc-t10/common/platform.c
+++ b/arch/mips/xburst/soc-t10/common/platform.c
@@ -45,6 +45,10 @@
 #ifdef CONFIG_I2C1_PB25_PB26
 	I2C1_PORTB,
 #endif
+
+#ifndef CONFIG_VIDEO_V4L2
+	MCLK_PORTA,
+#endif
 #ifdef CONFIG_SOC_MCLK
 	MCLK_PORTA,
 #endif
--- a/arch/mips/xburst/soc-t15/common/platform.c
+++ b/arch/mips/xburst/soc-t15/common/platform.c
@@ -91,6 +91,10 @@
 	#endif
 #endif
 
+#ifndef CONFIG_VIDEO_V4L2
+	MCLK_PORTE_BIT2,
+#endif
+
 #ifdef CONFIG_VIDEO_TX_ISP
 	MCLK_PORTE_BIT2,
 #endif
--- a/arch/mips/xburst/soc-t20/common/platform.c
+++ b/arch/mips/xburst/soc-t20/common/platform.c
@@ -45,6 +45,9 @@
 #ifdef CONFIG_I2C1_PB25_PB26
 	I2C1_PORTB,
 #endif
+#ifndef CONFIG_VIDEO_V4L2
+	MCLK_PORTA,
+#endif
 #ifdef CONFIG_SOC_MCLK
 	MCLK_PORTA,
 #endif
--- a/arch/mips/xburst/soc-t21/common/platform.c
+++ b/arch/mips/xburst/soc-t21/common/platform.c
@@ -49,7 +49,9 @@
 #ifdef CONFIG_I2C1_PC08_PC09
 	I2C1_PORTC,
 #endif
-
+#ifndef CONFIG_VIDEO_V4L2
+	MCLK_PORTA,
+#endif
 #ifdef CONFIG_SOC_MCLK
 	MCLK_PORTA,
 #endif
--- a/arch/mips/xburst/soc-t30/common/platform.c
+++ b/arch/mips/xburst/soc-t30/common/platform.c
@@ -49,7 +49,9 @@
 #ifdef CONFIG_I2C2_PC27_PC28
 		I2C2_PORTC,
 #endif
-
+#ifndef CONFIG_VIDEO_V4L2
+	MCLK_PORTA,
+#endif
 #ifdef CONFIG_SOC_MCLK
 	MCLK_PORTA,
 #endif
