--- a/drivers/net/fh_gmac/fh_gmac_phyt.c
+++ b/drivers/net/fh_gmac/fh_gmac_phyt.c
@@ -83,6 +83,7 @@
 	if (pGmac->phy_interface == PHY_INTERFACE_MODE_RMII) {
 		switch (pGmac->phydev->phy_id) {
 		case FH_GMAC_PHY_RTL8201:
+		case FH_GMAC_PHY_JL1101:
 			fh_mdio_write(bus, phyid,
 					gmac_phyt_rtl8201_page_select, 7);
 			fh_mdio_write(bus, phyid,
@@ -116,6 +117,7 @@
 	} else if (pGmac->phy_interface == PHY_INTERFACE_MODE_MII) {
 		switch (pGmac->phydev->phy_id) {
 		case FH_GMAC_PHY_RTL8201:
+		case FH_GMAC_PHY_JL1101:
 			fh_mdio_write(bus, phyid,
 					gmac_phyt_rtl8201_page_select, 7);
 			fh_mdio_write(bus, phyid,
--- a/drivers/net/fh_gmac/fh_gmac_phyt.h
+++ b/drivers/net/fh_gmac/fh_gmac_phyt.h
@@ -11,6 +11,7 @@
 #define FH_GMAC_PHY_IP101G	0x02430C54
 #define FH_GMAC_PHY_RTL8201	0x001CC816
 #define FH_GMAC_PHY_TI83848	0xFFFFFFFF
+#define FH_GMAC_PHY_JL1101	0x937c4023
 
 enum
 {
