diff --git a/buffer.c b/buffer.c
index 18eac7c9..3f18317f 100644
--- a/buffer.c
+++ b/buffer.c
@@ -63,6 +63,9 @@
 #include <sys/stat.h>
 #endif
 
+#if defined(BROKEN_MMAP)
+#include <sys/syscall.h>
+#endif
 
 #include <errno.h>
 #include <stdio.h>
@@ -2994,6 +2997,12 @@ done:
 	return result;
 }
 
+#if defined(BROKEN_MMAP)
+static void *mmap_ev(void *start, size_t len, int prot, int flags, int fd, off_t off) {
+	return (void *)syscall(SYS_mmap2, start, len, prot, flags, fd, off >> 12);
+}
+#endif
+
 /* TODO(niels): we may want to add to automagically convert to mmap, in
  * case evbuffer_remove() or evbuffer_pullup() are being used.
  */
@@ -3095,10 +3104,14 @@ evbuffer_file_segment_materialize(struct evbuffer_file_segment *seg)
 			offset_leftover = offset % page_size;
 			offset_rounded = offset - offset_leftover;
 		}
+#if defined(BROKEN_MMAP)
+		mapped = mmap_ev(NULL, length + offset_leftover,
+#else
 #if defined(EVENT__HAVE_MMAP64)
 		mapped = mmap64(NULL, length + offset_leftover,
 #else
 		mapped = mmap(NULL, length + offset_leftover,
+#endif
 #endif
 			PROT_READ, mmap_flags, fd, offset_rounded);
 		if (mapped == MAP_FAILED) {
