diff -drupN a/drivers/net/wireless/xr829/include/linux/pm_qos.h b/drivers/net/wireless/xr829/include/linux/pm_qos.h
--- a/drivers/net/wireless/xr829/include/linux/pm_qos.h	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/net/wireless/xr829/include/linux/pm_qos.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,12 @@
+#ifndef _COMPAT_LINUX_PM_QOS_H
+#define _COMPAT_LINUX_PM_QOS_H 1
+
+#include <generated/uapi/linux/version.h>
+
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(3, 2, 0))
+#include_next <linux/pm_qos.h>
+#else
+#include <linux/pm_qos_params.h>
+#endif /* (LINUX_VERSION_CODE >= KERNEL_VERSION(3, 2, 0)) */
+
+#endif	/* _COMPAT_LINUX_PM_QOS_H */
