diff -drupN a/drivers/net/wireless/rtl8822cs/include/rtw_version.h b/drivers/net/wireless/rtl8822cs/include/rtw_version.h
--- a/drivers/net/wireless/rtl8822cs/include/rtw_version.h	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/net/wireless/rtl8822cs/include/rtw_version.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,2 @@
+#define DRIVERVERSION	"v5.7.3_31881.20190102_COEX20181227-08_beta"
+#define BTCOEXVERSION	"COEX20181227-08"
