diff -drupN a/drivers/net/wireless/xradio/umac/debugfs_sta.h b/drivers/net/wireless/xradio/umac/debugfs_sta.h
--- a/drivers/net/wireless/xradio/umac/debugfs_sta.h	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/net/wireless/xradio/umac/debugfs_sta.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,14 @@
+#ifndef __XRMAC_DEBUGFS_STA_H
+#define __XRMAC_DEBUGFS_STA_H
+
+#include "sta_info.h"
+
+#ifdef CONFIG_XRMAC_DEBUGFS
+void mac80211_sta_debugfs_add(struct sta_info *sta);
+void mac80211_sta_debugfs_remove(struct sta_info *sta);
+#else
+static inline void mac80211_sta_debugfs_add(struct sta_info *sta) {}
+static inline void mac80211_sta_debugfs_remove(struct sta_info *sta) {}
+#endif
+
+#endif /* __XRMAC_DEBUGFS_STA_H */
