diff -drupN a/drivers/net/wireless/rtl8822cs/hal/rtl8822c/sdio/rtl8822cs_led.c b/drivers/net/wireless/rtl8822cs/hal/rtl8822c/sdio/rtl8822cs_led.c
--- a/drivers/net/wireless/rtl8822cs/hal/rtl8822c/sdio/rtl8822cs_led.c	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/net/wireless/rtl8822cs/hal/rtl8822c/sdio/rtl8822cs_led.c	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,100 @@
+/******************************************************************************
+ *
+ * Copyright(c) 2015 - 2017 Realtek Corporation.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of version 2 of the GNU General Public License as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+ * more details.
+ *
+ *****************************************************************************/
+#define _RTL8822CS_LED_C_
+
+#include <drv_types.h>		/* PADAPTER */
+#include <hal_data.h>		/* PHAL_DATA_TYPE */
+#include <hal_com_led.h>	/* PLED_SDIO */
+#ifdef CONFIG_RTW_SW_LED
+
+/*
+ * =============================================================================
+ * LED object.
+ * =============================================================================
+ */
+
+
+/*
+ * =============================================================================
+ *	Prototype of protected function.
+ * =============================================================================
+ */
+
+/*
+ * =============================================================================
+ * LED routines.
+ * =============================================================================
+ */
+
+/*
+ * Description:
+ *	Turn on LED according to LedPin specified.
+ */
+void swledon(PADAPTER adapter, PLED_SDIO pLed)
+{
+	u8 LedCfg;
+	PHAL_DATA_TYPE hal = GET_HAL_DATA(adapter);
+
+	if (RTW_CANNOT_RUN(adapter))
+		return;
+
+	pLed->bLedOn = _TRUE;
+}
+
+
+/*
+ * Description:
+ *	Turn off LED according to LedPin specified.
+ */
+void swledoff(PADAPTER adapter, PLED_SDIO pLed)
+{
+	u8 LedCfg;
+	PHAL_DATA_TYPE hal = GET_HAL_DATA(adapter);
+
+	if (RTW_CANNOT_RUN(adapter))
+		goto exit;
+
+exit:
+	pLed->bLedOn = _FALSE;
+}
+
+/*
+ * =============================================================================
+ * Interface to manipulate LED objects.
+ * =============================================================================
+ */
+
+/*
+ * =============================================================================
+ * Default LED behavior.
+ * =============================================================================
+ */
+
+/*
+ * Description:
+ *	Initialize all LED_871x objects.
+ */
+void rtl8822cs_initswleds(PADAPTER adapter)
+{
+}
+
+/*
+ * Description:
+ *	DeInitialize all LED_819xUsb objects.
+ */
+void rtl8822cs_deinitswleds(PADAPTER adapter)
+{
+}
+#endif
