diff -drupN a/init/main.c b/init/main.c
--- a/init/main.c	2018-08-06 17:23:04.000000000 +0300
+++ b/init/main.c	2022-06-12 05:28:14.000000000 +0300
@@ -121,6 +121,7 @@ void (*__initdata late_time_init)(void);
 char __initdata boot_command_line[COMMAND_LINE_SIZE];
 /* Untouched saved command line (eg. for /proc) */
 char *saved_command_line;
+EXPORT_SYMBOL(saved_command_line);
 /* Command line for parameter parsing */
 static char *static_command_line;
 /* Command line for per-initcall parameter parsing */
