diff -drupN a/drivers/video/fbdev/sunxi/disp2/disp/lcd/to20t20000.h b/drivers/video/fbdev/sunxi/disp2/disp/lcd/to20t20000.h
--- a/drivers/video/fbdev/sunxi/disp2/disp/lcd/to20t20000.h	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/video/fbdev/sunxi/disp2/disp/lcd/to20t20000.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,18 @@
+/*
+ * Allwinner SoCs display driver.
+ *
+ * Copyright (C) 2016 Allwinner.
+ *
+ * This file is licensed under the terms of the GNU General Public
+ * License version 2.  This program is licensed "as is" without any
+ * warranty of any kind, whether express or implied.
+ */
+
+#ifndef __TO20T20000_PANEL_H__
+#define  __TO20T20000_PANEL_H__
+
+#include "panels.h"
+
+extern struct __lcd_panel to20t20000_panel;
+
+#endif
