diff -drupN a/drivers/video/fbdev/sunxi/disp2/disp/de/lowlevel_sun8iw8/de_vep_table.h b/drivers/video/fbdev/sunxi/disp2/disp/de/lowlevel_sun8iw8/de_vep_table.h
--- a/drivers/video/fbdev/sunxi/disp2/disp/de/lowlevel_sun8iw8/de_vep_table.h	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/video/fbdev/sunxi/disp2/disp/de/lowlevel_sun8iw8/de_vep_table.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,22 @@
+/*
+ * Allwinner SoCs display driver.
+ *
+ * Copyright (C) 2016 Allwinner.
+ *
+ * This file is licensed under the terms of the GNU General Public
+ * License version 2.  This program is licensed "as is" without any
+ * warranty of any kind, whether express or implied.
+ */
+#ifndef __DE_VEP_TAB__
+#define __DE_VEP_TAB__
+
+extern int y2r[192];
+extern int r2y[128];
+extern int y2y[64];
+extern int r2r[32];
+extern int bypass_csc[12];
+extern unsigned int sin_cos[128];
+extern int fcc_range_gain[6];
+extern unsigned char ce_bypass_lut[256];
+
+#endif
