diff -drupN a/modules/nand/sun50iw1p1/nfd/nand_test_dev.h b/modules/nand/sun50iw1p1/nfd/nand_test_dev.h
--- a/modules/nand/sun50iw1p1/nfd/nand_test_dev.h	1970-01-01 03:00:00.000000000 +0300
+++ b/modules/nand/sun50iw1p1/nfd/nand_test_dev.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,45 @@
+/*
+************************************************************************************************************************
+*                                                      eNand
+*                                     Nand flash driver logic control module define
+*
+*                             Copyright(C), 2008-2009, SoftWinners Microelectronic Co., Ltd.
+*											       All Rights Reserved
+*
+* File Name : bsp_nand.h
+*
+* Author : Kevin.z
+*
+* Version : v0.1
+*
+* Date : 2008.03.25
+*
+* Description :
+*
+*
+* Others : None at present.
+*
+*
+* History :
+*
+*  <Author>        <time>       <version>      <description>
+*
+* Kevin.z         2008.03.25      0.1          build the file
+*
+************************************************************************************************************************
+*/
+#ifndef __BSP_NAND_TEST_H__
+#define __BSP_NAND_TEST_H__
+
+extern int nand_driver_test_init(void);
+extern int nand_driver_test_exit(void);
+extern unsigned int get_nftl_num(void);
+extern unsigned int get_nftl_cap(void);
+extern unsigned int get_first_nftl_cap(void);
+extern unsigned int nftl_test_read(unsigned int start_sector, unsigned int len,
+				   unsigned char *buf);
+extern unsigned int nftl_test_write(unsigned int start_sector, unsigned int len,
+				    unsigned char *buf);
+extern unsigned int nftl_test_flush_write_cache(void);
+
+#endif
