diff -drupN a/fs/notify/fanotify/fanotify_user.c b/fs/notify/fanotify/fanotify_user.c
--- a/fs/notify/fanotify/fanotify_user.c	2018-08-06 17:23:04.000000000 +0300
+++ b/fs/notify/fanotify/fanotify_user.c	2022-06-12 05:28:14.000000000 +0300
@@ -498,7 +498,7 @@ static int fanotify_find_path(int dfd, c
 	}
 
 	/* you can only watch an inode if you have read permissions on it */
-	ret = inode_permission(path->dentry->d_inode, MAY_READ);
+	ret = inode_permission2(path->mnt, path->dentry->d_inode, MAY_READ);
 	if (ret)
 		path_put(path);
 out:
