diff -drupN a/net/core/sock_diag.c b/net/core/sock_diag.c
--- a/net/core/sock_diag.c	2018-08-06 17:23:04.000000000 +0300
+++ b/net/core/sock_diag.c	2022-06-12 05:28:14.000000000 +0300
@@ -19,7 +19,7 @@ static int (*inet_rcv_compat)(struct sk_
 static DEFINE_MUTEX(sock_diag_table_mutex);
 static struct workqueue_struct *broadcast_wq;
 
-static u64 sock_gen_cookie(struct sock *sk)
+u64 sock_gen_cookie(struct sock *sk)
 {
 	while (1) {
 		u64 res = atomic64_read(&sk->sk_cookie);
