diff -drupN a/kernel/futex.c b/kernel/futex.c
--- a/kernel/futex.c	2018-08-06 17:23:04.000000000 +0300
+++ b/kernel/futex.c	2022-06-12 05:28:14.000000000 +0300
@@ -1989,6 +1989,9 @@ retry_private:
 		drop_count++;
 	}
 
+	if (nr_wake < 0 || nr_requeue < 0)
+		return -EINVAL;
+
 	/*
 	 * We took an extra initial reference to the pi_state either
 	 * in futex_proxy_trylock_atomic() or in lookup_pi_state(). We
