diff -drupN a/arch/arm/boot/dts/sun8iw8p1-soc.dts b/arch/arm/boot/dts/sun8iw8p1-soc.dts
--- a/arch/arm/boot/dts/sun8iw8p1-soc.dts	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/arm/boot/dts/sun8iw8p1-soc.dts	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,91 @@
+/*
+ * Allwinner Technology CO., Ltd. sunivw1p1 evb board.
+ * Modified by czy
+ * evb board support.
+ */
+
+/dts-v1/;
+#include "sun8iw8p1-soc.dtsi"
+
+/ {
+	aliases {
+		spinand = &spinand;
+		twi0 = &twi0;
+	};
+
+	memory@40000000 {
+		device_type = "memory";
+		reg = <0x00000000 0x40000000 0x00000000 0x8000000>;  /*128M*/
+	};
+
+	spinand: spi@01c68000 {
+		compatible = "allwinner,sunxi-spinand";
+		device_type = "spinand";
+		reg = <0x0 0x01c68000 0x0 0x1000>;
+		interrupts = <GIC_SPI 65 IRQ_TYPE_LEVEL_HIGH>;
+		clocks = <&clk_pll_periph0>, <&clk_spi0>;
+		pinctrl-names = "default", "sleep";
+		pinctrl-0 = <&spi0_pins_a &spi0_pins_b>;
+		pinctrl-1 = <&spi0_pins_c>;
+		nand0_regulator1 = "vcc-nand";
+		nand0_regulator2 = "none";
+		nand0_cache_level = <0x55aaaa55>;
+		nand0_flush_cache_num = <0x55aaaa55>;
+		nand0_capacity_level = <0x55aaaa55>;
+		nand0_id_number_ctl = <0x55aaaa55>;
+		nand0_print_level = <0x55aaaa55>;
+		nand0_p0 = <0x55aaaa55>;
+		nand0_p1 = <0x55aaaa55>;
+		nand0_p2 = <0x55aaaa55>;
+		nand0_p3 = <0x55aaaa55>;
+		status = "okay";
+	};
+
+	soc@01c00000 {
+	twi0: twi@0x01c2ac00{
+		status = "okay";
+		ac100: ac100@0{
+			#address-cells = <1>;
+			#size-cells = <0>;
+			compatible = "allwinner,sunxi-ac100";
+			reg = <0x1a>;
+			status = "okay";
+		};
+	};
+
+	ac100codec:ac100codec {
+		compatible = "allwinner,sunxi-ac100-codec";
+		gpio-spk = <&pio PB 9 1 1 1 1>;
+		speaker_double_used = <0x0>;
+		double_speaker_val = <0x1b>;
+		single_speaker_val = <0x1b>;
+		headset_val = <0x3b>;
+		earpiece_val = <0x1e>;
+		mainmic_val = <0x4>;
+		headsetmic_val = <0x4>;
+		dmic_used = <0x0>;
+		adc_digital_val = <0xc0c0>;
+		agc_used = <0x1>;
+		drc_used = <0x1>;
+		aif2_lrck_div = <0x100>;
+		aif2_bclk_div = <0xc>;
+		status = "okay";
+	};
+	bbdai:bbdai-controller@0x01c22c00 {
+		compatible = "allwinner,sunxi-bbdai";
+		reg = <0x0 0x01c22c00 0x0 0x2bc>;/*digital baseadress*/
+		status = "okay";
+	};
+	snddaudio0:sound@1 {
+		daudio_select = <0x01>;
+		daudio_master = <0x04>;
+		audio_format = <0x01>;
+		signal_inversion = <0x01>;
+		analog_bb = <0x00>;
+		digital_bb = <0x00>;
+		status = "okay";
+		sunxi,snddaudio-codec = "ac100-codec";
+		sunxi,bbdai-controller = <&bbdai>;
+	};
+};
+};
