diff -drupN a/mm/Makefile b/mm/Makefile
--- a/mm/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/mm/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -21,7 +21,9 @@ KCOV_INSTRUMENT_memcontrol.o := n
 KCOV_INSTRUMENT_mmzone.o := n
 KCOV_INSTRUMENT_vmstat.o := n
 
+ifneq ($(CONFIG_MMU),y)
 mmu-y			:= nommu.o
+endif
 mmu-$(CONFIG_MMU)	:= gup.o highmem.o memory.o mincore.o \
 			   mlock.o mmap.o mprotect.o mremap.o msync.o rmap.o \
 			   vmalloc.o pagewalk.o pgtable-generic.o
