diff -drupN a/drivers/rtc/Kconfig b/drivers/rtc/Kconfig
--- a/drivers/rtc/Kconfig	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/rtc/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -76,7 +76,11 @@ config RTC_DEBUG
 	help
 	  Say yes here to enable debugging support in the RTC framework
 	  and individual RTC drivers.
-
+config RTC_SHUTDOWN_ALARM
+	bool "RTC shutdown alarm support"
+	help
+	  Say yes here to enable shutdown alarm support in the RTC framework
+	  and individual RTC drivers.
 comment "RTC interfaces"
 
 config RTC_INTF_SYSFS
@@ -1431,20 +1435,36 @@ config RTC_DRV_SUN4V
 	  If you say Y here you will get support for the Hypervisor
 	  based RTC on SUN4V systems.
 
-config RTC_DRV_SUN6I
-	bool "Allwinner A31 RTC"
-	default MACH_SUN6I || MACH_SUN8I || COMPILE_TEST
+config RTC_DRV_SUNXI
+	tristate "Allwinner sunxi RTC"
 	depends on ARCH_SUNXI
 	help
-	  If you say Y here you will get support for the RTC found in
-	  some Allwinner SoCs like the A31 or the A64.
+	  If you say Y here you will get support for the RTC found on
+	  Allwinner sun8i/sun50i.
 
-config RTC_DRV_SUNXI
-	tristate "Allwinner sun4i/sun7i RTC"
-	depends on MACH_SUN4I || MACH_SUN7I || COMPILE_TEST
+config SUNXI_BOOTUP_EXTEND
+        depends on RTC_DRV_SUNXI
+        bool "Sunxi rtc reboot Feature"
+        help
+          Support sunxi rtc reboot, which can save state in rtc reg
+          according to the reboot arg.
+          This feature need support of reboot program in the userspace,
+          such as android reboot. Busybox reboot doesn't support this
+          feature.
+
+config SUNXI_TIMER_ALARM
+	tristate "Allwinner sunxi timer alarm"
 	help
-	  If you say Y here you will get support for the RTC found on
-	  Allwinner A10/A20.
+	  Use timer1 as an alarm which could wake up CPU from sleeping.
+	  The alarm interval can be set via /sys/class/timer_alarm/interval.
+	  This is useful on the Socs which does not have an RTC alarm.
+
+config SUNXI_RTC_BOOTCOUNT
+	depends on RTC_DRV_SUNXI
+	default no
+	bool "Sunxi rtc general register save bootcount"
+	help
+		Support sunxi rtc bootcount, which save bootcount in rtc reg
 
 config RTC_DRV_STARFIRE
 	bool "Starfire RTC"
