diff -drupN a/arch/x86/oprofile/nmi_int.c b/arch/x86/oprofile/nmi_int.c
--- a/arch/x86/oprofile/nmi_int.c	2018-08-06 17:23:04.000000000 +0300
+++ b/arch/x86/oprofile/nmi_int.c	2022-06-12 05:28:14.000000000 +0300
@@ -615,7 +615,7 @@ enum __force_cpu_type {
 
 static int force_cpu_type;
 
-static int set_cpu_type(const char *str, struct kernel_param *kp)
+static int set_cpu_type(const char *str, const struct kernel_param *kp)
 {
 	if (!strcmp(str, "timer")) {
 		force_cpu_type = timer;
