diff -drupN a/modules/nand/sun50iw11p1/phy-nand/spinand/Makefile b/modules/nand/sun50iw11p1/phy-nand/spinand/Makefile
--- a/modules/nand/sun50iw11p1/phy-nand/spinand/Makefile	1970-01-01 03:00:00.000000000 +0300
+++ b/modules/nand/sun50iw11p1/phy-nand/spinand/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,22 @@
+#
+#
+# Makefile for spinand
+#
+#
+
+#obj-$(CONFIG_SUNXI_NAND) += nand_format.o
+#obj-$(CONFIG_SUNXI_NAND) += nand_id.o
+#obj-$(CONFIG_SUNXI_NAND) += nand_phy.o
+#obj-$(CONFIG_SUNXI_NAND) += nand_physic_interface_spinand.o
+#obj-$(CONFIG_SUNXI_NAND) += nand_scan.o
+#obj-$(CONFIG_SUNXI_NAND) += spic_op.o
+obj-$(CONFIG_SUNXI_NAND) += spinand_boot.o
+obj-$(CONFIG_SUNXI_NAND) += spinand.o
+
+#obj-y += spinand0/
+#obj-y += spinand1/
+#obj-y += spinand2/
+#obj-y += spinand3/
+#obj-y += spinand4/
+#obj-y += spinand5/
+#obj-y += spinand6/
