diff -drupN a/arch/arm/boot/dts/sun8iw6p1-pinctrl.dtsi b/arch/arm/boot/dts/sun8iw6p1-pinctrl.dtsi
--- a/arch/arm/boot/dts/sun8iw6p1-pinctrl.dtsi	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/arm/boot/dts/sun8iw6p1-pinctrl.dtsi	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,501 @@
+/*
+ * Allwinner sun8iw6p1 pin config info.
+ */
+
+/ {
+	soc@01c00000{
+		pio: pinctrl@01c20800 {
+			compatible = "allwinner,sun8iw6p1-pinctrl";
+			reg = <0x0 0x01c20800 0x0 0x400>;
+			interrupts = <GIC_SPI 15 4>,
+				     <GIC_SPI 17 4>,
+				     <GIC_SPI 100 4>;
+			device_type = "pio";
+			clocks = <&clk_pio>;
+			gpio-controller;
+			interrupt-controller;
+			#interrupt-cells = <3>;
+			#size-cells = <0>;
+			#gpio-cells = <6>;
+
+			vdevice_pins_a: vdevice@0 {
+				allwinner,pins = "PB0", "PB1";
+				allwinner,function = "vdevice";
+				allwinner,muxsel = <5>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+				allwinner,data = <1>;
+			};
+
+			uart0_pins_a: uart0@0 {
+				allwinner,pins = "PB9", "PB10";
+				allwinner,pname = "uart0_tx", "uart0_rx";
+				allwinner,function = "uart0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart0_pins_b: uart0@1 {
+				allwinner,pins = "PB9", "PB10";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart1_pins_a: uart1@0 {
+				allwinner,pins = "PC6", "PC7", "PC8", "PC9";
+				allwinner,pname = "uart1_tx", "uart1_rx",
+						  "uart1_rts", "uart1_cts";
+				allwinner,function = "uart1";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart1_pins_b: uart1@1 {
+				allwinner,pins = "PC6", "PC7", "PC8",
+						 "PC9", "PC10", "PC11",
+						 "PC12", "PC13";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart2_pins_a: uart2@0 {
+				allwinner,pins = "PB0", "PB1", "PB2", "PB3";
+				allwinner,pname = "uart2_tx", "uart2_rx",
+						  "uart2_rts", "uart2_cts";
+				allwinner,function = "uart2";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart2_pins_b: uart2@1 {
+				allwinner,pins = "PB0", "PB1", "PB2", "PB3";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart3_pins_a: uart3@0 {
+				allwinner,pins = "PG10", "PG11", "PG12", "PG13";
+				allwinner,pname = "uart3_tx", "uart3_rx",
+						  "uart3_rts", "uart3_cts";
+				allwinner,function = "uart3";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart3_pins_b: uart3@1 {
+				allwinner,pins = "PG10", "PG11", "PG12", "PG13";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart4_pins_a: uart4@0 {
+				allwinner,pins = "PE10", "PE11", "PE12", "PE13";
+				allwinner,pname = "uart4_tx", "uart4_rx",
+						  "uart4_rts", "uart4_cts";
+				allwinner,function = "uart4";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart4_pins_b: uart4@1 {
+				allwinner,pins = "PE10", "PE11", "PE12", "PE13";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			twi0_pins_a: twi0@0 {
+				allwinner,pins = "PH0", "PH1";
+				allwinner,pname = "twi0_scl", "twi0_sda";
+				allwinner,function = "twi0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi0_pins_b: twi0@1 {
+				allwinner,pins = "PH0", "PH1";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi1_pins_a: twi1@0 {
+				allwinner,pins = "PH2", "PH3";
+				allwinner,pname = "twi1_scl", "twi1_sda";
+				allwinner,function = "twi1";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi1_pins_b: twi1@1 {
+				allwinner,pins = "PH2", "PH3";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi2_pins_a: twi2@0 {
+				allwinner,pins = "PH4", "PH5";
+				allwinner,pname = "twi2_scl", "twi2_sda";
+				allwinner,function = "twi2";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi2_pins_b: twi2@1 {
+				allwinner,pins = "PH4", "PH5";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			hdmi_pins_active: hdmi@0 {
+				allwinner,pins = "PH6", "PH7", "PH8";
+				allwinner,function = "hdmi";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			hdmi_pins_sleep: hdmi@1 {
+				allwinner,pins = "PH6", "PH7", "PH8";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			spi0_pins_a: spi0@0 {
+				allwinner,pins = "PC0", "PC1", "PC2";
+				allwinner,pname = "spi0_mosi", "spi0_miso",
+						  "spi0_sclk";
+				allwinner,function = "spi0";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			spi0_pins_b: spi0@1 {
+				allwinner,pins = "PC3";
+				allwinner,pname = "spi0_cs0";
+				allwinner,function = "spi0";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			spi0_pins_c: spi0@2 {
+				allwinner,pins = "PC0", "PC1", "PC2", "PC3";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			spi1_pins_a: spi1@0 {
+				allwinner,pins = "PG8", "PG9", "PG7";
+				allwinner,pname = "spi1_mosi", "spi1_miso",
+						  "spi1_sclk";
+				allwinner,function = "spi1";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			spi1_pins_b: spi1@1 {
+				allwinner,pins = "PG6";
+				allwinner,pname = "spi1_cs0";
+				allwinner,function = "spi1";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			spi1_pins_c: spi1@2 {
+				allwinner,pins = "PG8", "PG9", "PG7", "PG6";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			sdc0_pins_a: sdc0@0 {
+				allwinner,pins = "PF0", "PF1", "PF2",
+						 "PF3","PF4","PF5";
+				allwinner,function = "sdc0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			sdc0_pins_b: sdc0@1 {
+				allwinner,pins = "PF0", "PF1", "PF2",
+						 "PF3","PF4","PF5";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			sdc1_pins_a: sdc1@0 {
+				allwinner,pins = "PG0", "PG1", "PG2", "PG3",
+						 "PG4","PG5";
+				allwinner,function = "sdc1";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <3>;
+				allwinner,pull = <1>;
+			};
+
+			sdc1_pins_b: sdc1@1 {
+				allwinner,pins = "PG0", "PG1", "PG2", "PG3",
+						 "PG4","PG5";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			sdc2_pins_a: sdc2@0 {
+				allwinner,pins = "PC5","PC6", "PC8", "PC9",
+						 "PC10","PC11","PC12","PC13",
+						 "PC14","PC15","PC16";
+				allwinner,function = "sdc2";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <2>;
+				allwinner,pull = <1>;
+			};
+
+			sdc2_pins_b: sdc2@1 {
+				allwinner,pins = "PC5","PC6", "PC8", "PC9",
+						 "PC10","PC11","PC12","PC13",
+						 "PC14","PC15","PC16";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			daudio0_pins_a: daudio0@0 {
+				allwinner,pins = "PB6", "PB8", "PB4",
+						 "PB5", "PB7";
+				allwinner,function = "tdm0";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+			daudio0_pins_b: daudio0_sleep@0 {
+				allwinner,pins = "PB6", "PB8", "PB4",
+						 "PB5", "PB7";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+			i2s0_pins_a: i2s0@0 {
+				allwinner,pins = "PB4", "PB5", "PB6", "PB7", "PB8";
+				allwinner,function = "i2s0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+			i2s0_pins_b: i2s0_sleep@0 {
+				allwinner,pins = "PB4", "PB5", "PB6", "PB7", "PB8";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+			i2s1_pins_a: i2s1@0 {
+				allwinner,pins = "PG12", "PG11", "PG10", "PG13";
+				allwinner,function = "i2s1";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+			i2s1_pins_b: i2s1_sleep@0 {
+				allwinner,pins = "PG12", "PG11", "PG10", "PG13";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+			spdif_pins_a: spdif@0 {
+				allwinner,pins = "PE18";
+				allwinner,function = "spdif0";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+			spdif_pins_b: spdif_sleep@0 {
+				allwinner,pins = "PE18";/*pe18-spec*/
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			csi0_pins_a: csi0@0 {
+				allwinner,pins = "PE0", "PE1", "PE2", "PE3",
+						 "PE4", "PE5", "PE6", "PE7",
+						 "PE8", "PE9", "PE10", "PE11",
+						 "PE12", "PE13", "PE14","PE15";
+				allwinner,pname = "csi0_pck", "csi0_mck",
+						  "csi0_hsync", "csi0_vsync",
+						  "csi0_d0", "csi0_d1",
+						  "csi0_d2", "csi0_d3",
+						  "csi0_d4", "csi0_d5",
+						  "csi0_d6", "csi0_d7",
+						  "csi0_d8", "csi0_d9",
+						  "csi0_sck", "csi0_sda";
+				allwinner,function = "csi0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+			csi0_pins_b: csi0@1 {
+				allwinner,pins = "PE0", "PE1", "PE2", "PE3",
+						 "PE4", "PE5", "PE6", "PE7",
+						 "PE8", "PE9", "PE10", "PE11",
+						 "PE12", "PE13", "PE14", "PE15";
+				allwinner,pname = "csi0_pck", "csi0_mck",
+						  "csi0_hsync", "csi0_vsync",
+						  "csi0_d0", "csi0_d1",
+						  "csi0_d2", "csi0_d3",
+						  "csi0_d4", "csi0_d5",
+						  "csi0_d6", "csi0_d7",
+						  "csi0_d8", "csi0_d9",
+						  "csi0_sck", "csi0_sda";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+
+			scr_pins_a: smartcard@0 {
+				allwinner,pins = "PB1", "PB4", "PB5", "PB6",
+						 "PB7";
+				allwinner,function = "sim0";
+				allwinner,muxsel = <5>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			nand0_pins_a: nand0@0 {
+				allwinner,pins = "PC0", "PC1", "PC2", "PC4",
+						 "PC6", "PC7", "PC8", "PC9",
+						 "PC10", "PC11", "PC12",
+						 "PC13", "PC14";
+				allwinner,pname= "nand0_we", "nand0_ale",
+						 "nand0_cle", "nand0_nre",
+						 "nand0_d0", "nand0_d1",
+						 "nand0_d2", "nand0_d3",
+						 "nand0_d4", "nand0_d5",
+						 "nand0_d6", "nand0_d7",
+						 "nand0_ndqs";
+				allwinner,function = "nand0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			nand0_pins_b: nand0@1 {
+				allwinner,pins  = "PC3",       "PC5";
+				allwinner,pname = "nand0_ce0", "nand0_rb0";
+				allwinner,function = "nand0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			nand0_pins_c: nand0@2 {
+				allwinner,pins = "PC0", "PC1", "PC2", "PC3",
+						 "PC4", "PC5", "PC6", "PC7",
+						 "PC8", "PC9", "PC10", "PC11",
+						 "PC12", "PC13", "PC14";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel   = <7>;
+				allwinner,drive    = <1>;
+				allwinner,pull     = <0>;
+			};
+
+			gmac_pins_a: gmac@0 {
+				allwinner,pins = "PD8", "PD9", "PD10", "PD11",
+						"PD12", "PD13", "PD14", "PD15",
+						"PD16", "PD17", "PD18", "PD19",
+						"PD20", "PD21", "PD22", "PD23";
+				allwinner,function = "gmac0";
+				allwinner,muxsel   = <4>;
+				allwinner,drive    = <3>;
+				allwinner,pull     = <0>;
+			};
+		};
+
+		r_pio: pinctrl@01f02c00 {
+			compatible = "allwinner,sun8iw6p1-r-pinctrl";
+			reg = <0x0 0x01f02c00 0x0 0x400>;
+			interrupts = <GIC_SPI 45 4>;
+			clocks = <&clk_cpurpio>;
+			device_type = "r_pio";
+			gpio-controller;
+			interrupt-controller;
+			#interrupt-cells = <2>;
+			#size-cells = <0>;
+			#gpio-cells = <6>;
+
+			s_uart0_pins_a: s_uart0@0 {
+				allwinner,pins = "PL2", "PL3";
+				allwinner,function = "s_uart0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			s_rsb0_pins_a: s_rsb0@0 {
+				allwinner,pins = "PL0", "PL1";
+				allwinner,function = "s_rsb0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <2>;
+				allwinner,pull = <1>;
+			};
+
+			s_jtag0_pins_a: s_jtag0@0 {
+				allwinner,pins = "PL4", "PL5", "PL6", "PL7";
+				allwinner,function = "s_jtag0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <2>;
+				allwinner,pull = <1>;
+			};
+
+			s_cir0_pins_a: s_cir0@0 {
+				allwinner,pins = "PL12";
+				allwinner,function = "s_cir0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <2>;
+				allwinner,pull = <1>;
+			};
+		};
+	};
+
+};
