diff -drupN a/arch/arm/boot/dts/sun8iw8p1-soc-pinctrl.dtsi b/arch/arm/boot/dts/sun8iw8p1-soc-pinctrl.dtsi
--- a/arch/arm/boot/dts/sun8iw8p1-soc-pinctrl.dtsi	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/arm/boot/dts/sun8iw8p1-soc-pinctrl.dtsi	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,228 @@
+/*
+ * Allwinner sun8iw8p1 pin config info.
+ * Modified by czy
+ */
+
+/ {
+	soc: soc@01c00000 {
+		pio: pinctrl@01c20800 {
+			compatible = "allwinner,sun8iw8p1-pinctrl";
+			reg = <0x0 0x01c20800 0x0 0x400>;
+			device_type = "pio";
+			clocks = <&clk_pio>;
+			gpio-controller;
+			interrupt-controller;
+			#interrupt-cells = <2>;
+			#size-cells = <0>;
+			#gpio-cells = <6>;
+
+			interrupts = <GIC_SPI 15 4>,
+				     <GIC_SPI 17 4>;
+
+			uart0_pins_a: uart0@0 {
+				allwinner,pins = "PB8", "PB9";
+				allwinner,pname = "uart0_tx", "uart0_rx";
+				allwinner,function = "uart0";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart0_pins_b: uart0@1 {
+				allwinner,pins = "PB8", "PB9";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart1_pins_a: uart1@0 {
+				allwinner,pins = "PE21", "PE22";
+				allwinner,pname = "uart1_tx", "uart1_rx";
+				allwinner,function = "uart1";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart1_pins_b: uart1@1 {
+				allwinner,pins = "PE21", "PE22";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart2_pins_a: uart2@0 {
+				allwinner,pins = "PB0", "PB1";
+				allwinner,pname = "uart2_tx", "uart2_rx";
+				allwinner,function = "uart2";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart2_pins_b: uart2@1 {
+				allwinner,pins = "PB0", "PB1";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			twi0_pins_a: twi0@0 {
+				allwinner,pins = "PB6", "PB7";
+				allwinner,pname = "twi0_scl", "twi0_sda";
+				allwinner,function = "twi0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi0_pins_b: twi0@1 {
+				allwinner,pins = "PB6", "PB7";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi1_pins_a: twi1@0 {
+				allwinner,pins = "PE21", "PE22";
+				allwinner,pname = "twi1_scl", "twi1_sda";
+				allwinner,function = "twi1";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi1_pins_b: twi1@1 {
+				allwinner,pins = "PE21", "PE22";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			spi0_pins_a: spi0@0 {
+				allwinner,pins = "PC3", "PC0", "PC1";
+				allwinner,pname = "spi0_mosi", "spi0_miso", "spi0_sclk";
+				allwinner,function = "spi0";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			spi0_pins_b: spi0@1 {
+				allwinner,pins = "PC2";
+				allwinner,pname = "spi0_cs0";
+				allwinner,function = "spi0";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>; 	// only CS should be pulled up
+			};
+
+			spi0_pins_c: spi0@2 {
+				allwinner,pins = "PC3", "PC0", "PC1","PC2";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <3>;
+				allwinner,pull = <1>;
+			};
+
+			csi0_pins_a: csi0@0 {
+				allwinner,pins = "PE21","PE22";
+				allwinner,pname = "csi0_sck","csi0_sda";
+				allwinner,function = "csi0";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+			csi0_pins_b: csi0_sleep@0 {
+				allwinner,pins = "PE21","PE22";
+				allwinner,pname = "csi0_sck","csi0_sda";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			sdc0_pins_a: sdc0@0 {
+				allwinner,pins = "PF0","PF1","PF2","PF3","PF4","PF5";
+				allwinner,function = "sdc0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			sdc0_pins_b: sdc0@1 {
+				allwinner,pins = "PF0","PF1","PF2","PF3","PF4","PF5";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			sdc1_pins_a: sdc1@0 {
+				allwinner,pins = "PG0","PG1","PG2","PG3","PG4","PG5";
+				allwinner,function = "sdc1";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			sdc1_pins_b: sdc1@1 {
+				allwinner,pins = "PG0","PG1","PG2","PG3","PG4","PG5";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+			daudio0_pins_a:daudio0@0 {
+				allwinner,pins = "PG10","PG11","PG12","PG13";
+				allwinner,function = "tdm0";
+				allwinner,muxsel = <2>;
+				allwinner,driver = <1>;
+				allwinner,pull = <1>;
+			};
+			daudio0_pins_b:daudio0_sleep@0 {
+				allwinner,pins = "PG10","PG11","PG12","PG13";
+				allwinner,function = "io_disable";
+				allwinner,muxsel = <7>;
+				allwinner,driver = <1>;
+				allwinner,pull = <1>;
+			};
+			pwm0_pins_a: pwm0@0 {
+				allwinner,pins = "PB4";
+				allwinner,function = "pwm0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			pwm0_pins_b: pwm0@1 {
+				allwinner,pins = "PB4";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			pwm1_pins_a: pwm1@0 {
+				allwinner,pins = "PB5";
+				allwinner,function = "pwm1";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			pwm1_pins_b: pwm1@1 {
+				allwinner,pins = "PB5";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+		};
+	};
+};
