diff -drupN a/drivers/hwtracing/coresight/coresight-priv.h b/drivers/hwtracing/coresight/coresight-priv.h
--- a/drivers/hwtracing/coresight/coresight-priv.h	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/hwtracing/coresight/coresight-priv.h	2022-06-12 05:28:14.000000000 +0300
@@ -111,7 +111,9 @@ static inline void CS_UNLOCK(void __iome
 void coresight_disable_path(struct list_head *path);
 int coresight_enable_path(struct list_head *path, u32 mode);
 struct coresight_device *coresight_get_sink(struct list_head *path);
-struct list_head *coresight_build_path(struct coresight_device *csdev);
+struct coresight_device *coresight_get_enabled_sink(bool reset);
+struct list_head *coresight_build_path(struct coresight_device *csdev,
+				       struct coresight_device *sink);
 void coresight_release_path(struct list_head *path);
 
 #ifdef CONFIG_CORESIGHT_SOURCE_ETM3X
