diff -drupN a/drivers/gpu/drm/drm_fops.c b/drivers/gpu/drm/drm_fops.c
--- a/drivers/gpu/drm/drm_fops.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/gpu/drm/drm_fops.c	2022-06-12 05:28:14.000000000 +0300
@@ -664,6 +664,10 @@ void drm_event_cancel_free(struct drm_de
 		list_del(&p->pending_link);
 	}
 	spin_unlock_irqrestore(&dev->event_lock, flags);
+
+	if (p->fence)
+		fence_put(p->fence);
+
 	kfree(p);
 }
 EXPORT_SYMBOL(drm_event_cancel_free);
