diff -drupN a/arch/arm/mach-sunxi/include/mach/timex.h b/arch/arm/mach-sunxi/include/mach/timex.h
--- a/arch/arm/mach-sunxi/include/mach/timex.h	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/arm/mach-sunxi/include/mach/timex.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,24 @@
+/*
+ *  arch/arm/mach-sunxi/include/mach/timex.h
+ *
+ * Copyright(c) 2013-2016 Allwinnertech Co., Ltd.
+ * Author: Benn Huang <benn@allwinnertech.com>
+ *	   Sugar <shuge@allwinnertech.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#ifndef __SUNXI_ARCH_TIMEX_H
+#define __SUNXI_ARCH_TIMEX_H
+
+#define CLOCK_TICK_RATE 24000000
+
+#endif
