diff -drupN a/drivers/isdn/hardware/mISDN/speedfax.c b/drivers/isdn/hardware/mISDN/speedfax.c
--- a/drivers/isdn/hardware/mISDN/speedfax.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/isdn/hardware/mISDN/speedfax.c	2022-06-12 05:28:14.000000000 +0300
@@ -94,7 +94,7 @@ _set_debug(struct sfax_hw *card)
 }
 
 static int
-set_debug(const char *val, struct kernel_param *kp)
+set_debug(const char *val, const struct kernel_param *kp)
 {
 	int ret;
 	struct sfax_hw *card;
