diff -drupN a/drivers/thermal/sunxi_thermal/sunxi_cooling_device/Kconfig b/drivers/thermal/sunxi_thermal/sunxi_cooling_device/Kconfig
--- a/drivers/thermal/sunxi_thermal/sunxi_cooling_device/Kconfig	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/thermal/sunxi_thermal/sunxi_cooling_device/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,38 @@
+#
+# SUNXI cooling device drivers configuration
+#
+config SUNXI_BUDGET_COOLING
+	tristate "sunxi budget cooling"
+	depends on ARCH_SUNXI
+	depends on CPU_FREQ
+	help
+	  Adds sunxi cpu budget cooling devices, and these cooling devices can be
+	  bound to thermal zone trip points. When a trip point reached, the
+	  bound cpufreq cooling device turns active to set CPU frequency low to
+	  cool down the CPU.
+
+config SUNXI_BUDGET_COOLING_CPU_DVFS
+	tristate "sunxi budget cooling about CPU-DVFS"
+	depends on THERMAL_OF
+	depends on CPU_FREQ
+	depends on SUNXI_BUDGET_COOLING
+	help
+	  Adds CPU cooling device about CPU dynamic-voltage-frequence-scaling.
+
+config SUNXI_BUDGET_COOLING_HOTPLUG
+	tristate "sunxi budget cooling about Hotplug"
+	depends on THERMAL_OF
+	depends on HOTPLUG_CPU
+	depends on SUNXI_BUDGET_COOLING
+	help
+	  Adds CPU cooling device use hotplug limit.
+
+config SUNXI_GPU_COOLING
+	tristate "sunxi gpu cooling"
+	depends on THERMAL_OF
+	depends on SUNXI_BUDGET_COOLING
+	help
+	  Adds sunxi gpu cooling devices, and these cooling devices can be
+	  bound to thermal zone trip points. When a trip point reached, the
+	  gpu cooling device turns active to set gpu frequency low to
+	  cool down the CPU.
