diff -drupN a/drivers/iio/imu/inv_mpu6050/inv_mpu_iio.h b/drivers/iio/imu/inv_mpu6050/inv_mpu_iio.h
--- a/drivers/iio/imu/inv_mpu6050/inv_mpu_iio.h	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/iio/imu/inv_mpu6050/inv_mpu_iio.h	2022-06-12 05:28:14.000000000 +0300
@@ -69,6 +69,7 @@ struct inv_mpu6050_reg_map {
 /*device enum */
 enum inv_devices {
 	INV_MPU6050,
+	INV_MPU6050A,
 	INV_MPU6500,
 	INV_MPU6000,
 	INV_MPU9150,
@@ -140,6 +141,7 @@ struct inv_mpu6050_state {
 	struct iio_mount_matrix orientation;
 	DECLARE_KFIFO(timestamps, long long, TIMESTAMP_FIFO_SIZE);
 	struct regmap *map;
+	struct work_struct		resume_work;
 	int irq;
 };
 
@@ -219,14 +221,16 @@ struct inv_mpu6050_state {
 /* init parameters */
 #define INV_MPU6050_INIT_FIFO_RATE           50
 #define INV_MPU6050_TIME_STAMP_TOR           5
-#define INV_MPU6050_MAX_FIFO_RATE            1000
+#define INV_MPU6050_MAX_FIFO_RATE            8000
 #define INV_MPU6050_MIN_FIFO_RATE            4
 #define INV_MPU6050_ONE_K_HZ                 1000
+#define INV_MPU6050_8K_HZ                 8000
 
 #define INV_MPU6050_REG_WHOAMI			117
 
 #define INV_MPU6000_WHOAMI_VALUE		0x68
 #define INV_MPU6050_WHOAMI_VALUE		0x68
+#define INV_MPU6050A_WHOAMI_VALUE		0x98
 #define INV_MPU6500_WHOAMI_VALUE		0x70
 #define INV_MPU9150_WHOAMI_VALUE		0x68
 #define INV_ICM20608_WHOAMI_VALUE		0xAF
