diff -drupN a/drivers/pinctrl/vt8500/pinctrl-wmt.c b/drivers/pinctrl/vt8500/pinctrl-wmt.c
--- a/drivers/pinctrl/vt8500/pinctrl-wmt.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/pinctrl/vt8500/pinctrl-wmt.c	2022-06-12 05:28:14.000000000 +0300
@@ -429,7 +429,7 @@ static int wmt_pinconf_set(struct pinctr
 {
 	struct wmt_pinctrl_data *data = pinctrl_dev_get_drvdata(pctldev);
 	enum pin_config_param param;
-	u16 arg;
+	u32 arg;
 	u32 bank = WMT_BANK_FROM_PIN(pin);
 	u32 bit = WMT_BIT_FROM_PIN(pin);
 	u32 reg_pull_en = data->banks[bank].reg_pull_en;
