diff -drupN a/drivers/staging/android/uapi/ion.h b/drivers/staging/android/uapi/ion.h
--- a/drivers/staging/android/uapi/ion.h	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/staging/android/uapi/ion.h	2022-06-12 05:28:14.000000000 +0300
@@ -40,12 +40,18 @@ enum ion_heap_type {
 	ION_HEAP_TYPE_CARVEOUT,
 	ION_HEAP_TYPE_CHUNK,
 	ION_HEAP_TYPE_DMA,
-	ION_HEAP_TYPE_CUSTOM, /*
+	ION_HEAP_TYPE_CUSTOM,
+	ION_HEAP_TYPE_SECURE, /* allwinner add */
+			       /*
 			       * must be last so device specific heaps always
 			       * are at the end of this enum
 			       */
 };
 
+#define ION_HEAP_SYSTEM_MASK		(1 << ION_HEAP_TYPE_SYSTEM)
+#define ION_HEAP_SYSTEM_CONTIG_MASK	(1 << ION_HEAP_TYPE_SYSTEM_CONTIG)
+#define ION_HEAP_CARVEOUT_MASK		(1 << ION_HEAP_TYPE_CARVEOUT)
+#define ION_HEAP_TYPE_DMA_MASK		(1 << ION_HEAP_TYPE_DMA)
 #define ION_NUM_HEAP_IDS		(sizeof(unsigned int) * 8)
 
 /**
