diff -drupN a/drivers/media/platform/sunxi-vin/vin_test/mplane_image/Makefile b/drivers/media/platform/sunxi-vin/vin_test/mplane_image/Makefile
--- a/drivers/media/platform/sunxi-vin/vin_test/mplane_image/Makefile	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/media/platform/sunxi-vin/vin_test/mplane_image/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,15 @@
+#CC:=/home/zequnzheng/longan/out/gcc-linaro-5.3.1-2016.05-x86_64_aarch64-linux-gnu/bin/aarch64-linux-gnu-gcc
+CC:=/home/zequnzheng/longan/out/gcc-linaro-5.3.1-2016.05-x86_64_arm-linux-gnueabi/bin/arm-linux-gnueabi-gcc
+ARCH_DIR=/home/zequnzheng/longan/kernel/linux-4.9/arch/arm64/include/
+CFLAGS := -I $(ARCH_DIR)
+TARGET := csi_test_mplane
+
+.PHONY: all clean
+
+all: $(TARGET)
+
+csi_test_mplane: csi_test_mplane.c
+	$(CC) $(CFLAGS) -static  $^  -o  $@
+
+clean:
+	rm -rf $(TARGET)
