diff -drupN a/mm/mlock.c b/mm/mlock.c
--- a/mm/mlock.c	2018-08-06 17:23:04.000000000 +0300
+++ b/mm/mlock.c	2022-06-12 05:28:14.000000000 +0300
@@ -529,7 +529,7 @@ static int mlock_fixup(struct vm_area_st
 	pgoff = vma->vm_pgoff + ((start - vma->vm_start) >> PAGE_SHIFT);
 	*prev = vma_merge(mm, *prev, start, end, newflags, vma->anon_vma,
 			  vma->vm_file, pgoff, vma_policy(vma),
-			  vma->vm_userfaultfd_ctx);
+			  vma->vm_userfaultfd_ctx, vma_get_anon_name(vma));
 	if (*prev) {
 		vma = *prev;
 		goto success;
