diff -drupN a/drivers/mtd/aw-spinand/Kconfig b/drivers/mtd/aw-spinand/Kconfig
--- a/drivers/mtd/aw-spinand/Kconfig	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/mtd/aw-spinand/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,54 @@
+config AW_SPINAND_PHYSICAL_LAYER
+	bool
+	help
+	  Enable support for Allwinner's physical layer for spinand.
+	  It's a command physical layer, used by both AW-nand with nftl and
+	  MTD-nand with ubi.
+
+config AW_SPINAND_SECURE_STORAGE
+	bool
+	help
+	  Enable secure storage for Allwinner's spinand.
+
+	  If unsure, say no.
+choice
+	prompt "AW SPINAND"
+config AW_NFTL_SPINAND
+	tristate "Allwinner NFTL SPINAND Device Support"
+	depends on ARCH_SUNXI
+	select AW_SPINAND_PHYSICAL_LAYER
+	select SUNXI_NAND
+	help
+	Enable support for SPINAND Flash chips on Allwinner SoCs.
+	Tt's use Allwinner's NFTL system.
+
+config AW_MTD_SPINAND
+	tristate "Allwinner MTD SPINAND Device Support"
+	depends on MTD
+	depends on ARCH_SUNXI
+	select AW_SPINAND_PHYSICAL_LAYER
+	select AW_SPINAND_SECURE_STORAGE
+	select MTD_UBI
+	help
+	  Enables support for SPINAND Flash chips on Allwinner SoCs.
+	  It's different with Allwinner's privately nand driver that it use
+	  ubi system rather than Allwinner's NFTL.
+endchoice
+config AW_SPINAND_PSTORE_MTD_PART
+	bool "create pstore mtd partition for aw ubi spinand"
+	depends on AW_MTD_SPINAND
+	help
+	  Whether create pstore mtd partition, which is need by pstroe-blk.
+	  If you want linux kernel dump log to spinand when oops/panic, you
+	  should create pstreo mtd partition by this configure.
+
+	  If unsure, say no.
+
+config AW_SPINAND_ENABLE_PHY_CRC16
+	bool "check crc16 for each page on spinand physical layer"
+	depends on AW_SPINAND_PHYSICAL_LAYER
+	help
+	  It is experimental.
+	  To check crc16 for each page on spinand physical layer.
+
+	  If unsure, say no.
