diff -drupN a/drivers/platform/goldfish/Makefile b/drivers/platform/goldfish/Makefile
--- a/drivers/platform/goldfish/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/platform/goldfish/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -2,4 +2,5 @@
 # Makefile for Goldfish platform specific drivers
 #
 obj-$(CONFIG_GOLDFISH_BUS)	+= pdev_bus.o
-obj-$(CONFIG_GOLDFISH_PIPE)	+= goldfish_pipe.o
+obj-$(CONFIG_GOLDFISH_PIPE)	+= goldfish_pipe_all.o
+goldfish_pipe_all-objs := goldfish_pipe.o goldfish_pipe_v2.o
