diff -drupN a/include/sound/soc-dai.h b/include/sound/soc-dai.h
--- a/include/sound/soc-dai.h	2018-08-06 17:23:04.000000000 +0300
+++ b/include/sound/soc-dai.h	2022-06-12 05:28:14.000000000 +0300
@@ -68,7 +68,7 @@ struct snd_compr_stream;
  *
  * "Negative" FSYNC polarity is the one opposite of "normal" polarity.
  */
-#define SND_SOC_DAIFMT_NB_NF		(0 << 8) /* normal bit clock + frame */
+#define SND_SOC_DAIFMT_NB_NF		(1 << 8) /* normal bit clock + frame */
 #define SND_SOC_DAIFMT_NB_IF		(2 << 8) /* normal BCLK + inv FRM */
 #define SND_SOC_DAIFMT_IB_NF		(3 << 8) /* invert BCLK + nor FRM */
 #define SND_SOC_DAIFMT_IB_IF		(4 << 8) /* invert BCLK + FRM */
