diff -drupN a/scripts/dtc/script_parser/script_to_dts.h b/scripts/dtc/script_parser/script_to_dts.h
--- a/scripts/dtc/script_parser/script_to_dts.h	1970-01-01 03:00:00.000000000 +0300
+++ b/scripts/dtc/script_parser/script_to_dts.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,9 @@
+#ifndef _SCRIPT_TO_DTS_H_
+#define _SCRIPT_TO_DTS_H_
+
+#include "iniparser.h"
+#include "script.h"
+
+int script_parse(const char *script_name);
+
+#endif
