diff -drupN a/kernel/Kconfig.hz b/kernel/Kconfig.hz
--- a/kernel/Kconfig.hz	2018-08-06 17:23:04.000000000 +0300
+++ b/kernel/Kconfig.hz	2022-06-12 05:28:14.000000000 +0300
@@ -15,6 +15,10 @@ choice
 	 environment leading to NR_CPUS * HZ number of timer interrupts
 	 per second.
 
+	config HZ_15
+		bool "15 HZ"
+	help
+	  15Hz is for cpuidle test.
 
 	config HZ_100
 		bool "100 HZ"
@@ -49,6 +53,7 @@ endchoice
 
 config HZ
 	int
+	default 15 if HZ_15
 	default 100 if HZ_100
 	default 250 if HZ_250
 	default 300 if HZ_300
