diff -drupN a/mm/madvise.c b/mm/madvise.c
--- a/mm/madvise.c	2018-08-06 17:23:04.000000000 +0300
+++ b/mm/madvise.c	2022-06-12 05:28:14.000000000 +0300
@@ -109,7 +109,7 @@ static long madvise_behavior(struct vm_a
 	pgoff = vma->vm_pgoff + ((start - vma->vm_start) >> PAGE_SHIFT);
 	*prev = vma_merge(mm, *prev, start, end, new_flags, vma->anon_vma,
 			  vma->vm_file, pgoff, vma_policy(vma),
-			  vma->vm_userfaultfd_ctx);
+			  vma->vm_userfaultfd_ctx, vma_get_anon_name(vma));
 	if (*prev) {
 		vma = *prev;
 		goto success;
