diff -drupN a/drivers/base/core.c b/drivers/base/core.c
--- a/drivers/base/core.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/base/core.c	2022-06-12 05:28:14.000000000 +0300
@@ -430,6 +430,7 @@ static ssize_t online_show(struct device
 	bool val;
 
 	device_lock(dev);
+	dev->offline = !(!!cpu_online(dev->id));
 	val = !dev->offline;
 	device_unlock(dev);
 	return sprintf(buf, "%u\n", val);
@@ -449,6 +450,8 @@ static ssize_t online_store(struct devic
 	if (ret)
 		return ret;
 
+	dev->offline = !(!!cpu_online(dev->id));
+
 	ret = val ? device_online(dev) : device_offline(dev);
 	unlock_device_hotplug();
 	return ret < 0 ? ret : count;
