diff -drupN a/include/uapi/linux/tcp.h b/include/uapi/linux/tcp.h
--- a/include/uapi/linux/tcp.h	2018-08-06 17:23:04.000000000 +0300
+++ b/include/uapi/linux/tcp.h	2022-06-12 05:28:14.000000000 +0300
@@ -116,6 +116,7 @@ enum {
 #define TCP_SAVE_SYN		27	/* Record SYN headers for new connections */
 #define TCP_SAVED_SYN		28	/* Get SYN headers recorded for connection */
 #define TCP_REPAIR_WINDOW	29	/* Get/set window parameters */
+#define TCP_FASTOPEN_CONNECT	30	/* Attempt FastOpen with connect */
 
 struct tcp_repair_opt {
 	__u32	opt_code;
