diff -drupN a/net/ipv4/inet_fragment.c b/net/ipv4/inet_fragment.c
--- a/net/ipv4/inet_fragment.c	2018-08-06 17:23:04.000000000 +0300
+++ b/net/ipv4/inet_fragment.c	2022-06-12 05:28:14.000000000 +0300
@@ -25,9 +25,6 @@
 #include <net/inet_frag.h>
 #include <net/inet_ecn.h>
 
-#define INETFRAGS_EVICT_BUCKETS   128
-#define INETFRAGS_EVICT_MAX	  512
-
 /* don't rebuild inetfrag table with new secret more often than this */
 #define INETFRAGS_MIN_REBUILD_INTERVAL (5 * HZ)
 
