diff -drupN a/fs/proc/Makefile b/fs/proc/Makefile
--- a/fs/proc/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/fs/proc/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -5,7 +5,9 @@
 obj-y   += proc.o
 
 CFLAGS_task_mmu.o	+= $(call cc-option,-Wno-override-init,)
+ifneq ($(CONFIG_MMU),y)
 proc-y			:= nommu.o task_nommu.o
+endif
 proc-$(CONFIG_MMU)	:= task_mmu.o
 
 proc-y       += inode.o root.o base.o generic.o array.o \
@@ -25,6 +27,7 @@ proc-y	+= softirqs.o
 proc-y	+= namespaces.o
 proc-y	+= self.o
 proc-y	+= thread_self.o
+proc-$(CONFIG_PROC_UID)  += uid.o
 proc-$(CONFIG_PROC_SYSCTL)	+= proc_sysctl.o
 proc-$(CONFIG_NET)		+= proc_net.o
 proc-$(CONFIG_PROC_KCORE)	+= kcore.o
