diff -drupN a/arch/arm64/boot/dts/sunxi/sun50iw8p1-pinctrl.dtsi b/arch/arm64/boot/dts/sunxi/sun50iw8p1-pinctrl.dtsi
--- a/arch/arm64/boot/dts/sunxi/sun50iw8p1-pinctrl.dtsi	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/arm64/boot/dts/sunxi/sun50iw8p1-pinctrl.dtsi	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,737 @@
+/*
+ * Allwinner sun50iw8 pin config info.
+ */
+
+/ {
+	soc@03000000{
+		r_pio: pinctrl@07022000 {
+			compatible = "allwinner,sun50iw8p1-r-pinctrl";
+			reg = <0x0 0x07022000 0x0 0x400>;
+			interrupts = <GIC_SPI 9 4>;
+			clocks = <&clk_cpurpio>;
+			device_type = "r_pio";
+			gpio-controller;
+			interrupt-controller;
+			#interrupt-cells = <3>;
+			#size-cells = <0>;
+			#gpio-cells = <6>;
+
+			s_uart0_pins_a: s_uart0@0 {
+				allwinner,pins = "PL2", "PL3";
+				allwinner,function = "s_uart0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			s_rsb0_pins_a: s_rsb0@0 {
+				allwinner,pins = "PL0", "PL1";
+				allwinner,function = "s_rsb0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <2>;
+				allwinner,pull = <1>;
+			};
+
+			s_twi0_pins_a: s_twi0@0 {
+				allwinner,pins = "PL0", "PL1";
+				allwinner,function = "s_twi0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <2>;
+				allwinner,pull = <1>;
+			};
+
+			s_jtag0_pins_a: s_jtag0@0 {
+				allwinner,pins = "PL4", "PL5", "PL6", "PL7";
+				allwinner,function = "s_jtag0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <2>;
+				allwinner,pull = <1>;
+			};
+
+			s_cir0_pins_a: s_cir0@0 {
+				allwinner,pins = "PL14";
+				allwinner,function = "s_ir_rx";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <2>;
+				allwinner,pull = <1>;
+			};
+		};
+
+		pio: pinctrl@0300b000 {
+			compatible = "allwinner,sun50iw8p1-pinctrl";
+			reg = <0x0 0x0300b000 0x0 0x400>;
+			interrupts = <GIC_SPI 43 4>,
+				     <GIC_SPI 44 4>,
+				     <GIC_SPI 45 4>,
+				     <GIC_SPI 46 4>,
+				     <GIC_SPI 47 4>;
+			device_type = "pio";
+			clocks = <&clk_pio>;
+			gpio-controller;
+			interrupt-controller;
+			#interrupt-cells = <3>;
+			#size-cells = <0>;
+			#gpio-cells = <6>;
+
+			vdevice_pins_a: vdevice@0 {
+				allwinner,pins = "PB2", "PB3";
+				allwinner,function = "vdevice";
+				allwinner,muxsel = <5>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart0_pins_a: uart0@0 {
+				allwinner,pins = "PB8", "PB9";
+				allwinner,pname = "uart0_tx", "uart0_rx";
+				allwinner,function = "uart0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart0_pins_b: uart0@1 {
+				allwinner,pins = "PB8", "PB9";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			uart1_pins_a: uart1@0 {
+				allwinner,pins = "PG6", "PG7", "PG8", "PG9";
+				allwinner,pname = "uart1_tx", "uart1_rx",
+						  "uart1_rts", "uart1_cts";
+				allwinner,function = "uart1";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart1_pins_b: uart1@1 {
+				allwinner,pins = "PG6", "PG7", "PG8", "PG9";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			uart2_pins_a: uart2@0 {
+				allwinner,pins = "PB0", "PB1", "PB2", "PB3";
+				allwinner,pname = "uart2_tx", "uart2_rx",
+						  "uart2_rts", "uart2_cts";
+				allwinner,function = "uart2";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart2_pins_b: uart2@1 {
+				allwinner,pins = "PB0", "PB1", "PB2", "PB3";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			uart3_pins_a: uart3@0 {
+				allwinner,pins = "PH4", "PH5", "PH6", "PH7";
+				allwinner,pname = "uart3_tx", "uart3_rx",
+						  "uart3_rts", "uart3_cts";
+				allwinner,function = "uart3";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart3_pins_b: uart3@1 {
+				allwinner,pins = "PH4", "PH5", "PH6", "PH7";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			uart4_pins_a: uart4@0 {
+				allwinner,pins = "PD18", "PD19", "PD20", "PD21";
+				allwinner,pname = "uart4_tx", "uart4_rx",
+						  "uart4_rts", "uart4_cts";
+				allwinner,function = "uart4";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart4_pins_b: uart4@1 {
+				allwinner,pins = "PD18", "PD19", "PD20", "PD21";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi0_pins_a: twi0@0 {
+				allwinner,pins = "PD14", "PD15";
+				allwinner,pname = "twi0_scl", "twi0_sda";
+				allwinner,function = "twi0";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi0_pins_b: twi0@1 {
+				allwinner,pins = "PD14", "PD15";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi1_pins_a: twi1@0 {
+				allwinner,pins = "PD16", "PD17";
+				allwinner,pname = "twi1_scl", "twi1_sda";
+				allwinner,function = "twi1";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi1_pins_b: twi1@1 {
+				allwinner,pins = "PD16", "PD17";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi2_pins_a: twi2@0 {
+				allwinner,pins = "PE1", "PE2";
+				allwinner,pname = "twi2_scl", "twi2_sda";
+				allwinner,function = "twi2";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi2_pins_b: twi2@1 {
+				allwinner,pins = "PE1", "PE2";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			ts0_pins_a: ts0@0 {
+				allwinner,pins = "PE0", "PE1", "PE2", "PE3",
+						 "PE4", "PE5", "PE6", "PE7",
+						 "PE8", "PE9", "PE10", "PE11";
+				allwinner,pname = "ts0_clk", "ts0_err",
+						  "ts0_sync", "ts0_dvld",
+						  "ts0_d0", "ts0_d1",
+						  "ts0_d2", "ts0_d3",
+						  "ts0_d4", "ts0_d5",
+						  "ts0_d6", "ts0_d7";
+				allwinner,function = "ts0";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			ts0_pins_b: ts0_sleep@0 {
+				allwinner,pins = "PE0", "PE1", "PE2", "PE3",
+						 "PE4", "PE5", "PE6", "PE7",
+						 "PE8", "PE9", "PE10", "PE11";
+				allwinner,pname = "ts0_clk", "ts0_err",
+						  "ts0_sync", "ts0_dvld",
+						  "ts0_d0", "ts0_d1",
+						  "ts0_d2", "ts0_d3",
+						  "ts0_d4", "ts0_d5",
+						  "ts0_d6", "ts0_d7";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			ts1_pins_a: ts1@0 {
+				allwinner,pins = "PE7", "PE8", "PE9",
+						 "PE10", "PE11";
+				allwinner,pname = "ts1_clk", "ts1_err",
+						  "ts1_sync", "ts1_dvld",
+						  "ts1_d0";
+				allwinner,function = "ts1";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			ts1_pins_b: ts1_sleep@0 {
+				allwinner,pins = "PE7", "PE8", "PE9",
+						 "PE10", "PE11";
+				allwinner,pname = "ts1_clk", "ts1_err",
+						  "ts1_sync", "ts1_dvld",
+						  "ts1_d0";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			ts2_pins_a: ts2@0 {
+				allwinner,pins = "PD0", "PD1", "PD2", "PD3",
+						 "PD4", "PD5", "PD6", "PD7",
+						 "PD8", "PD9", "PD10", "PD11";
+				allwinner,pname = "ts2_clk", "ts2_err",
+						  "ts2_sync", "ts2_dvld",
+						  "ts2_d0", "ts2_d1",
+						  "ts2_d2", "ts2_d3",
+						  "ts2_d4", "ts2_d5",
+						  "ts2_d6", "ts2_d7";
+				allwinner,function = "ts2";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			ts2_pins_b: ts2_sleep@0 {
+				allwinner,pins = "PD0", "PD1", "PD2", "PD3",
+						 "PD4", "PD5", "PD6", "PD7",
+						 "PD8", "PD9", "PD10", "PD11";
+				allwinner,pname = "ts2_clk", "ts2_err",
+						  "ts2_sync", "ts2_dvld",
+						  "ts2_d0", "ts2_d1",
+						  "ts2_d2", "ts2_d3",
+						  "ts2_d4", "ts2_d5",
+						  "ts2_d6", "ts2_d7";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			ts3_pins_a: ts3@0 {
+				allwinner,pins = "PD7", "PD8", "PD9",
+						 "PD10", "PD11";
+				allwinner,pname = "ts3_clk", "ts3_err",
+						  "ts3_sync", "ts3_dvld",
+						  "ts3_d0";
+				allwinner,function = "ts3";
+				allwinner,muxsel = <5>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			ts3_pins_b: ts3_sleep@0 {
+				allwinner,pins = "PD7", "PD8", "PD9",
+						 "PD10", "PD11";
+				allwinner,pname = "ts3_clk", "ts3_err",
+						  "ts3_sync", "ts3_dvld",
+						  "ts3_d0";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			spi0_pins_a: spi0@0 {
+				allwinner,pins = "PC0", "PC2", "PC4", "PC16",
+						 "PC15";
+				allwinner,pname = "spi0_sclk", "spi0_mosi",
+						  "spi0_miso", "spi0_hold",
+						  "spi0_wp";
+				allwinner,function = "spi0";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			spi0_pins_b: spi0@1 {
+				allwinner,pins = "PC3";
+				allwinner,pname = "spi0_cs0";
+				allwinner,function = "spi0";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>; 	// only CS should be pulled up
+			};
+
+			spi0_pins_c: spi0@2 {
+				allwinner,pins = "PC0", "PC2", "PC3", "PC4",
+						 "PC15", "PC16";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			spi1_pins_a: spi1@0 {
+				allwinner,pins = "PH1", "PH2", "PH3";
+				allwinner,pname = "spi1_sclk", "spi1_mosi",
+						  "spi1_miso";
+				allwinner,function = "spi1";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			spi1_pins_b: spi1@1 {
+				allwinner,pins = "PH0";
+				allwinner,pname = "spi1_cs0";
+				allwinner,function = "spi1";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>; // only CS should be pulled up
+			};
+
+			spi1_pins_c: spi1@2 {
+				allwinner,pins = "PH0", "PH1", "PH2", "PH3";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			sdc0_pins_a: sdc0@0 {
+				allwinner,pins = "PF0", "PF1", "PF2",
+						 "PF3", "PF4", "PF5";
+				allwinner,function = "sdc0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			sdc0_pins_b: sdc0@1 {
+				allwinner,pins = "PF0", "PF1", "PF2",
+						 "PF3", "PF4", "PF5";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			sdc1_pins_a: sdc1@0 {
+				allwinner,pins = "PG0", "PG1", "PG2",
+						 "PG3", "PG4", "PG5";
+				allwinner,function = "sdc1";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <3>;
+				allwinner,pull = <1>;
+			};
+
+			sdc1_pins_b: sdc1@1 {
+				allwinner,pins = "PG0", "PG1", "PG2",
+						 "PG3", "PG4", "PG5";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			sdc2_pins_a: sdc2@0 {
+				allwinner,pins = "PC0", "PC1", "PC5", "PC6",
+						 "PC8", "PC9", "PC10", "PC11",
+						 "PC13", "PC14", "PC15", "PC16";
+				allwinner,function = "sdc2";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <2>;
+				allwinner,pull = <1>;
+			};
+
+			sdc2_pins_b: sdc2@1 {
+				allwinner,pins = "PC0", "PC1", "PC5", "PC6",
+						 "PC8", "PC9", "PC10", "PC11",
+						 "PC13", "PC14", "PC15", "PC16";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			daudio0_pins_a: daudio0@0 {
+				allwinner,pins = "PB23", "PB24", "PB25",
+						 "PB26", "PB27", "PB28", "PB29";
+				allwinner,function = "i2s0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			daudio0_pins_b: daudio0_sleep@0 {
+				allwinner,pins = "PB23", "PB24", "PB25",
+						 "PB26", "PB27", "PB28", "PB29";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			daudio1_pins_a: daudio1@0 {
+				allwinner,pins = "PB23", "PB24", "PB25",
+						 "PB26", "PB27", "PB28", "PB29";
+				allwinner,function = "i2s1";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			daudio1_pins_b: daudio1_sleep@0 {
+				allwinner,pins = "PB23", "PB24", "PB25",
+						 "PB26", "PB27", "PB28", "PB29";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			daudio2_pins_a: daudio2@0 {
+				allwinner,pins = "PB23", "PB24", "PB25",
+						 "PB26", "PB27", "PB28", "PB29";
+				allwinner,function = "i2s2";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			daudio2_pins_b: daudio2_sleep@0 {
+				allwinner,pins = "PB23", "PB24", "PB25",
+						 "PB26", "PB27", "PB28", "PB29";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			dmic_pins_a: dmic@0 {
+				allwinner,pins = "PF9";
+				allwinner,function = "dmic";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			dmic_pins_b: dmic_sleep@0 {
+				allwinner,pins = "PF9";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			spdif_pins_a: spdif@0 {
+				allwinner,pins = "PB6", "PB7";
+				allwinner,function = "spdif";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			spdif_pins_b: spdif_sleep@0 {
+				allwinner,pins = "PB6", "PB7";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+
+			aif2_pins_a: aif2@0 {
+				allwinner,pins = "PB6", "PB4", "PB5", "PB7";
+				allwinner,function = "aif2";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			aif2_pins_b: aif2_sleep@0 {
+				allwinner,pins = "PB6", "PB4", "PB5", "PB7";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			aif3_pins_a: aif3@0 {
+				allwinner,pins = "PG10", "PG11", "PG12", "PG13";
+				allwinner,function = "aif3";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			aif3_pins_b: aif3_sleep@0 {
+				allwinner,pins = "PG10", "PG11", "PG12", "PG13";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			csi_mclk0_pins_a: csi_mck0@0 {
+				allwinner,pins = "PE0";
+				allwinner,pname = "csi0_mck";
+				allwinner,function = "csi_mck0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+			csi_mclk0_pins_b: csi_mck0@1 {
+				allwinner,pins = "PE0";
+				allwinner,pname = "csi0_mck";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+			csi_cci0_pins_a: csi_cci0@0 {
+				allwinner,pins = "PE1", "PE2";
+				allwinner,pname = "csi0_sck", "csi0_sda";
+				allwinner,function = "csi_cci0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+			csi_cci0_pins_b: csi_cci0@1 {
+				allwinner,pins = "PE1", "PE2";
+				allwinner,pname = "csi0_sck", "csi0_sda";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+
+			scr0_pins_a: scr0@0 {
+				allwinner,pins = "PH21", "PH24", "PH25", "PH26", "PH27";
+				allwinner,pname = "scr0_rst", "scr0_det", "scr0_vccen", "scr0_sck", "scr0_sda";
+				allwinner,function = "sim0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <0>;
+				allwinner,pull = <1>;
+			};
+
+			scr0_pins_b: scr0@1 {
+				allwinner,pins = "PH22", "PH23";
+				allwinner,pname = "scr0_vppen", "scr0_vppp";
+				allwinner,function = "sim0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <0>;
+				allwinner,pull = <1>;
+			};
+
+			scr0_pins_c: scr0@2 {
+				allwinner,pins = "PH21", "PH22", "PH23", "PH24", "PH25", "PH26", "P27";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <0>;
+				allwinner,pull = <0>;
+			};
+
+			scr1_pins_a: scr1@0 {
+				allwinner,pins = "PD15", "PD16", "PD12",
+						 "PD13", "PD14";
+				allwinner,pname = "scr1_rst", "scr1_det",
+						  "scr1_vccen", "scr1_sck",
+						  "scr1_sda";
+				allwinner,function = "sim1";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			scr1_pins_b: scr1@1 {
+				allwinner,pins = "PE14", "PE15";
+				allwinner,pname = "scr1_vppen", "scr1_vppp";
+				allwinner,function = "sim1";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			scr1_pins_c: scr1@2 {
+				allwinner,pins = "PD15", "PD16", "PD12",
+						 "PD13", "PD14", "PE14",
+						 "PE15";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			nand0_pins_a: nand0@0 {
+				allwinner,pins = "PC0", "PC1", "PC2", "PC4",
+						 "PC6", "PC7", "PC8", "PC9",
+						 "PC10", "PC11", "PC12", "PC13",
+						 "PC14";
+				allwinner,pname= "nand0_we", "nand0_ale","nand0_cle", "nand0_nre",
+						 "nand0_d0", "nand0_d1", "nand0_d2", "nand0_d3",
+						 "nand0_d4", "nand0_d5", "nand0_d6", "nand0_d7",
+						 "nand0_ndqs";
+				allwinner,function = "nand0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			nand0_pins_b: nand0@1 {
+				allwinner,pins = "PC3", "PC5", "PC15", "PC16";
+				allwinner,pname= "nand0_ce0", "nand0_rb0", "nand0_ce1", "nand0_rb1";
+				allwinner,function = "nand0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;// only RB&CE should be pulled up
+			};
+
+			nand0_pins_c: nand0@2 {
+				allwinner,pins = "PC0", "PC1", "PC2", "PC3",
+						 "PC4", "PC5", "PC6", "PC7",
+						 "PC8", "PC9", "PC10", "PC11",
+						 "PC12", "PC13", "PC14", "PC15",
+						 "PC16";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			gmac_pins_a: gmac@0 {
+				allwinner,pins = "PA0", "PA1", "PA2", "PA3",
+						 "PA4", "PA5", "PA6", "PA7",
+						 "PA8", "PA9", "PA10", "PA11",
+						 "PA12", "PA13", "PA14", "PA15",
+						 "PA16", "PA17", "PA18", "PA19",
+						 "PA20", "PA21", "PA22", "PA23",
+						 "PA24", "PA25", "PA26", "PA27",
+						 "PA28", "PA29", "PA30", "PA31";
+				allwinner,function = "gmac0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <3>;
+				allwinner,pull = <0>;
+			};
+
+			gmac_pins_b: gmac@1 {
+				allwinner,pins = "PA0", "PA1", "PA2", "PA3",
+						 "PA4", "PA5", "PA6", "PA7",
+						 "PA8", "PA9", "PA10", "PA11",
+						 "PA12", "PA13", "PA14", "PA15",
+						 "PA16", "PA17", "PA18", "PA19",
+						 "PA20", "PA21", "PA22", "PA23",
+						 "PA24", "PA25", "PA26", "PA27",
+						 "PA28", "PA29", "PA30", "PA31";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <3>;
+				allwinner,pull = <0>;
+			};
+
+			ledc_pins_a: ledc@0 {
+				allwinner,pins = "PG0";
+				allwinner,function = "ledc";
+				allwinner,muxsel = <5>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+		};
+	};
+};
