diff -drupN a/drivers/mmc/core/mmc_ops.h b/drivers/mmc/core/mmc_ops.h
--- a/drivers/mmc/core/mmc_ops.h	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/mmc/core/mmc_ops.h	2022-06-12 05:28:14.000000000 +0300
@@ -32,5 +32,7 @@ int __mmc_switch(struct mmc_card *card,
 		unsigned int timeout_ms, bool use_busy_signal, bool send_status,
 		bool ignore_crc);
 
+#define MMC_OPS_TIMEOUT_MS	(10 * 60 * 1000) /* 10 minute timeout */
+
 #endif
 
