diff -drupN a/drivers/staging/android/ion/Makefile b/drivers/staging/android/ion/Makefile
--- a/drivers/staging/android/ion/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/staging/android/ion/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -1,6 +1,9 @@
 obj-$(CONFIG_ION) +=	ion.o ion-ioctl.o ion_heap.o \
-			ion_page_pool.o ion_system_heap.o \
-			ion_carveout_heap.o ion_chunk_heap.o ion_cma_heap.o
+			ion_page_pool.o ion_system_heap.o ion_secure_heap.o \
+			ion_carveout_heap.o ion_chunk_heap.o
+ifdef CONFIG_CMA
+obj-$(CONFIG_ION) +=	ion_cma_heap.o
+endif
 obj-$(CONFIG_ION_TEST) += ion_test.o
 ifdef CONFIG_COMPAT
 obj-$(CONFIG_ION) += compat_ion.o
@@ -9,5 +12,6 @@ endif
 obj-$(CONFIG_ION_DUMMY) += ion_dummy_driver.o
 obj-$(CONFIG_ION_TEGRA) += tegra/
 obj-$(CONFIG_ION_HISI) += hisilicon/
+obj-$(CONFIG_ION_SUNXI) += sunxi/
 obj-$(CONFIG_ION_OF) += ion_of.o
 
