diff -drupN a/arch/arm64/boot/dts/sunxi/Makefile b/arch/arm64/boot/dts/sunxi/Makefile
--- a/arch/arm64/boot/dts/sunxi/Makefile	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/arm64/boot/dts/sunxi/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,17 @@
+dtb-$(CONFIG_ARCH_SUN50IW3P1) += board.dtb
+
+dtb-$(CONFIG_ARCH_SUN50IW1P1) += sun50iw1p1-fpga.dtb  sun50iw1p1-perf1_v1_0.dtb \
+				sun50iw1p1-perf2_v1_0.dtb sun50iw1p1-t1.dtb \
+				sun50iw1p1-sata.dtb sun50iw1p1-soc.dtb
+
+dtb-$(CONFIG_ARCH_SUN50IW6P1) += sun50iw6p1-soc.dtb
+
+dtb-$(CONFIG_ARCH_SUN50IW8P1) += sun50iw8p1-fpga.dtb sun50iw8p1-soc.dtb
+dtb-$(CONFIG_ARCH_SUN50IW9P1) += board.dtb
+dtb-$(CONFIG_ARCH_SUN50IW10P1) += sun50iw10p1-fpga.dtb sun50iw10p1-soc.dtb
+dtb-$(CONFIG_ARCH_SUN50IW11P1) += board.dtb
+dtb-$(CONFIG_ARCH_SUN50IW5T) += sun50iw5t-fpga.dtb
+
+always		:= $(dtb-y)
+subdir-y	:= $(dts-dirs)
+clean-files	:= *.dtb
