diff -drupN a/drivers/devfreq/Makefile b/drivers/devfreq/Makefile
--- a/drivers/devfreq/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/devfreq/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -5,11 +5,15 @@ obj-$(CONFIG_DEVFREQ_GOV_PERFORMANCE)	+=
 obj-$(CONFIG_DEVFREQ_GOV_POWERSAVE)	+= governor_powersave.o
 obj-$(CONFIG_DEVFREQ_GOV_USERSPACE)	+= governor_userspace.o
 obj-$(CONFIG_DEVFREQ_GOV_PASSIVE)	+= governor_passive.o
+obj-$(CONFIG_DEVFREQ_GOV_ADAPTIVE)	+= governor_adaptive.o
 
 # DEVFREQ Drivers
 obj-$(CONFIG_ARM_EXYNOS_BUS_DEVFREQ)	+= exynos-bus.o
 obj-$(CONFIG_ARM_RK3399_DMC_DEVFREQ)	+= rk3399_dmc.o
 obj-$(CONFIG_ARM_TEGRA_DEVFREQ)		+= tegra-devfreq.o
+obj-$(CONFIG_DEVFREQ_DRAM_FREQ)		+= dramfreq/
+obj-$(CONFIG_SUNXI_DRAM_SIMPLIFIED_DEVFREQ)	+= sunxi-dramfreq_simplify.o
+obj-$(CONFIG_ARM_SUNXI_DRAM_DEVFREQ)	+= sunxi-dramfreq.o
 
 # DEVFREQ Event Drivers
 obj-$(CONFIG_PM_DEVFREQ_EVENT)		+= event/
