diff -drupN a/drivers/media/v4l2-core/Makefile b/drivers/media/v4l2-core/Makefile
--- a/drivers/media/v4l2-core/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/media/v4l2-core/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -20,7 +20,7 @@ videodev-$(CONFIG_MEDIA_CONTROLLER) += v
 
 obj-$(CONFIG_VIDEO_V4L2) += videodev.o
 obj-$(CONFIG_VIDEO_V4L2) += v4l2-common.o
-obj-$(CONFIG_VIDEO_V4L2) += v4l2-dv-timings.o
+obj-$(CONFIG_VIDEO_DV_TIMINGS) += v4l2-dv-timings.o
 
 obj-$(CONFIG_VIDEO_TUNER) += tuner.o
 
