diff -drupN a/sound/soc/sunxi/sun8iw18-sndcodec.h b/sound/soc/sunxi/sun8iw18-sndcodec.h
--- a/sound/soc/sunxi/sun8iw18-sndcodec.h	1970-01-01 03:00:00.000000000 +0300
+++ b/sound/soc/sunxi/sun8iw18-sndcodec.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,37 @@
+/*
+ * sound\soc\sunxi\sun8iw18-sndcodec.h
+ * (C) Copyright 2014-2018
+ * Allwinner Technology Co., Ltd. <www.allwinnertech.com>
+ * yumingfeng <yumingfeng@allwinertech.com>
+ *
+ * some simple description for this code
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of
+ * the License, or (at your option) any later version.
+ *
+ */
+
+#ifndef	__SUN8IW18_SNDCODEC_H_
+#define	__SUN8IW18_SNDCODEC_H_
+
+#ifdef CONFIG_SND_SUNXI_MAD
+#include "sunxi-mad.h"
+#endif
+
+#include "sun8iw18-codec.h"
+enum sunxi_sndcodec_div_id {
+	CODEC_DIV_PLAYBACK = 0,
+	CODEC_DIV_CAPTURE = 1,
+};
+
+struct sunxi_sndcodec_priv {
+	struct snd_soc_card *card;
+
+#ifdef CONFIG_SND_SUNXI_MAD
+	struct sunxi_mad_priv mad_priv;
+#endif
+};
+
+#endif
