diff -drupN a/include/linux/tcp.h b/include/linux/tcp.h
--- a/include/linux/tcp.h	2018-08-06 17:23:04.000000000 +0300
+++ b/include/linux/tcp.h	2022-06-12 05:28:14.000000000 +0300
@@ -219,7 +219,7 @@ struct tcp_sock {
 	} rack;
 	u16	advmss;		/* Advertised MSS			*/
 	u8	rate_app_limited:1,  /* rate_{delivered,interval_us} limited? */
-		is_sack_reneg:1,    /* in recovery from loss with SACK reneg? */
+		fastopen_connect:1, /* FASTOPEN_CONNECT sockopt */
 		unused:6;
 	u8	nonagle     : 4,/* Disable Nagle algorithm?             */
 		thin_lto    : 1,/* Use linear timeouts for thin streams */
