diff -drupN a/drivers/amba/bus.c b/drivers/amba/bus.c
--- a/drivers/amba/bus.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/amba/bus.c	2022-06-12 05:28:14.000000000 +0300
@@ -85,7 +85,7 @@ static ssize_t driver_override_store(str
 	char *driver_override, *old, *cp;
 
 	/* We need to keep extra room for a newline */
-	if (count >= (PAGE_SIZE - 1))
+	if (count > (PAGE_SIZE) - 1)
 		return -EINVAL;
 
 	driver_override = kstrndup(buf, count, GFP_KERNEL);
