diff -drupN a/drivers/char/sunxi_tr/Makefile b/drivers/char/sunxi_tr/Makefile
--- a/drivers/char/sunxi_tr/Makefile	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/char/sunxi_tr/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,10 @@
+# Makefile for the Linux transform drivers on sunxi platform.
+# 23 jan 2015, Tyle, <mailto:tyle@allwinnertech.com>
+# Rewritten to use lists instead of if-statements.
+
+obj-$(CONFIG_SUNXI_TRANSFORM) += sunxi_tr.o
+
+sunxi_tr-y := sunxi_transform.o
+sunxi_tr-$(CONFIG_ARCH_SUN50IW1) += sun50iw1_transform.o
+sunxi_tr-$(CONFIG_ARCH_SUN8IW11) += sun8iw11_transform.o
+sunxi_tr-$(CONFIG_ARCH_SUN50IW3) += sun8iw11_transform.o
