diff -drupN a/include/linux/suspend.h b/include/linux/suspend.h
--- a/include/linux/suspend.h	2018-08-06 17:23:04.000000000 +0300
+++ b/include/linux/suspend.h	2022-06-12 05:28:14.000000000 +0300
@@ -436,6 +436,7 @@ extern bool pm_get_wakeup_count(unsigned
 extern bool pm_save_wakeup_count(unsigned int count);
 extern void pm_wakep_autosleep_enabled(bool set);
 extern void pm_print_active_wakeup_sources(void);
+extern void pm_get_active_wakeup_sources(char *pending_sources, size_t max);
 
 static inline void lock_system_sleep(void)
 {
