diff -drupN a/include/linux/blk_types.h b/include/linux/blk_types.h
--- a/include/linux/blk_types.h	2018-08-06 17:23:04.000000000 +0300
+++ b/include/linux/blk_types.h	2022-06-12 05:28:14.000000000 +0300
@@ -25,6 +25,7 @@ typedef void (bio_end_io_t) (struct bio
 struct bio {
 	struct bio		*bi_next;	/* request queue link */
 	struct block_device	*bi_bdev;
+	unsigned short		bi_write_hint;
 	int			bi_error;
 	unsigned int		bi_opf;		/* bottom bits req flags,
 						 * top bits REQ_OP. Use
