diff -drupN a/drivers/gator/gator_events_mali_4xx.h b/drivers/gator/gator_events_mali_4xx.h
--- a/drivers/gator/gator_events_mali_4xx.h	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/gator/gator_events_mali_4xx.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,18 @@
+/**
+ * Copyright (C) ARM Limited 2011-2016. All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ *
+ */
+
+/*
+ * Header contains common definitions for the Mali-4xx processors.
+ */
+#if !defined(GATOR_EVENTS_MALI_4xx_H)
+#define GATOR_EVENTS_MALI_4xx_H
+
+extern void gator_events_mali_log_dvfs_event(unsigned int d0, unsigned int d1);
+
+#endif /* GATOR_EVENTS_MALI_4xx_H */
