diff -drupN a/drivers/char/ipmi/ipmi_poweroff.c b/drivers/char/ipmi/ipmi_poweroff.c
--- a/drivers/char/ipmi/ipmi_poweroff.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/char/ipmi/ipmi_poweroff.c	2022-06-12 05:28:14.000000000 +0300
@@ -66,7 +66,7 @@ static void (*specific_poweroff_func)(ip
 /* Holds the old poweroff function so we can restore it on removal. */
 static void (*old_poweroff_func)(void);
 
-static int set_param_ifnum(const char *val, struct kernel_param *kp)
+static int set_param_ifnum(const char *val, const struct kernel_param *kp)
 {
 	int rv = param_set_int(val, kp);
 	if (rv)
