diff -drupN a/drivers/misc/Makefile b/drivers/misc/Makefile
--- a/drivers/misc/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/misc/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -54,6 +54,10 @@ obj-$(CONFIG_VEXPRESS_SYSCFG)	+= vexpres
 obj-$(CONFIG_CXL_BASE)		+= cxl/
 obj-$(CONFIG_PANEL)             += panel.o
 
+obj-$(CONFIG_UID_SYS_STATS)	+= uid_sys_stats.o
+obj-$(CONFIG_MEMORY_STATE_TIME)	+= memory_state_time.o
+obj-$(CONFIG_SUNXI_SST_STORAGE)	+= sst_storage/sst_storage.o
+
 lkdtm-$(CONFIG_LKDTM)		+= lkdtm_core.o
 lkdtm-$(CONFIG_LKDTM)		+= lkdtm_bugs.o
 lkdtm-$(CONFIG_LKDTM)		+= lkdtm_heap.o
@@ -63,6 +67,7 @@ lkdtm-$(CONFIG_LKDTM)		+= lkdtm_usercopy
 
 KCOV_INSTRUMENT_lkdtm_rodata.o	:= n
 
+CFLAGS_lkdtm_rodata.o		+= $(DISABLE_LTO)
 OBJCOPYFLAGS :=
 OBJCOPYFLAGS_lkdtm_rodata_objcopy.o := \
 			--set-section-flags .text=alloc,readonly \
@@ -70,3 +75,6 @@ OBJCOPYFLAGS_lkdtm_rodata_objcopy.o := \
 targets += lkdtm_rodata.o lkdtm_rodata_objcopy.o
 $(obj)/lkdtm_rodata_objcopy.o: $(obj)/lkdtm_rodata.o FORCE
 	$(call if_changed,objcopy)
+
+obj-$(CONFIG_SUNXI_RFKILL)      += sunxi-rf/
+obj-y				+= leds/
