diff -drupN a/drivers/staging/nanohub/Makefile b/drivers/staging/nanohub/Makefile
--- a/drivers/staging/nanohub/Makefile	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/staging/nanohub/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,10 @@
+#
+# Makefile for nanohub
+#
+
+obj-$(CONFIG_NANOHUB) += nanohub.o
+nanohub-y := main.o bl.o comms.o
+nanohub-$(CONFIG_NANOHUB_SPI)  += spi.o
+nanohub-$(CONFIG_NANOHUB_I2C) += i2c.o
+nanohub-$(CONFIG_NANOHUB_SHAREMEM) += sharemem.o
+ccflags-y +=-DLINUX_KERNEL_4_9
