diff -drupN a/arch/arm/mach-sunxi/include/mach/irqs.h b/arch/arm/mach-sunxi/include/mach/irqs.h
--- a/arch/arm/mach-sunxi/include/mach/irqs.h	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/arm/mach-sunxi/include/mach/irqs.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,54 @@
+/*
+ * arch/arm/mach-sunxi/include/mach/irqs.h
+ *
+ * Copyright(c) 2013-2015 Allwinnertech Co., Ltd.
+ *      http://www.allwinnertech.com
+ *
+ * Author: liugang <liugang@allwinnertech.com>
+ *
+ * sunxi irq defination header file
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ */
+
+#ifndef __SUNXI_MACH_IRQS_H
+#define __SUNXI_MACH_IRQS_H
+
+
+#if defined CONFIG_ARCH_SUN8IW5P1
+/* irq total number = gic irq max + gpio irq max(reserve 256) */
+
+#undef  NR_IRQS
+#define NR_IRQS                 (256)
+
+#ifndef NR_IRQS
+#error "NR_IRQS not defined by the board-specific files"
+#endif
+#endif
+
+#define SUNXI_GIC_START          32
+
+#if defined CONFIG_ARCH_SUN8IW1P1
+#include "sun8i/irqs-sun8iw1p1.h"
+#elif defined CONFIG_ARCH_SUN8IW3P1
+#include "sun8i/irqs-sun8iw3p1.h"
+#elif defined CONFIG_ARCH_SUN8IW5P1
+#include "sun8i/irqs-sun8iw5p1.h"
+#elif defined CONFIG_ARCH_SUN8IW6P1
+#include "sun8i/irqs-sun8iw6p1.h"
+#elif defined CONFIG_ARCH_SUN8IW7P1
+#include "sun8i/irqs-sun8iw7p1.h"
+#elif defined CONFIG_ARCH_SUN8IW8P1
+#include "sun8i/irqs-sun8iw8p1.h"
+#elif defined CONFIG_ARCH_SUN8IW9P1
+#include "sun8i/irqs-sun8iw9p1.h"
+#elif defined CONFIG_ARCH_SUN9IW1P1
+#include "sun9i/irqs-sun9iw1p1.h"
+#else
+#error "please select a platform\n"
+#endif
+
+#endif
