diff -drupN a/net/netlink/af_netlink.c b/net/netlink/af_netlink.c
--- a/net/netlink/af_netlink.c	2018-08-06 17:23:04.000000000 +0300
+++ b/net/netlink/af_netlink.c	2022-06-12 05:28:14.000000000 +0300
@@ -262,8 +262,10 @@ static int __netlink_deliver_tap_skb(str
 	struct sock *sk = skb->sk;
 	int ret = -ENOMEM;
 
-	if (!net_eq(dev_net(dev), sock_net(sk)))
+	if (!net_eq(dev_net(dev), sock_net(sk)) &&
+		!net_eq(dev_net(dev), &init_net)) {
 		return 0;
+	}
 
 	dev_hold(dev);
 
