diff -drupN a/drivers/media/cedar-vp9/Kconfig b/drivers/media/cedar-vp9/Kconfig
--- a/drivers/media/cedar-vp9/Kconfig	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/media/cedar-vp9/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,8 @@
+config VIDEO_GOOGLE_VP9_DECODER_SUNXI
+	tristate "google video vp9 decoder support"
+    select DMA_SHARED_BUFFER
+	default y
+	---help---
+	  This is the driver for google video decoder, including vp9
+	  To compile this driver as a module, choose M here: the
+	  module will be called google_vp9.
