diff -drupN a/drivers/usb/phy/Makefile b/drivers/usb/phy/Makefile
--- a/drivers/usb/phy/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/usb/phy/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -3,6 +3,8 @@
 #
 obj-$(CONFIG_USB_PHY)			+= phy.o
 obj-$(CONFIG_OF)			+= of.o
+obj-$(CONFIG_USB_OTG_WAKELOCK)		+= otg-wakelock.o
+obj-$(CONFIG_DUAL_ROLE_USB_INTF)	+= class-dual-role.o
 
 # transceiver drivers, keep the list sorted
 
