diff -drupN a/fs/utimes.c b/fs/utimes.c
--- a/fs/utimes.c	2018-08-06 17:23:04.000000000 +0300
+++ b/fs/utimes.c	2022-06-12 05:28:14.000000000 +0300
@@ -91,7 +91,7 @@ static int utimes_common(struct path *pa
 	}
 retry_deleg:
 	inode_lock(inode);
-	error = notify_change(path->dentry, &newattrs, &delegated_inode);
+	error = notify_change2(path->mnt, path->dentry, &newattrs, &delegated_inode);
 	inode_unlock(inode);
 	if (delegated_inode) {
 		error = break_deleg_wait(&delegated_inode);
