diff -drupN a/drivers/char/Makefile b/drivers/char/Makefile
--- a/drivers/char/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/char/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -59,4 +59,12 @@ js-rtc-y = rtc.o
 
 obj-$(CONFIG_TILE_SROM)		+= tile-srom.o
 obj-$(CONFIG_XILLYBUS)		+= xillybus/
+obj-y                           += sunxi-scr/
 obj-$(CONFIG_POWERNV_OP_PANEL)	+= powernv-op-panel.o
+obj-$(CONFIG_ARCH_SUNXI)	+= sunxi-sysinfo/
+
+obj-$(CONFIG_SUNXI_TIMER_TEST)  += timer_test/
+obj-$(CONFIG_DUMP_REG)          += dump_reg/
+obj-$(CONFIG_SUNXI_TRANSFORM)   += sunxi_tr/
+obj-y                           += sunxi-di/
+obj-$(CONFIG_SUNXI_G2D)         += sunxi_g2d/
