diff -drupN a/drivers/pinctrl/sirf/pinctrl-atlas7.c b/drivers/pinctrl/sirf/pinctrl-atlas7.c
--- a/drivers/pinctrl/sirf/pinctrl-atlas7.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/pinctrl/sirf/pinctrl-atlas7.c	2022-06-12 05:28:14.000000000 +0300
@@ -5322,7 +5322,8 @@ static int atlas7_pin_config_set(struct
 				unsigned pin, unsigned long *configs,
 				unsigned num_configs)
 {
-	u16 param, arg;
+	u16 param;
+	u32 arg;
 	int idx, err;
 
 	for (idx = 0; idx < num_configs; idx++) {
