diff -drupN a/drivers/mfd/Kconfig b/drivers/mfd/Kconfig
--- a/drivers/mfd/Kconfig	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/mfd/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -18,6 +18,32 @@ config MFD_CS5535
 	  This is the core driver for CS5535/CS5536 MFD functions.  This is
           necessary for using the board's GPIO and MFGPT functionality.
 
+config MFD_ACX00
+	bool "Support Allwinnertech ACX00"
+	select MFD_CORE
+	select REGMAP_I2C
+	select REGMAP_IRQ
+	depends on I2C=y
+	default n
+	help
+	  Say yes here if you have ac100/ac200 chip on your board.
+	  .This driver provides core support for the ac100,ac200,
+	  in order to use the actual functionaltiy of the device
+	  other drivers must be enabled.
+
+config MFD_AC100
+	bool "Support Allwinnertech AC100"
+	select MFD_CORE
+	select REGMAP_I2C
+	select REGMAP_IRQ
+	depends on I2C=y
+	default n
+	help
+	  Say yes here if you have ac100 chip on your board.
+	  .This driver provides core support for the ac100,
+	  in order to use the actual functionaltiy of the device
+	  other drivers must be enabled.
+
 config MFD_ALTERA_A10SR
 	bool "Altera Arria10 DevKit System Resource chip"
 	depends on ARCH_SOCFPGA && SPI_MASTER=y && OF
@@ -112,7 +138,7 @@ config MFD_BCM590XX
 	help
 	  Support for the BCM590xx PMUs from Broadcom
 
-config MFD_AC100
+config XP_MFD_AC100
 	tristate "X-Powers AC100"
 	select MFD_CORE
 	depends on SUNXI_RSB
@@ -122,6 +148,23 @@ config MFD_AC100
 	  This driver include only the core APIs. You have to select individual
 	  components like codecs or RTC under the corresponding menus.
 
+config MFD_AXP2101
+	tristate
+	select MFD_CORE
+	select REGMAP_IRQ
+
+config MFD_AXP2101_I2C
+	tristate "X-Powers AXP2101 PMICs with I2C"
+	select MFD_AXP2101
+	select REGMAP_I2C
+	depends on I2C
+	help
+	  If you say Y here you get support for the X-Powers AXP2101 power
+	  management ICs (PMICs) controlled with I2C.
+	  This driver include only the core APIs. You have to select individual
+	  components like regulators or the PEK (Power Enable Key) under the
+	  corresponding menus.
+
 config MFD_AXP20X
 	tristate
 	select MFD_CORE
