diff -drupN a/drivers/media/pci/tw686x/tw686x-core.c b/drivers/media/pci/tw686x/tw686x-core.c
--- a/drivers/media/pci/tw686x/tw686x-core.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/media/pci/tw686x/tw686x-core.c	2022-06-12 05:28:14.000000000 +0300
@@ -72,12 +72,12 @@ static const char *dma_mode_name(unsigne
 	}
 }
 
-static int tw686x_dma_mode_get(char *buffer, struct kernel_param *kp)
+static int tw686x_dma_mode_get(char *buffer, const struct kernel_param *kp)
 {
 	return sprintf(buffer, dma_mode_name(dma_mode));
 }
 
-static int tw686x_dma_mode_set(const char *val, struct kernel_param *kp)
+static int tw686x_dma_mode_set(const char *val, const struct kernel_param *kp)
 {
 	if (!strcasecmp(val, dma_mode_name(TW686X_DMA_MODE_MEMCPY)))
 		dma_mode = TW686X_DMA_MODE_MEMCPY;
