diff -drupN a/drivers/w1/w1_int.c b/drivers/w1/w1_int.c
--- a/drivers/w1/w1_int.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/w1/w1_int.c	2022-06-12 05:28:14.000000000 +0300
@@ -32,7 +32,7 @@
 #include "w1_netlink.h"
 #include "w1_int.h"
 
-static int w1_search_count = -1; /* Default is continual scan */
+static int w1_search_count = 1; /* Default is continual scan */
 module_param_named(search_count, w1_search_count, int, 0);
 
 static int w1_enable_pullup = 1;
