diff -drupN a/fs/fs-writeback.c b/fs/fs-writeback.c
--- a/fs/fs-writeback.c	2018-08-06 17:23:04.000000000 +0300
+++ b/fs/fs-writeback.c	2022-06-12 05:28:14.000000000 +0300
@@ -2112,7 +2112,7 @@ void __mark_inode_dirty(struct inode *in
 	    (dirtytime && (inode->i_state & I_DIRTY_INODE)))
 		return;
 
-	if (unlikely(block_dump))
+	if (unlikely(block_dump > 1))
 		block_dump___mark_inode_dirty(inode);
 
 	spin_lock(&inode->i_lock);
