diff -drupN a/drivers/power/supply/Kconfig b/drivers/power/supply/Kconfig
--- a/drivers/power/supply/Kconfig	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/power/supply/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -8,6 +8,113 @@ menuconfig POWER_SUPPLY
 
 if POWER_SUPPLY
 
+menuconfig AW_AXP
+	bool "AXP Power drivers"
+	depends on REGULATOR
+	depends on ARCH_SUNXI
+	select MFD_CORE
+	help
+	  This option allows you to select AXP power driver
+
+if AW_AXP
+
+config AXP_TWI_USED
+	bool "AXP use twi as transfer channel"
+	depends on I2C
+	help
+	  To use twi as transfer channel
+
+config AXP_ARISC_TWI_USED
+	bool "AXP use arisc twi as transfer channel"
+	help
+	  To use arisc twi as transfer channel
+
+config DUAL_AXP_USED
+	bool "Support dual axp online"
+	depends on ARCH_SUN8IW10 || ARCH_SUN8IW15
+
+config AW_AXP20X
+	bool "AXP20x driver"
+	depends on ARCH_SUN8IW8
+	help
+	  AXP20x driver
+
+config AW_AXP2101
+	def_bool n
+	prompt "AXP2101 driver"
+	help
+	  axp2101 driver
+
+config AW_AXP22X
+	bool "AXP22x driver"
+	depends on ARCH_SUN8IW11 || ARCH_SUN8IW12 || ARCH_SUN8IW15 \
+		|| ARCH_SUN8IW16
+	help
+	  AXP22x driver
+
+if AW_AXP22X
+config AW_AXP22X_BAT_CAPACITY_DAEMON
+       bool "add timer irq to capture battery capacity"
+       depends on AW_AXP22X
+       default y
+       help
+         if axp chip capacity not correct,you can use it.
+         it will calculate the capacity according to ocv
+         and battery curve and adjuse axp capacity to system.
+endif
+
+
+config AW_PMU1736
+    bool "PMU1736 driver"
+    depends on ARCH_SUN8IW15 || ARCH_SUN8IW17
+    help
+        PMU1736 driver
+
+config AW_BMU1760
+	bool "BMU1760 driver"
+	depends on ARCH_SUN8IW15
+	help
+	BMU1760 driver, BMU1760 is a BMU,which is just for battery power manager.
+	We could use a BMU and a PMU for high voltage project.The BMU is good
+	at manage battery, and the PMU could be a supplier of  periphrals.
+	Say Y here to enable BMU1760 driver.
+
+config TYPE_C
+	bool "Support Type-C cc logic"
+	depends on AW_BMU1760
+	help
+	  To use Type-C cc logic function
+
+config AW_AXP259
+	bool "AXP259 driver"
+	depends on ARCH_SUN8IW11 || ARCH_SUN8IW12 || ARCH_SUN8IW16
+	help
+	AXP259 driver, AXP259 is a BMU,which is just for battery power manager.
+	We could use a BMU and a PMU for high voltage project.The BMU is good
+	at manager battery, and the PMU could supply for periphrals.
+	Say Y here to enable AXP259 driver.
+
+config AW_AXP233
+	bool "AXP233 driver"
+	depends on ARCH_SUN8IW11 || ARCH_SUN8IW12 || ARCH_SUN8IW16
+	default n
+	help
+	  AXP233 driver, compatible with AXP22X.
+
+config AW_AXP80X
+	bool "AXP80x driver"
+	depends on ARCH_SUN8IW12 || ARCH_SUN50IW6 || ARCH_SUN8IW16
+	help
+	  AXP80x driver
+
+config AW_AXP803
+	bool "AXP803 driver"
+	depends on ARCH_SUN50IW1 || ARCH_SUN50IW3 || ARCH_SUN8IW6
+	help
+	  AXP803 driver
+
+endif # AW_AXP
+
 config POWER_SUPPLY_DEBUG
 	bool "Power supply debug"
 	help
@@ -511,4 +618,17 @@ config AXP20X_POWER
 	  This driver provides support for the power supply features of
 	  AXP20x PMIC.
 
+config AXP2101_POWER
+	tristate "AXP2101 power supply driver"
+	depends on MFD_AXP2101
+	help
+	  This driver provides support for the power supply features of
+	  AXP2101 PMIC.
+
+config AXP152_VBUS_POWER
+	tristate "AXP152 vbus check"
+	depends on MFD_AXP2101
+	help
+	  This driver provides support for the power supply features of
+	  AXP152 PMIC.
 endif # POWER_SUPPLY
