diff -drupN a/include/linux/cgroup_subsys.h b/include/linux/cgroup_subsys.h
--- a/include/linux/cgroup_subsys.h	2018-08-06 17:23:04.000000000 +0300
+++ b/include/linux/cgroup_subsys.h	2022-06-12 05:28:14.000000000 +0300
@@ -20,6 +20,10 @@ SUBSYS(cpu)
 SUBSYS(cpuacct)
 #endif
 
+#if IS_ENABLED(CONFIG_CGROUP_SCHEDTUNE)
+SUBSYS(schedtune)
+#endif
+
 #if IS_ENABLED(CONFIG_BLK_CGROUP)
 SUBSYS(io)
 #endif
