diff -drupN a/include/linux/efi.h b/include/linux/efi.h
--- a/include/linux/efi.h	2018-08-06 17:23:04.000000000 +0300
+++ b/include/linux/efi.h	2022-06-12 05:28:14.000000000 +0300
@@ -1427,7 +1427,7 @@ efi_status_t handle_cmdline_files(efi_sy
 				  unsigned long *load_addr,
 				  unsigned long *load_size);
 
-efi_status_t efi_parse_options(char *cmdline);
+efi_status_t efi_parse_options(char const *cmdline);
 
 efi_status_t efi_setup_gop(efi_system_table_t *sys_table_arg,
 			   struct screen_info *si, efi_guid_t *proto,
