diff -drupN a/sound/soc/codecs/Makefile b/sound/soc/codecs/Makefile
--- a/sound/soc/codecs/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/sound/soc/codecs/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -220,6 +220,15 @@ snd-soc-max9877-objs := max9877.o
 snd-soc-max98504-objs := max98504.o
 snd-soc-tpa6130a2-objs := tpa6130a2.o
 snd-soc-tas2552-objs := tas2552.o
+# Sunxi
+snd-soc-acx00-objs := acx00.o
+snd-soc-ac100-objs := ac100_dapm.o
+snd-soc-ac102-objs := ac102.o
+snd-soc-ac107-objs := ac107.o
+snd-soc-ac108-objs := ac108.o
+# add by Sunxi
+snd-soc-ad82584f-objs := ad82584f.o
+snd-soc-es7243-objs := es7243.o
 
 obj-$(CONFIG_SND_SOC_88PM860X)	+= snd-soc-88pm860x.o
 obj-$(CONFIG_SND_SOC_AB8500_CODEC)	+= snd-soc-ab8500-codec.o
@@ -440,3 +449,14 @@ obj-$(CONFIG_SND_SOC_WM_HUBS)	+= snd-soc
 obj-$(CONFIG_SND_SOC_MAX9877)	+= snd-soc-max9877.o
 obj-$(CONFIG_SND_SOC_MAX98504)	+= snd-soc-max98504.o
 obj-$(CONFIG_SND_SOC_TPA6130A2)	+= snd-soc-tpa6130a2.o
+
+# ACX00
+obj-$(CONFIG_SND_SOC_ACX00)	+= snd-soc-acx00.o
+obj-$(CONFIG_SND_SOC_AC100)	+= snd-soc-ac100.o
+obj-$(CONFIG_SND_SOC_AC102)	+= snd-soc-ac102.o
+obj-$(CONFIG_SND_SOC_AC107)	+= snd-soc-ac107.o
+obj-$(CONFIG_SND_SOC_AC108)	+= snd-soc-ac108.o
+
+# add by Sunxi
+obj-$(CONFIG_SND_SOC_ES7243)	+= snd-soc-es7243.o
+obj-$(CONFIG_SND_SOC_AD82584F)	+= snd-soc-ad82584f.o
