diff -drupN a/drivers/media/rc/keymaps/Makefile b/drivers/media/rc/keymaps/Makefile
--- a/drivers/media/rc/keymaps/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/media/rc/keymaps/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -106,4 +106,5 @@ obj-$(CONFIG_RC_MAP) += rc-adstech-dvb-t
 			rc-videomate-tv-pvr.o \
 			rc-winfast.o \
 			rc-winfast-usbii-deluxe.o \
-			rc-su3000.o
+			rc-su3000.o \
+			rc-aw-nec.o
\ В конце файла нет новой строки
