diff -drupN a/include/net/ip_fib.h b/include/net/ip_fib.h
--- a/include/net/ip_fib.h	2018-08-06 17:23:04.000000000 +0300
+++ b/include/net/ip_fib.h	2022-06-12 05:28:14.000000000 +0300
@@ -70,7 +70,11 @@ struct fnhe_hash_bucket {
 	struct fib_nh_exception __rcu	*chain;
 };
 
+#if CONFIG_BASE_SMALL > 0
 #define FNHE_HASH_SHIFT		11
+#else
+#define FNHE_HASH_SHIFT		4
+#endif
 #define FNHE_HASH_SIZE		(1 << FNHE_HASH_SHIFT)
 #define FNHE_RECLAIM_DEPTH	5
 
@@ -171,6 +175,7 @@ struct fib_result_nl {
 #ifdef CONFIG_IP_MULTIPLE_TABLES
 #define FIB_TABLE_HASHSZ 256
 #else
+/* Can we use 1 for BASE_SMALL? */
 #define FIB_TABLE_HASHSZ 2
 #endif
 
