diff -drupN a/drivers/input/sensor/bma253/Makefile b/drivers/input/sensor/bma253/Makefile
--- a/drivers/input/sensor/bma253/Makefile	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/input/sensor/bma253/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,14 @@
+#
+# Makefile for Bosch sensor driver.
+#
+#obj-y += bstclass.o
+
+obj-$(CONFIG_SENSORS_BMA2X2)    += bma250.o
+bma250-objs = bma2x2.o bstclass.o
+ifeq ($(CONFIG_SENSORS_BMA2X2_ENABLE_INT1),y)
+	EXTRA_CFLAGS += -DBMA2X2_ENABLE_INT1
+endif
+
+ifeq ($(CONFIG_BOSCH_BMA2X2_ENABLE_INT2),y)
+	EXTRA_CFLAGS += -DBMA2X2_ENABLE_INT2
+endif
