diff -drupN a/drivers/gpu/drm/drm_plane.c b/drivers/gpu/drm/drm_plane.c
--- a/drivers/gpu/drm/drm_plane.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/gpu/drm/drm_plane.c	2022-06-12 05:28:14.000000000 +0300
@@ -137,6 +137,7 @@ int drm_universal_plane_init(struct drm_
 
 	if (drm_core_check_feature(dev, DRIVER_ATOMIC)) {
 		drm_object_attach_property(&plane->base, config->prop_fb_id, 0);
+		drm_object_attach_property(&plane->base, config->prop_in_fence_fd, -1);
 		drm_object_attach_property(&plane->base, config->prop_crtc_id, 0);
 		drm_object_attach_property(&plane->base, config->prop_crtc_x, 0);
 		drm_object_attach_property(&plane->base, config->prop_crtc_y, 0);
