diff -drupN a/drivers/perf/arm_pmu.c b/drivers/perf/arm_pmu.c
--- a/drivers/perf/arm_pmu.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/perf/arm_pmu.c	2022-06-12 05:28:14.000000000 +0300
@@ -1018,7 +1018,7 @@ int arm_pmu_device_probe(struct platform
 			 const struct pmu_probe_info *probe_table)
 {
 	const struct of_device_id *of_id;
-	const int (*init_fn)(struct arm_pmu *);
+	int (*init_fn)(struct arm_pmu *);
 	struct device_node *node = pdev->dev.of_node;
 	struct arm_pmu *pmu;
 	int ret = -ENODEV;
