diff -drupN a/arch/arm64/include/asm/perf_event.h b/arch/arm64/include/asm/perf_event.h
--- a/arch/arm64/include/asm/perf_event.h	2018-08-06 17:23:04.000000000 +0300
+++ b/arch/arm64/include/asm/perf_event.h	2022-06-12 05:28:14.000000000 +0300
@@ -17,6 +17,8 @@
 #ifndef __ASM_PERF_EVENT_H
 #define __ASM_PERF_EVENT_H
 
+#include <asm/stack_pointer.h>
+
 #define	ARMV8_PMU_MAX_COUNTERS	32
 #define	ARMV8_PMU_COUNTER_MASK	(ARMV8_PMU_MAX_COUNTERS - 1)
 
