diff -drupN a/fs/lockd/svc.c b/fs/lockd/svc.c
--- a/fs/lockd/svc.c	2018-08-06 17:23:04.000000000 +0300
+++ b/fs/lockd/svc.c	2022-06-12 05:28:14.000000000 +0300
@@ -600,7 +600,7 @@ static struct ctl_table nlm_sysctl_root[
  */
 
 #define param_set_min_max(name, type, which_strtol, min, max)		\
-static int param_set_##name(const char *val, struct kernel_param *kp)	\
+static int param_set_##name(const char *val, const struct kernel_param *kp) \
 {									\
 	char *endp;							\
 	__typeof__(type) num = which_strtol(val, &endp, 0);		\
