diff -drupN a/drivers/Makefile b/drivers/Makefile
--- a/drivers/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -124,6 +124,7 @@ obj-$(CONFIG_EISA)		+= eisa/
 obj-y				+= lguest/
 obj-$(CONFIG_CPU_FREQ)		+= cpufreq/
 obj-$(CONFIG_CPU_IDLE)		+= cpuidle/
+obj-$(CONFIG_SUNXI_MPP)		+= mpp/
 obj-y				+= mmc/
 obj-$(CONFIG_MEMSTICK)		+= memstick/
 obj-$(CONFIG_NEW_LEDS)		+= leds/
@@ -175,3 +176,5 @@ obj-$(CONFIG_STM)		+= hwtracing/stm/
 obj-$(CONFIG_ANDROID)		+= android/
 obj-$(CONFIG_NVMEM)		+= nvmem/
 obj-$(CONFIG_FPGA)		+= fpga/
+obj-$(CONFIG_TEE)		+= tee/
+obj-$(CONFIG_GATOR)		+= gator/
\ В конце файла нет новой строки
