diff -drupN a/drivers/tty/serial/serial_core.c b/drivers/tty/serial/serial_core.c
--- a/drivers/tty/serial/serial_core.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/tty/serial/serial_core.c	2022-06-12 05:28:14.000000000 +0300
@@ -131,6 +131,9 @@ static void __uart_start(struct tty_stru
 	struct uart_state *state = tty->driver_data;
 	struct uart_port *port = state->uart_port;
 
+	if (port && port->ops->wake_peer)
+		port->ops->wake_peer(port);
+
 	if (port && !uart_tx_stopped(port))
 		port->ops->start_tx(port);
 }
