diff -drupN a/arch/arm/boot/dts/sun8iw17p1-fpga.dts b/arch/arm/boot/dts/sun8iw17p1-fpga.dts
--- a/arch/arm/boot/dts/sun8iw17p1-fpga.dts	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/arm/boot/dts/sun8iw17p1-fpga.dts	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,26 @@
+/*
+ * Allwinner Technology CO., Ltd. sun8iw17p1 fpga board.
+ *
+ * fpga board support.
+ */
+
+/dts-v1/;
+
+/memreserve/ 0x78000000 0x08000000; /* ion carvout heap revserve  : [0x38000000~0x40000000], size = 128M  */
+
+#include "sun8iw17p1.dtsi"
+
+/{
+	ion {
+		carvout{
+			type = <2>;
+			name = "carvout";
+			base = <0x78000000>;
+			size = <0x08000000>;
+		};
+	};
+	soc@03000000 {
+
+	};
+
+};
