diff -drupN a/drivers/mtd/devices/block2mtd.c b/drivers/mtd/devices/block2mtd.c
--- a/drivers/mtd/devices/block2mtd.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/mtd/devices/block2mtd.c	2022-06-12 05:28:14.000000000 +0300
@@ -431,7 +431,7 @@ static int block2mtd_setup2(const char *
 }
 
 
-static int block2mtd_setup(const char *val, struct kernel_param *kp)
+static int block2mtd_setup(const char *val, const struct kernel_param *kp)
 {
 #ifdef MODULE
 	return block2mtd_setup2(val);
