diff -drupN a/arch/arm/include/asm/mach/arch.h b/arch/arm/include/asm/mach/arch.h
--- a/arch/arm/include/asm/mach/arch.h	2018-08-06 17:23:04.000000000 +0300
+++ b/arch/arm/include/asm/mach/arch.h	2022-06-12 05:28:14.000000000 +0300
@@ -18,7 +18,7 @@ struct pt_regs;
 struct smp_operations;
 #ifdef CONFIG_SMP
 #define smp_ops(ops) (&(ops))
-#define smp_init_ops(ops) (&(ops))
+#define smp_init_ops(ops) ((void *)&(ops))
 #else
 #define smp_ops(ops) (struct smp_operations *)NULL
 #define smp_init_ops(ops) (bool (*)(void))NULL
