diff -drupN a/drivers/video/fbdev/sunxi/disp2/disp/de/lowlevel_sun8iw8/de_vep.h b/drivers/video/fbdev/sunxi/disp2/disp/de/lowlevel_sun8iw8/de_vep.h
--- a/drivers/video/fbdev/sunxi/disp2/disp/de/lowlevel_sun8iw8/de_vep.h	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/video/fbdev/sunxi/disp2/disp/de/lowlevel_sun8iw8/de_vep.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,20 @@
+/*
+ * Allwinner SoCs display driver.
+ *
+ * Copyright (C) 2016 Allwinner.
+ *
+ * This file is licensed under the terms of the GNU General Public
+ * License version 2.  This program is licensed "as is" without any
+ * warranty of any kind, whether express or implied.
+ */
+
+#ifndef __DE_VEP_H__
+#define __DE_VEP_H__
+
+int de_fcc_init(unsigned int sel, unsigned int reg_base);
+int de_fcc_update_regs(unsigned int sel);
+int de_fcc_set_reg_base(unsigned int sel, unsigned int chno, void *base);
+int de_fcc_csc_set(unsigned int sel, unsigned int chno, unsigned int en,
+		   unsigned int mode);
+
+#endif
