diff -drupN a/scripts/Makefile.clean b/scripts/Makefile.clean
--- a/scripts/Makefile.clean	2018-08-06 17:23:04.000000000 +0300
+++ b/scripts/Makefile.clean	2022-06-12 05:28:14.000000000 +0300
@@ -11,7 +11,7 @@ include scripts/Kbuild.include
 
 # The filename Kbuild has precedence over Makefile
 kbuild-dir := $(if $(filter /%,$(src)),$(src),$(srctree)/$(src))
-include $(if $(wildcard $(kbuild-dir)/Kbuild), $(kbuild-dir)/Kbuild, $(kbuild-dir)/Makefile)
+-include $(if $(wildcard $(kbuild-dir)/Kbuild), $(kbuild-dir)/Kbuild, $(kbuild-dir)/Makefile)
 
 # Figure out what we need to build from the various variables
 # ==========================================================================
