diff -drupN a/drivers/video/fbdev/sunxi/disp2/disp/de/disp_enhance.h b/drivers/video/fbdev/sunxi/disp2/disp/de/disp_enhance.h
--- a/drivers/video/fbdev/sunxi/disp2/disp/de/disp_enhance.h	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/video/fbdev/sunxi/disp2/disp/de/disp_enhance.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,23 @@
+/*
+ * Allwinner SoCs display driver.
+ *
+ * Copyright (C) 2016 Allwinner.
+ *
+ * This file is licensed under the terms of the GNU General Public
+ * License version 2.  This program is licensed "as is" without any
+ * warranty of any kind, whether express or implied.
+ */
+
+#ifndef _DISP_ENHANCE_H_
+#define _DISP_ENHANCE_H_
+
+#include "disp_private.h"
+
+static s32 disp_enhance_shadow_protect(struct disp_enhance *enhance,
+				       bool protect);
+s32 disp_init_enhance(struct disp_bsp_init_para *para);
+
+s32 disp_enhance_set_para(struct disp_enhance *enhance,
+			  struct disp_enhance_para *para);
+
+#endif
