diff -drupN a/drivers/media/Kconfig b/drivers/media/Kconfig
--- a/drivers/media/Kconfig	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/media/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -219,5 +219,8 @@ source "drivers/media/i2c/Kconfig"
 source "drivers/media/spi/Kconfig"
 source "drivers/media/tuners/Kconfig"
 source "drivers/media/dvb-frontends/Kconfig"
-
+source "drivers/media/cedar-ve/Kconfig"
+source "drivers/media/cedar-vp9/Kconfig"
+source "drivers/media/tsc/Kconfig"
+source "drivers/media/sunxi_nna/Kconfig"
 endif # MEDIA_SUPPORT
