diff -drupN a/fs/proc/internal.h b/fs/proc/internal.h
--- a/fs/proc/internal.h	2018-08-06 17:23:04.000000000 +0300
+++ b/fs/proc/internal.h	2022-06-12 05:28:14.000000000 +0300
@@ -256,6 +256,15 @@ static inline void sysctl_head_put(struc
 #endif
 
 /*
+ * uid.c
+ */
+#ifdef CONFIG_PROC_UID
+extern int proc_uid_init(void);
+#else
+static inline void proc_uid_init(void) { }
+#endif
+
+/*
  * proc_tty.c
  */
 #ifdef CONFIG_TTY
