diff -drupN a/drivers/cpufreq/Kconfig.arm b/drivers/cpufreq/Kconfig.arm
--- a/drivers/cpufreq/Kconfig.arm	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/cpufreq/Kconfig.arm	2022-06-12 05:28:14.000000000 +0300
@@ -1,6 +1,22 @@
 #
 # ARM CPU Frequency scaling drivers
 #
+config ARM_SUNXI_IKS_CPUFREQ
+	bool "SUNXI CPUFreq IKS driver support"
+	depends on ARCH_SUN9IW1 || ARCH_SUN8IW6 || ARCH_SUN8IW9 || ARCH_SUN8IW17
+	help
+	  This adds the CPUFreq driver for sunxi big.little soc.
+	  If in doubt, say N.
+
+config ARM_SUNXI_AVS
+	bool "Use adaptive voltage scaling"
+	depends on CPU_FREQ && SUNXI_ARISC && (ARCH_SUN8IW16P1 || ARCH_SUN50IW6)
+	default y if ARCH_SUN50IW6
+	default n if !ARCH_SUN50IW6
+	help
+	  use adaptive voltage scaling to reduce power consumption.
+
+
 
 # big LITTLE core layer and glue drivers
 config ARM_BIG_LITTLE_CPUFREQ
@@ -258,3 +274,80 @@ config ACPI_CPPC_CPUFREQ
 	  support for its operation.
 
 	  If in doubt, say N.
+config ARM_SUNXI_CPUFREQ
+	bool "SUNXI CPUFreq support"
+	select PM_OPP
+	depends on ARCH_SUNXI
+	help
+	  This adds the CPUFreq driver for the Allwinner SUNXI SoC.
+
+	  If in doubt, say N.
+config ARM_SUNXI_CPUFREQ_PWM
+	bool "SUNXI PWM CPUFreq support"
+	depends on ARCH_SUNXI
+	depends on HAVE_CLK && OF
+	select PM_OPP
+	select PWM
+	help
+	  This adds a generic DT based pwm cpufreq driver for frequency
+	  management. It supports both uniprocessor (UP) and symmetric
+	  multiprocessor (SMP) systems without pmu.
+
+	  If in doubt, say N.
+config SUNXI_CPUFREQ_ASYN
+	bool "CPUFreq support asynchronous work"
+	depends on ARM_SUNXI_CPUFREQ || ARM_SUNXI_IKS_CPUFREQ
+	default n
+	help
+		support asynchronous work for cpufreq
+
+config ARM_SUNXI_PSENSOR_BIN
+	tristate "use psensor to get soc bin"
+	depends on ARCH_SUN50IW3P1 || ARCH_SUN8IW17P1 || ARCH_SUN8IW15P1
+	help
+	Use the psensor to get the soc bin which used to choose the
+	suitable vf table.
+
+	If in doubt, say N.
+
+config CPU_AUTOHOTPLUG
+	tristate "CPU autohotplug support"
+	depends on SCHED_HMP || SCHED_SMP_DCMP || SMP
+	depends on HOTPLUG_CPU && !BL_SWITCHER
+	default n
+	help
+	  This add the CPU AUTOHOTPLUG support for sunxi.
+	  The driver will up/down cpu core according to cpu load.
+
+	  If in doubt, say N.
+
+config CPU_AUTOHOTPLUG_ROOMAGE
+	bool "CPU autohotplug support roomage limit"
+	depends on CPU_AUTOHOTPLUG
+	default y
+	help
+	 This add the CPU AUTOHOTPLUG ROOMAGE support for sunxi.
+	 The driver will add roomage check when cpu hotplug.
+
+	 If in doubt, say N.
+
+config CPU_AUTOHOTPLUG_STATS
+	bool "CPU autohotplug statistics"
+	depends on CPU_AUTOHOTPLUG
+	default y
+	help
+	  Enable up/down count and total time plugged statistics per CPU.
+	  just debug.
+
+	  If in doubt say N.
+
+config CPU_AUTOHOTPLUG_INPUT_EVNT_NOTIFY
+	bool "CPU autohotplug input event notify"
+	depends on INPUT && CPU_AUTOHOTPLUG
+	default y
+	help
+	  This will allow input sub-system notify autohotplug
+	  that some user event has happened, need change cpu up threshold
+	  to response it.
+
+	  If in doubt, say N.
