diff -drupN a/drivers/crypto/Kconfig b/drivers/crypto/Kconfig
--- a/drivers/crypto/Kconfig	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/crypto/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -538,6 +538,19 @@ config CRYPTO_DEV_SUN4I_SS
 	  To compile this driver as a module, choose M here: the module
 	  will be called sun4i-ss.
 
+config CRYPTO_DEV_SUNXI
+	tristate "Support for Allwinner Sunxi CryptoEngine"
+	depends on ARCH_SUNXI
+	select CRYPTO_MD5
+	select CRYPTO_SHA1
+	select CRYPTO_AES
+	select CRYPTO_DES
+	select CRYPTO_BLKCIPHER
+	help
+	  Allwinner Sunxi SoC have a crypto accelerator named
+	  CryptoEngine. It support some SoC of sun8i/sun50i series.
+	  Select this if you want to use it.
+
 config CRYPTO_DEV_ROCKCHIP
 	tristate "Rockchip's Cryptographic Engine driver"
 	depends on OF && ARCH_ROCKCHIP
@@ -554,5 +567,6 @@ config CRYPTO_DEV_ROCKCHIP
 	  Supporting cbc/ecb chainmode, and aes/des/des3_ede cipher mode.
 
 source "drivers/crypto/chelsio/Kconfig"
+source "drivers/crypto/sunxi-emce/Kconfig"
 
 endif # CRYPTO_HW
