diff -drupN a/drivers/pinctrl/pinctrl-single.c b/drivers/pinctrl/pinctrl-single.c
--- a/drivers/pinctrl/pinctrl-single.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/pinctrl/pinctrl-single.c	2022-06-12 05:28:14.000000000 +0300
@@ -630,7 +630,7 @@ static int pcs_pinconf_set(struct pinctr
 	struct pcs_device *pcs = pinctrl_dev_get_drvdata(pctldev);
 	struct pcs_function *func;
 	unsigned offset = 0, shift = 0, i, data, ret;
-	u16 arg;
+	u32 arg;
 	int j;
 
 	ret = pcs_get_function(pctldev, pin, &func);
