diff -drupN a/include/linux/proc_fs.h b/include/linux/proc_fs.h
--- a/include/linux/proc_fs.h	2018-08-06 17:23:04.000000000 +0300
+++ b/include/linux/proc_fs.h	2022-06-12 05:28:14.000000000 +0300
@@ -74,6 +74,12 @@ static inline int remove_proc_subtree(co
 
 #endif /* CONFIG_PROC_FS */
 
+#ifdef CONFIG_PROC_UID
+extern void proc_register_uid(kuid_t uid);
+#else
+static inline void proc_register_uid(kuid_t uid) {}
+#endif
+
 struct net;
 
 static inline struct proc_dir_entry *proc_net_mkdir(
