diff -drupN a/arch/arm64/kernel/vdso/Makefile b/arch/arm64/kernel/vdso/Makefile
--- a/arch/arm64/kernel/vdso/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/arch/arm64/kernel/vdso/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -14,6 +14,7 @@ obj-vdso := $(addprefix $(obj)/, $(obj-v
 ccflags-y := -shared -fno-common -fno-builtin
 ccflags-y += -nostdlib -Wl,-soname=linux-vdso.so.1 \
 		$(call cc-ldoption, -Wl$(comma)--hash-style=sysv)
+ccflags-y += $(DISABLE_LTO)
 
 # Disable gcov profiling for VDSO code
 GCOV_PROFILE := n
