diff -drupN a/include/linux/cryptohash.h b/include/linux/cryptohash.h
--- a/include/linux/cryptohash.h	2018-08-06 17:23:04.000000000 +0300
+++ b/include/linux/cryptohash.h	2022-06-12 05:28:14.000000000 +0300
@@ -15,6 +15,8 @@ void sha_transform(__u32 *digest, const
 
 void md5_transform(__u32 *hash, __u32 const *in);
 
+#ifdef CONFIG_HALFMD4
 __u32 half_md4_transform(__u32 buf[4], __u32 const in[8]);
+#endif
 
 #endif
