diff -drupN a/drivers/soc/sunxi/Makefile b/drivers/soc/sunxi/Makefile
--- a/drivers/soc/sunxi/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/soc/sunxi/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -1 +1,7 @@
 obj-$(CONFIG_SUNXI_SRAM) +=	sunxi_sram.o
+obj-y +=        sunxi-sid.o
+ifeq ($(strip $(CONFIG_ARCH_SUN8IW19)$(CONFIG_ARCH_SUN8IW16)$(CONFIG_ARCH_SUN50IW10)$(CONFIG_ARCH_SUN50IW11)),)
+obj-$(CONFIG_PM) += pm.o
+endif
+obj-y += arisc/
+obj-y += sunxi-dump.o
