diff -drupN a/drivers/video/fbdev/sunxi/Makefile b/drivers/video/fbdev/sunxi/Makefile
--- a/drivers/video/fbdev/sunxi/Makefile	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/video/fbdev/sunxi/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,25 @@
+# Makefile for the Linux video drivers on sunxi platform.
+# 30 Sep 2016, Tyle, <mailto:tyle@allwinnertech.com>
+# Rewritten to use lists instead of if-statements.
+#
+# Each configuration option enables a list of files.
+
+obj-$(CONFIG_DISP2_SUNXI) += disp2/disp/
+
+obj-$(CONFIG_HDMI_DISP2_SUNXI) += disp2/hdmi/
+
+obj-$(CONFIG_HDMI2_DISP2_SUNXI) += disp2/hdmi2/
+
+obj-$(CONFIG_HDMI_EP952_DISP2_SUNXI) += disp2/hdmi_ep952/
+
+obj-$(CONFIG_TV_DISP2_SUNXI) += disp2/tv/
+
+obj-$(CONFIG_DISP2_TV_AC200) += disp2/tv/
+
+obj-$(CONFIG_VDPO_DISP2_SUNXI) += disp2/vdpo/
+
+obj-$(CONFIG_EDP_DISP2_SUNXI) += disp2/edp/
+
+obj-$(CONFIG_EINK200_SUNXI) += disp2/eink200/
+
+obj-$(CONFIG_LCD_FB) += lcd_fb/
