diff -drupN a/modules/nand/sun50iw9p1/phy-nand/rawnand/controller/ndfc_timings.h b/modules/nand/sun50iw9p1/phy-nand/rawnand/controller/ndfc_timings.h
--- a/modules/nand/sun50iw9p1/phy-nand/rawnand/controller/ndfc_timings.h	1970-01-01 03:00:00.000000000 +0300
+++ b/modules/nand/sun50iw9p1/phy-nand/rawnand/controller/ndfc_timings.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,21 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+
+#ifndef __NAND_CONTROLLER_COMMON_H__
+#define __NAND_CONTROLLER_COMMON_H__
+
+#include "ndfc_base.h"
+
+int add_to_nctri(struct nand_controller_info *node);
+struct nand_controller_info *nctri_get(struct nand_controller_info *head, unsigned int num);
+int before_update_nctri_interface(struct nand_controller_info *nctri);
+int start_update_nctri_interface(struct nand_controller_info *nctri, u32 input_timing_para, u32 ddr_type, u32 ddr_type_old, u32 sclk0, u32 sclk1, u32 sdr_edo_i, u32 ddr_edo_i, u32 ddr_delay_i);
+int ddr_mode_to_sdr(struct nand_controller_info *nctri);
+int update_1st_nctri_interface(struct nand_controller_info *nctri);
+int update_other_nctri_interface(struct nand_controller_info *nctri);
+s32 _update_ext_accessfreq_para(struct nand_controller_info *nctri);
+int set_nand_script_frequence(void);
+int update_nctri(struct nand_controller_info *nctri);
+int check_nctri(struct nand_controller_info *nctri);
+int update_boot0_nctri_interface(struct nand_controller_info *nctri);
+
+#endif /*NAND_CONTROLLER_COMMON_H*/
