diff -drupN a/arch/arm64/kernel/insn.c b/arch/arm64/kernel/insn.c
--- a/arch/arm64/kernel/insn.c	2018-08-06 17:23:04.000000000 +0300
+++ b/arch/arm64/kernel/insn.c	2022-06-12 05:28:14.000000000 +0300
@@ -97,7 +97,7 @@ static void __kprobes *patch_map(void *a
 	if (module && IS_ENABLED(CONFIG_DEBUG_SET_MODULE_RONX))
 		page = vmalloc_to_page(addr);
 	else if (!module)
-		page = pfn_to_page(PHYS_PFN(__pa(addr)));
+		page = phys_to_page(__pa_symbol(addr));
 	else
 		return addr;
 
