diff -drupN a/drivers/thermal/sunxi_thermal/sunxi_thermal_sensor/sunxi_ths_efuse.h b/drivers/thermal/sunxi_thermal/sunxi_thermal_sensor/sunxi_ths_efuse.h
--- a/drivers/thermal/sunxi_thermal/sunxi_thermal_sensor/sunxi_ths_efuse.h	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/thermal/sunxi_thermal/sunxi_thermal_sensor/sunxi_ths_efuse.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,47 @@
+/*
+ * drivers/thermal/sunxi_thermal/sunxi_ths_efuse.h
+ *
+ * Copyright (C) 2013-2024 allwinner.
+ *	JiaRui Xiao<xiaojiarui@allwinnertech.com>
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License version 2 as
+ *  published by the Free Software Foundation.
+ */
+
+#ifndef __SUNXI_THS_EFUSE_H__
+#define __SUNXI_THS_EFUSE_H__
+
+/* GET TMEP RETURN VALUE */
+#define WRONG_EFUSE_REG_DATA		(0xffff)
+#define THS_EFUSE_DEFAULT_VALUE		(0x800)
+#define THS_EFUSE_ENVIRONMENT_MASK	(0x0fff)
+#define THS_EFUSE_CP_FT_MASK		(0x3000)
+#define THS_EFUSE_CP_FT_BIT		(12)/* depend on ths calibration doc */
+#define THS_CALIBRATION_IN_FT		(1)
+
+
+
+#if defined(CONFIG_ARCH_SUN8IW17)
+/*Tcdata = 0x800 - [(T-envT)/0.1172) */
+#define SENSOR_CP_EUFSE_PER_REG_TO_TEMP (1172)/* this value is 0.1172 */
+#define CONST_DIV			(1000)
+#else
+/*Tcdata = 0x800 - [(T-envT)/0.0672) */
+#define SENSOR_CP_EUFSE_PER_REG_TO_TEMP (672)/* this value is 0.0672 */
+#define CONST_DIV			(1000)
+#endif
+
+#if defined(CONFIG_ARCH_SUN50IW6)
+#define FT_CALIBRATION_DEVIATION	(7)/* degrees celsius */
+#else
+#define FT_CALIBRATION_DEVIATION	(0)/* degrees celsius */
+#endif
+
+#if defined CONFIG_ARCH_SUN8IW7
+#define CALIBRATION_DEVIATION  15
+#endif
+
+
+#endif/* __SUNXI_THS_EFUSE_H__ */
+
