diff -drupN a/drivers/thermal/Kconfig b/drivers/thermal/Kconfig
--- a/drivers/thermal/Kconfig	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/thermal/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -381,6 +381,17 @@ config MTK_THERMAL
 	  Enable this option if you want to have support for thermal management
 	  controller present in Mediatek SoCs
 
+
+menu "allwinner(SUNXI) thermal drivers"
+depends on ARCH_SUNXI
+source "drivers/thermal/sunxi_thermal/Kconfig"
+endmenu
+
+menu "Texas Instruments thermal drivers"
+depends on ARCH_HAS_BANDGAP || COMPILE_TEST
+source "drivers/thermal/ti-soc-thermal/Kconfig"
+endmenu
+
 menu "Texas Instruments thermal drivers"
 depends on ARCH_HAS_BANDGAP || COMPILE_TEST
 depends on HAS_IOMEM
