diff -drupN a/drivers/mtd/ubi/build.c b/drivers/mtd/ubi/build.c
--- a/drivers/mtd/ubi/build.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/mtd/ubi/build.c	2022-06-12 05:28:14.000000000 +0300
@@ -1403,7 +1403,7 @@ static int __init bytes_str_to_int(const
  * This function returns zero in case of success and a negative error code in
  * case of error.
  */
-static int __init ubi_mtd_param_parse(const char *val, struct kernel_param *kp)
+static int __init ubi_mtd_param_parse(const char *val, const struct kernel_param *kp)
 {
 	int i, len;
 	struct mtd_dev_param *p;
