diff -drupN a/drivers/staging/goldfish/Kconfig b/drivers/staging/goldfish/Kconfig
--- a/drivers/staging/goldfish/Kconfig	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/staging/goldfish/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -4,6 +4,14 @@ config GOLDFISH_AUDIO
 	---help---
 	  Emulated audio channel for the Goldfish Android Virtual Device
 
+config GOLDFISH_SYNC
+    tristate "Goldfish AVD Sync Driver"
+    depends on GOLDFISH
+    depends on SW_SYNC
+    depends on SYNC_FILE
+	---help---
+	  Emulated sync fences for the Goldfish Android Virtual Device
+
 config MTD_GOLDFISH_NAND
 	tristate "Goldfish NAND device"
 	depends on GOLDFISH
