diff -drupN a/drivers/video/fbdev/sunxi/disp2/hdmi2/hdmi_core/api/hdcp22/include/ESMHostLibDriverErrors.h b/drivers/video/fbdev/sunxi/disp2/hdmi2/hdmi_core/api/hdcp22/include/ESMHostLibDriverErrors.h
--- a/drivers/video/fbdev/sunxi/disp2/hdmi2/hdmi_core/api/hdcp22/include/ESMHostLibDriverErrors.h	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/video/fbdev/sunxi/disp2/hdmi2/hdmi_core/api/hdcp22/include/ESMHostLibDriverErrors.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,21 @@
+/*
+ * Allwinner SoCs hdmi2.0 driver.
+ *
+ * Copyright (C) 2016 Allwinner.
+ *
+ * This file is licensed under the terms of the GNU General Public
+ * License version 2.  This program is licensed "as is" without any
+ * warranty of any kind, whether express or implied.
+ */
+#ifndef _ESMDRIVERERROR_H_
+#define _ESMDRIVERERROR_H_
+
+#define ESM_HL_DRIVER_SUCCESS                0
+#define ESM_HL_DRIVER_FAILED               (-1)
+#define ESM_HL_DRIVER_NO_MEMORY            (-2)
+#define ESM_HL_DRIVER_NO_ACCESS            (-3)
+#define ESM_HL_DRIVER_INVALID_PARAM        (-4)
+#define ESM_HL_DRIVER_TOO_MANY_ESM_DEVICES (-5)
+#define ESM_HL_DRIVER_USER_DEFINED_ERROR   (-6)
+
+#endif
