diff -drupN a/drivers/soc/sunxi/arisc/Kconfig b/drivers/soc/sunxi/arisc/Kconfig
--- a/drivers/soc/sunxi/arisc/Kconfig	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/soc/sunxi/arisc/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,20 @@
+#
+# sunxi arisc driver configuration.
+#
+config SUNXI_ARISC
+	tristate "sunxi arisc driver"
+	depends on ARCH_SUNXI && HWSPINLOCK_SUNXI
+	help
+	  Arisc driver support sunxi family SOCs.
+	  Previously known as "ar100" or "cpus".
+	  Set y here to support the sunxi arisc driver.
+	  If unsure, set N.
+
+config MCPM_WITH_ARISC_DVFS_SUPPORT
+	tristate "mcpm with arisc dvfs support"
+	depends on MCPM && CPU_FREQ
+	depends on EVB_PLATFORM && SUNXI_ARISC
+	help
+	  Arisc driver support dvfs when mcpm init.
+	  Set y here to support mcpm with arisc dvfs.
+	  If unsure, set N.
