diff -drupN a/modules/nand/common1/phy-nand/nand-partition/build_nand_partition.h b/modules/nand/common1/phy-nand/nand-partition/build_nand_partition.h
--- a/modules/nand/common1/phy-nand/nand-partition/build_nand_partition.h	1970-01-01 03:00:00.000000000 +0300
+++ b/modules/nand/common1/phy-nand/nand-partition/build_nand_partition.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,15 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+#ifndef __BUILD_NAND_PARTITION_H__
+#define __BUILD_NAND_PARTITION_H__
+
+#include "phy.h"
+
+void change_block_addr(struct _nand_partition *nand, struct _nand_super_block *super_block, unsigned short nBlkNum);
+int nand_erase_superblk(struct _nand_partition *nand, struct _physic_par *p);
+int nand_read_page(struct _nand_partition *nand, struct _physic_par *p);
+int nand_write_page(struct _nand_partition *nand, struct _physic_par *p);
+int nand_is_blk_good(struct _nand_partition *nand, struct _physic_par *p);
+int nand_mark_bad_blk(struct _nand_partition *nand, struct _physic_par *p);
+struct _nand_partition *build_nand_partition(struct _nand_phy_partition *phy_partition);
+int free_nand_partition(struct _nand_partition *nand_partition);
+#endif /*BUILD_NAND_PARTITION_H*/
