diff -drupN a/drivers/md/dm-verity-fec.h b/drivers/md/dm-verity-fec.h
--- a/drivers/md/dm-verity-fec.h	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/md/dm-verity-fec.h	2022-06-12 05:28:14.000000000 +0300
@@ -12,6 +12,8 @@
 #ifndef DM_VERITY_FEC_H
 #define DM_VERITY_FEC_H
 
+#include "dm.h"
+#include "dm-core.h"
 #include "dm-verity.h"
 #include <linux/rslib.h>
 
@@ -51,6 +53,8 @@ struct dm_verity_fec {
 	mempool_t *extra_pool;	/* mempool for extra buffers */
 	mempool_t *output_pool;	/* mempool for output */
 	struct kmem_cache *cache;	/* cache for buffers */
+	atomic_t corrected;		/* corrected errors */
+	struct dm_kobject_holder kobj_holder;	/* for sysfs attributes */
 };
 
 /* per-bio data */
