diff -drupN a/drivers/pinctrl/pinctrl-max77620.c b/drivers/pinctrl/pinctrl-max77620.c
--- a/drivers/pinctrl/pinctrl-max77620.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/pinctrl/pinctrl-max77620.c	2022-06-12 05:28:14.000000000 +0300
@@ -402,7 +402,7 @@ static int max77620_pinconf_set(struct p
 	struct device *dev = mpci->dev;
 	struct max77620_fps_config *fps_config;
 	int param;
-	u16 param_val;
+	u32 param_val;
 	unsigned int val;
 	unsigned int pu_val;
 	unsigned int pd_val;
