diff -drupN a/drivers/video/fbdev/sunxi/disp2/hdmi2/hdmi_core/api/hdcp22/include/esm_hdcp_hdmi_id.h b/drivers/video/fbdev/sunxi/disp2/hdmi2/hdmi_core/api/hdcp22/include/esm_hdcp_hdmi_id.h
--- a/drivers/video/fbdev/sunxi/disp2/hdmi2/hdmi_core/api/hdcp22/include/esm_hdcp_hdmi_id.h	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/video/fbdev/sunxi/disp2/hdmi2/hdmi_core/api/hdcp22/include/esm_hdcp_hdmi_id.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,20 @@
+/*
+ * Allwinner SoCs hdmi2.0 driver.
+ *
+ * Copyright (C) 2016 Allwinner.
+ *
+ * This file is licensed under the terms of the GNU General Public
+ * License version 2.  This program is licensed "as is" without any
+ * warranty of any kind, whether express or implied.
+ */
+#ifndef _ESM_HDCP_HDMI_ID_H_
+#define _ESM_HDCP_HDMI_ID_H_
+
+/* ID's used to version information */
+#define ESM_HDCP_HDMI_UNKNOWN_ID 0x00
+#define ESM_HDCP_HDMI_RX_ID      0x01
+#define ESM_HDCP_HDMI_TX_ID      0x02
+#define ESM_HDCP_HDMI_RPTX_ID    0x04
+#define ESM_HDCP_HDMI_RPRX_ID    0x08
+
+#endif /* _ESM_HDCP_HDMI_ID_H_ */
