diff -drupN a/include/linux/mmc/pm.h b/include/linux/mmc/pm.h
--- a/include/linux/mmc/pm.h	2018-08-06 17:23:04.000000000 +0300
+++ b/include/linux/mmc/pm.h	2022-06-12 05:28:14.000000000 +0300
@@ -26,5 +26,6 @@ typedef unsigned int mmc_pm_flag_t;
 
 #define MMC_PM_KEEP_POWER	(1 << 0)	/* preserve card power during suspend */
 #define MMC_PM_WAKE_SDIO_IRQ	(1 << 1)	/* wake up host system on SDIO IRQ assertion */
+#define MMC_PM_IGNORE_PM_NOTIFY	(1 << 2)	/* ignore mmc pm notify */
 
 #endif /* LINUX_MMC_PM_H */
