diff -drupN a/arch/arm/boot/dts/sun8iw17p1-piano-chimera.dts b/arch/arm/boot/dts/sun8iw17p1-piano-chimera.dts
--- a/arch/arm/boot/dts/sun8iw17p1-piano-chimera.dts	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/arm/boot/dts/sun8iw17p1-piano-chimera.dts	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,352 @@
+/*
+ * Allwinner Technology CO., Ltd. sun8iw17p1 soc board.
+ *
+ * soc board support.
+ */
+
+/dts-v1/;
+
+#include "sun8iw17p1.dtsi"
+
+/{
+	soc@03000000 {
+			ac108@35{
+				compatible = "Allwinnertech,MicArray_1";
+				reg = <0x35>;
+				gpio-power = <&pio PE 2 1 1 1 1>;
+				gpio-reset = <&pio PE 6 1 1 1 1>;
+				twi_bus = <1>;
+				debug_mode = <0>;
+				pga_gain = <0x32>;
+				slot_width = <0x20>;
+			};
+
+			is31fl3736: led-controller@50{
+				compatible = "issi,is31fl3736";
+				reg = <0x50>;
+				gpio-power = <&pio PE 3 1 1 1 1>;
+				gpio-shdn = <&pio PE 7 1 1 1 1>;
+				gpio-int = <&pio PH 0 0 1 1 1>;
+				twi_bus = <1>;
+				/*------------------sw1---------------*/
+				led@1 {
+					reg = <1>;
+					label = "sw1:red:led1";
+					sw_location = <1>;
+					cs_location = <1>;
+				};
+				led@2 {
+					reg = <2>;
+					label = "sw1:red:led2";
+					sw_location = <1>;
+					cs_location = <2>;
+				};
+				led@3 {
+					reg = <3>;
+					label = "sw1:red:led3";
+					sw_location = <1>;
+					cs_location = <3>;
+				};
+				led@4 {
+					reg = <4>;
+					label = "sw1:red:led4";
+					sw_location = <1>;
+					cs_location = <4>;
+				};
+				led@5 {
+					reg = <5>;
+					label = "sw1:red:led5";
+					sw_location = <1>;
+					cs_location = <5>;
+				};
+				led@6 {
+					reg = <6>;
+					label = "sw1:red:led6";
+					sw_location = <1>;
+					cs_location = <6>;
+				};
+				led@7 {
+					reg = <7>;
+					label = "sw1:red:led7";
+					sw_location = <1>;
+					cs_location = <7>;
+				};
+				led@8 {
+					reg = <8>;
+					label = "sw1:red:led8";
+					sw_location = <1>;
+					cs_location = <8>;
+				};
+				/*------------------sw2---------------*/
+				led@9 {
+					reg = <9>;
+					label = "sw2:green:led1";
+					sw_location = <2>;
+					cs_location = <1>;
+				};
+				led@10 {
+					reg = <10>;
+					label = "sw2:green:led2";
+					sw_location = <2>;
+					cs_location = <2>;
+				};
+				led@11 {
+					reg = <11>;
+					label = "sw2:green:led3";
+					sw_location = <2>;
+					cs_location = <3>;
+				};
+				led@12 {
+					reg = <12>;
+					label = "sw2:green:led4";
+					sw_location = <2>;
+					cs_location = <4>;
+				};
+				led@13 {
+					reg = <13>;
+					label = "sw2:green:led5";
+					sw_location = <2>;
+					cs_location = <5>;
+				};
+				led@14 {
+					reg = <14>;
+					label = "sw2:green:led6";
+					sw_location = <2>;
+					cs_location = <6>;
+				};
+				led@15 {
+					reg = <15>;
+					label = "sw2:green:led7";
+					sw_location = <2>;
+					cs_location = <7>;
+				};
+				led@16 {
+					reg = <16>;
+					label = "sw2:green:led8";
+					sw_location = <2>;
+					cs_location = <8>;
+				};
+				/*------------------sw3---------------*/
+				led@17 {
+					reg = <17>;
+					label = "sw3:blue:led1";
+					sw_location = <3>;
+					cs_location = <1>;
+				};
+				led@18 {
+					reg = <18>;
+					label = "sw3:blue:led2";
+					sw_location = <3>;
+					cs_location = <2>;
+				};
+				led@19 {
+					reg = <19>;
+					label = "sw3:blue:led3";
+					sw_location = <3>;
+					cs_location = <3>;
+				};
+				led@20 {
+					reg = <20>;
+					label = "sw3:blue:led4";
+					sw_location = <3>;
+					cs_location = <4>;
+				};
+				led@21 {
+					reg = <21>;
+					label = "sw3:blue:led5";
+					sw_location = <3>;
+					cs_location = <5>;
+				};
+				led@22 {
+					reg = <22>;
+					label = "sw3:blue:led6";
+					sw_location = <3>;
+					cs_location = <6>;
+				};
+				led@23 {
+					reg = <23>;
+					label = "sw3:blue:led7";
+					sw_location = <3>;
+					cs_location = <7>;
+				};
+				led@24 {
+					reg = <24>;
+					label = "sw3:blue:led8";
+					sw_location = <3>;
+					cs_location = <8>;
+				};
+				/*------------------sw4---------------*/
+				led@25 {
+					reg = <25>;
+					label = "sw4:red:led9";
+					sw_location = <4>;
+					cs_location = <1>;
+				};
+				led@26 {
+					reg = <26>;
+					label = "sw4:red:led10";
+					sw_location = <4>;
+					cs_location = <2>;
+				};
+				led@27 {
+					reg = <27>;
+					label = "sw4:red:led11";
+					sw_location = <4>;
+					cs_location = <3>;
+				};
+				led@28 {
+					reg = <28>;
+					label = "sw4:red:led12";
+					sw_location = <4>;
+					cs_location = <4>;
+				};
+				led@29 {
+					reg = <29>;
+					label = "sw4:red:led13";
+					sw_location = <4>;
+					cs_location = <5>;
+				};
+				led@30 {
+					reg = <30>;
+					label = "sw4:red:led14";
+					sw_location = <4>;
+					cs_location = <6>;
+				};
+				led@31 {
+					reg = <31>;
+					label = "sw4:red:led15";
+					sw_location = <4>;
+					cs_location = <7>;
+				};
+				led@32 {
+					reg = <32>;
+					label = "sw4:red:led16";
+					sw_location = <4>;
+					cs_location = <8>;
+				};
+				/*-----------------sw5------------------*/
+				led@33 {
+					reg = <33>;
+					label = "sw5:green:led9";
+					sw_location = <5>;
+					cs_location = <1>;
+				};
+				led@34 {
+					reg = <34>;
+					label = "sw5:green:led10";
+					sw_location = <5>;
+					cs_location = <2>;
+				};
+				led@35 {
+					reg = <35>;
+					label = "sw5:green:led11";
+					sw_location = <5>;
+					cs_location = <3>;
+				};
+				led@36 {
+					reg = <36>;
+					label = "sw5:green:led12";
+					sw_location = <5>;
+					cs_location = <4>;
+				};
+				led@37 {
+					reg = <37>;
+					label = "sw5:green:led13";
+					sw_location = <5>;
+					cs_location = <5>;
+				};
+				led@38 {
+					reg = <38>;
+					label = "sw5:green:led14";
+					sw_location = <5>;
+					cs_location = <6>;
+				};
+				led@39 {
+					reg = <39>;
+					label = "sw5:green:led15";
+					sw_location = <5>;
+					cs_location = <7>;
+				};
+				led@40 {
+					reg = <40>;
+					label = "sw5:green:led16";
+					sw_location = <5>;
+					cs_location = <8>;
+				};
+				/*------------------sw6---------------*/
+				led@41 {
+					reg = <41>;
+					label = "sw6:blue:led9";
+					sw_location = <6>;
+					cs_location = <1>;
+				};
+				led@42 {
+					reg = <42>;
+					label = "sw6:blue:led10";
+					sw_location = <6>;
+					cs_location = <2>;
+				};
+				led@43 {
+					reg = <43>;
+					label = "sw6:blue:led11";
+					sw_location = <6>;
+					cs_location = <3>;
+				};
+				led@44 {
+					reg = <44>;
+					label = "sw6:blue:led12";
+					sw_location = <6>;
+					cs_location = <4>;
+				};
+				led@45 {
+					reg = <45>;
+					label = "sw6:blue:led13";
+					sw_location = <6>;
+					cs_location = <5>;
+				};
+				led@46 {
+					reg = <46>;
+					label = "sw6:blue:led14";
+					sw_location = <6>;
+					cs_location = <6>;
+				};
+				led@47 {
+					reg = <47>;
+					label = "sw6:blue:led15";
+					sw_location = <6>;
+					cs_location = <7>;
+				};
+				led@48 {
+					reg = <48>;
+					label = "sw6:blue:led16";
+					sw_location = <6>;
+					cs_location = <8>;
+				};
+			};
+		};
+
+		snddaudio0:sound@1{
+			sunxi,snddaudio-codec = "ac108.1-003b";
+			sunxi,snddaudio-codec-dai = "ac108-pcm0";
+			status = "okay";
+		};
+	};
+
+	wlan:wlan {
+		compatible = "allwinner,sunxi-wlan";
+		clocks = <&clk_losc_out>;
+		wlan_busnum = <1>;
+		status = "okay";
+	};
+
+	bt:bt {
+		compatible = "allwinner,sunxi-bt";
+		clocks = <&clk_losc_out>;
+		status = "okay";
+	};
+
+	btlpm:btlpm {
+		compatible = "allwinner,sunxi-btlpm";
+		status = "okay";
+	};
+};
