diff -drupN a/drivers/mmc/core/mmc_ops.c b/drivers/mmc/core/mmc_ops.c
--- a/drivers/mmc/core/mmc_ops.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/mmc/core/mmc_ops.c	2022-06-12 05:28:14.000000000 +0300
@@ -22,8 +22,6 @@
 #include "host.h"
 #include "mmc_ops.h"
 
-#define MMC_OPS_TIMEOUT_MS	(10 * 60 * 1000) /* 10 minute timeout */
-
 static const u8 tuning_blk_pattern_4bit[] = {
 	0xff, 0x0f, 0xff, 0x00, 0xff, 0xcc, 0xc3, 0xcc,
 	0xc3, 0x3c, 0xcc, 0xff, 0xfe, 0xff, 0xfe, 0xef,
@@ -87,6 +85,7 @@ int mmc_send_status(struct mmc_card *car
 {
 	return __mmc_send_status(card, status, false);
 }
+EXPORT_SYMBOL_GPL(mmc_send_status);
 
 static int _mmc_select_card(struct mmc_host *host, struct mmc_card *card)
 {
@@ -175,6 +174,7 @@ int mmc_go_idle(struct mmc_host *host)
 
 	return err;
 }
+EXPORT_SYMBOL_GPL(mmc_go_idle);
 
 int mmc_send_op_cond(struct mmc_host *host, u32 ocr, u32 *rocr)
 {
@@ -215,6 +215,7 @@ int mmc_send_op_cond(struct mmc_host *ho
 
 	return err;
 }
+EXPORT_SYMBOL_GPL(mmc_send_op_cond);
 
 int mmc_all_send_cid(struct mmc_host *host, u32 *cid)
 {
