diff -drupN a/drivers/media/platform/sunxi-vin/vin-mipi/combo_csi/combo_csi_reg.h b/drivers/media/platform/sunxi-vin/vin-mipi/combo_csi/combo_csi_reg.h
--- a/drivers/media/platform/sunxi-vin/vin-mipi/combo_csi/combo_csi_reg.h	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/media/platform/sunxi-vin/vin-mipi/combo_csi/combo_csi_reg.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,43 @@
+/*
+ * combo csi module
+ *
+ * Copyright (c) 2019 by Allwinnertech Co., Ltd.  http://www.allwinnertech.com
+ *
+ * Authors:  Zheng Zequn <zequnzheng@allwinnertech.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ */
+
+#include "combo_csi_reg_i.h"
+
+#ifndef __COMBO_CSI_REG__H__
+#define __COMBO_CSI_REG__H__
+
+int cmb_csi_set_base_addr(unsigned int sel, unsigned long addr);
+void cmb_phy_top_enable(unsigned int sel);
+void cmb_phy_top_disable(unsigned int sel);
+void cmb_phy_power_enable(unsigned int sel);
+void cmb_phy_power_disable(unsigned int sel);
+void cmb_phy0_en(unsigned int sel, unsigned int en);
+void cmb_phy_lane_num_enable(unsigned int sel);
+void cmb_phy0_ibias_en(unsigned int sel, unsigned int en);
+void cmb_phy0_term_dly(unsigned int sel, unsigned int dly);
+void cmb_phy0_hs_dly(unsigned int sel, unsigned int dly);
+void cmb_phy0_s2p_width(unsigned int sel, unsigned int width);
+void cmb_phy0_s2p_dly(unsigned int sel, unsigned int dly);
+void cmb_phy_mipi_lpnum_enable(unsigned int sel);
+void cmb_phy0_mipilp_dbc_en(unsigned int sel, unsigned int en);
+void cmb_port_enable(unsigned int sel);
+void cmb_port_disable(unsigned int sel);
+void cmb_port_lane_num(unsigned int sel, unsigned int num);
+void cmb_port_out_num(unsigned int sel, unsigned int num);
+void cmb_port_lane_map(unsigned int sel);
+void cmb_port_mipi_enpack_enable(unsigned int sel);
+void cmb_port_mipi_enpack_disable(unsigned int sel);
+void cmb_port_mipi_yuv_seq(unsigned int sel, unsigned int seq);
+void cmb_port_mipi_ch0_dt(unsigned int sel, unsigned int type);
+void cmb_port_mipi_ch_trig_en(unsigned int sel, unsigned int en);
+
+#endif /*__COMBO_CSI_REG__H__*/
