diff -drupN a/drivers/bus/Makefile b/drivers/bus/Makefile
--- a/drivers/bus/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/bus/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -17,6 +17,7 @@ obj-$(CONFIG_OMAP_INTERCONNECT)	+= omap_
 obj-$(CONFIG_OMAP_OCP2SCP)	+= omap-ocp2scp.o
 obj-$(CONFIG_QCOM_EBI2)		+= qcom-ebi2.o
 obj-$(CONFIG_SUNXI_RSB)		+= sunxi-rsb.o
+obj-$(CONFIG_SUNXI_MBUS)	+= sunxi_mbus.o
 obj-$(CONFIG_SIMPLE_PM_BUS)	+= simple-pm-bus.o
 obj-$(CONFIG_TEGRA_ACONNECT)	+= tegra-aconnect.o
 obj-$(CONFIG_UNIPHIER_SYSTEM_BUS)	+= uniphier-system-bus.o
