diff -drupN a/sound/soc/sunxi/sun8iw8/Kconfig b/sound/soc/sunxi/sun8iw8/Kconfig
--- a/sound/soc/sunxi/sun8iw8/Kconfig	1970-01-01 03:00:00.000000000 +0300
+++ b/sound/soc/sunxi/sun8iw8/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,51 @@
+config SND_SUNXI_SOC_AUDIOCODEC
+	tristate "Audiocodec for the SUNXI chips"
+	depends on  ARCH_SUN8IW8
+	select SND_SOC_DMAENGINE_PCM
+	default n
+	help
+		Say Y or M if you want to add support for audiocodec
+
+config SND_SUN8IW8_DAC_TRIGGER_SYNC_WITH_OTHER_CODEC
+	bool
+	depends on ARCH_SUN8IW8 && SND_SUNXI_SOC_AUDIOCODEC
+	default n
+
+config SND_SUN8IW8_ADC_TRIGGER_SYNC_WITH_OTHER_CODEC
+	bool
+	depends on ARCH_SUN8IW8 && SND_SUNXI_SOC_AUDIOCODEC
+	default n
+
+config SND_SUN8IW8_TRIGGER_SYNC_WITH_OTHER_CODEC
+	bool "Audiocodec triggered synchronously with other codecs"
+	depends on ARCH_SUN8IW8 && SND_SUNXI_SOC_AUDIOCODEC
+	default n
+	help
+	  Say Y if you want the audiocodec is triggered synchronously with
+	  other codecs.
+choice
+	prompt "Select synchronous codecs"
+	depends on SND_SUN8IW8_TRIGGER_SYNC_WITH_OTHER_CODEC
+	help
+	  Select the codecs that audiocodec is triggered synchronously with
+	config SND_SUN8IW8_TRIGGER_SYNC_WITH_ALOOP
+		bool "Loopback"
+		depends on SND_ALOOP
+		select SND_SUN8IW8_DAC_TRIGGER_SYNC_WITH_OTHER_CODEC
+		select SND_SUN8IW8_ADC_TRIGGER_SYNC_WITH_OTHER_CODEC
+	config SND_SUN8IW8_TRIGGER_SYNC_WITH_DAUDIO_CAPTURE
+		bool "Daudio capture"
+		depends on SND_SUNXI_SOC_SUN8IW8_DAUDIO || (ARCH_SUN8IW8 && SND_SUNXI_SOC_SUNXI_DAUDIO)
+		select SND_SUN8IW8_ADC_TRIGGER_SYNC_WITH_OTHER_CODEC
+endchoice
+
+config SND_SUNXI_SOC_SUN8IW8_DAUDIO
+	tristate "allwinner digital audio machine support"
+        depends on ARCH_SUN8IW8
+	select SND_SUNXI_SOC
+	select REGMAP_MMIO
+	select SND_SUNXI_SOC_DAUDIO
+	select SND_SOC_AC100
+    help
+        Select Y or M to add support for I2S/PCM/TDM Module in the
+        Allwinner SoCs.
