diff -drupN a/drivers/block/panicpart/Kconfig b/drivers/block/panicpart/Kconfig
--- a/drivers/block/panicpart/Kconfig	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/block/panicpart/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,8 @@
+config SUNXI_PANICPART
+	bool "sunxi panic partition"
+	select PSTORE
+	select PSTORE_BLK
+	select PSTORE_BLKOOPS
+	help
+	  This enable sunxi panic operations for blkoops.
+	  If you want blkoops to record panic log, say Y, otherwise, say N.
