diff -drupN a/drivers/gpio/Makefile b/drivers/gpio/Makefile
--- a/drivers/gpio/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/gpio/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -141,3 +141,4 @@ obj-$(CONFIG_GPIO_ZEVIO)	+= gpio-zevio.o
 obj-$(CONFIG_GPIO_ZYNQ)		+= gpio-zynq.o
 obj-$(CONFIG_GPIO_ZX)		+= gpio-zx.o
 obj-$(CONFIG_GPIO_LOONGSON1)	+= gpio-loongson1.o
+obj-$(CONFIG_GPIO_SUNXI)	+= gpio-sunxi.o
