diff -drupN a/fs/fs_struct.c b/fs/fs_struct.c
--- a/fs/fs_struct.c	2018-08-06 17:23:04.000000000 +0300
+++ b/fs/fs_struct.c	2022-06-12 05:28:14.000000000 +0300
@@ -44,6 +44,7 @@ void set_fs_pwd(struct fs_struct *fs, co
 	if (old_pwd.dentry)
 		path_put(&old_pwd);
 }
+EXPORT_SYMBOL(set_fs_pwd);
 
 static inline int replace_path(struct path *p, const struct path *old, const struct path *new)
 {
@@ -89,6 +90,7 @@ void free_fs_struct(struct fs_struct *fs
 	path_put(&fs->pwd);
 	kmem_cache_free(fs_cachep, fs);
 }
+EXPORT_SYMBOL(free_fs_struct);
 
 void exit_fs(struct task_struct *tsk)
 {
@@ -127,6 +129,7 @@ struct fs_struct *copy_fs_struct(struct
 	}
 	return fs;
 }
+EXPORT_SYMBOL_GPL(copy_fs_struct);
 
 int unshare_fs_struct(void)
 {
