diff -drupN a/arch/x86/boot/string.c b/arch/x86/boot/string.c
--- a/arch/x86/boot/string.c	2018-08-06 17:23:04.000000000 +0300
+++ b/arch/x86/boot/string.c	2022-06-12 05:28:14.000000000 +0300
@@ -16,6 +16,15 @@
 #include "ctype.h"
 #include "string.h"
 
+/*
+ * Undef these macros so that the functions that we provide
+ * here will have the correct names regardless of how string.h
+ * may have chosen to #define them.
+ */
+#undef memcpy
+#undef memset
+#undef memcmp
+
 int memcmp(const void *s1, const void *s2, size_t len)
 {
 	bool diff;
