diff -drupN a/drivers/rpmsg/Kconfig b/drivers/rpmsg/Kconfig
--- a/drivers/rpmsg/Kconfig	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/rpmsg/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -4,6 +4,14 @@ menu "Rpmsg drivers"
 config RPMSG
 	tristate
 
+config SUNXI_MSGBOX
+	tristate "allwinnertech msgbox"
+	select RPMSG
+	help
+	  Say y here to enable support for the allwinnertech providing
+	  communication channels to remote processors in allwinnertech
+	  platforms.
+
 config RPMSG_QCOM_SMD
 	tristate "Qualcomm Shared Memory Driver (SMD)"
 	depends on QCOM_SMEM
