diff -drupN a/fs/exec.c b/fs/exec.c
--- a/fs/exec.c	2018-08-06 17:23:04.000000000 +0300
+++ b/fs/exec.c	2022-06-12 05:28:14.000000000 +0300
@@ -1303,7 +1303,7 @@ EXPORT_SYMBOL(flush_old_exec);
 void would_dump(struct linux_binprm *bprm, struct file *file)
 {
 	struct inode *inode = file_inode(file);
-	if (inode_permission(inode, MAY_READ) < 0) {
+	if (inode_permission2(file->f_path.mnt, inode, MAY_READ) < 0) {
 		struct user_namespace *old, *user_ns;
 		bprm->interp_flags |= BINPRM_FLAGS_ENFORCE_NONDUMP;
 
