diff -drupN a/arch/arm/common/mcpm_platsmp.c b/arch/arm/common/mcpm_platsmp.c
--- a/arch/arm/common/mcpm_platsmp.c	2018-08-06 17:23:04.000000000 +0300
+++ b/arch/arm/common/mcpm_platsmp.c	2022-06-12 05:28:14.000000000 +0300
@@ -60,6 +60,7 @@ static int mcpm_cpu_kill(unsigned int cp
 {
 	unsigned int pcpu, pcluster;
 
+	pr_debug("%s: cpu %u\n", __func__, cpu);
 	cpu_to_pcpu(cpu, &pcpu, &pcluster);
 
 	return !mcpm_wait_for_cpu_powerdown(pcpu, pcluster);
@@ -67,6 +68,7 @@ static int mcpm_cpu_kill(unsigned int cp
 
 static bool mcpm_cpu_can_disable(unsigned int cpu)
 {
+	pr_debug("%s: cpu %u\n", __func__, cpu);
 	/* We assume all CPUs may be shut down. */
 	return true;
 }
@@ -74,6 +76,7 @@ static bool mcpm_cpu_can_disable(unsigne
 static void mcpm_cpu_die(unsigned int cpu)
 {
 	unsigned int mpidr, pcpu, pcluster;
+	pr_debug("%s: cpu %u\n", __func__, cpu);
 	mpidr = read_cpuid_mpidr();
 	pcpu = MPIDR_AFFINITY_LEVEL(mpidr, 0);
 	pcluster = MPIDR_AFFINITY_LEVEL(mpidr, 1);
