diff -drupN a/drivers/char/sunxi-di/drv_div3x/di_fops.h b/drivers/char/sunxi-di/drv_div3x/di_fops.h
--- a/drivers/char/sunxi-di/drv_div3x/di_fops.h	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/char/sunxi-di/drv_div3x/di_fops.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,21 @@
+/*
+ * Copyright (c) 2007-2018 Allwinnertech Co., Ltd.
+ *
+ * This software is licensed under the terms of the GNU General Public
+ * License version 2, as published by the Free Software Foundation, and
+ * may be copied, distributed, and modified under those terms.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ */
+
+#ifndef _DI_FOPS_H_
+#define _DI_FOPS_H_
+
+#include <linux/fs.h>
+
+extern const struct file_operations di_fops;
+
+#endif /* #ifndef _DI_FOPS_H_ */
\ В конце файла нет новой строки
