diff -drupN a/drivers/dma/Makefile b/drivers/dma/Makefile
--- a/drivers/dma/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/dma/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -25,6 +25,7 @@ obj-$(CONFIG_DMA_OMAP) += omap-dma.o
 obj-$(CONFIG_DMA_SA11X0) += sa11x0-dma.o
 obj-$(CONFIG_DMA_SUN4I) += sun4i-dma.o
 obj-$(CONFIG_DMA_SUN6I) += sun6i-dma.o
+obj-$(CONFIG_DMA_SUNXI) += sunxi-dma.o
 obj-$(CONFIG_DW_DMAC_CORE) += dw/
 obj-$(CONFIG_EP93XX_DMA) += ep93xx_dma.o
 obj-$(CONFIG_FSL_DMA) += fsldma.o
