diff -drupN a/include/linux/scatterlist.h b/include/linux/scatterlist.h
--- a/include/linux/scatterlist.h	2018-08-06 17:23:04.000000000 +0300
+++ b/include/linux/scatterlist.h	2022-06-12 05:28:14.000000000 +0300
@@ -136,10 +136,10 @@ static inline struct page *sg_page(struc
 static inline void sg_set_buf(struct scatterlist *sg, const void *buf,
 			      unsigned int buflen)
 {
-#ifdef CONFIG_DEBUG_SG
-	BUG_ON(!virt_addr_valid(buf));
-#endif
-	sg_set_page(sg, virt_to_page(buf), buflen, offset_in_page(buf));
+	if (virt_addr_valid(buf))
+		sg_set_page(sg, virt_to_page(buf), buflen, offset_in_page(buf));
+	else
+		sg_set_page(sg, vmalloc_to_page(buf), buflen, offset_in_page(buf));
 }
 
 /*
