diff -drupN a/drivers/video/fbdev/sunxi/disp2/disp/de/disp_vdpo.h b/drivers/video/fbdev/sunxi/disp2/disp/de/disp_vdpo.h
--- a/drivers/video/fbdev/sunxi/disp2/disp/de/disp_vdpo.h	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/video/fbdev/sunxi/disp2/disp/de/disp_vdpo.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,23 @@
+/*
+ * Allwinner SoCs display driver.
+ *
+ * Copyright (C) 2016 Allwinner.
+ *
+ * This file is licensed under the terms of the GNU General Public
+ * License version 2.  This program is licensed "as is" without any
+ * warranty of any kind, whether express or implied.
+ */
+#ifndef _DISP_VDPO_H
+#define _DISP_VDPO_H
+
+#include "disp_private.h"
+
+#if defined(SUPPORT_VDPO)
+s32 disp_init_vdpo(struct disp_bsp_init_para *para);
+
+
+s32 disp_vdpo_set_config(struct disp_device *p_vdpo,
+			 struct disp_vdpo_config *p_cfg);
+#endif /*endif SUPPORT_VDPO */
+
+#endif /*End of file*/
