diff -drupN a/tools/aw/memtool/readme b/tools/aw/memtool/readme
--- a/tools/aw/memtool/readme	1970-01-01 03:00:00.000000000 +0300
+++ b/tools/aw/memtool/readme	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,7 @@
+Usage : memtool [option]\n"
+	option: -k     : show kernel space mem used;
+			-u     : show user space mem used;
+			-a     : show mem used of kernel sapce and user space;
+			-p pid : show process mem and get detail mem msg for this process;
+			-p all : show all process mem and get detail mem msg, equal to '-u';
+			-v     : print vesion info;
