diff -drupN a/drivers/pwm/Kconfig b/drivers/pwm/Kconfig
--- a/drivers/pwm/Kconfig	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/pwm/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -471,4 +471,36 @@ config PWM_VT8500
 	  To compile this driver as a module, choose M here: the module
 	  will be called pwm-vt8500.
 
+choice
+	prompt "SUNXI PWM SELECT."
+
+config PWM_SUNXI
+	tristate "Sunxi PWM support"
+	depends on PWM && ARCH_SUNXI
+	help
+	  Generic PWM framework driver for sunxi.
+
+	  To compile this driver as a module, choose M here: the module
+	  will be called pwm-sunxi.
+
+config PWM_SUNXI_NEW
+	tristate "Sunxi Enhance PWM support"
+	depends on PWM && ARCH_SUNXI && (ARCH_SUN8IW10 || ARCH_SUN8IW11 || \
+			ARCH_SUN8IW12 || ARCH_SUN50IW8 || ARCH_SUN8IW17 ) || \
+			ARCH_SUN8IW18 || ARCH_SUN8IW16 || ARCH_SUN50IW9
+	help
+	  Enhance PWM framework driver for sunxi.
+
+	  To compile this driver as a module, choose M here: the module
+	  will be called pwm-sunxi.
+
+config PWM_SUNXI_GROUP
+	tristate "Sunxi PWM group support"
+	depends on PWM && ARCH_SUNXI && (ARCH_SUN8IW19 || ARCH_SUN50IW10)
+	help
+	  Enhance PWM framework driver for sunxi.
+
+	  To compile this driver as a module, choose M here: the module
+	  will be called pwm-sunxi.
+endchoice
 endif
