diff -drupN a/drivers/i2c/busses/Kconfig b/drivers/i2c/busses/Kconfig
--- a/drivers/i2c/busses/Kconfig	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/i2c/busses/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -714,6 +714,24 @@ config I2C_OMAP
 	  Like OMAP1510/1610/1710/5912 and OMAP242x.
 	  For details see http://www.ti.com/omap.
 
+config I2C_SUNXI
+	tristate "SUNXI I2C controller"
+	depends on ARCH_SUNXI
+	default y if MACH_SUNXI
+	help
+	  If you say yes to this option, support will be included for the
+	  I2C interface from Allwinner Technology SUNXI platform.
+
+	  This driver can also be built as a module. If so, the module
+	  will be called i2c-sunxi.
+
+config I2C_SUNXI_TEST
+	tristate "SUNXI I2C controller test by EEPROM"
+	depends on I2C_SUNXI
+	default n
+	help
+	  It can Only be used in FPGA board.
+
 config I2C_PASEMI
 	tristate "PA Semi SMBus interface"
 	depends on PPC_PASEMI && PCI
