diff -drupN a/drivers/char/sunxi-sysinfo/sunxi-sysinfo-user.h b/drivers/char/sunxi-sysinfo/sunxi-sysinfo-user.h
--- a/drivers/char/sunxi-sysinfo/sunxi-sysinfo-user.h	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/char/sunxi-sysinfo/sunxi-sysinfo-user.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,24 @@
+/*
+ * The interface define of sysinfo dev for userspace.
+ *
+ * Copyright (C) 2016 Allwinner.
+ *
+ * Matteo <duanmintao@allwinnertech.com>
+ *
+ * This file is licensed under the terms of the GNU General Public
+ * License version 2.  This program is licensed "as is" without any
+ * warranty of any kind, whether express or implied.
+ */
+
+#ifndef __SUNXI_SYSINFO_USER_H__
+#define __SUNXI_SYSINFO_USER_H__
+
+#define CHECK_SOC_SECURE_ATTR 0x00
+#define CHECK_SOC_VERSION     0x01
+#define CHECK_SOC_BONDING     0x03
+#define CHECK_SOC_CHIPID      0x04
+#define CHECK_SOC_FT_ZONE     0x05
+#define CHECK_SOC_ROTPK_STATUS     0x06
+
+#endif /* end of __SUNXI_SYSINFO_USER_H__ */
+
