diff -drupN a/drivers/scsi/mpt3sas/mpt3sas_base.c b/drivers/scsi/mpt3sas/mpt3sas_base.c
--- a/drivers/scsi/mpt3sas/mpt3sas_base.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/scsi/mpt3sas/mpt3sas_base.c	2022-06-12 05:28:14.000000000 +0300
@@ -105,7 +105,7 @@ _base_get_ioc_facts(struct MPT3SAS_ADAPT
  *
  */
 static int
-_scsih_set_fwfault_debug(const char *val, struct kernel_param *kp)
+_scsih_set_fwfault_debug(const char *val, const struct kernel_param *kp)
 {
 	int ret = param_set_int(val, kp);
 	struct MPT3SAS_ADAPTER *ioc;
