diff -drupN a/drivers/edac/edac_module.c b/drivers/edac/edac_module.c
--- a/drivers/edac/edac_module.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/edac/edac_module.c	2022-06-12 05:28:14.000000000 +0300
@@ -19,7 +19,8 @@
 
 #ifdef CONFIG_EDAC_DEBUG
 
-static int edac_set_debug_level(const char *buf, struct kernel_param *kp)
+static int edac_set_debug_level(const char *buf,
+				const struct kernel_param *kp)
 {
 	unsigned long val;
 	int ret;
