diff -drupN a/arch/arm/boot/dts/sun8iw12p1-pinctrl.dtsi b/arch/arm/boot/dts/sun8iw12p1-pinctrl.dtsi
--- a/arch/arm/boot/dts/sun8iw12p1-pinctrl.dtsi	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/arm/boot/dts/sun8iw12p1-pinctrl.dtsi	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,965 @@
+/*
+ * Allwinner sun8iw12 pin config info.
+ */
+
+/ {
+	soc@03000000{
+		r_pio: pinctrl@07022000 {
+			compatible = "allwinner,sun8iw12p1-r-pinctrl";
+			reg = <0x0 0x07022000 0x0 0x400>;
+			interrupts = <GIC_SPI 110 4>;
+			clocks = <&clk_cpurpio>;
+			device_type = "r_pio";
+			gpio-controller;
+			interrupt-controller;
+			#interrupt-cells = <3>;
+			#size-cells = <0>;
+			#gpio-cells = <6>;
+
+			s_uart0_pins_a: s_uart0@0 {
+				allwinner,pins = "PL2", "PL3";
+				allwinner,function = "s_uart0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+			s_rsb0_pins_a: s_rsb0@0 {
+				allwinner,pins = "PL0", "PL1";
+				allwinner,function = "s_rsb0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <2>;
+				allwinner,pull = <1>;
+			};
+
+			s_twi0_pins_a: s_twi0@0 {
+				allwinner,pins = "PL0", "PL1";
+				allwinner,function = "s_twi0";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <2>;
+				allwinner,pull = <1>;
+			};
+
+			s_jtag0_pins_a: s_jtag0@0 {
+				allwinner,pins = "PL4", "PL5", "PL6", "PL7";
+				allwinner,function = "s_jtag0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <2>;
+				allwinner,pull = <1>;
+			};
+
+			s_cir0_pins_a: s_cir0@0 {
+				allwinner,pins = "PL13";
+				allwinner,function = "s_cir0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <2>;
+				allwinner,pull = <1>;
+			};
+		};
+
+		pio: pinctrl@0300b000 {
+			compatible = "allwinner,sun8iw12p1-pinctrl";
+			reg = <0x0 0x0300b000 0x0 0x400>;
+			interrupts = <GIC_SPI 69 4>,
+				     <GIC_SPI 70 4>,
+				     <GIC_SPI 71 4>,
+				     <GIC_SPI 72 4>,
+				     <GIC_SPI 73 4>;
+			device_type = "pio";
+			clocks = <&clk_pio>;
+			gpio-controller;
+			interrupt-controller;
+			#interrupt-cells = <3>;
+			#size-cells = <0>;
+			#gpio-cells = <6>;
+
+			vdevice_pins_a: vdevice@0 {
+				allwinner,pins = "PB1", "PB2";
+				allwinner,function = "vdevice";
+				allwinner,muxsel = <5>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart0_pins_a: uart0@0 {
+				allwinner,pins = "PB9", "PB10";
+				allwinner,pname = "uart0_tx", "uart0_rx";
+				allwinner,function = "uart0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart0_pins_b: uart0@1 {
+				allwinner,pins = "PB9", "PB10";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			uart1_pins_a: uart1@0 {
+				allwinner,pins = "PG6", "PG7", "PG8", "PG9";
+				allwinner,pname = "uart1_tx", "uart1_rx",
+						  "uart1_rts", "uart1_cts";
+				allwinner,function = "uart1";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart1_pins_b: uart1@1 {
+				allwinner,pins = "PG6", "PG7", "PG8", "PG9";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			uart2_pins_a: uart2@0 {
+				allwinner,pins = "PB0", "PB1", "PB2", "PB3";
+				allwinner,pname = "uart2_tx", "uart2_rx",
+						  "uart2_rts", "uart2_cts";
+				allwinner,function = "uart2";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart2_pins_b: uart2@1 {
+				allwinner,pins = "PB0", "PB1", "PB2", "PB3";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			uart3_pins_a: uart3@0 {
+				allwinner,pins = "PG10", "PG11", "PG12", "PG13";
+				allwinner,pname = "uart3_tx", "uart3_rx",
+						  "uart3_rts", "uart3_cts";
+				allwinner,function = "uart3";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart3_pins_b: uart3@1 {
+				allwinner,pins = "PG10", "PG11", "PG12", "PG13";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			uart4_pins_a: uart4@0 {
+				allwinner,pins = "PH0", "PH1", "PH15", "PH14";
+				allwinner,pname = "uart4_tx", "uart4_rx",
+						  "uart4_rts", "uart4_cts";
+				allwinner,function = "uart4";
+				allwinner,muxsel = <5>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart4_pins_b: uart4@1 {
+				allwinner,pins = "PH0", "PH1", "PH15", "PH14";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi0_pins_a: twi0@0 {
+				allwinner,pins = "PH0", "PH1";
+				allwinner,pname = "twi0_scl", "twi0_sda";
+				allwinner,function = "twi0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi0_pins_b: twi0@1 {
+				allwinner,pins = "PH0", "PH1";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi1_pins_a: twi1@0 {
+				allwinner,pins = "PH2", "PH3";
+				allwinner,pname = "twi1_scl", "twi1_sda";
+				allwinner,function = "twi1";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi1_pins_b: twi1@1 {
+				allwinner,pins = "PH2", "PH3";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi2_pins_a: twi2@0 {
+				allwinner,pins = "PE16", "PE17";
+				allwinner,pname = "twi2_scl", "twi2_sda";
+				allwinner,function = "twi2";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi2_pins_b: twi2@1 {
+				allwinner,pins = "PE16", "PE17";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi3_pins_a: twi3@0 {
+				allwinner,pins = "PI8", "PI9";
+				allwinner,pname = "twi3_scl", "twi3_sda";
+				allwinner,function = "twi3";
+				allwinner,muxsel = <5>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi3_pins_b: twi3@1 {
+				allwinner,pins = "PI8", "PI9";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			ts0_pins_a: ts0@0 {
+				allwinner,pins = "PE0", "PE1", "PE2", "PE3",
+						 "PE4", "PE5", "PE6", "PE7",
+						 "PE8", "PE9", "PE10", "PE11";
+				allwinner,pname = "ts0_clk", "ts0_err",
+						  "ts0_sync", "ts0_dvld",
+						  "ts0_d0", "ts0_d1",
+						  "ts0_d2", "ts0_d3",
+						  "ts0_d4", "ts0_d5",
+						  "ts0_d6", "ts0_d7";
+				allwinner,function = "ts0";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			ts0_pins_b: ts0_sleep@0 {
+				allwinner,pins = "PE0", "PE1", "PE2", "PE3",
+						 "PE4", "PE5", "PE6", "PE7",
+						 "PE8", "PE9", "PE10", "PE11";
+				allwinner,pname = "ts0_clk", "ts0_err",
+						  "ts0_sync", "ts0_dvld",
+						  "ts0_d0", "ts0_d1",
+						  "ts0_d2", "ts0_d3",
+						  "ts0_d4", "ts0_d5",
+						  "ts0_d6", "ts0_d7";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			ts1_pins_a: ts1@0 {
+				allwinner,pins = "PE7", "PE8", "PE9",
+						 "PE10", "PE11";
+				allwinner,pname = "ts1_clk", "ts1_err",
+						  "ts1_sync", "ts1_dvld",
+						  "ts1_d0";
+				allwinner,function = "ts1";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			ts1_pins_b: ts1_sleep@0 {
+				allwinner,pins = "PE7", "PE8", "PE9",
+						 "PE10", "PE11";
+				allwinner,pname = "ts1_clk", "ts1_err",
+						  "ts1_sync", "ts1_dvld",
+						  "ts1_d0";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			ts2_pins_a: ts2@0 {
+				allwinner,pins = "PD0", "PD1", "PD2", "PD3",
+						 "PD4", "PD5", "PD6", "PD7",
+						 "PD8", "PD9", "PD10", "PD11";
+				allwinner,pname = "ts2_clk", "ts2_err",
+						  "ts2_sync", "ts2_dvld",
+						  "ts2_d0", "ts2_d1",
+						  "ts2_d2", "ts2_d3",
+						  "ts2_d4", "ts2_d5",
+						  "ts2_d6", "ts2_d7";
+				allwinner,function = "ts2";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			ts2_pins_b: ts2_sleep@0 {
+				allwinner,pins = "PD0", "PD1", "PD2", "PD3",
+						 "PD4", "PD5", "PD6", "PD7",
+						 "PD8", "PD9", "PD10", "PD11";
+				allwinner,pname = "ts2_clk", "ts2_err",
+						  "ts2_sync", "ts2_dvld",
+						  "ts2_d0", "ts2_d1",
+						  "ts2_d2", "ts2_d3",
+						  "ts2_d4", "ts2_d5",
+						  "ts2_d6", "ts2_d7";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			ts3_pins_a: ts3@0 {
+				allwinner,pins = "PD7", "PD8", "PD9",
+						 "PD10", "PD11";
+				allwinner,pname = "ts3_clk", "ts3_err",
+						  "ts3_sync", "ts3_dvld",
+						  "ts3_d0";
+				allwinner,function = "ts3";
+				allwinner,muxsel = <5>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			ts3_pins_b: ts3_sleep@0 {
+				allwinner,pins = "PD7", "PD8", "PD9",
+						 "PD10", "PD11";
+				allwinner,pname = "ts3_clk", "ts3_err",
+						  "ts3_sync", "ts3_dvld",
+						  "ts3_d0";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			spi0_pins_a: spi0@0 {
+				allwinner,pins = "PC0", "PC2", "PC3",
+						"PC6", "PC7";
+				allwinner,pname = "spi0_sclk", "spi0_mosi",
+						  "spi0_miso", "spi0_hold", "spi0_wp";
+				allwinner,function = "spi0";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			spi0_pins_b: spi0@1 {
+				allwinner,pins = "PC15", "PC16";
+				allwinner,pname = "spi0_cs0", "spi0_cs1";
+				allwinner,function = "spi0";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>; 	// only CS should be pulled up
+			};
+
+			spi0_pins_c: spi0@2 {
+				allwinner,pins = "PC0", "PC2", "PC3", "PC6",
+						"PC7", "PC15", "PC16";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			spi1_pins_a: spi1@0 {
+				allwinner,pins = "PH15", "PH14", "PH6";
+				allwinner,pname = "spi1_sclk", "spi1_mosi",
+						  "spi1_miso";
+				allwinner,function = "spi1";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			spi1_pins_b: spi1@1 {
+				allwinner,pins = "PH7", "PH8";
+				allwinner,pname = "spi1_cs0", "spi1_cs1";
+				allwinner,function = "spi1";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>; // only CS should be pulled up
+			};
+
+			spi1_pins_c: spi1@2 {
+				allwinner,pins = "PH15", "PH14", "PH6", "PH7", "PH8";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			spi2_pins_a: spi2@0 {
+				allwinner,pins = "PI1", "PI2", "PI3";
+				allwinner,pname = "spi2_sclk", "spi2_mosi",
+						  "spi2_miso";
+				allwinner,function = "spi2";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			spi2_pins_b: spi2@1 {
+				allwinner,pins = "PI4";
+				allwinner,pname = "spi2_cs0";
+				allwinner,function = "spi2";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>; // only CS should be pulled up
+			};
+
+			spi2_pins_c: spi2@2 {
+				allwinner,pins = "PI1", "PI2", "PI3", "PI4";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+			spi3_pins_a: spi3@0 {
+				allwinner,pins = "PI8", "PI9", "PI10";
+				allwinner,pname = "spi3_sclk", "spi3_mosi",
+						  "spi3_miso";
+				allwinner,function = "spi3";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			spi3_pins_b: spi3@1 {
+				allwinner,pins = "PI11";
+				allwinner,pname = "spi3_cs0";
+				allwinner,function = "spi3";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>; // only CS should be pulled up
+			};
+
+			spi3_pins_c: spi3@2 {
+				allwinner,pins = "PI8", "PI9", "PI10", "PI11";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			sdc0_pins_a: sdc0@0 {
+				allwinner,pins = "PF0", "PF1", "PF2",
+						 "PF3", "PF4", "PF5";
+				allwinner,function = "sdc0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			sdc0_pins_b: sdc0@1 {
+				allwinner,pins = "PF0", "PF1", "PF2",
+						 "PF3", "PF4", "PF5";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+
+			sdc0_pins_c: sdc0@2 {
+				allwinner,pins = "PF0", "PF1", "PF2",
+						 "PF3", "PF4", "PF5";
+				allwinner,function = "uart0_jtag";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			sdc1_pins_a: sdc1@0 {
+				allwinner,pins = "PG0", "PG1", "PG2",
+						 "PG3", "PG4", "PG5";
+				allwinner,function = "sdc1";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <3>;
+				allwinner,pull = <1>;
+			};
+
+			sdc1_pins_b: sdc1@1 {
+				allwinner,pins = "PG0", "PG1", "PG2",
+						 "PG3", "PG4", "PG5";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			sdc2_pins_a: sdc2@0 {
+				allwinner,pins = "PC1", "PC4", "PC5", "PC6",
+						 "PC7", "PC8", "PC9", "PC10",
+						 "PC11", "PC12", "PC13", "PC14";
+				allwinner,function = "sdc2";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <2>;
+				allwinner,pull = <1>;
+			};
+
+			sdc2_pins_b: sdc2@1 {
+				allwinner,pins = "PC1", "PC4", "PC5", "PC6",
+						 "PC7", "PC8", "PC9", "PC10",
+						 "PC11", "PC12", "PC13", "PC14";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			aif2_pins_a: aif2@0 {
+				allwinner,pins = "PB6", "PB4", "PB5", "PB7";
+				allwinner,function = "aif2";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			aif3_pins_a: aif3@0 {
+				allwinner,pins = "PG10", "PG11", "PG12", "PG13";
+				allwinner,function = "aif3";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			aif2_pins_b: aif2_sleep@0 {
+				allwinner,pins = "PB6", "PB4", "PB5", "PB7";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			aif3_pins_b: aif3_sleep@0 {
+				allwinner,pins = "PG10", "PG11", "PG12", "PG13";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			daudio0_pins_a: daudio0@0 {
+				allwinner,pins = "PB4", "PB5", "PB6",
+						 "PB7", "PB8";
+				allwinner,function = "pcm0";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			daudio0_pins_b: daudio0_sleep@0 {
+				allwinner,pins = "PB4", "PB5", "PB6",
+						 "PB7", "PB8";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			daudio2_pins_a: daudio1@0 {
+				allwinner,pins = "PG10", "PG11", "PG12", "PG13";
+				allwinner,function = "pcm2";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			daudio2_pins_b: daudio1_sleep@0 {
+				allwinner,pins = "PG10", "PG11", "PG12", "PG13";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			dmic0_pins_a: dmic@0 {
+				allwinner,pins = "PG9", "PG10", "PG11",
+						"PG12", "PG13";
+				allwinner,function = "dmic0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			dmic0_pins_b: dmic_sleep@0 {
+				allwinner,pins = "PG9", "PG10", "PG11",
+						"PG12", "PG13";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			dmic1_pins_a: dmic@1 {
+				allwinner,pins = "PH4", "PH5", "PH6",
+						"PH7", "PH8";
+				allwinner,function = "dmic1";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			dmic1_pins_b: dmic_sleep@1 {
+				allwinner,pins = "PH4", "PH5", "PH6",
+						"PH7", "PH8";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			csi_cci0_pins_a: csi_cci0@0 {
+				allwinner,pins = "PE16", "PE17";
+				allwinner,pname = "csi_cci0_sck", "csi_cci0_sda";
+				allwinner,function = "csi_cci0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+
+			csi_cci0_pins_b: csi_cci0@1 {
+				allwinner,pins = "PE16", "PE17";
+				allwinner,pname = "csi_cci0_sck", "csi_cci0_sda";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+
+			csi_cci1_pins_a: csi_cci1@0 {
+				allwinner,pins = "PJ16", "PJ17";
+				allwinner,pname = "csi_cci1_sck", "csi_cci1_sda";
+				allwinner,function = "csi_cci1";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+
+			csi_cci1_pins_b: csi_cci1@1 {
+				allwinner,pins = "PJ16", "PJ17";
+				allwinner,pname = "csi_cci1_sck", "csi_cci1_sda";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+			csi_cci2_pins_a: csi_cci2@0 {
+				allwinner,pins = "PI1", "PI2";
+				allwinner,pname = "csi_cci2_sck", "csi_cci2_sda";
+				allwinner,function = "csi_cci2";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+
+			csi_cci2_pins_b: csi_cci2@1 {
+				allwinner,pins = "PI1", "PI2";
+				allwinner,pname = "csi_cci2_sck", "csi_cci2_sda";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+
+			csi_cci3_pins_a: csi_cci3@0 {
+				allwinner,pins = "PI8", "PI9";
+				allwinner,pname = "csi_cci3_sck", "csi_cci3_sda";
+				allwinner,function = "csi_cci3";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+			csi_cci3_pins_b: csi_cci3@1 {
+				allwinner,pins = "PI8", "PI9";
+				allwinner,pname = "csi_cci3_sck", "csi_cci3_sda";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+			csi_mclk0_pins_a: csi_mclk0@0 {
+				allwinner,pins = "PE1";
+				allwinner,pname = "csi_mclk0";
+				allwinner,function = "csi_mclk0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+
+			csi_mclk0_pins_b: csi_mclk0@1 {
+				allwinner,pins = "PE1";
+				allwinner,pname = "csi_mclk0";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+
+			csi_mclk1_pins_a: csi_mclk1@0 {
+				allwinner,pins = "PJ1";
+				allwinner,pname = "csi_mclk1";
+				allwinner,function = "csi_mclk1";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+
+			csi_mclk1_pins_b: csi_mclk1@1 {
+				allwinner,pins = "PJ1";
+				allwinner,pname = "csi_mclk1";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+
+			csi_mclk2_pins_a: csi_mclk2@0 {
+				allwinner,pins = "PI0";
+				allwinner,pname = "csi_mclk2";
+				allwinner,function = "csi_mclk2";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+
+			csi_mclk2_pins_b: csi_mclk2@1 {
+				allwinner,pins = "PI0";
+				allwinner,pname = "csi_mclk2";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+
+			csi_mclk3_pins_a: csi_mclk3@0 {
+				allwinner,pins = "PI7";
+				allwinner,pname = "csi_mclk3";
+				allwinner,function = "csi_mclk3";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+			csi_mclk3_pins_b: csi_mclk3@1 {
+				allwinner,pins = "PI7";
+				allwinner,pname = "csi_mclk3";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+
+			csi2_pins_a: csi2@0 {
+				allwinner,pins = "PE0", "PE2", "PE3",
+						 "PE4", "PE5", "PE6", "PE7",
+						 "PE8", "PE9", "PE10", "PE11",
+						 "PE12", "PE13", "PE14", "PE15";
+				allwinner,pname = "csi2_pck", "csi2_hsync", "csi2_vsync",
+						  "csi2_d0", "csi2_d1", "csi2_d2", "csi2_d3",
+						  "csi2_d4", "csi2_d5", "csi2_d6", "csi2_d7",
+						  "csi2_d8", "csi2_d9", "csi2_d10", "csi2_d11";
+				allwinner,function = "csi2";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+
+			csi2_pins_b: csi2@1 {
+				allwinner,pins = "PE0", "PE2", "PE3",
+						 "PE4", "PE5", "PE6", "PE7",
+						 "PE8", "PE9", "PE10", "PE11",
+						 "PE12", "PE13", "PE14", "PE15";
+				allwinner,pname = "csi2_pck", "csi2_hsync", "csi2_vsync",
+						  "csi2_d0", "csi2_d1", "csi2_d2", "csi2_d3",
+						  "csi2_d4", "csi2_d5", "csi2_d6", "csi2_d7",
+						  "csi2_d8", "csi2_d9", "csi2_d10", "csi2_d11";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+
+			csi3_pins_a: csi3@0 {
+				allwinner,pins = "PJ0", "PJ2", "PJ3",
+						 "PJ4", "PJ5", "PJ6", "PJ7",
+						 "PJ8", "PJ9", "PJ10", "PJ11",
+						 "PJ12", "PJ13", "PJ14", "PJ15";
+				allwinner,pname = "csi3_pck", "csi3_hsync", "csi3_vsync",
+						  "csi3_d0", "csi3_d1", "csi3_d2", "csi3_d3",
+						  "csi3_d4", "csi3_d5", "csi3_d6", "csi3_d7",
+						  "csi3_d8", "csi3_d9", "csi3_d10", "csi3_d11";
+				allwinner,function = "csi3";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+
+			csi3_pins_b: csi3@1 {
+				allwinner,pins = "PJ0", "PJ2", "PJ3",
+						 "PJ4", "PJ5", "PJ6", "PJ7",
+						 "PJ8", "PJ9", "PJ10", "PJ11",
+						 "PJ12", "PJ13", "PJ14", "PJ15";
+				allwinner,pname = "csi3_pck", "csi3_hsync", "csi3_vsync",
+						  "csi3_d0", "csi3_d1", "csi3_d2", "csi3_d3",
+						  "csi3_d4", "csi3_d5", "csi3_d6", "csi3_d7",
+						  "csi3_d8", "csi3_d9", "csi3_d10", "csi3_d11";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+
+			scr0_pins_a: scr0@0 {
+				allwinner,pins = "PH21", "PH24", "PH25", "PH26", "PH27";
+				allwinner,pname = "scr0_rst", "scr0_det", "scr0_vccen", "scr0_sck", "scr0_sda";
+				allwinner,function = "sim0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <0>;
+				allwinner,pull = <1>;
+			};
+
+			scr0_pins_b: scr0@1 {
+				allwinner,pins = "PH22", "PH23";
+				allwinner,pname = "scr0_vppen", "scr0_vppp";
+				allwinner,function = "sim0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <0>;
+				allwinner,pull = <1>;
+			};
+
+			scr0_pins_c: scr0@2 {
+				allwinner,pins = "PH21", "PH22", "PH23", "PH24", "PH25", "PH26", "P27";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <0>;
+				allwinner,pull = <0>;
+			};
+
+			scr1_pins_a: scr1@0 {
+				allwinner,pins = "PD15", "PD16", "PD12",
+						 "PD13", "PD14";
+				allwinner,pname = "scr1_rst", "scr1_det",
+						  "scr1_vccen", "scr1_sck",
+						  "scr1_sda";
+				allwinner,function = "sim1";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			scr1_pins_b: scr1@1 {
+				allwinner,pins = "PE14", "PE15";
+				allwinner,pname = "scr1_vppen", "scr1_vppp";
+				allwinner,function = "sim1";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			scr1_pins_c: scr1@2 {
+				allwinner,pins = "PD15", "PD16", "PD12",
+						 "PD13", "PD14", "PE14",
+						 "PE15";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			nand0_pins_a: nand0@0 {
+				allwinner,pins = "PC0", "PC1", "PC2", "PC4",
+						 "PC6", "PC7", "PC8", "PC9",
+						 "PC10", "PC11", "PC12", "PC13",
+						 "PC14";
+				allwinner,pname= "nand0_we", "nand0_ale","nand0_cle", "nand0_nre",
+						 "nand0_d0", "nand0_d1", "nand0_d2", "nand0_d3",
+						 "nand0_d4", "nand0_d5", "nand0_d6", "nand0_d7",
+						 "nand0_ndqs";
+				allwinner,function = "nand0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			nand0_pins_b: nand0@1 {
+				allwinner,pins = "PC3", "PC5", "PC15", "PC16";
+				allwinner,pname= "nand0_ce0", "nand0_rb0", "nand0_ce1", "nand0_rb1";
+				allwinner,function = "nand0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;// only RB&CE should be pulled up
+			};
+
+			nand0_pins_c: nand0@2 {
+				allwinner,pins = "PC0", "PC1", "PC2", "PC3",
+						 "PC4", "PC5", "PC6", "PC7",
+						 "PC8", "PC9", "PC10", "PC11",
+						 "PC12", "PC13", "PC14", "PC15",
+						 "PC16";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			gmac_pins_a: gmac@0 {
+				allwinner,pins = "PA0", "PA1", "PA2", "PA3",
+						 "PA4", "PA5", "PA6", "PA7",
+						 "PA8", "PA9", "PA10", "PA11",
+						 "PA12", "PA13", "PA14", "PA15";
+				allwinner,function = "gmac0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <3>;
+				allwinner,pull = <0>;
+			};
+
+			gmac_pins_b: gmac@1 {
+				allwinner,pins = "PA0", "PA1", "PA2", "PA3",
+						 "PA4", "PA5", "PA6", "PA7",
+						 "PA8", "PA9", "PA10", "PA11",
+						 "PA12", "PA13", "PA14", "PA15";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <3>;
+				allwinner,pull = <0>;
+			};
+		};
+	};
+};
