diff -drupN a/modules/nand/sun50iw9p1/phy-nand/rawnand/Makefile b/modules/nand/sun50iw9p1/phy-nand/rawnand/Makefile
--- a/modules/nand/sun50iw9p1/phy-nand/rawnand/Makefile	1970-01-01 03:00:00.000000000 +0300
+++ b/modules/nand/sun50iw9p1/phy-nand/rawnand/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,27 @@
+#
+# Makefile for rawnand
+#
+#
+#
+
+obj-$(CONFIG_SUNXI_RAWNAND) += rawnand_ids.o
+obj-$(CONFIG_SUNXI_RAWNAND) += rawnand_base.o
+obj-$(CONFIG_SUNXI_RAWNAND) += rawnand_boot0.o
+obj-$(CONFIG_SUNXI_RAWNAND) += rawnand_ops.o
+obj-$(CONFIG_SUNXI_RAWNAND) += rawnand_readretry.o
+obj-$(CONFIG_SUNXI_RAWNAND) += rawnand_boot1.o
+#obj-$(CONFIG_SUNXI_RAWNAND) += version.o
+obj-$(CONFIG_SUNXI_RAWNAND) += rawnand_cfg.o
+obj-$(CONFIG_SUNXI_RAWNAND) += rawnand_cfg.o
+obj-$(CONFIG_SUNXI_RAWNAND) += rawnand_debug.o
+obj-$(CONFIG_SUNXI_RAWNAND) += rawnand_chip.o
+obj-$(CONFIG_SUNXI_RAWNAND) += rawnand_timings.o
+#obj-y += chip/
+obj-y += controller/
+#obj-$(CONFIG_SUNXI_NAND) += nand_id2.o
+#obj-$(CONFIG_SUNXI_NAND) += nand_phy_init.o
+#obj-$(CONFIG_SUNXI_NAND) += nand_physic_interface_rawnand.o
+#obj-$(CONFIG_SUNXI_NAND) += nand_platform_data.o
+#obj-y += adapter/
+#obj-y += hynix/
+#obj-y += universal/
