diff -drupN a/drivers/spi/Kconfig b/drivers/spi/Kconfig
--- a/drivers/spi/Kconfig	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/spi/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -648,6 +648,12 @@ config SPI_TEGRA20_SLINK
 	help
 	  SPI driver for Nvidia Tegra20/Tegra30 SLINK Controller interface.
 
+config SPI_SUNXI
+	tristate "SUNXI SPI Controller"
+	depends on ARCH_SUNXI && OF
+	help
+	  Allwinner Soc SPI controller, present on SUNXI chips.
+
 config SPI_THUNDERX
 	tristate "Cavium ThunderX SPI controller"
 	depends on PCI && 64BIT && (ARM64 || COMPILE_TEST)
