diff -drupN a/arch/arm64/boot/dts/sunxi/sun50iw1p1-clk.dtsi b/arch/arm64/boot/dts/sunxi/sun50iw1p1-clk.dtsi
--- a/arch/arm64/boot/dts/sunxi/sun50iw1p1-clk.dtsi	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/arm64/boot/dts/sunxi/sun50iw1p1-clk.dtsi	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,612 @@
+/{
+clocks {
+	compatible = "allwinner,clk-init";
+	device_type = "clocks";
+	#address-cells = <2>;
+	#size-cells = <2>;
+	ranges;
+	reg = <0x0 0x01c20000 0x0 0x0320> ,  /*cpux space*/
+		  <0x0 0x01f01400 0x0 0x01C0> , /*cpus space*/
+		  <0x0 0x01f00060 0x0 0x4>;
+/* register fixed rate clock*/
+		clk_losc: losc {
+			#clock-cells = <0>;
+			compatible = "allwinner,fixed-clock";
+			clock-frequency = <32768>;
+			clock-output-names = "losc";
+		};
+
+		clk_iosc: iosc {
+			#clock-cells = <0>;
+			compatible = "allwinner,fixed-clock";
+			clock-frequency = <16000000>;
+			clock-output-names = "iosc";
+		};
+
+		clk_hosc: hosc {
+			#clock-cells = <0>;
+			compatible = "allwinner,fixed-clock";
+			clock-frequency = <24000000>;
+			clock-output-names = "hosc";
+		};
+/* register allwinner,pll-clock */
+		clk_pll_cpu: pll_cpu {
+			#clock-cells = <0>;
+			compatible = "allwinner,pll-clock";
+			lock-mode = "new";
+			clock-output-names = "pll_cpu";
+		};
+		clk_pll_audio: pll_audio {
+			#clock-cells = <0>;
+			compatible = "allwinner,pll-clock";
+			lock-mode = "new";
+			assigned-clock-rates = <24576000>;
+			clock-output-names = "pll_audio";
+		};
+		clk_pll_video0: pll_video0 {
+			#clock-cells = <0>;
+			compatible = "allwinner,pll-clock";
+			lock-mode = "new";
+			assigned-clock-rates = <297000000>;
+			clock-output-names = "pll_video0";
+		};
+		clk_pll_ve: pll_ve {
+			#clock-cells = <0>;
+			compatible = "allwinner,pll-clock";
+			lock-mode = "new";
+			clock-output-names = "pll_ve";
+		};
+		clk_pll_ddr0: pll_ddr0 {
+			#clock-cells = <0>;
+			compatible = "allwinner,pll-clock";
+			lock-mode = "new";
+			clock-output-names = "pll_ddr0";
+		};
+		clk_pll_periph0: pll_periph0 {
+			#clock-cells = <0>;
+			compatible = "allwinner,pll-clock";
+			lock-mode = "new";
+			clock-output-names = "pll_periph0";
+		};
+		clk_pll_periph1: pll_periph1 {
+			#clock-cells = <0>;
+			compatible = "allwinner,pll-clock";
+			lock-mode = "new";
+			clock-output-names = "pll_periph1";
+		};
+		clk_pll_video1: pll_video1 {
+			#clock-cells = <0>;
+			compatible = "allwinner,pll-clock";
+			lock-mode = "new";
+			assigned-clock-rates = <297000000>;
+			clock-output-names = "pll_video1";
+		};
+		clk_pll_gpu: pll_gpu {
+			#clock-cells = <0>;
+			compatible = "allwinner,pll-clock";
+			lock-mode = "new";
+			clock-output-names = "pll_gpu";
+		};
+		clk_pll_mipi: pll_mipi {
+			#clock-cells = <0>;
+			compatible = "allwinner,pll-clock";
+			lock-mode = "new";
+			clock-output-names = "pll_mipi";
+		};
+		clk_pll_hsic: pll_hsic {
+			#clock-cells = <0>;
+			compatible = "allwinner,pll-clock";
+			lock-mode = "new";
+			clock-output-names = "pll_hsic";
+		};
+		clk_pll_de: pll_de {
+			#clock-cells = <0>;
+			compatible = "allwinner,pll-clock";
+			lock-mode = "new";
+			assigned-clock-rates = <297000000>;
+			clock-output-names = "pll_de";
+		};
+		clk_pll_ddr1: pll_ddr1 {
+			#clock-cells = <0>;
+			compatible = "allwinner,pll-clock";
+			lock-mode = "new";
+			clock-output-names = "pll_ddr1";
+		};
+/* register fixed factor clock*/
+		clk_pll_audiox8: pll_audiox8 {
+			#clock-cells = <0>;
+			compatible = "allwinner,fixed-factor-clock";
+			clocks = <&clk_pll_audio>;
+			clock-mult = <8>;
+			clock-div = <1>;
+			clock-output-names = "pll_audiox8";
+		};
+		clk_pll_audiox4: pll_audiox4 {
+			#clock-cells = <0>;
+			compatible = "allwinner,fixed-factor-clock";
+			clocks = <&clk_pll_audio>;
+			clock-mult = <8>;
+			clock-div = <2>;
+			clock-output-names = "pll_audiox4";
+		};
+		clk_pll_audiox2: pll_audiox2 {
+			#clock-cells = <0>;
+			compatible = "allwinner,fixed-factor-clock";
+			clocks = <&clk_pll_audio>;
+			clock-mult = <8>;
+			clock-div = <4>;
+			clock-output-names = "pll_audiox2";
+		};
+		clk_pll_video0x2: pll_video0x2 {
+			#clock-cells = <0>;
+			compatible = "allwinner,fixed-factor-clock";
+			clocks = <&clk_pll_video0>;
+			clock-mult = <2>;
+			clock-div = <1>;
+			clock-output-names = "pll_video0x2";
+		};
+		clk_pll_periph0x2: pll_periph0x2 {
+			#clock-cells = <0>;
+			compatible = "allwinner,fixed-factor-clock";
+			clocks = <&clk_pll_periph0>;
+			clock-mult = <2>;
+			clock-div = <1>;
+			clock-output-names = "pll_periph0x2";
+		};
+		clk_pll_periph1x2: pll_periph1x2 {
+			#clock-cells = <0>;
+			compatible = "allwinner,fixed-factor-clock";
+			clocks = <&clk_pll_periph1>;
+			clock-mult = <2>;
+			clock-div = <1>;
+			clock-output-names = "pll_periph1x2";
+		};
+		clk_pll_periph0d2: pll_periph0d2 {
+			#clock-cells = <0>;
+			compatible = "allwinner,fixed-factor-clock";
+			clocks = <&clk_pll_periph0>;
+			clock-mult = <1>;
+			clock-div = <2>;
+			clock-output-names = "pll_periph0d2";
+		};
+		clk_hoscd2: hoscd2 {
+			#clock-cells = <0>;
+			compatible = "allwinner,fixed-factor-clock";
+			clocks = <&clk_hosc>;
+			clock-mult = <1>;
+			clock-div = <2>;
+			clock-output-names = "hoscd2";
+		};
+		clk_hoscx2: hoscx2 {
+			#clock-cells = <0>;
+			compatible = "allwinner,fixed-factor-clock";
+			clocks = <&clk_hosc>;
+			clock-mult = <2>;
+			clock-div = <1>;
+			clock-output-names = "hoscx2";
+		};
+/* register allwinner,periph-clock */
+		clk_cpu: cpu {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "cpu";
+		};
+		clk_cpuapb: cpuapb {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "cpuapb";
+		};
+		clk_axi: axi {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "axi";
+		};
+		clk_pll_periphahb0: pll_periphahb0 {
+			#clock-cells = <0>;
+			assigned-clock-rates = <200000000>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "pll_periphahb0";
+		};
+		clk_ahb1: ahb1 {
+			#clock-cells = <0>;
+			assigned-clock-parents = <&clk_pll_periphahb0>;
+			assigned-clock-rates = <200000000>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "ahb1";
+		};
+		clk_apb1: apb1 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "apb1";
+		};
+		clk_apb2: apb2 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "apb2";
+		};
+		clk_ahb2: ahb2 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "ahb2";
+		};
+		clk_ths: ths {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "ths";
+		};
+		clk_nand: nand {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "nand";
+		};
+		clk_sdmmc0_mod: sdmmc0_mod {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "sdmmc0_mod";
+		};
+		clk_sdmmc0_bus: sdmmc0_bus {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "sdmmc0_bus";
+		};
+		clk_sdmmc0_rst: sdmmc0_rst {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "sdmmc0_rst";
+		};
+
+		clk_sdmmc1_mod: sdmmc1_mod {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "sdmmc1_mod";
+		};
+		clk_sdmmc1_bus: sdmmc1_bus {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "sdmmc1_bus";
+		};
+		clk_sdmmc1_rst: sdmmc1_rst {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "sdmmc1_rst";
+		};
+		clk_sdmmc2_mod: sdmmc2_mod {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "sdmmc2_mod";
+		};
+		clk_sdmmc2_bus: sdmmc2_bus {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "sdmmc2_bus";
+		};
+		clk_sdmmc2_rst: sdmmc2_rst {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "sdmmc2_rst";
+		};
+		clk_ts: ts {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "ts";
+		};
+		clk_ce: ce {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "ce";
+		};
+		clk_spi0: spi0 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "spi0";
+		};
+		clk_spi1: spi1 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "spi1";
+		};
+		clk_i2s0: i2s0 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "i2s0";
+		};
+		clk_i2s1: i2s1 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "i2s1";
+		};
+		clk_i2s2: i2s2 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "i2s2";
+		};
+		clk_spdif: spdif {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "spdif";
+		};
+		clk_usbphy0: usbphy0 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "usbphy0";
+		};
+		clk_usbphy1: usbphy1 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "usbphy1";
+		};
+		clk_usbhsic: usbhsic {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "usbhsic";
+		};
+		clk_usbhsic12m: usbhsic12m {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "usbhsic12m";
+		};
+		clk_usbohci112m: usbohci112m {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "usbohci112m";
+		};
+		clk_usbohci012m: usbohci012m {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "usbohci012m";
+		};
+		clk_usbohci1: usbohci1 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "usbohci1";
+		};
+		clk_usbohci0: usbohci0 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "usbohci0";
+		};
+		clk_de: de {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			assigned-clocks = <&clk_de>;
+			assigned-clock-parents = <&clk_pll_de>;
+			assigned-clock-rates = <297000000>;
+			clock-output-names = "de";
+		};
+		clk_tcon0: tcon0 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			assigned-clocks = <&clk_tcon0>;
+			assigned-clock-parents = <&clk_pll_mipi>;
+			clock-output-names = "tcon0";
+		};
+		clk_tcon1: tcon1 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			assigned-clocks = <&clk_tcon1>;
+			assigned-clock-parents = <&clk_pll_video1>;
+			clock-output-names = "tcon1";
+		};
+		clk_deinterlace: deinterlace {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "deinterlace";
+		};
+		clk_csi_s: csi_s {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "csi_s";
+		};
+		clk_csi_m: csi_m {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "csi_m";
+		};
+		clk_csi_misc: csi_misc {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "csi_misc";
+		};
+		clk_ve: ve {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "ve";
+		};
+		clk_adda: adda {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "adda";
+		};
+		clk_addax4: addax4 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "addax4";
+		};
+		clk_avs: avs {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "avs";
+		};
+		clk_hdmi: hdmi {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			assigned-clocks = <&clk_hdmi>;
+			assigned-clock-parents = <&clk_pll_video1>;
+			clock-output-names = "hdmi";
+		};
+		clk_hdmi_slow: hdmi_slow {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			assigned-clocks = <&clk_hdmi_slow>;
+			clock-output-names = "hdmi_slow";
+		};
+		clk_mbus: mbus {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "mbus";
+		};
+		clk_mipidsi: mipidsi {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "mipidsi";
+		};
+		clk_gpu: gpu {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "gpu";
+		};
+		clk_usbohci_16: usbehci_16 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "usbohci_16";
+		};
+		clk_usbehci1: usbehci1 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "usbehci1";
+		};
+		clk_usbehci0: usbehci0 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "usbehci0";
+		};
+		clk_usbotg: usbotg {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "usbotg";
+		};
+		clk_gmac: gmac {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "gmac";
+		};
+		clk_sdram: sdram {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "sdram";
+		};
+		clk_dma: dma {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "dma";
+		};
+		clk_hwspinlock_rst: hwspinlock_rst {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "hwspinlock_rst";
+		};
+		clk_hwspinlock_bus: hwspinlock_bus {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "hwspinlock_bus";
+		};
+		clk_msgbox: msgbox {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "msgbox";
+		};
+		clk_lvds: lvds {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			assigned-clocks = <&clk_lvds>;
+			clock-output-names = "lvds";
+		};
+		clk_uart0: uart0 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "uart0";
+		};
+		clk_uart1: uart1 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "uart1";
+		};
+		clk_uart2: uart2 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "uart2";
+		};
+		clk_uart3: uart3 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "uart3";
+		};
+		clk_uart4: uart4 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "uart4";
+		};
+		clk_scr: scr {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "scr";
+		};
+		clk_twi0: twi0 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "twi0";
+		};
+		clk_twi1: twi1 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "twi1";
+		};
+		clk_twi2: twi2 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "twi2";
+		};
+		clk_twi3: twi3 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "twi3";
+		};
+		clk_pio: pio {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-clock";
+			clock-output-names = "pio";
+		};
+/*cpus space clocks from PRCM-SPEC*/
+		clk_cpurcir: cpurcir {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-cpus-clock";
+			clock-output-names = "cpurcir";
+		};
+		clk_cpurpio: cpurpio {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-cpus-clock";
+			clock-output-names = "cpurpio";
+		};
+		/*   "cpurpll_peri0 cpurcpus cpurahbs cpurapbs" are read only , just to keep a clock tree */
+		clk_cpurpll_peri0: cpurpll_peri0 {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-cpus-clock";
+			clock-output-names = "cpurpll_peri0";
+		};
+		clk_cpurcpus: cpurcpus {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-cpus-clock";
+			clock-output-names = "cpurcpus";
+		};
+		clk_cpurahbs: cpurahbs {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-cpus-clock";
+			clock-output-names = "cpurahbs";
+		};
+		clk_cpurapbs: cpurapbs {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-cpus-clock";
+			clock-output-names = "cpurapbs";
+		};
+		clk_losc_out: losc_out {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-cpus-clock";
+			clock-output-names = "losc_out";
+		};
+		clk_adda_com: adda_com {
+			#clock-cells = <0>;
+			compatible = "allwinner,periph-cpus-clock";
+			clock-output-names = "adda_com";
+		};
+	};/*clocks end*/
+};
