diff -drupN a/fs/fscache/object-list.c b/fs/fscache/object-list.c
--- a/fs/fscache/object-list.c	2018-08-06 17:23:04.000000000 +0300
+++ b/fs/fscache/object-list.c	2022-06-12 05:28:14.000000000 +0300
@@ -329,7 +329,7 @@ static void fscache_objlist_config(struc
 	config = 0;
 	rcu_read_lock();
 
-	confkey = user_key_payload(key);
+	confkey = user_key_payload_rcu(key);
 	if (!confkey) {
 		/* key was revoked */
 		rcu_read_unlock();
