diff -drupN a/drivers/video/fbdev/core/modedb.c b/drivers/video/fbdev/core/modedb.c
--- a/drivers/video/fbdev/core/modedb.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/video/fbdev/core/modedb.c	2022-06-12 05:28:14.000000000 +0300
@@ -689,9 +689,11 @@ int fb_find_mode(struct fb_var_screeninf
 	if (!default_bpp)
 		default_bpp = 8;
 
+#ifdef CONFIG_FB_CMDLINE
 	/* Did the user specify a video mode? */
 	if (!mode_option)
 		mode_option = fb_mode_option;
+#endif
 	if (mode_option) {
 		const char *name = mode_option;
 		unsigned int namelen = strlen(name);
