diff -drupN a/drivers/iio/imu/inv_mpu6050/inv_mpu_ring.c b/drivers/iio/imu/inv_mpu6050/inv_mpu_ring.c
--- a/drivers/iio/imu/inv_mpu6050/inv_mpu_ring.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/iio/imu/inv_mpu6050/inv_mpu_ring.c	2022-06-12 05:28:14.000000000 +0300
@@ -106,8 +106,10 @@ irqreturn_t inv_mpu6050_irq_handler(int
 	struct iio_dev *indio_dev = pf->indio_dev;
 	struct inv_mpu6050_state *st = iio_priv(indio_dev);
 	s64 timestamp;
+	struct timespec ts;
 
-	timestamp = iio_get_time_ns(indio_dev);
+	ktime_get_ts(&ts);
+	timestamp = timespec_to_ns(&ts);
 	kfifo_in_spinlocked(&st->timestamps, &timestamp, 1,
 			    &st->time_stamp_lock);
 
@@ -128,7 +130,11 @@ irqreturn_t inv_mpu6050_read_fifo(int ir
 	u16 fifo_count;
 	s64 timestamp;
 
-	mutex_lock(&indio_dev->mlock);
+	if (!mutex_trylock(&indio_dev->mlock)) {
+		iio_trigger_notify_done(indio_dev->trig);
+		return IRQ_NONE;
+	}
+
 	if (!(st->chip_config.accl_fifo_enable |
 		st->chip_config.gyro_fifo_enable))
 		goto end_session;
