diff -drupN a/drivers/pci/host/Makefile b/drivers/pci/host/Makefile
--- a/drivers/pci/host/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/pci/host/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -1,5 +1,6 @@
 obj-$(CONFIG_PCIE_DW) += pcie-designware.o
 obj-$(CONFIG_PCIE_DW_PLAT) += pcie-designware-plat.o
+obj-$(CONFIG_PCIE_SUNXI_PLAT) += pcie-sunxi-plat.o
 obj-$(CONFIG_PCI_DRA7XX) += pci-dra7xx.o
 obj-$(CONFIG_PCI_EXYNOS) += pci-exynos.o
 obj-$(CONFIG_PCI_IMX6) += pci-imx6.o
