diff -drupN a/include/linux/platform_data/media/sunxi-gpio-ir.h b/include/linux/platform_data/media/sunxi-gpio-ir.h
--- a/include/linux/platform_data/media/sunxi-gpio-ir.h	1970-01-01 03:00:00.000000000 +0300
+++ b/include/linux/platform_data/media/sunxi-gpio-ir.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,27 @@
+/* Copyright (c) 2012, Code Aurora Forum. All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 and
+ * only version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#ifndef __SUNXI_GPIO_IR_H__
+#define __SUNXI_GPIO_IR_H__
+
+struct gpio_ir_platform_data {
+	int gpio_rx_nr;
+	int gpio_tx_nr;
+	bool pwm_tx_used;
+	int pwm_port;
+	bool active_low;
+	bool tx_capable;
+	u64 allowed_protos;
+	const char *map_name;
+};
+
+#endif /* __SUNXI_GPIO_IR_H__ */
