diff -drupN a/drivers/acpi/button.c b/drivers/acpi/button.c
--- a/drivers/acpi/button.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/acpi/button.c	2022-06-12 05:28:14.000000000 +0300
@@ -556,7 +556,8 @@ static int acpi_button_remove(struct acp
 	return 0;
 }
 
-static int param_set_lid_init_state(const char *val, struct kernel_param *kp)
+static int param_set_lid_init_state(const char *val,
+				    const struct kernel_param *kp)
 {
 	int result = 0;
 
@@ -574,7 +575,8 @@ static int param_set_lid_init_state(cons
 	return result;
 }
 
-static int param_get_lid_init_state(char *buffer, struct kernel_param *kp)
+static int param_get_lid_init_state(char *buffer,
+				    const struct kernel_param *kp)
 {
 	switch (lid_init_state) {
 	case ACPI_BUTTON_LID_INIT_OPEN:
