diff -drupN a/lib/debugobjects.c b/lib/debugobjects.c
--- a/lib/debugobjects.c	2018-08-06 17:23:04.000000000 +0300
+++ b/lib/debugobjects.c	2022-06-12 05:28:14.000000000 +0300
@@ -293,10 +293,13 @@ static void debug_object_is_on_stack(voi
 		return;
 
 	limit++;
-	if (is_on_stack)
-		pr_warn("object is on stack, but not annotated\n");
-	else
-		pr_warn("object is not on stack, but annotated\n");
+	if (is_on_stack) {
+		pr_warn("object %p is on stack %p, but NOT annotated\n", addr,
+				task_stack_page(current));
+	} else {
+		pr_warn("object %p is NOT on stack %p, but annotated\n", addr,
+				task_stack_page(current));
+	}
 	WARN_ON(1);
 }
 
