diff -drupN a/arch/arm64/include/asm/traps.h b/arch/arm64/include/asm/traps.h
--- a/arch/arm64/include/asm/traps.h	2018-08-06 17:23:04.000000000 +0300
+++ b/arch/arm64/include/asm/traps.h	2022-06-12 05:28:14.000000000 +0300
@@ -37,18 +37,11 @@ void unregister_undef_hook(struct undef_
 
 void arm64_notify_segfault(struct pt_regs *regs, unsigned long addr);
 
-#ifdef CONFIG_FUNCTION_GRAPH_TRACER
 static inline int __in_irqentry_text(unsigned long ptr)
 {
 	return ptr >= (unsigned long)&__irqentry_text_start &&
 	       ptr < (unsigned long)&__irqentry_text_end;
 }
-#else
-static inline int __in_irqentry_text(unsigned long ptr)
-{
-	return 0;
-}
-#endif
 
 static inline int in_exception_text(unsigned long ptr)
 {
