diff -drupN a/arch/arm64/boot/dts/sunxi/sun50iw1p1-t1.dts b/arch/arm64/boot/dts/sunxi/sun50iw1p1-t1.dts
--- a/arch/arm64/boot/dts/sunxi/sun50iw1p1-t1.dts	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/arm64/boot/dts/sunxi/sun50iw1p1-t1.dts	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,59 @@
+/*
+ * Allwinner Technology CO., Ltd. sun50iw1p1 fpga board.
+ *
+ * fpga board support.
+ */
+
+/dts-v1/;
+
+#include "sun50iw1p1.dtsi"
+
+/{
+
+	soc@01c00000 {
+		twi0: twi@0x01c2ac00{
+			status = "okay";
+		};
+
+		twi1: twi@0x01c2b000{
+			status = "okay";
+		};
+		s_cir0: s_cir@1f02000 {
+			status = "disable";
+		};
+	};
+
+	gpu_mali400_0: gpu@0x01c40000 {
+		compatible = "arm,mali-400", "arm,mali-utgard";
+		reg = <0x0 0x01c40000 0x0 0x10000>;
+		interrupts = <GIC_SPI 97 4>, <GIC_SPI 98 4>, <GIC_SPI 99 4>, <GIC_SPI 100 4>, <GIC_SPI 102 4>, <GIC_SPI 103 4>;
+		interrupt-names = "IRQGP", "IRQGPMMU", "IRQPP0", "IRQPPMMU0", "IRQPP1", "IRQPPMMU1";
+		clocks = <&clk_pll_gpu>, <&clk_gpu>;
+	};
+
+	wlan:wlan {
+		compatible = "allwinner,sunxi-wlan";
+		/*wlan_power = "vcc-wifi";*/
+		wlan_io_regulator = "vcc-wifi-io";
+		wlan_busnum = <1>;
+		wlan_regon = <&r_pio PL 2 1 1 1 0>;
+		wlan_hostwake = <&r_pio PL 3 6 0 0 0>;
+		status = "okay";
+	};
+
+	bt:bt {
+		compatible = "allwinner,sunxi-bt";
+		/*bt_power = "vcc-wifi";*/
+		bt_io_regulator = "vcc-wifi-io";
+		bt_rst_n = <&r_pio PL 4 1 1 1 0>;
+		status = "okay";
+		};
+
+	btlpm:btlpm {
+		compatible = "allwinner,sunxi-btlpm";
+		uart_index = <1>;
+		bt_wake = <&r_pio PL 6 1 1 1 1>;
+		bt_hostwake = <&r_pio PL 5 6 0 0 0>;
+		status = "okay";
+	};
+};
