diff -drupN a/arch/arm64/kvm/hyp/Makefile b/arch/arm64/kvm/hyp/Makefile
--- a/arch/arm64/kvm/hyp/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/arch/arm64/kvm/hyp/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -4,6 +4,10 @@
 
 ccflags-y += -fno-stack-protector -DDISABLE_BRANCH_PROFILING
 
+ifeq ($(cc-name),clang)
+ccflags-y += -fno-jump-tables
+endif
+
 KVM=../../../../virt/kvm
 
 obj-$(CONFIG_KVM_ARM_HOST) += $(KVM)/arm/hyp/vgic-v2-sr.o
