diff -drupN a/drivers/scsi/ufs/ufshcd.h b/drivers/scsi/ufs/ufshcd.h
--- a/drivers/scsi/ufs/ufshcd.h	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/scsi/ufs/ufshcd.h	2022-06-12 05:28:14.000000000 +0300
@@ -573,6 +573,9 @@ struct ufs_hba {
 
 	enum bkops_status urgent_bkops_lvl;
 	bool is_urgent_bkops_lvl_checked;
+	int latency_hist_enabled;
+	struct io_latency_state io_lat_read;
+	struct io_latency_state io_lat_write;
 
 	struct ufs_desc_size desc_size;
 };
