diff -drupN a/drivers/rtc/rtc-omap.c b/drivers/rtc/rtc-omap.c
--- a/drivers/rtc/rtc-omap.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/rtc/rtc-omap.c	2022-06-12 05:28:14.000000000 +0300
@@ -610,7 +610,7 @@ static int rtc_pinconf_set(struct pinctr
 	struct omap_rtc *rtc = pinctrl_dev_get_drvdata(pctldev);
 	u32 val;
 	unsigned int param;
-	u16 param_val;
+	u32 param_val;
 	int i;
 
 	rtc->type->unlock(rtc);
