diff -drupN a/drivers/pinctrl/intel/pinctrl-cherryview.c b/drivers/pinctrl/intel/pinctrl-cherryview.c
--- a/drivers/pinctrl/intel/pinctrl-cherryview.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/pinctrl/intel/pinctrl-cherryview.c	2022-06-12 05:28:14.000000000 +0300
@@ -1060,7 +1060,7 @@ static int chv_config_get(struct pinctrl
 }
 
 static int chv_config_set_pull(struct chv_pinctrl *pctrl, unsigned pin,
-			       enum pin_config_param param, u16 arg)
+			       enum pin_config_param param, u32 arg)
 {
 	void __iomem *reg = chv_padreg(pctrl, pin, CHV_PADCTRL0);
 	unsigned long flags;
@@ -1152,7 +1152,7 @@ static int chv_config_set(struct pinctrl
 	struct chv_pinctrl *pctrl = pinctrl_dev_get_drvdata(pctldev);
 	enum pin_config_param param;
 	int i, ret;
-	u16 arg;
+	u32 arg;
 
 	if (chv_pad_locked(pctrl, pin))
 		return -EBUSY;
