diff -drupN a/drivers/video/fbdev/sunxi/disp2/hdmi/dev_hdmi.h b/drivers/video/fbdev/sunxi/disp2/hdmi/dev_hdmi.h
--- a/drivers/video/fbdev/sunxi/disp2/hdmi/dev_hdmi.h	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/video/fbdev/sunxi/disp2/hdmi/dev_hdmi.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,42 @@
+/*
+ * Allwinner SoCs hdmi driver.
+ *
+ * Copyright (C) 2016 Allwinner.
+ *
+ * This file is licensed under the terms of the GNU General Public
+ * License version 2.  This program is licensed "as is" without any
+ * warranty of any kind, whether express or implied.
+ */
+
+#ifndef __DEV_DISPLAY_H__
+#define __DEV_DISPLAY_H__
+
+#include "drv_hdmi_i.h"
+
+int hdmi_open(struct inode *inode, struct file *file);
+int hdmi_release(struct inode *inode, struct file *file);
+ssize_t hdmi_read(struct file *file, char __user *buf,
+		size_t count, loff_t *ppos);
+ssize_t hdmi_write(struct file *file, const char __user *buf,
+		size_t count, loff_t *ppos);
+int hdmi_mmap(struct file *file, struct vm_area_struct *vma);
+long hdmi_ioctl(struct file *file, unsigned int cmd, unsigned long arg);
+
+extern s32 hdmi_init(struct platform_device *pdev);
+extern s32 hdmi_exit(void);
+extern s32 Fb_Init(u32 from);
+s32 hdmi_hpd_state(u32 state);
+s32 hdmi_hpd_event(void);
+
+typedef struct {
+	struct device           *dev;
+	bool                    bopen;
+	enum disp_tv_mode            mode;/* vic */
+	u32                   base_hdmi;
+	struct work_struct      hpd_work;
+} hdmi_info_t;
+
+extern hdmi_info_t ghdmi;
+extern struct disp_video_timings video_timing[];
+
+#endif
