diff -drupN a/drivers/hid/hid-core.c b/drivers/hid/hid-core.c
--- a/drivers/hid/hid-core.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/hid/hid-core.c	2022-06-12 05:28:14.000000000 +0300
@@ -50,7 +50,7 @@ module_param_named(debug, hid_debug, int
 MODULE_PARM_DESC(debug, "toggle HID debugging messages");
 EXPORT_SYMBOL_GPL(hid_debug);
 
-static int hid_ignore_special_drivers = 0;
+static int hid_ignore_special_drivers = 1;
 module_param_named(ignore_special_drivers, hid_ignore_special_drivers, int, 0600);
 MODULE_PARM_DESC(ignore_special_drivers, "Ignore any special drivers and handle all devices by generic driver");
 
