diff -drupN a/drivers/gpu/drm/drm_atomic_helper.c b/drivers/gpu/drm/drm_atomic_helper.c
--- a/drivers/gpu/drm/drm_atomic_helper.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/gpu/drm/drm_atomic_helper.c	2022-06-12 05:28:14.000000000 +0300
@@ -3166,6 +3166,9 @@ void __drm_atomic_helper_plane_destroy_s
 {
 	if (state->fb)
 		drm_framebuffer_unreference(state->fb);
+
+	if (state->fence)
+		fence_put(state->fence);
 }
 EXPORT_SYMBOL(__drm_atomic_helper_plane_destroy_state);
 
