diff -drupN a/arch/arm/boot/dts/sun8iw8p1-bassoon-std.dts b/arch/arm/boot/dts/sun8iw8p1-bassoon-std.dts
--- a/arch/arm/boot/dts/sun8iw8p1-bassoon-std.dts	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/arm/boot/dts/sun8iw8p1-bassoon-std.dts	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,43 @@
+/*
+ * Allwinner Technology CO., Ltd. sunivw1p1 evb board.
+ * Modified by czy
+ * evb board support.
+ */
+
+/dts-v1/;
+#include "sun8iw8p1.dtsi"
+
+/ {
+	aliases {
+		spinand = &spinand;
+	};
+
+	chosen {
+		bootargs = "earlyprintk=sunxi-uart,0x01c28800 loglevel=8 initcall_debug=1 console=ttyS1 init=/init root=/dev/nandd";
+		linux,initrd-start = <0x0 0x0>;
+		linux,initrd-end = <0x0 0x0>;
+	};
+
+	spinand: spi@01c68000 {
+		compatible = "allwinner,sunxi-spinand";
+		device_type = "spinand";
+		reg = <0x0 0x01c68000 0x0 0x1000>;
+		interrupts = <GIC_SPI 65 IRQ_TYPE_LEVEL_HIGH>;
+		clocks = <&clk_pll_periph0>, <&clk_spi0>;
+		pinctrl-names = "default", "sleep";
+		pinctrl-0 = <&spi0_pins_a &spi0_pins_b>;
+		pinctrl-1 = <&spi0_pins_c>;
+		nand0_regulator1 = "vcc-nand";
+		nand0_regulator2 = "none";
+		nand0_cache_level = <0x55aaaa55>;
+		nand0_flush_cache_num = <0x55aaaa55>;
+		nand0_capacity_level = <0x55aaaa55>;
+		nand0_id_number_ctl = <0x55aaaa55>;
+		nand0_print_level = <0x55aaaa55>;
+		nand0_p0 = <0x55aaaa55>;
+		nand0_p1 = <0x55aaaa55>;
+		nand0_p2 = <0x55aaaa55>;
+		nand0_p3 = <0x55aaaa55>;
+		status = "okay";
+	};
+};
