diff -drupN a/drivers/usb/Makefile b/drivers/usb/Makefile
--- a/drivers/usb/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/usb/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -13,6 +13,7 @@ obj-$(CONFIG_USB_ISP1760)	+= isp1760/
 
 obj-$(CONFIG_USB_MON)		+= mon/
 
+obj-$(CONFIG_USB_SUNXI_HCI)	+= host/
 obj-$(CONFIG_PCI)		+= host/
 obj-$(CONFIG_USB_EHCI_HCD)	+= host/
 obj-$(CONFIG_USB_ISP116X_HCD)	+= host/
@@ -61,3 +62,4 @@ obj-$(CONFIG_USB_GADGET)	+= gadget/
 obj-$(CONFIG_USB_COMMON)	+= common/
 
 obj-$(CONFIG_USBIP_CORE)	+= usbip/
+obj-$(CONFIG_USB_SUNXI_USB)     += sunxi_usb/
