diff -drupN a/modules/nand/sun8iw18p1/nfd/nand_class.h b/modules/nand/sun8iw18p1/nfd/nand_class.h
--- a/modules/nand/sun8iw18p1/nfd/nand_class.h	1970-01-01 03:00:00.000000000 +0300
+++ b/modules/nand/sun8iw18p1/nfd/nand_class.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,30 @@
+/*
+ * nand_class.h for  SUNXI NAND .
+ *
+ * Copyright (C) 2016 Allwinner.
+ *
+ *
+ * This file is licensed under the terms of the GNU General Public
+ * License version 2.  This program is licensed "as is" without any
+ * warranty of any kind, whether express or implied.
+ */
+#ifndef __NAND_CLASS_H__
+#define __NAND_CLASS_H__
+
+#include "nand_blk.h"
+#include "nand_dev.h"
+
+struct nand_kobject {
+	struct kobject kobj;
+	struct _nftl_blk *nftl_blk;
+};
+
+extern int debug_data;
+extern int nand_debug_init(struct _nftl_blk *nftl_blk, int part_num);
+extern int _dev_nand_read2(char *name, unsigned int start_sector,
+				unsigned int len, unsigned char *buf);
+extern int _dev_nand_write2(char *name, unsigned int start_sector,
+				unsigned int len, unsigned char *buf);
+extern int NAND_Get_Version(int *ver_main, int *ver_sub, int *date, int *time);
+extern int nftl_get_gc_info(void *zone, char *buf, int size);
+#endif
