diff -drupN a/drivers/net/wireless/Kconfig b/drivers/net/wireless/Kconfig
--- a/drivers/net/wireless/Kconfig	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/net/wireless/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -32,6 +32,9 @@ source "drivers/net/wireless/rsi/Kconfig
 source "drivers/net/wireless/st/Kconfig"
 source "drivers/net/wireless/ti/Kconfig"
 source "drivers/net/wireless/zydas/Kconfig"
+#source "drivers/net/wireless/xradio/Kconfig"
+#source "drivers/net/wireless/xr819/Kconfig"
+#source "drivers/net/wireless/xr829/Kconfig"
 
 config PCMCIA_RAYCS
 	tristate "Aviator/Raytheon 2.4GHz wireless support"
@@ -100,4 +103,13 @@ config USB_NET_RNDIS_WLAN
 
 	  If you choose to build a module, it'll be called rndis_wlan.
 
+#source "drivers/net/wireless/bcmdhd/Kconfig"
+#source "drivers/net/wireless/rtl8189es/Kconfig"
+#source "drivers/net/wireless/rtl8189fs/Kconfig"
+#source "drivers/net/wireless/rtl8723cs/Kconfig"
+#source "drivers/net/wireless/rtl8723ds/Kconfig"
+#source "drivers/net/wireless/rtl8723bs_vq0/Kconfig"
+#source "drivers/net/wireless/rtl8821cs/Kconfig"
+#source "drivers/net/wireless/rtl8822cs/Kconfig"
+
 endif # WLAN
