diff -drupN a/drivers/media/platform/Kconfig b/drivers/media/platform/Kconfig
--- a/drivers/media/platform/Kconfig	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/media/platform/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -121,6 +121,34 @@ source "drivers/media/platform/xilinx/Kc
 source "drivers/media/platform/rcar-vin/Kconfig"
 source "drivers/media/platform/atmel/Kconfig"
 
+config VIDEO_SUNXI_EVE
+	tristate "eve"
+	default n
+
+if VIDEO_SUNXI_EVE
+source "drivers/media/platform/sunxi-eve/Kconfig"
+endif
+
+config VIDEO_SUNXI_VIN
+	tristate "sunxi video input (camera csi/mipi isp vipp)driver"
+	default n
+
+if VIDEO_SUNXI_VIN
+source "drivers/media/platform/sunxi-vin/Kconfig"
+endif
+
+config SENSOR_POWER
+	tristate "sensor power in front of vin"
+	default n
+
+config VIDEO_SUNXI_VFE
+	tristate "sunxi video front end (camera and etc)driver"
+	default n
+
+if VIDEO_SUNXI_VFE
+source "drivers/media/platform/sunxi-vfe/Kconfig"
+endif
+
 config VIDEO_TI_CAL
 	tristate "TI CAL (Camera Adaptation Layer) driver"
 	depends on VIDEO_DEV && VIDEO_V4L2 && VIDEO_V4L2_SUBDEV_API
@@ -134,6 +162,18 @@ config VIDEO_TI_CAL
 	  In TI Technical Reference Manual this module is referred as
 	  Camera Interface Subsystem (CAMSS).
 
+config VIDEO_SUNXI_TVD
+	tristate "sunxi tvd driver"
+    depends on ARCH_SUN8IW11 || ARCH_SUN8IW17
+    select VIDEOBUF2_DMA_CONTIG
+	default m
+
+config VIDEO_SUNXI_CAR_REVERSE
+	tristate "sunxi fast car reverse driver"
+	depends on ARCH_SUN8IW11 || ARCH_SUN8IW17
+	depends on VIDEO_SUNXI_TVD
+	default n
+
 endif # V4L_PLATFORM_DRIVERS
 
 menuconfig V4L_MEM2MEM_DRIVERS
@@ -376,3 +416,5 @@ menuconfig DVB_PLATFORM_DRIVERS
 if DVB_PLATFORM_DRIVERS
 source "drivers/media/platform/sti/c8sectpfe/Kconfig"
 endif #DVB_PLATFORM_DRIVERS
+source "drivers/media/platform/sunxi-ise/Kconfig"
+source "drivers/media/platform/sunxi-eise/Kconfig"
\ В конце файла нет новой строки
