diff -drupN a/net/bluetooth/hci_core.c b/net/bluetooth/hci_core.c
--- a/net/bluetooth/hci_core.c	2018-08-06 17:23:04.000000000 +0300
+++ b/net/bluetooth/hci_core.c	2022-06-12 05:28:14.000000000 +0300
@@ -814,8 +814,9 @@ static int __hci_init(struct hci_dev *hd
 		return err;
 
 	err = __hci_req_sync(hdev, hci_init4_req, 0, HCI_INIT_TIMEOUT, NULL);
-	if (err < 0)
-		return err;
+	if (err < 0) {
+		pr_err("%s(%d), err = %d, EBADRQC.\n", __func__, __LINE__, err);
+	}
 
 	/* This function is only called when the controller is actually in
 	 * configured state. When the controller is marked as unconfigured,
