diff -drupN a/arch/powerpc/platforms/pseries/cmm.c b/arch/powerpc/platforms/pseries/cmm.c
--- a/arch/powerpc/platforms/pseries/cmm.c	2018-08-06 17:23:04.000000000 +0300
+++ b/arch/powerpc/platforms/pseries/cmm.c	2022-06-12 05:28:14.000000000 +0300
@@ -708,7 +708,7 @@ static void cmm_exit(void)
  * Return value:
  * 	0 on success / other on failure
  **/
-static int cmm_set_disable(const char *val, struct kernel_param *kp)
+static int cmm_set_disable(const char *val, const struct kernel_param *kp)
 {
 	int disable = simple_strtoul(val, NULL, 10);
 
