diff -drupN a/arch/arm/boot/dts/sun8iw17p1-soc.dts b/arch/arm/boot/dts/sun8iw17p1-soc.dts
--- a/arch/arm/boot/dts/sun8iw17p1-soc.dts	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/arm/boot/dts/sun8iw17p1-soc.dts	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,16 @@
+/*
+ * Allwinner Technology CO., Ltd. sun8iw17p1 soc board.
+ *
+ * soc board support.
+ */
+
+/dts-v1/;
+
+#include "sun8iw17p1.dtsi"
+
+/{
+
+	soc@03000000 {
+
+	};
+};
