diff -drupN a/drivers/video/fbdev/sunxi/lcd_fb/panels/kld35512/kld35512.h b/drivers/video/fbdev/sunxi/lcd_fb/panels/kld35512/kld35512.h
--- a/drivers/video/fbdev/sunxi/lcd_fb/panels/kld35512/kld35512.h	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/video/fbdev/sunxi/lcd_fb/panels/kld35512/kld35512.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,24 @@
+/*
+ * drivers/video/fbdev/sunxi/lcd_fb/panels/kld39501/kld39501/kld39501.h
+ *
+ * Copyright (c) 2007-2019 Allwinnertech Co., Ltd.
+ * Author: zhengxiaobin <zhengxiaobin@allwinnertech.com>
+ *
+ * This software is licensed under the terms of the GNU General Public
+ * License version 2, as published by the Free Software Foundation, and
+ * may be copied, distributed, and modified under those terms.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ */
+#ifndef _KLD35512_H
+#define _KLD35512_H
+
+#include "../panels.h"
+
+struct __lcd_panel kld35512_panel;
+
+#endif /*End of file*/
