diff -drupN a/drivers/leds/Makefile b/drivers/leds/Makefile
--- a/drivers/leds/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/leds/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -69,8 +69,13 @@ obj-$(CONFIG_LEDS_POWERNV)		+= leds-powe
 obj-$(CONFIG_LEDS_SEAD3)		+= leds-sead3.o
 obj-$(CONFIG_LEDS_IS31FL319X)		+= leds-is31fl319x.o
 obj-$(CONFIG_LEDS_IS31FL32XX)		+= leds-is31fl32xx.o
+obj-$(CONFIG_LEDS_IS31FL3736)		+= leds-is31fl3736.o
 obj-$(CONFIG_LEDS_PM8058)		+= leds-pm8058.o
 obj-$(CONFIG_LEDS_MLXCPLD)		+= leds-mlxcpld.o
+obj-$(CONFIG_LEDS_SUNXI)		+= leds-sunxi.o
+
+#for AW2016 LED Driver
+obj-$(CONFIG_LEDS_AW2016) += leds-aw2016.o
 
 # LED SPI Drivers
 obj-$(CONFIG_LEDS_DAC124S085)		+= leds-dac124s085.o
