diff -drupN a/security/selinux/Kconfig b/security/selinux/Kconfig
--- a/security/selinux/Kconfig	2018-08-06 17:23:04.000000000 +0300
+++ b/security/selinux/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -2,6 +2,7 @@ config SECURITY_SELINUX
 	bool "NSA SELinux Support"
 	depends on SECURITY_NETWORK && AUDIT && NET && INET
 	select NETWORK_SECMARK
+	select FLEX_ARRAY
 	default n
 	help
 	  This selects NSA Security-Enhanced Linux (SELinux).
