diff -drupN a/security/inode.c b/security/inode.c
--- a/security/inode.c	2018-08-06 17:23:04.000000000 +0300
+++ b/security/inode.c	2022-06-12 05:28:14.000000000 +0300
@@ -100,7 +100,7 @@ struct dentry *securityfs_create_file(co
 	dir = d_inode(parent);
 
 	inode_lock(dir);
-	dentry = lookup_one_len(name, parent, strlen(name));
+	dentry = lookup_one_len2(name, mount, parent, strlen(name));
 	if (IS_ERR(dentry))
 		goto out;
 
