diff -drupN a/scripts/dtc/Makefile.dtc b/scripts/dtc/Makefile.dtc
--- a/scripts/dtc/Makefile.dtc	2018-08-06 17:23:04.000000000 +0300
+++ b/scripts/dtc/Makefile.dtc	2022-06-12 05:28:14.000000000 +0300
@@ -12,7 +12,11 @@ DTC_SRCS = \
 	livetree.c \
 	srcpos.c \
 	treesource.c \
-	util.c
+	util.c \
+	script_parser/iniparser.c \
+	script_parser/dictionary.c \
+	script_parser/script.c
+	script_parser/script_to_dts.c
 
 DTC_GEN_SRCS = dtc-lexer.lex.c dtc-parser.tab.c
 DTC_OBJS = $(DTC_SRCS:%.c=%.o) $(DTC_GEN_SRCS:%.c=%.o)
