diff -drupN a/drivers/misc/sunxi-rf/Kconfig b/drivers/misc/sunxi-rf/Kconfig
--- a/drivers/misc/sunxi-rf/Kconfig	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/misc/sunxi-rf/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,9 @@
+#
+# Allwinner rfkill driver.
+#
+config SUNXI_RFKILL
+  tristate "Allwinner rfkill driver"
+  depends on WIRELESS || RFKILL
+  help
+    allwinner rfkill driver
+
