diff -drupN a/drivers/thermal/sunxi_thermal/sunxi_cooling_device/Makefile b/drivers/thermal/sunxi_thermal/sunxi_cooling_device/Makefile
--- a/drivers/thermal/sunxi_thermal/sunxi_cooling_device/Makefile	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/thermal/sunxi_thermal/sunxi_cooling_device/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,9 @@
+#
+# Makefile for the sunxi cpu budget cooling drivers.
+#
+#obj-$(CONFIG_SUNXI_BUDGET_COOLING) += sunxi-budget.o
+obj-$(CONFIG_SUNXI_BUDGET_COOLING) += sunxi-budget-cooling.o
+#sunxi-budget-y += sunxi-budget-cooling.o
+obj-$(CONFIG_SUNXI_BUDGET_COOLING_CPU_DVFS) += sunxi-budget-cooling-dvfs.o
+obj-$(CONFIG_SUNXI_BUDGET_COOLING_HOTPLUG) += sunxi-budget-cooling-hotplug.o
+obj-$(CONFIG_SUNXI_GPU_COOLING) += sunxi_gpu_cooling.o
