diff -drupN a/arch/arm/include/asm/cacheflush.h b/arch/arm/include/asm/cacheflush.h
--- a/arch/arm/include/asm/cacheflush.h	2018-08-06 17:23:04.000000000 +0300
+++ b/arch/arm/include/asm/cacheflush.h	2022-06-12 05:28:14.000000000 +0300
@@ -163,6 +163,8 @@ extern void dmac_flush_range(const void
 
 #endif
 
+#define __dma_flush_area(a, s) dmac_flush_range((a), (a) + ((s) - 1))
+
 /*
  * Copy user data from/to a page which is mapped into a different
  * processes address space.  Really, we want to allow our "user
