diff -drupN a/arch/arm64/include/asm/elf.h b/arch/arm64/include/asm/elf.h
--- a/arch/arm64/include/asm/elf.h	2018-08-06 17:23:04.000000000 +0300
+++ b/arch/arm64/include/asm/elf.h	2022-06-12 05:28:14.000000000 +0300
@@ -169,7 +169,7 @@ extern int arch_setup_additional_pages(s
 #ifdef CONFIG_COMPAT
 
 /* PIE load location for compat arm. Must match ARM ELF_ET_DYN_BASE. */
-#define COMPAT_ELF_ET_DYN_BASE		0x000400000UL
+#define COMPAT_ELF_ET_DYN_BASE		(2 * TASK_SIZE_32 / 3)
 
 /* AArch32 registers. */
 #define COMPAT_ELF_NGREG		18
