diff -drupN a/security/keys/dh.c b/security/keys/dh.c
--- a/security/keys/dh.c	2018-08-06 17:23:04.000000000 +0300
+++ b/security/keys/dh.c	2022-06-12 05:28:14.000000000 +0300
@@ -55,7 +55,7 @@ static ssize_t mpi_from_key(key_serial_t
 		if (status == 0) {
 			const struct user_key_payload *payload;
 
-			payload = user_key_payload(key);
+			payload = user_key_payload_locked(key);
 
 			if (maxlen == 0) {
 				*mpi = NULL;
