diff -drupN a/net/netfilter/nf_nat_irc.c b/net/netfilter/nf_nat_irc.c
--- a/net/netfilter/nf_nat_irc.c	2018-08-06 17:23:04.000000000 +0300
+++ b/net/netfilter/nf_nat_irc.c	2022-06-12 05:28:14.000000000 +0300
@@ -107,7 +107,7 @@ static int __init nf_nat_irc_init(void)
 }
 
 /* Prior to 2.6.11, we had a ports param.  No longer, but don't break users. */
-static int warn_set(const char *val, struct kernel_param *kp)
+static int warn_set(const char *val, const struct kernel_param *kp)
 {
 	printk(KERN_INFO KBUILD_MODNAME
 	       ": kernel >= 2.6.10 only uses 'ports' for conntrack modules\n");
