diff -drupN a/include/linux/pm.h b/include/linux/pm.h
--- a/include/linux/pm.h	2018-08-06 17:23:04.000000000 +0300
+++ b/include/linux/pm.h	2022-06-12 05:28:14.000000000 +0300
@@ -563,6 +563,7 @@ struct dev_pm_info {
 	bool			is_suspended:1;	/* Ditto */
 	bool			is_noirq_suspended:1;
 	bool			is_late_suspended:1;
+	bool			is_runtime_disabled:1;
 	bool			early_init:1;	/* Owned by the PM core */
 	bool			direct_complete:1;	/* Owned by the PM core */
 	spinlock_t		lock;
