diff -drupN a/drivers/video/fbdev/sunxi/disp2/disp/de/disp_vga.h b/drivers/video/fbdev/sunxi/disp2/disp/de/disp_vga.h
--- a/drivers/video/fbdev/sunxi/disp2/disp/de/disp_vga.h	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/video/fbdev/sunxi/disp2/disp/de/disp_vga.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,42 @@
+/*
+ * Allwinner SoCs display driver.
+ *
+ * Copyright (C) 2016 Allwinner.
+ *
+ * This file is licensed under the terms of the GNU General Public
+ * License version 2.  This program is licensed "as is" without any
+ * warranty of any kind, whether express or implied.
+ */
+
+#ifndef __DISP_VGA_H__
+#define __DISP_VGA_H__
+
+#include "disp_private.h"
+#include "disp_display.h"
+
+struct disp_vga_private_data {
+	bool enabled;
+	bool suspended;
+
+	enum disp_tv_mode vga_mode;
+	struct disp_tv_func tv_func;
+	struct disp_video_timings *video_info;
+
+	struct disp_clk_info lcd_clk;
+	struct clk *clk;
+	struct clk *clk_parent;
+
+	s32 frame_per_sec;
+	u32 usec_per_line;
+	u32 judge_line;
+
+	u32 irq_no;
+	spinlock_t data_lock;
+	struct mutex mlock;
+};
+
+s32 disp_init_vga(void);
+
+
+extern void sync_event_proc(u32 disp, bool timeout);
+#endif
