diff -drupN a/net/dns_resolver/dns_query.c b/net/dns_resolver/dns_query.c
--- a/net/dns_resolver/dns_query.c	2018-08-06 17:23:04.000000000 +0300
+++ b/net/dns_resolver/dns_query.c	2022-06-12 05:28:14.000000000 +0300
@@ -70,7 +70,7 @@ int dns_query(const char *type, const ch
 	      const char *options, char **_result, time64_t *_expiry)
 {
 	struct key *rkey;
-	const struct user_key_payload *upayload;
+	struct user_key_payload *upayload;
 	const struct cred *saved_cred;
 	size_t typelen, desclen;
 	char *desc, *cp;
@@ -141,7 +141,7 @@ int dns_query(const char *type, const ch
 	if (ret)
 		goto put;
 
-	upayload = user_key_payload(rkey);
+	upayload = user_key_payload_locked(rkey);
 	len = upayload->datalen;
 
 	ret = -ENOMEM;
