diff -drupN a/drivers/input/keyboard/Kconfig b/drivers/input/keyboard/Kconfig
--- a/drivers/input/keyboard/Kconfig	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/input/keyboard/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -734,4 +734,24 @@ config KEYBOARD_BCM
 	  To compile this driver as a module, choose M here: the
 	  module will be called bcm-keypad.
 
+config KEYBOARD_SUNXI
+	tristate "softwinnner KEY BOARD support"
+	help
+	  Say Y here to enable the keyboard, support 7 keys.
+	  based on verify board.
+
+config IR_TX_SUNXI
+	tristate "sunxi IR TX support"
+	help
+	  Say Y here to enable the IR TX, support NEC code.
+
+config GPIO_IR_RX_SUNXI
+	tristate "sunxi GPIO IR RX support"
+	help
+	  Say Y here to enable the IR RX, support NEC code or send time to user.
+
+config GPIO_IR_TX_SUNXI
+	tristate "sunxi GPIO IR TX support"
+	help
+	  Say Y here to enable the IR TX, support NEC protocol.
 endif
