diff -drupN a/kernel/sched/idle.c b/kernel/sched/idle.c
--- a/kernel/sched/idle.c	2018-08-06 17:23:04.000000000 +0300
+++ b/kernel/sched/idle.c	2022-06-12 05:28:14.000000000 +0300
@@ -23,9 +23,10 @@ extern char __cpuidle_text_start[], __cp
  * sched_idle_set_state - Record idle state for the current CPU.
  * @idle_state: State to record.
  */
-void sched_idle_set_state(struct cpuidle_state *idle_state)
+void sched_idle_set_state(struct cpuidle_state *idle_state, int index)
 {
 	idle_set_state(this_rq(), idle_state);
+	idle_set_state_idx(this_rq(), index);
 }
 
 static int __read_mostly cpu_idle_force_poll;
