diff -drupN a/drivers/media/v4l2-core/v4l2-subdev.c b/drivers/media/v4l2-core/v4l2-subdev.c
--- a/drivers/media/v4l2-core/v4l2-subdev.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/media/v4l2-core/v4l2-subdev.c	2022-06-12 05:28:14.000000000 +0300
@@ -115,7 +115,7 @@ static int subdev_close(struct file *fil
 	if (sd->internal_ops && sd->internal_ops->close)
 		sd->internal_ops->close(sd, subdev_fh);
 #if defined(CONFIG_MEDIA_CONTROLLER)
-	if (sd->v4l2_dev->mdev)
+	if (sd->v4l2_dev && sd->v4l2_dev->mdev)
 		media_entity_put(&sd->entity);
 #endif
 	v4l2_fh_del(vfh);
