diff -drupN a/drivers/video/fbdev/sunxi/disp2/hdmi/Makefile b/drivers/video/fbdev/sunxi/disp2/hdmi/Makefile
--- a/drivers/video/fbdev/sunxi/disp2/hdmi/Makefile	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/video/fbdev/sunxi/disp2/hdmi/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,19 @@
+obj-$(CONFIG_HDMI_DISP2_SUNXI) += hdmi.o
+
+hdmi-y := drv_hdmi.o hdmi_core.o hdmi_edid.o
+ifeq ($(CONFIG_ARM64),y)
+$(shell cp $(obj)/libhdmi_sun50iw1 $(obj)/libhdmi_sun50iw1.a)
+hdmi-$(CONFIG_ARCH_SUN50IW1P1) += libhdmi_sun50iw1.a
+hdmi-$(CONFIG_ARCH_SUN50IW2P1) += libhdmi_sun50iw1.a
+endif
+
+ifeq ($(CONFIG_ARM),y)
+$(shell cp $(obj)/libhdmi_sun8iw11 $(obj)/libhdmi_sun8iw11.a)
+hdmi-$(CONFIG_ARCH_SUN50IW1P1) += libhdmi_sun8iw11.a
+hdmi-$(CONFIG_ARCH_SUN50IW2P1) += libhdmi_sun8iw11.a
+hdmi-$(CONFIG_ARCH_SUN8IW11)   += libhdmi_sun8iw11.a
+hdmi-$(CONFIG_ARCH_SUN8IW12) += hdmi_bsp_sun8iw12.o
+hdmi-$(CONFIG_ARCH_SUN8IW7) += hdmi_bsp_sun8iw7.o
+hdmi-$(CONFIG_ARCH_SUN8IW6) += hdmi_bsp_sun8iw6.o
+endif
+
