diff -drupN a/drivers/power/supply/axp/axp80x/axp80x.h b/drivers/power/supply/axp/axp80x/axp80x.h
--- a/drivers/power/supply/axp/axp80x/axp80x.h	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/power/supply/axp/axp80x/axp80x.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,76 @@
+/*
+ * drivers/power/axp/axp80x/axp80x.h
+ * (C) Copyright 2010-2016
+ * Allwinner Technology Co., Ltd. <www.allwinnertech.com>
+ * Pannan <pannan@allwinnertech.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of
+ * the License, or (at your option) any later version.
+ */
+
+#ifndef __AXP80X_H_
+#define __AXP80X_H_
+
+#include "../axp-core.h"
+
+#define AXP80X_STARTUP_SOURCE      (0x00)
+#define AXP80X_IC_TYPE             (0x03)
+#define AXP80X_DATA_BUFFER1        (0x04)
+#define AXP80X_DATA_BUFFER2        (0x05)
+#define AXP80X_DATA_BUFFER3        (0x06)
+#define AXP80X_DATA_BUFFER4        (0x07)
+#define AXP80X_ONOFF_CTRL1         (0x10)
+#define AXP80X_ONOFF_CTRL2         (0x11)
+#define AXP80X_DCAOUT_VOL          (0x12)
+#define AXP80X_DCBOUT_VOL          (0x13)
+#define AXP80X_DCCOUT_VOL          (0x14)
+#define AXP80X_DCDOUT_VOL          (0x15)
+#define AXP80X_DCEOUT_VOL          (0x16)
+#define AXP80X_ALDO1OUT_VOL        (0x17)
+#define AXP80X_ALDO2OUT_VOL        (0x18)
+#define AXP80X_ALDO3OUT_VOL        (0x19)
+#define AXP80X_DCDC_DVM_CTRL       (0x1A)
+#define AXP80X_DCDC_MODE_CTRL      (0x1B)
+#define AXP80X_DCDC_FREQSET        (0x1C)
+#define AXP80X_DCDC_MON_CTRL       (0x1D)
+#define AXP80X_IFQ_PWROK_SET       (0x1F)
+#define AXP80X_BLDO1OUT_VOL        (0x20)
+#define AXP80X_BLDO2OUT_VOL        (0x21)
+#define AXP80X_BLDO3OUT_VOL        (0x22)
+#define AXP80X_BLDO4OUT_VOL        (0x23)
+#define AXP80X_CLDO1OUT_VOL        (0x24)
+#define AXP80X_CLDO2OUT_VOL        (0x25)
+#define AXP80X_CLDO3OUT_VOL        (0x26)
+#define AXP80X_VOFF_SET            (0x31)
+#define AXP80X_OFF_CTL             (0x32)
+#define AXP80X_WAKEUP_PIN_CTRL     (0x35)
+#define AXP80X_POK_SET             (0x36)
+#define AXP80X_INTERFACE_MODE      (0x3E)
+#define AXP80X_SPECIAL_CTRL        (0x3F)
+
+#define AXP80X_INTEN1              (0x40)
+#define AXP80X_INTEN2              (0x41)
+#define AXP80X_INTSTS1             (0x48)
+#define AXP80X_INTSTS2             (0x49)
+
+/* bit definitions for AXP events ,irq event */
+/*  AXP80X  */
+#define AXP80X_IRQ_OVER_WARN1      (0)
+#define AXP80X_IRQ_OVER_WARN2      (1)
+#define AXP80X_IRQ_DCDCA_LOW_85P   (3)
+#define AXP80X_IRQ_DCDCB_LOW_85P   (4)
+#define AXP80X_IRQ_DCDCC_LOW_85P   (5)
+#define AXP80X_IRQ_DCDCD_LOW_85P   (6)
+#define AXP80X_IRQ_DCDCE_LOW_85P   (7)
+
+#define AXP80X_IRQ_PEKLO           (8)
+#define AXP80X_IRQ_PEKSH           (9)
+#define AXP80X_IRQ_WS_EN_WAKEUPIN  (12)
+#define AXP80X_IRQ_PEKFE           (13)
+#define AXP80X_IRQ_PEKRE           (14)
+
+extern struct axp_config_info axp80x_config;
+
+#endif /* __AXP80X_H_ */
