diff -drupN a/drivers/gpu/drm/drm_crtc_internal.h b/drivers/gpu/drm/drm_crtc_internal.h
--- a/drivers/gpu/drm/drm_crtc_internal.h	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/gpu/drm/drm_crtc_internal.h	2022-06-12 05:28:14.000000000 +0300
@@ -41,6 +41,8 @@ int drm_crtc_check_viewport(const struct
 			    const struct drm_display_mode *mode,
 			    const struct drm_framebuffer *fb);
 
+struct fence *drm_crtc_create_fence(struct drm_crtc *crtc);
+
 void drm_fb_release(struct drm_file *file_priv);
 
 /* dumb buffer support IOCTLs */
