diff -drupN a/fs/internal.h b/fs/internal.h
--- a/fs/internal.h	2018-08-06 17:23:04.000000000 +0300
+++ b/fs/internal.h	2022-06-12 05:28:14.000000000 +0300
@@ -88,9 +88,11 @@ extern struct file *get_empty_filp(void)
  * super.c
  */
 extern int do_remount_sb(struct super_block *, int, void *, int);
+extern int do_remount_sb2(struct vfsmount *, struct super_block *, int,
+								void *, int);
 extern bool trylock_super(struct super_block *sb);
 extern struct dentry *mount_fs(struct file_system_type *,
-			       int, const char *, void *);
+			       int, const char *, struct vfsmount *, void *);
 extern struct super_block *user_get_super(dev_t);
 
 /*
