diff -drupN a/sound/soc/sunxi/sun50iw11/Kconfig b/sound/soc/sunxi/sun50iw11/Kconfig
--- a/sound/soc/sunxi/sun50iw11/Kconfig	1970-01-01 03:00:00.000000000 +0300
+++ b/sound/soc/sunxi/sun50iw11/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,124 @@
+# CPU DAI module
+config SND_SUN50IW11_SOC_CPUDAI
+	tristate
+
+# Digital Audio(I2S/PCM/TDM)
+config SND_SUN50IW11_SOC_DAUDIO
+	tristate
+
+# SPDIF Audio Interface
+config SND_SUN50IW11_SOC_SPDIF
+	tristate
+
+# DMIC Audio Interface
+config SND_SUN50IW11_SOC_DMIC
+	tristate
+
+# MAD Audio Interface
+config SND_SUN50IW11_MAD
+	bool
+
+# SUNXI Platform DMA
+config SND_SUN50IW11_SOC
+	tristate
+
+# SUNXI Codec Embedded
+config SND_SUN50IW11_CODEC
+	tristate
+
+# MAD module
+config SND_SUN50IW11_SOC_MAD
+	tristate "Allwinner Mad Support"
+	select REGMAP_MMIO
+        select SND_SUN50IW11_MAD
+	depends on ARCH_SUN50IW11
+	help
+	  Say Y or M if you want to using mad module.
+
+# Card Machine Select
+config SND_SUNXI_SOC_SUN50IW11_CODEC
+	tristate "Allwinner Audio Codec Support"
+	select SND_SUN50IW11_SOC
+	select SND_SOC_GENERIC_DMAENGINE_PCM
+	select REGMAP_MMIO
+	select SND_SUN50IW11_CODEC
+	select SND_SUNXI_SOC_RWFUNC
+	select SND_SUN50IW11_SOC_CPUDAI
+	depends on ARCH_SUN50IW11
+	help
+	Select Y or M to add support for the Codec embedded in the Allwinner
+	SUN50IW11 and affiliated SoCs.
+
+# Daudio Machine Select
+config SND_SUNXI_SOC_SUN50IW11_DAUDIO
+	tristate "Allwinner Digital Audio Support"
+	select SND_SUN50IW11_SOC
+	select SND_SOC_GENERIC_DMAENGINE_PCM
+	select SND_SUN50IW11_SOC_DAUDIO
+	select REGMAP_MMIO
+	depends on ARCH_SUN50IW11
+	help
+	Select Y or M to add support for I2S/PCM/TDM Module in the
+	Allwinner SoCs.
+
+# SPDIF Machine Select
+config SND_SUNXI_SOC_SUN50IW11_SPDIF
+	tristate "Allwinner SPDIF Support"
+	select SND_SUN50IW11_SOC
+	select SND_SOC_GENERIC_DMAENGINE_PCM
+	select SND_SUN50IW11_SOC_SPDIF
+	select REGMAP_MMIO
+	depends on ARCH_SUN50IW11
+	help
+	Select Y or M to add support SPDIF Module in Allwinner
+	SoCs.
+
+# DMIC Machine Select
+config SND_SUNXI_SOC_SUN50IW11_DMIC
+	tristate "Allwinner DMIC Support"
+	select SND_SUN50IW11_SOC
+	select SND_SOC_GENERIC_DMAENGINE_PCM
+	select SND_SUN50IW11_SOC_DMIC
+        select SND_SOC_DMIC
+	select REGMAP_MMIO
+	depends on ARCH_SUN50IW11
+	help
+	Select Y or M to add support DMIC Module in Allwinner
+	SoCs.
+
+config SUN50IW11_AUDIO_DEBUG
+	bool "Allwinner Audio Debug Support"
+        default n
+	depends on SND_SUN50IW11_SOC
+	help
+	  Select Y to enable debug information of allwinner audio driver
+	  If you not sure, Just not select it.
+
+config SND_SUN50IW11_DAUDIO_DEBUG
+	bool "Allwinner DAUDIO Debug"
+        default n
+	depends on SUN50IW11_AUDIO_DEBUG && SND_SUN50IW11_SOC_DAUDIO
+	help
+	  Say Y if you want to debug daudio driver.
+
+config SND_SUN50IW11_SPDIF_DEBUG
+	bool "Allwinner SPDIF Debug"
+        default n
+	depends on SUN50IW11_AUDIO_DEBUG && SND_SUN50IW11_SOC_SPDIF
+	help
+	  Say Y if you want to debug spdif driver.
+
+config SND_SUN50IW11_DMIC_DEBUG
+	bool "Allwinner DMIC Debug"
+        default n
+	depends on SUN50IW11_AUDIO_DEBUG && SND_SUN50IW11_SOC_DMIC
+	help
+	  Say Y if you want to debug dmic driver.
+
+config SND_SUN50IW11_MAD_DEBUG
+	bool "Allwinner MAD Debug"
+        default n
+	depends on SUN50IW11_AUDIO_DEBUG && SND_SUN50IW11_SOC_MAD
+	help
+	  Say Y if you want to debug mad driver.
+
