diff -drupN a/drivers/media/platform/sunxi-vin/utility/config.h b/drivers/media/platform/sunxi-vin/utility/config.h
--- a/drivers/media/platform/sunxi-vin/utility/config.h	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/media/platform/sunxi-vin/utility/config.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,24 @@
+
+/*
+ * config.h for device tree and sensor list parser.
+ *
+ * Copyright (c) 2017 by Allwinnertech Co., Ltd.  http://www.allwinnertech.com
+ *
+ * Authors:  Zhao Wei <zhaowei@allwinnertech.com>
+ *	Yang Feng <yangfeng@allwinnertech.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ */
+
+#ifndef __CONFIG__H__
+#define __CONFIG__H__
+
+#include "../vin-video/vin_core.h"
+#include "../vin.h"
+#include "cfg_op.h"
+
+int parse_modules_from_device_tree(struct vin_md *vind);
+
+#endif /*__CONFIG__H__*/
