diff -drupN a/scripts/dtc/srcpos.h b/scripts/dtc/srcpos.h
--- a/scripts/dtc/srcpos.h	2018-08-06 17:23:04.000000000 +0300
+++ b/scripts/dtc/srcpos.h	2022-06-12 05:28:14.000000000 +0300
@@ -105,7 +105,6 @@ extern struct srcpos srcpos_empty;
 extern void srcpos_update(struct srcpos *pos, const char *text, int len);
 extern struct srcpos *srcpos_copy(struct srcpos *pos);
 extern char *srcpos_string(struct srcpos *pos);
-extern void srcpos_dump(struct srcpos *pos);
 
 extern void srcpos_verror(struct srcpos *pos, const char *prefix,
 			  const char *fmt, va_list va)
