diff -drupN a/include/linux/mount.h b/include/linux/mount.h
--- a/include/linux/mount.h	2018-08-06 17:23:04.000000000 +0300
+++ b/include/linux/mount.h	2022-06-12 05:28:14.000000000 +0300
@@ -67,6 +67,7 @@ struct vfsmount {
 	struct dentry *mnt_root;	/* root of the mounted tree */
 	struct super_block *mnt_sb;	/* pointer to superblock */
 	int mnt_flags;
+	void *data;
 };
 
 struct file; /* forward dec */
