diff -drupN a/include/linux/usb/gadget.h b/include/linux/usb/gadget.h
--- a/include/linux/usb/gadget.h	2018-08-06 17:23:04.000000000 +0300
+++ b/include/linux/usb/gadget.h	2022-06-12 05:28:14.000000000 +0300
@@ -109,6 +109,10 @@ struct usb_request {
 	void			*context;
 	struct list_head	list;
 
+#if IS_ENABLED(CONFIG_USB_SUNXI_UDC0)
+	int			dma_flag;
+#endif
+
 	int			status;
 	unsigned		actual;
 };
@@ -494,7 +498,11 @@ static inline int gadget_avoids_skb_rese
  */
 static inline int gadget_is_dualspeed(struct usb_gadget *g)
 {
+#if IS_ENABLED(CONFIG_USB_SUNXI_UDC0)
+	return 1;
+#else
 	return g->max_speed >= USB_SPEED_HIGH;
+#endif
 }
 
 /**
