diff -drupN a/modules/nand/sun50iw9p1/phy-nand/nand-partition/Makefile b/modules/nand/sun50iw9p1/phy-nand/nand-partition/Makefile
--- a/modules/nand/sun50iw9p1/phy-nand/nand-partition/Makefile	1970-01-01 03:00:00.000000000 +0300
+++ b/modules/nand/sun50iw9p1/phy-nand/nand-partition/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,10 @@
+#
+# Makefile for nand-partition
+#
+#
+#
+
+
+obj-$(CONFIG_SUNXI_NAND) += build_nand_partition.o
+obj-$(CONFIG_SUNXI_NAND) += build_phy_partition.o
+obj-$(CONFIG_SUNXI_NAND) += nand_info_init.o
