diff -drupN a/drivers/md/md.c b/drivers/md/md.c
--- a/drivers/md/md.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/md/md.c	2022-06-12 05:28:14.000000000 +0300
@@ -5098,7 +5098,7 @@ static struct kobject *md_probe(dev_t de
 	return NULL;
 }
 
-static int add_named_array(const char *val, struct kernel_param *kp)
+static int add_named_array(const char *val, const struct kernel_param *kp)
 {
 	/* val must be "md_*" where * is not all digits.
 	 * We allocate an array with a large free minor number, and
@@ -8989,11 +8989,11 @@ static __exit void md_exit(void)
 subsys_initcall(md_init);
 module_exit(md_exit)
 
-static int get_ro(char *buffer, struct kernel_param *kp)
+static int get_ro(char *buffer, const struct kernel_param *kp)
 {
 	return sprintf(buffer, "%d", start_readonly);
 }
-static int set_ro(const char *val, struct kernel_param *kp)
+static int set_ro(const char *val, const struct kernel_param *kp)
 {
 	return kstrtouint(val, 10, (unsigned int *)&start_readonly);
 }
