diff -drupN a/arch/arm64/boot/dts/sunxi/sun50iw1p1-fpga.dts b/arch/arm64/boot/dts/sunxi/sun50iw1p1-fpga.dts
--- a/arch/arm64/boot/dts/sunxi/sun50iw1p1-fpga.dts	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/arm64/boot/dts/sunxi/sun50iw1p1-fpga.dts	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,43 @@
+/*
+ * Allwinner Technology CO., Ltd. sun50iw1p1 fpga board.
+ *
+ * fpga board support.
+ */
+
+/dts-v1/;
+#include "sun50iw1p1.dtsi"
+
+/{
+
+	soc@01c00000 {
+		uart0: uart@01c28000 {
+			status = "okay";
+		};
+
+		twi0: twi@0x01c2ac00{
+			status = "disabled";
+
+			eeprom@50 {
+				compatible = "atmel,24c16";
+				reg = <0x50>;
+			};
+		};
+
+		spi0: spi@01c68000 {
+			status = "disabled";
+
+			nor_flash@0 {
+				#address-cells = <1>;
+				#size-cells = <1>;
+				compatible = "atmel,at25df641";
+				spi-max-frequency = <50000000>;
+				reg = <0>;	/* Chip select 0 */
+
+				partition@0 {
+					label = "NorFlash part0";
+					reg = <0x00000000 0x00800000>;
+				};
+			};
+		};
+	};
+};
