diff -drupN a/drivers/media/rc/Makefile b/drivers/media/rc/Makefile
--- a/drivers/media/rc/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/media/rc/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -35,5 +35,8 @@ obj-$(CONFIG_IR_IGORPLUGUSB) += igorplug
 obj-$(CONFIG_IR_IGUANA) += iguanair.o
 obj-$(CONFIG_IR_TTUSBIR) += ttusbir.o
 obj-$(CONFIG_RC_ST) += st_rc.o
-obj-$(CONFIG_IR_SUNXI) += sunxi-cir.o
+obj-$(CONFIG_IR_SUNXI) += sunxi-ir-rx.o
+obj-$(CONFIG_IR_SUNXI_GPIO) += sunxi-gpio-ir.o
 obj-$(CONFIG_IR_IMG) += img-ir/
+sunxi-ir-rx-objs :=  rc-sunxi-keymaps.o sunxi-ir-dev.o
+# sunxi-cir.o
