diff -drupN a/net/ipv4/udp_offload.c b/net/ipv4/udp_offload.c
--- a/net/ipv4/udp_offload.c	2018-08-06 17:23:04.000000000 +0300
+++ b/net/ipv4/udp_offload.c	2022-06-12 05:28:14.000000000 +0300
@@ -169,7 +169,13 @@ struct sk_buff *skb_udp_tunnel_segment(s
 		gso_inner_segment = skb_mac_gso_segment;
 		break;
 	case ENCAP_TYPE_IPPROTO:
+#if IS_ENABLED(CONFIG_INET) && IS_ENABLED(CONFIG_IPV6)
 		offloads = is_ipv6 ? inet6_offloads : inet_offloads;
+#else
+		offloads = is_ipv6 ? NULL : inet_offloads;
+#endif
+		if (!offloads)
+			goto out_unlock;
 		ops = rcu_dereference(offloads[skb->inner_ipproto]);
 		if (!ops || !ops->callbacks.gso_segment)
 			goto out_unlock;
