diff -drupN a/include/uapi/linux/netfilter/xt_socket.h b/include/uapi/linux/netfilter/xt_socket.h
--- a/include/uapi/linux/netfilter/xt_socket.h	2018-08-06 17:23:04.000000000 +0300
+++ b/include/uapi/linux/netfilter/xt_socket.h	2022-06-12 05:28:14.000000000 +0300
@@ -26,4 +26,11 @@ struct xt_socket_mtinfo3 {
 			   | XT_SOCKET_NOWILDCARD \
 			   | XT_SOCKET_RESTORESKMARK)
 
+struct sock *xt_socket_lookup_slow_v4(struct net *net,
+				      const struct sk_buff *skb,
+				      const struct net_device *indev);
+struct sock *xt_socket_lookup_slow_v6(struct net *net,
+				      const struct sk_buff *skb,
+				      const struct net_device *indev);
+
 #endif /* _XT_SOCKET_H */
