diff -drupN a/net/openvswitch/Kconfig b/net/openvswitch/Kconfig
--- a/net/openvswitch/Kconfig	2018-08-06 17:23:04.000000000 +0300
+++ b/net/openvswitch/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -11,6 +11,7 @@ config OPENVSWITCH
 				     (!NF_NAT_IPV4 || NF_NAT_IPV4) && \
 				     (!NF_NAT_IPV6 || NF_NAT_IPV6)))
 	select LIBCRC32C
+	select FLEX_ARRAY
 	select MPLS
 	select NET_MPLS_GSO
 	select DST_CACHE
