diff -drupN a/arch/arm/boot/dts/sun8iw16p1-pro.dts b/arch/arm/boot/dts/sun8iw16p1-pro.dts
--- a/arch/arm/boot/dts/sun8iw16p1-pro.dts	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/arm/boot/dts/sun8iw16p1-pro.dts	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,73 @@
+/*
+ * Allwinner Technology CO., Ltd. sun8iw16p1 soc board.
+ *
+ * soc board support.
+ */
+
+/dts-v1/;
+/* carveout used */
+/memreserve/ 0x43200000 0x04e00000; /* carveout space range       : [0x43200000~0x48000000], size = 78M */
+
+#include "sun8iw16p1.dtsi"
+
+/{
+	ion {
+		heap_carveout@0{
+			compatible = "allwinner,carveout";
+			heap-name  = "carveout";
+			heap-id    = <0x2>;
+			heap-base  = <0x43200000>;
+			heap-size  = <0x04e00000>;
+			heap-type  = "ion_carveout";
+		};
+	};
+
+	soc@03000000 {
+		pmu0: pmu@0{
+			interrupts = <GIC_SPI 104 IRQ_TYPE_LEVEL_HIGH>;
+			status = "okay";
+			wakeup-source;
+
+			powerkey0: powerkey@0{
+				status = "okay";
+			};
+
+			regulator0: regulator@0{
+				status = "okay";
+
+				regulator2 {
+					regulator-name = "vdd-cpua";
+				};
+
+				regulator9 {
+					regulator-name = "vcc-pll";
+				};
+
+			};
+
+			axp_gpio0: axp_gpio@0{
+				gpio-controller;
+				#size-cells = <0>;
+				#gpio-cells = <6>;
+				status = "okay";
+			};
+
+			charger0: charger@0{
+				status = "okay";
+			};
+		};
+	};
+
+	wlan:wlan {
+		compatible = "allwinner,sunxi-wlan";
+		clocks = <&clk_losc_out>;
+		wlan_busnum = <1>;
+		wlan_usbnum = <3>;
+		wlan_power;
+		wlan_io_regulator;
+		wlan_en;
+		wlan_regon;
+		wlan_hostwake;
+		status = "disabled";
+	};
+};
