diff -drupN a/drivers/soc/sunxi/arisc/interfaces/Makefile b/drivers/soc/sunxi/arisc/interfaces/Makefile
--- a/drivers/soc/sunxi/arisc/interfaces/Makefile	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/soc/sunxi/arisc/interfaces/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,6 @@
+#
+# Makefile for the linux kernel.
+#
+
+obj-y=arisc_dvfs.o arisc_standby.o arisc_rsb.o arisc_axp.o arisc_debug_level.o arisc_loopback.o arisc_dram_crc.o arisc_twi.o
+obj-y += arisc_sharemem.o
