diff -drupN a/modules/nand/sun50iw9p1/phy-nand/rawnand/controller/Makefile b/modules/nand/sun50iw9p1/phy-nand/rawnand/controller/Makefile
--- a/modules/nand/sun50iw9p1/phy-nand/rawnand/controller/Makefile	1970-01-01 03:00:00.000000000 +0300
+++ b/modules/nand/sun50iw9p1/phy-nand/rawnand/controller/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,14 @@
+#
+#
+# Makefile for controller
+#
+#
+
+obj-$(CONFIG_SUNXI_RAWNAND) += ndfc_base.o
+
+obj-$(CONFIG_SUNXI_RAWNAND) += ndfc_timings.o
+ifdef CONFIG_ARCH_SUN8IW18
+obj-$(CONFIG_SUNXI_RAWNAND) += ndfc_v2px.o
+else
+obj-$(CONFIG_SUNXI_RAWNAND) += ndfc_v1px.o
+endif
