diff -drupN a/security/Makefile b/security/Makefile
--- a/security/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/security/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -29,3 +29,8 @@ obj-$(CONFIG_CGROUP_DEVICE)		+= device_c
 # Object integrity file lists
 subdir-$(CONFIG_INTEGRITY)		+= integrity
 obj-$(CONFIG_INTEGRITY)			+= integrity/
+# File integrity check module
+obj-y					+= fivm.o
+ifeq ($(CONFIG_FILE_INTEGRITY),m)
+subdir-$(CONFIG_FILE_INTEGRITY) += fivm
+endif
