diff -drupN a/drivers/usb/gadget/udc/core.c b/drivers/usb/gadget/udc/core.c
--- a/drivers/usb/gadget/udc/core.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/usb/gadget/udc/core.c	2022-06-12 05:28:14.000000000 +0300
@@ -945,7 +945,11 @@ int usb_gadget_ep_match_desc(struct usb_
 		/* Bulk endpoints handle interrupt transfers,
 		 * except the toggle-quirky iso-synch kind
 		 */
+#if IS_ENABLED(CONFIG_USB_SUNXI_UDC0)
+		if (!ep->caps.type_int)
+#else
 		if (!ep->caps.type_int && !ep->caps.type_bulk)
+#endif
 			return 0;
 		/* INT:  limit 64 bytes full speed, 1024 high/super speed */
 		if (!gadget_is_dualspeed(gadget) && max > 64)
