diff -drupN a/drivers/usb/sunxi_usb/udc/sunxi_udc_config.h b/drivers/usb/sunxi_usb/udc/sunxi_udc_config.h
--- a/drivers/usb/sunxi_usb/udc/sunxi_udc_config.h	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/usb/sunxi_usb/udc/sunxi_udc_config.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,50 @@
+/*
+ * drivers/usb/sunxi_usb/udc/sunxi_udc_config.h
+ * (C) Copyright 2010-2015
+ * Allwinner Technology Co., Ltd. <www.allwinnertech.com>
+ * javen, 2010-3-3, create this file
+ *
+ * usb udc config.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of
+ * the License, or (at your option) any later version.
+ *
+ */
+
+#ifndef __SUNXI_UDC_CONFIG_H__
+#define __SUNXI_UDC_CONFIG_H__
+
+#include <linux/slab.h>
+#include <linux/list.h>
+#include <linux/interrupt.h>
+#include <linux/errno.h>
+#include <linux/clk.h>
+#include <linux/device.h>
+#include <linux/usb/ch9.h>
+
+#define  SW_UDC_DOUBLE_FIFO       /* double FIFO          */
+#define  SW_UDC_DMA
+
+/**
+ * only SUN8IW5 and later ic support inner dma,
+ * former ic(eg. SUN8IW1, SUN8IW3, SUN8IW2 etc) use outer dma.
+ */
+#ifdef SW_UDC_DMA
+#define  SW_UDC_DMA_INNER
+#endif
+
+#define  SW_UDC_HS_TO_FS          /* support HS to FS */
+#define  SW_UDC_DEBUG
+
+/* sw udc debug print */
+#if	0
+#define DMSG_DBG_UDC	DMSG_MSG
+#else
+#define DMSG_DBG_UDC(...)
+#endif
+
+#include  "../include/sunxi_usb_config.h"
+
+#endif /* __SUNXI_UDC_CONFIG_H__ */
