diff -drupN a/drivers/pinctrl/bcm/pinctrl-iproc-gpio.c b/drivers/pinctrl/bcm/pinctrl-iproc-gpio.c
--- a/drivers/pinctrl/bcm/pinctrl-iproc-gpio.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/pinctrl/bcm/pinctrl-iproc-gpio.c	2022-06-12 05:28:14.000000000 +0300
@@ -619,7 +619,7 @@ static int iproc_pin_config_set(struct p
 {
 	struct iproc_gpio *chip = pinctrl_dev_get_drvdata(pctldev);
 	enum pin_config_param param;
-	u16 arg;
+	u32 arg;
 	unsigned i, gpio = iproc_pin_to_gpio(pin);
 	int ret = -ENOTSUPP;
 
