diff -drupN a/drivers/watchdog/Kconfig b/drivers/watchdog/Kconfig
--- a/drivers/watchdog/Kconfig	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/watchdog/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -71,6 +71,14 @@ config SOFT_WATCHDOG
 	  To compile this driver as a module, choose M here: the
 	  module will be called softdog.
 
+config AXP2XX_WATCHDOG
+	tristate "axp2xx watchdog"
+	default y
+	depends on MFD_AXP2101
+	select WATCHDOG_CORE
+	help
+	  axp2xx serial watchdog. say Y to enablee it
+
 config DA9052_WATCHDOG
         tristate "Dialog DA9052 Watchdog"
         depends on PMIC_DA9052
@@ -462,6 +470,12 @@ config SUNXI_WATCHDOG
 	  To compile this driver as a module, choose M here: the
 	  module will be called sunxi_wdt.
 
+config SUNXI_WDOG_BOOTON
+	bool "Open sunxi wdog during initialization"
+	depends on SUNXI_WATCHDOG
+	help
+	 Say Y here to open wdog.
+
 config COH901327_WATCHDOG
 	bool "ST-Ericsson COH 901 327 watchdog"
 	depends on ARCH_U300
