diff -drupN a/drivers/tty/serial/Makefile b/drivers/tty/serial/Makefile
--- a/drivers/tty/serial/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/tty/serial/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -93,6 +93,7 @@ obj-$(CONFIG_SERIAL_STM32)	+= stm32-usar
 obj-$(CONFIG_SERIAL_MVEBU_UART)	+= mvebu-uart.o
 obj-$(CONFIG_SERIAL_PIC32)	+= pic32_uart.o
 obj-$(CONFIG_SERIAL_MPS2_UART)	+= mps2-uart.o
+obj-$(CONFIG_SERIAL_SUNXI)	+= sunxi-uart.o
 
 # GPIOLIB helpers for modem control lines
 obj-$(CONFIG_SERIAL_MCTRL_GPIO)	+= serial_mctrl_gpio.o
