diff -drupN a/drivers/char/sunxi_g2d/Makefile b/drivers/char/sunxi_g2d/Makefile
--- a/drivers/char/sunxi_g2d/Makefile	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/char/sunxi_g2d/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,16 @@
+#
+# Makefile for the kernel mixer device drivers.
+#
+
+include ${obj}/g2d_rcq/Makefile
+
+obj-$(CONFIG_SUNXI_G2D) += g2d_sunxi.o
+g2d_sunxi-${CONFIG_ARCH_SUN8IW12P1} += g2d_driver.o g2d_bsp_v2.o
+g2d_sunxi-${CONFIG_ARCH_SUN8IW17P1} += g2d_driver.o g2d_bsp_v2.o
+g2d_sunxi-${CONFIG_ARCH_SUN8IW11P1} += g2d_driver.o g2d_bsp_sun8iw11.o
+g2d_sunxi-${CONFIG_ARCH_SUN8IW15P1} += g2d_driver.o g2d_bsp_v2.o
+g2d_sunxi-${CONFIG_ARCH_SUN8IW16P1} += g2d_driver.o g2d_bsp_v2.o
+g2d_sunxi-${CONFIG_ARCH_SUN50IW9P1} += g2d_driver.o g2d_bsp_v2.o
+g2d_sunxi-${CONFIG_ARCH_SUN50IW10P1} += g2d_driver.o g2d_bsp_v2.o
+
+g2d_sunxi-$(CONFIG_ARCH_SUN8IW19P1) += ${rcq_obj}
