diff -drupN a/sound/soc/codecs/Kconfig b/sound/soc/codecs/Kconfig
--- a/sound/soc/codecs/Kconfig	2018-08-06 17:23:04.000000000 +0300
+++ b/sound/soc/codecs/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -1089,4 +1089,61 @@ config SND_SOC_TPA6130A2
 	tristate "Texas Instruments TPA6130A2 headphone amplifier"
 	depends on I2C
 
+# Sunxi
+config SND_SOC_ACX00
+        tristate "Sunxi ACX00 Codec"
+	select MFD_ACX00
+        depends on ARCH_SUN50IW6
+        default n
+        help
+          ACX00 now used as SUN50IW6 internal Codec, Connect Through I2S3.
+	  Say Y or M if you want to add support internal audio codec.
+
+config SND_SOC_AC100
+        tristate "Sunxi AC100 Codec"
+	select MFD_AC100
+        depends on ARCH_SUN8IW8
+        default n
+        help
+          ACX00 now used as SUN8IW8 External Codec, Connect Through I2S0.
+	  Say Y or M if you want to add support internal audio codec.
+
+config SND_SOC_AC102
+        tristate "Sunxi AC102 Codec"
+        depends on ARCH_SUN8IW8
+        default n
+        help
+          AC102 now used as SUN8IW8 External Codec, Connect Through I2S0.
+	  Say Y or M if you want to add support capture.
+
+config SND_SOC_AC107
+	tristate "Sunxi AC107 Codec"
+	depends on I2C
+	default n
+	help
+	say Y or M if you want to add support for AC107.
+
+config SND_SOC_AC108
+        tristate "Sunxi AC108 Codec"
+        depends on I2C
+        default n
+        help
+          AC108 now used as SUN8IW18 external codec, Connect Through I2S.
+          Say Y or M if you want to add support external audio codec.
+
+# add by Sunxi
+config SND_SOC_AD82584F
+        tristate "ad82584f codec"
+        depends on ARCH_SUN8IW15 || ARCH_SUN8IW8
+        default n
+        help
+          Say Y or M if you want to add support ad82584f.
+
+config SND_SOC_ES7243
+	tristate "ES7243 Codec"
+	depends on ARCH_SUN8IW8 || ARCH_SUN8IW18
+	default n
+	help
+	  Say Y or M if you want to add support for ES7243.
+
 endmenu
