diff -drupN a/drivers/net/phy/Kconfig b/drivers/net/phy/Kconfig
--- a/drivers/net/phy/Kconfig	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/net/phy/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -5,6 +5,7 @@
 menuconfig PHYLIB
 	tristate "PHY Device support and infrastructure"
 	depends on NETDEVICES
+	select NET_ETHTOOL
 	help
 	  Ethernet controllers are usually attached to PHY
 	  devices.  This option provides infrastructure for
@@ -149,6 +150,10 @@ config MDIO_XGENE
 
 comment "MII PHY device drivers"
 
+config SUNXI_EPHY
+	tristate "Drivers for Allwinnertech EPHY"
+	depends on MFD_ACX00
+
 config AMD_PHY
 	tristate "AMD PHYs"
 	---help---
