diff -drupN a/drivers/thermal/sunxi_thermal/sunxi_thermal_sensor/Kconfig b/drivers/thermal/sunxi_thermal/sunxi_thermal_sensor/Kconfig
--- a/drivers/thermal/sunxi_thermal/sunxi_thermal_sensor/Kconfig	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/thermal/sunxi_thermal/sunxi_thermal_sensor/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,12 @@
+config SUNXI_THERMAL_SENSOR
+	tristate "Thermal sensor driver"
+	depends on ARCH_SUNXI
+	help
+	  Thermal sensor on SUNXI platform.
+
+config SUNXI_THS_DRIVER
+	tristate "Thermal sensor driver for SUNXI platform"
+	depends on SUNXI_THERMAL_SENSOR
+	help
+	  Temperature sensor on SUNXI platform. The thermal sensor driver used to
+	  get temp. It call the thermal driver core interface.
