diff -drupN a/drivers/video/fbdev/sunxi/disp2/hdmi2/hdmi_core/api/hdcp22/include/elliptic_system_types.h b/drivers/video/fbdev/sunxi/disp2/hdmi2/hdmi_core/api/hdcp22/include/elliptic_system_types.h
--- a/drivers/video/fbdev/sunxi/disp2/hdmi2/hdmi_core/api/hdcp22/include/elliptic_system_types.h	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/video/fbdev/sunxi/disp2/hdmi2/hdmi_core/api/hdcp22/include/elliptic_system_types.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,19 @@
+/*
+ * Allwinner SoCs hdmi2.0 driver.
+ *
+ * Copyright (C) 2016 Allwinner.
+ *
+ * This file is licensed under the terms of the GNU General Public
+ * License version 2.  This program is licensed "as is" without any
+ * warranty of any kind, whether express or implied.
+ */
+#ifndef __ELLIPTIC_SYSTEM_TYPE_H__
+#define __ELLIPTIC_SYSTEM_TYPE_H__
+
+/* System types definitions. */
+#define ELP_STATUS                      int16_t
+#define ELP_SYSTEM_INFO                 int16_t
+#define ELP_SYSTEM_BIG_ENDIAN           0
+#define ELP_SYSTEM_LITLLE_ENDIAN        1
+
+#endif
