diff -drupN a/drivers/w1/masters/Makefile b/drivers/w1/masters/Makefile
--- a/drivers/w1/masters/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/w1/masters/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -8,5 +8,7 @@ obj-$(CONFIG_W1_MASTER_DS2482)		+= ds248
 obj-$(CONFIG_W1_MASTER_MXC)		+= mxc_w1.o
 
 obj-$(CONFIG_W1_MASTER_DS1WM)		+= ds1wm.o
+obj-$(CONFIG_W1_MASTER_SUNXI)		+= sunxi-owc.o
 obj-$(CONFIG_W1_MASTER_GPIO)		+= w1-gpio.o
 obj-$(CONFIG_HDQ_MASTER_OMAP)		+= omap_hdq.o
+
