diff -drupN a/modules/nand/sun50iw9p1/aw_nand_type.h b/modules/nand/sun50iw9p1/aw_nand_type.h
--- a/modules/nand/sun50iw9p1/aw_nand_type.h	1970-01-01 03:00:00.000000000 +0300
+++ b/modules/nand/sun50iw9p1/aw_nand_type.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,193 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+#ifndef __AW_NAND_TYPE_H__
+#define __AW_NAND_TYPE_H__
+
+#ifndef __s8
+typedef signed char __s8;
+#endif
+#ifndef _s8
+typedef signed char _s8;
+#endif
+#ifndef s8
+typedef signed char s8;
+#endif
+#ifndef S8
+typedef signed char S8;
+#endif
+#ifndef sint8
+typedef signed char sint8;
+#endif
+#ifndef SINT8
+typedef signed char SINT8;
+#endif
+#ifndef int8
+typedef signed char int8;
+#endif
+#ifndef INT8
+typedef signed char INT8;
+#endif
+#ifndef __u8
+typedef unsigned char __u8;
+#endif
+#ifndef _u8
+typedef unsigned char _u8;
+#endif
+#ifndef u8
+typedef unsigned char u8;
+#endif
+#ifndef U8
+typedef unsigned char U8;
+#endif
+#ifndef uint8
+typedef unsigned char uint8;
+#endif
+#ifndef UINT8
+typedef unsigned char UINT8;
+#endif
+#ifndef uchar
+typedef unsigned char uchar;
+#endif
+#ifndef UCHAR
+typedef unsigned char UCHAR;
+#endif
+#ifndef __s16
+typedef signed short __s16;
+#endif
+#ifndef _s16
+typedef signed short _s16;
+#endif
+#ifndef s16
+typedef signed short s16;
+#endif
+#ifndef S16
+typedef signed short S16;
+#endif
+#ifndef sint16
+typedef signed short sint16;
+#endif
+#ifndef int16
+typedef signed short int16;
+#endif
+#ifndef INT16
+typedef signed short INT16;
+#endif
+#ifndef __u16
+typedef unsigned short __u16;
+#endif
+#ifndef _u16
+typedef unsigned short _u16;
+#endif
+#ifndef u16
+typedef unsigned short u16;
+#endif
+#ifndef U16
+typedef unsigned short U16;
+#endif
+#ifndef uint16
+typedef unsigned short uint16;
+#endif
+#ifndef UINT16
+typedef unsigned short UINT16;
+#endif
+#ifndef __s32
+typedef signed int __s32;
+#endif
+#ifndef _s32
+typedef signed int _s32;
+#endif
+#ifndef s32
+typedef signed int s32;
+#endif
+#ifndef S32
+typedef signed int S32;
+#endif
+#ifndef int32
+typedef signed int int32;
+#endif
+#ifndef INT32
+typedef signed int INT32;
+#endif
+#ifndef sint32
+typedef signed int sint32;
+#endif
+#ifndef SINT32
+typedef signed int SINT32;
+#endif
+#ifndef __u32
+typedef unsigned int __u32;
+#endif
+#ifndef _u32
+typedef unsigned int _u32;
+#endif
+#ifndef u32
+typedef unsigned int u32;
+#endif
+#ifndef U32
+typedef unsigned int U32;
+#endif
+#ifndef uint32
+typedef unsigned int uint32;
+#endif
+#ifndef UINT32
+typedef unsigned int UINT32;
+#endif
+#ifndef __s64
+typedef signed long long __s64;
+#endif
+#ifndef _s64
+typedef signed long long _s64;
+#endif
+#ifndef s64
+typedef signed long long s64;
+#endif
+#ifndef S64
+typedef signed long long S64;
+#endif
+#ifndef sint64
+typedef signed long long sint64;
+#endif
+#ifndef SINT64
+typedef signed long long SINT64;
+#endif
+#ifndef int64
+typedef signed long long int64;
+#endif
+#ifndef INT64
+typedef signed long long INT64;
+#endif
+#ifndef __u64
+typedef unsigned long long __u64;
+#endif
+#ifndef _u64
+typedef unsigned long long _u64;
+#endif
+#ifndef u64
+typedef unsigned long long u64;
+#endif
+#ifndef U64
+typedef unsigned long long U64;
+#endif
+#ifndef uint64
+typedef unsigned long long uint64;
+#endif
+#ifndef UINT64
+typedef unsigned long long UINT64;
+#endif
+
+#ifndef size_t
+typedef unsigned long size_t;
+#endif
+
+#ifndef __kernel_loff_t
+typedef long long	__kernel_loff_t;
+#endif
+
+#ifndef loff_t
+typedef __kernel_loff_t		loff_t;
+#endif
+
+#ifndef NULL
+#define NULL ((void *)0)
+#endif
+
+#endif /*AW_NAND_TYPE_H*/
