diff -drupN a/drivers/video/fbdev/sunxi/disp2/disp/lcd/t30p106.h b/drivers/video/fbdev/sunxi/disp2/disp/lcd/t30p106.h
--- a/drivers/video/fbdev/sunxi/disp2/disp/lcd/t30p106.h	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/video/fbdev/sunxi/disp2/disp/lcd/t30p106.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,19 @@
+/* drivers/video/sunxi/disp2/disp/lcd/he0801a068.h
+ *
+ * Copyright (c) 2017 Allwinnertech Co., Ltd.
+ * Author: zhengxiaobin <zhengxiaobin@allwinnertech.com>
+ *
+ * he0801a-068 panel driver
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+*/
+#ifndef __T30P106_H_
+#define __T30P106_H_
+
+#include "panels.h"
+
+extern struct __lcd_panel t30p106_panel;
+
+#endif /*End of file*/
