diff -drupN a/drivers/base/regmap/regmap-debugfs.c b/drivers/base/regmap/regmap-debugfs.c
--- a/drivers/base/regmap/regmap-debugfs.c	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/base/regmap/regmap-debugfs.c	2022-06-12 05:28:14.000000000 +0300
@@ -269,7 +269,7 @@ static ssize_t regmap_map_read_file(stru
 				   count, ppos);
 }
 
-#undef REGMAP_ALLOW_WRITE_DEBUGFS
+#define REGMAP_ALLOW_WRITE_DEBUGFS
 #ifdef REGMAP_ALLOW_WRITE_DEBUGFS
 /*
  * This can be dangerous especially when we have clients such as
