diff -drupN a/include/linux/kernel.h b/include/linux/kernel.h
--- a/include/linux/kernel.h	2018-08-06 17:23:04.000000000 +0300
+++ b/include/linux/kernel.h	2022-06-12 05:28:14.000000000 +0300
@@ -453,6 +453,8 @@ extern int sysctl_panic_on_stackoverflow
 
 extern bool crash_kexec_post_notifiers;
 
+extern int sunxi_dump;
+
 /*
  * panic_cpu is used for synchronizing panic() and crash_kexec() execution. It
  * holds a CPU number which is executing panic() currently. A value of
