diff -drupN a/modules/nand/sun50iw1p1/nfd/nand_dev.h b/modules/nand/sun50iw1p1/nfd/nand_dev.h
--- a/modules/nand/sun50iw1p1/nfd/nand_dev.h	1970-01-01 03:00:00.000000000 +0300
+++ b/modules/nand/sun50iw1p1/nfd/nand_dev.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,17 @@
+
+#ifndef __NAND_DEV_H__
+#define __NAND_DEV_H__
+
+struct nand_kobject {
+	struct kobject kobj;
+	struct _nftl_blk *nftl_blk;
+	char name[32];
+};
+
+extern unsigned int do_static_wear_leveling(void *zone);
+extern unsigned int static_wear_leveling(void *zone);
+extern unsigned short nftl_get_zone_write_cache_nums(void *_zone);
+extern unsigned int garbage_collect(void *zone);
+extern unsigned int do_prio_gc(void *zone);
+
+#endif
\ В конце файла нет новой строки
