diff -drupN a/include/linux/pagemap.h b/include/linux/pagemap.h
--- a/include/linux/pagemap.h	2018-08-06 17:23:04.000000000 +0300
+++ b/include/linux/pagemap.h	2022-06-12 05:28:14.000000000 +0300
@@ -225,7 +225,7 @@ static inline gfp_t readahead_gfp_mask(s
 				  __GFP_COLD | __GFP_NORETRY | __GFP_NOWARN;
 }
 
-typedef int filler_t(void *, struct page *);
+typedef int filler_t(struct file *, struct page *);
 
 pgoff_t page_cache_next_hole(struct address_space *mapping,
 			     pgoff_t index, unsigned long max_scan);
@@ -369,7 +369,7 @@ extern int read_cache_pages(struct addre
 static inline struct page *read_mapping_page(struct address_space *mapping,
 				pgoff_t index, void *data)
 {
-	filler_t *filler = (filler_t *)mapping->a_ops->readpage;
+	filler_t *filler = mapping->a_ops->readpage;
 	return read_cache_page(mapping, index, filler, data);
 }
 
