diff -drupN a/drivers/mtd/aw-spinand/Makefile b/drivers/mtd/aw-spinand/Makefile
--- a/drivers/mtd/aw-spinand/Makefile	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/mtd/aw-spinand/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,15 @@
+#
+# Makefile for the SPINAND MTD
+#
+
+obj-$(CONFIG_AW_MTD_SPINAND) += aw-ubi-spinand.o
+
+aw-ubi-spinand-objs += sunxi-core.o sunxi-debug.o
+
+#obj-$(CONFIG_AW_NFTL_SPINAND) += aw-nftl-spinand.o
+
+#aw-nftl-spinand-objs += sunxi-common.o sunxi-nftl-core.o
+
+obj-$(CONFIG_AW_SPINAND_PHYSICAL_LAYER) += physic/
+
+obj-$(CONFIG_AW_SPINAND_SECURE_STORAGE) += secure-storage.o
