diff -drupN a/include/linux/task_io_accounting_ops.h b/include/linux/task_io_accounting_ops.h
--- a/include/linux/task_io_accounting_ops.h	2018-08-06 17:23:04.000000000 +0300
+++ b/include/linux/task_io_accounting_ops.h	2022-06-12 05:28:14.000000000 +0300
@@ -96,6 +96,7 @@ static inline void task_chr_io_accountin
 	dst->wchar += src->wchar;
 	dst->syscr += src->syscr;
 	dst->syscw += src->syscw;
+	dst->syscfs += src->syscfs;
 }
 #else
 static inline void task_chr_io_accounting_add(struct task_io_accounting *dst,
