diff -drupN a/drivers/pwm/Makefile b/drivers/pwm/Makefile
--- a/drivers/pwm/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/pwm/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -46,3 +46,7 @@ obj-$(CONFIG_PWM_TIPWMSS)	+= pwm-tipwmss
 obj-$(CONFIG_PWM_TWL)		+= pwm-twl.o
 obj-$(CONFIG_PWM_TWL_LED)	+= pwm-twl-led.o
 obj-$(CONFIG_PWM_VT8500)	+= pwm-vt8500.o
+obj-$(CONFIG_PWM_SUNXI)		+= pwm-sunxi.o
+obj-$(CONFIG_PWM_SUNXI_NEW)	+= pwm-sunxi-new.o
+obj-$(CONFIG_PWM_SUNXI_GROUP)	+= pwm-sunxi-group.o
+obj-$(CONFIG_PWM_SUNXI_GROUP)	+= pwm-sunxi-g_dev.o
\ В конце файла нет новой строки
