diff -drupN a/drivers/media/platform/sunxi-vin/modules/sensor/nvp6324/i2c/Makefile b/drivers/media/platform/sunxi-vin/modules/sensor/nvp6324/i2c/Makefile
--- a/drivers/media/platform/sunxi-vin/modules/sensor/nvp6324/i2c/Makefile	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/media/platform/sunxi-vin/modules/sensor/nvp6324/i2c/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,65 @@
+#########################################################################
+
+SUB_DIR = `find ../ -type d`
+
+#########################################################################
+#	Toolchain.
+#########################################################################
+
+CROSS_COMPILE := /opt/freescale/usr/local/gcc-4.6.2-glibc-2.13-linaro-multilib-2011.12/fsl-linaro-toolchain/bin/arm-fsl-linux-gnueabi-
+
+CC:= $(CROSS_COMPILE)gcc
+CPP:= $(CROSS_COMPILE)g++
+AR:= $(CROSS_COMPILE)ar
+AS:= $(CROSS_COMPILE)as
+LD:= $(CROSS_COMPILE)ld
+NM:= $(CROSS_COMPILE)nm
+RANLIB:= $(CROSS_COMPILE)ranlib
+OBJCOPY:= $(CROSS_COMPILE)objcopy
+STRIP:= $(CROSS_COMPILE)strip
+
+######################################################################
+# NFS path
+NFS_PATH = /nfs/$(shell whoami)/imx6
+
+# Build options
+######################################################################
+# Target
+DDOBJS += i2c_dd.c \
+	strfunc.c \
+
+DECOBJS += i2c_dec.c \
+	strfunc.c \
+
+DD := dd
+DEC := dec
+
+######################################################################
+# Build
+
+all:
+	$(Q)$(CC) -g -Wall -o $(DD) $(DDOBJS)
+	$(Q)$(CC) -g -Wall -o $(DEC) $(DECOBJS)
+
+clean:
+	@for SDIR in $(SUB_DIR) ; do	\
+		rm -f $${SDIR}/*.o;			\
+		rm -f $${SDIR}/*.bak;		\
+		rm -f *.o;		\
+	done;
+
+	@rm -f $(DD)
+	@rm -f $(DEC)
+
+	@echo "Cleaning......   [OK]"
+
+install: all
+	cp -a $(DD) $(NFS_PATH)
+	cp -a $(DEC) $(NFS_PATH)
+
+#########################################################################
+# Dependency
+ifeq (.depend,$(wildcard .depend))
+include .depend
+endif
+
