diff -drupN a/drivers/video/fbdev/sunxi/disp2/disp/de/lowlevel_sun8iw8/Makefile b/drivers/video/fbdev/sunxi/disp2/disp/de/lowlevel_sun8iw8/Makefile
--- a/drivers/video/fbdev/sunxi/disp2/disp/de/lowlevel_sun8iw8/Makefile	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/video/fbdev/sunxi/disp2/disp/de/lowlevel_sun8iw8/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,15 @@
+#
+# Copyright (c) 2015 Allwinnertech, Inc.
+# Tyle <tyle at allwinnertech.com>
+#
+#
+
+mydir = de/lowlevel_sun8iw8
+
+obj_low = $(mydir)/de_fcc.o $(mydir)/de_gsu.o $(mydir)/de_hal.o $(mydir)/disp_al.o \
+					$(mydir)/de_peak.o $(mydir)/de_rtmx.o $(mydir)/de_scaler_table.o\
+					$(mydir)/de_vep_table.o $(mydir)/de_vsu.o $(mydir)/de_feat.o \
+					$(mydir)/de_lcd.o $(mydir)/de_lti.o $(mydir)/de_fce.o \
+					$(mydir)/de_ase.o $(mydir)/de_bws.o $(mydir)/de_enhance.o \
+					$(mydir)/de_smbl.o $(mydir)/de_ccsc.o $(mydir)/de_dcsc.o \
+					$(mydir)/de_wb.o $(mydir)/de_clock.o
