diff -drupN a/drivers/power/supply/axp/axp803/axp803-regulator.h b/drivers/power/supply/axp/axp803/axp803-regulator.h
--- a/drivers/power/supply/axp/axp803/axp803-regulator.h	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/power/supply/axp/axp803/axp803-regulator.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,92 @@
+/*
+ * drivers/power/axp/axp803/axp803-regu.h
+ * (C) Copyright 2010-2016
+ * Allwinner Technology Co., Ltd. <www.allwinnertech.com>
+ * Pannan <pannan@allwinnertech.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of
+ * the License, or (at your option) any later version.
+ */
+
+#ifndef AXP803_REGULATOR_H
+#define AXP803_REGULATOR_H
+
+enum {
+	AXP803_ID_LDO1,   /* RTCLDO */
+	AXP803_ID_LDO2,   /* ALDO1 */
+	AXP803_ID_LDO3,   /* ALDO2 */
+	AXP803_ID_LDO4,   /* ALDO3 */
+	AXP803_ID_LDO5,   /* DLDO1 */
+	AXP803_ID_LDO6,   /* DLDO2 */
+	AXP803_ID_LDO7,   /* DLDO3 */
+	AXP803_ID_LDO8,   /* DLDO4 */
+	AXP803_ID_LDO9,   /* ELDO1 */
+	AXP803_ID_LDO10,  /* ELDO2 */
+	AXP803_ID_LDO11,  /* ELDO3 */
+	AXP803_ID_LDO12,  /* FLDO1 */
+	AXP803_ID_LDO13,  /* FLDO2 */
+	AXP803_ID_LDOIO0 = AXP_LDOIO_ID_START,
+	AXP803_ID_LDOIO1,
+	AXP803_ID_SW1,
+	AXP803_ID_DCDC1 = AXP_DCDC_ID_START,
+	AXP803_ID_DCDC2,
+	AXP803_ID_DCDC3,
+	AXP803_ID_DCDC4,
+	AXP803_ID_DCDC5,
+	AXP803_ID_DCDC6,
+	AXP803_ID_DCDC7,
+};
+
+#define AXP803_DCDC1         AXP803_DC1OUT_VOL
+#define AXP803_DCDC2         AXP803_DC2OUT_VOL
+#define AXP803_DCDC3         AXP803_DC3OUT_VOL
+#define AXP803_DCDC4         AXP803_DC4OUT_VOL
+#define AXP803_DCDC5         AXP803_DC5OUT_VOL
+#define AXP803_DCDC6         AXP803_DC6OUT_VOL
+#define AXP803_DCDC7         AXP803_DC7OUT_VOL
+#define AXP803_LDO1          AXP803_STATUS
+#define AXP803_LDO2          AXP803_ALDO1OUT_VOL
+#define AXP803_LDO3          AXP803_ALDO2OUT_VOL
+#define AXP803_LDO4          AXP803_ALDO3OUT_VOL
+#define AXP803_LDO5          AXP803_DLDO1OUT_VOL
+#define AXP803_LDO6          AXP803_DLDO2OUT_VOL
+#define AXP803_LDO7          AXP803_DLDO3OUT_VOL
+#define AXP803_LDO8          AXP803_DLDO4OUT_VOL
+#define AXP803_LDO9          AXP803_ELDO1OUT_VOL
+#define AXP803_LDO10         AXP803_ELDO2OUT_VOL
+#define AXP803_LDO11         AXP803_ELDO3OUT_VOL
+#define AXP803_LDO12         AXP803_FLDO1OUT_VOL
+#define AXP803_LDO13         AXP803_FLDO2OUT_VOL
+#define AXP803_LDOIO0        AXP803_GPIO0LDOOUT_VOL
+#define AXP803_LDOIO1        AXP803_GPIO1LDOOUT_VOL
+#define AXP803_DC1SW         AXP803_DC1OUT_VOL
+
+#define AXP803_DCDC1EN       AXP803_LDO_DC_EN1
+#define AXP803_DCDC2EN       AXP803_LDO_DC_EN1
+#define AXP803_DCDC3EN       AXP803_LDO_DC_EN1
+#define AXP803_DCDC4EN       AXP803_LDO_DC_EN1
+#define AXP803_DCDC5EN       AXP803_LDO_DC_EN1
+#define AXP803_DCDC6EN       AXP803_LDO_DC_EN1
+#define AXP803_DCDC7EN       AXP803_LDO_DC_EN1
+#define AXP803_LDO1EN        AXP803_STATUS
+#define AXP803_LDO2EN        AXP803_LDO_DC_EN3
+#define AXP803_LDO3EN        AXP803_LDO_DC_EN3
+#define AXP803_LDO4EN        AXP803_LDO_DC_EN3
+#define AXP803_LDO5EN        AXP803_LDO_DC_EN2
+#define AXP803_LDO6EN        AXP803_LDO_DC_EN2
+#define AXP803_LDO7EN        AXP803_LDO_DC_EN2
+#define AXP803_LDO8EN        AXP803_LDO_DC_EN2
+#define AXP803_LDO9EN        AXP803_LDO_DC_EN2
+#define AXP803_LDO10EN       AXP803_LDO_DC_EN2
+#define AXP803_LDO11EN       AXP803_LDO_DC_EN2
+#define AXP803_LDO12EN       AXP803_LDO_DC_EN3
+#define AXP803_LDO13EN       AXP803_LDO_DC_EN3
+#define AXP803_LDOIO0EN      AXP803_GPIO0_CTL
+#define AXP803_LDOIO1EN      AXP803_GPIO1_CTL
+#define AXP803_DC1SWEN       AXP803_LDO_DC_EN2
+#define AXP803_BUCKMODE      AXP803_DCDC_MODESET
+#define AXP803_BUCKFREQ      AXP803_DCDC_FREQSET
+
+#endif /* AXP803_REGULATOR_H */
