diff -drupN a/drivers/block/Kconfig b/drivers/block/Kconfig
--- a/drivers/block/Kconfig	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/block/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -15,6 +15,8 @@ menuconfig BLK_DEV
 
 if BLK_DEV
 
+source "drivers/block/panicpart/Kconfig"
+
 config BLK_DEV_NULL_BLK
 	tristate "Null test block driver"
 
@@ -110,6 +112,7 @@ source "drivers/block/mtip32xx/Kconfig"
 
 source "drivers/block/zram/Kconfig"
 
+source "modules/nand/Kconfig"
 config BLK_CPQ_CISS_DA
 	tristate "Compaq Smart Array 5xxx support"
 	depends on PCI
