diff -drupN a/arch/sparc/include/uapi/asm/socket.h b/arch/sparc/include/uapi/asm/socket.h
--- a/arch/sparc/include/uapi/asm/socket.h	2018-08-06 17:23:04.000000000 +0300
+++ b/arch/sparc/include/uapi/asm/socket.h	2022-06-12 05:28:14.000000000 +0300
@@ -86,6 +86,8 @@
 
 #define SO_CNX_ADVICE		0x0037
 
+#define SO_COOKIE		0x003b
+
 /* Security levels - as per NRL IPv6 - don't actually do anything */
 #define SO_SECURITY_AUTHENTICATION		0x5001
 #define SO_SECURITY_ENCRYPTION_TRANSPORT	0x5002
