diff -drupN a/drivers/media/platform/sunxi-vin/modules/sensor/nvp6324/Makefile b/drivers/media/platform/sunxi-vin/modules/sensor/nvp6324/Makefile
--- a/drivers/media/platform/sunxi-vin/modules/sensor/nvp6324/Makefile	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/media/platform/sunxi-vin/modules/sensor/nvp6324/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,8 @@
+obj-m := nvp6324.o
+nvp6324-objs := nvp6324_mipi.o mipi_dev_nvp6324.o jaguar1_video.o jaguar1_coax_protocol.o jaguar1_motion.o jaguar1_video_eq.o jaguar1_mipi.o
+#nvp6324-objs := nvp6324_mipi.o mipi_dev_nvp6324.o jaguar1_i2c.o jaguar1_video.o jaguar1_coax_protocol.o jaguar1_motion.o jaguar1_video_eq.o jaguar1_mipi.o
+
+clean:
+	@make -C $(KERNEL_PATH) M=$(PWD) clean
+
+
