diff -drupN a/arch/arm/boot/dts/Makefile b/arch/arm/boot/dts/Makefile
--- a/arch/arm/boot/dts/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/arch/arm/boot/dts/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -1,3 +1,5 @@
+dtstree		:= $(srctree)/$(src)
+
 ifeq ($(CONFIG_OF),y)
 
 dtb-$(CONFIG_ARCH_ALPINE) += \
@@ -832,6 +834,34 @@ dtb-$(CONFIG_MACH_SUN8I) += \
 dtb-$(CONFIG_MACH_SUN9I) += \
 	sun9i-a80-optimus.dtb \
 	sun9i-a80-cubieboard4.dtb
+dtb-$(CONFIG_ARCH_SUN8IW6P1) += \
+	sun8iw6p1-soc.dtb
+dtb-$(CONFIG_ARCH_SUN8IW7P1) += \
+	sun8iw7p1-soc.dtb
+dtb-$(CONFIG_ARCH_SUN8IW8P1) += \
+	board.dtb
+dtb-$(CONFIG_ARCH_SUN8IW11P1) += \
+	sun8iw11p1-soc.dtb \
+	sun8iw11p1-v40_perf3_v1_0_axpv1.dtb
+dtb-$(CONFIG_ARCH_SUN8IW12P1) += \
+	sun8iw12p1-soc.dtb
+dtb-$(CONFIG_ARCH_SUN8IW15P1) += \
+	board.dtb
+
+ifneq (strip $(CONFIG_ARCH_SUN8IW16P1),)
+DTS_SRC := $(patsubst $(dtstree)/sun8iw16p1-%.dts,sun8iw16p1-%.dtb, $(wildcard $(dtstree)/sun8iw16p1-*.dts))
+dtb-$(CONFIG_ARCH_SUN8IW16P1) += \
+	sun8iw16p1-soc.dtb \
+	$(DTS_SRC)
+endif
+dtb-$(CONFIG_ARCH_SUN8IW17P1) += \
+	board.dtb
+dtb-$(CONFIG_ARCH_SUN8IW18P1) += \
+	board.dtb
+
+dtb-$(CONFIG_ARCH_SUN8IW19P1) += \
+	board.dtb
+
 dtb-$(CONFIG_ARCH_TANGO) += \
 	tango4-vantage-1172.dtb
 dtb-$(CONFIG_ARCH_TEGRA_2x_SOC) += \
@@ -957,8 +987,17 @@ dtb-$(CONFIG_ARCH_ASPEED) += aspeed-bmc-
 	aspeed-ast2500-evb.dtb
 endif
 
-dtstree		:= $(srctree)/$(src)
 dtb-$(CONFIG_OF_ALL_DTBS) := $(patsubst $(dtstree)/%.dts,%.dtb, $(wildcard $(dtstree)/*.dts))
 
-always		:= $(dtb-y)
+DTB_NAMES := $(subst $\",,$(CONFIG_BUILD_ARM_APPENDED_DTB_IMAGE_NAMES))
+ifneq ($(DTB_NAMES),)
+DTB_LIST := $(addsuffix .dtb,$(DTB_NAMES))
+else
+DTB_LIST := $(dtb-y)
+endif
+
+targets += dtbs dtbs_install
+targets += $(DTB_LIST)
+
+always		:= $(DTB_LIST)
 clean-files	:= *.dtb
