diff -drupN a/drivers/staging/android/ion/compat_ion.h b/drivers/staging/android/ion/compat_ion.h
--- a/drivers/staging/android/ion/compat_ion.h	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/staging/android/ion/compat_ion.h	2022-06-12 05:28:14.000000000 +0300
@@ -21,6 +21,13 @@
 
 long compat_ion_ioctl(struct file *filp, unsigned int cmd, unsigned long arg);
 
+#define COMPAT_ION_IOC_SUNXI_FLUSH_RANGE       5
+
+typedef struct {
+		compat_long_t   start;
+		compat_long_t   end;
+} compat_sunxi_cache_range;
+
 #else
 
 #define compat_ion_ioctl  NULL
