diff -drupN a/drivers/usb/gadget/function/f_mass_storage.h b/drivers/usb/gadget/function/f_mass_storage.h
--- a/drivers/usb/gadget/function/f_mass_storage.h	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/usb/gadget/function/f_mass_storage.h	2022-06-12 05:28:14.000000000 +0300
@@ -4,6 +4,16 @@
 #include <linux/usb/composite.h>
 #include "storage_common.h"
 
+#if IS_ENABLED(CONFIG_USB_SUNXI_UDC0)
+#define DO_VFS_STAT              1
+#define DO_VFS_END               2
+
+extern atomic_t vfs_read_flag;
+extern atomic_t vfs_write_flag;
+extern unsigned int vfs_amount;
+extern loff_t vfs_file_offset;
+#endif
+
 struct fsg_module_parameters {
 	char		*file[FSG_MAX_LUNS];
 	bool		ro[FSG_MAX_LUNS];
