diff -drupN a/drivers/media/platform/sunxi-eve/Makefile b/drivers/media/platform/sunxi-eve/Makefile
--- a/drivers/media/platform/sunxi-eve/Makefile	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/media/platform/sunxi-eve/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,5 @@
+#
+# Makefile for the embedded vision engine device.
+#
+obj-$(CONFIG_SUNXI_EVE) += sunxi_eve.o
+sunxi_eve-objs := eve.o
\ В конце файла нет новой строки
