diff -drupN a/include/net/netfilter/nf_conntrack.h b/include/net/netfilter/nf_conntrack.h
--- a/include/net/netfilter/nf_conntrack.h	2018-08-06 17:23:04.000000000 +0300
+++ b/include/net/netfilter/nf_conntrack.h	2022-06-12 05:28:14.000000000 +0300
@@ -306,7 +306,7 @@ static inline bool nf_ct_should_gc(const
 
 struct kernel_param;
 
-int nf_conntrack_set_hashsize(const char *val, struct kernel_param *kp);
+int nf_conntrack_set_hashsize(const char *val, const struct kernel_param *kp);
 int nf_conntrack_hash_resize(unsigned int hashsize);
 
 extern struct hlist_nulls_head *nf_conntrack_hash;
