diff -drupN a/net/ipv6/syncookies.c b/net/ipv6/syncookies.c
--- a/net/ipv6/syncookies.c	2018-08-06 17:23:04.000000000 +0300
+++ b/net/ipv6/syncookies.c	2022-06-12 05:28:14.000000000 +0300
@@ -228,6 +228,7 @@ struct sock *cookie_v6_check(struct sock
 		fl6.flowi6_mark = ireq->ir_mark;
 		fl6.fl6_dport = ireq->ir_rmt_port;
 		fl6.fl6_sport = inet_sk(sk)->inet_sport;
+		fl6.flowi6_uid = sk->sk_uid;
 		security_req_classify_flow(req, flowi6_to_flowi(&fl6));
 
 		dst = ip6_dst_lookup_flow(sk, &fl6, final_p);
