diff -drupN a/drivers/mfd/Makefile b/drivers/mfd/Makefile
--- a/drivers/mfd/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/mfd/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -116,6 +116,9 @@ obj-$(CONFIG_PMIC_DA9052)	+= da9052-core
 obj-$(CONFIG_MFD_DA9052_SPI)	+= da9052-spi.o
 obj-$(CONFIG_MFD_DA9052_I2C)	+= da9052-i2c.o
 
+obj-$(CONFIG_MFD_AXP2101)	+= axp2101.o
+obj-$(CONFIG_MFD_AXP2101_I2C)	+= axp2101-i2c.o
+
 obj-$(CONFIG_MFD_AC100)		+= ac100.o
 obj-$(CONFIG_MFD_AXP20X)	+= axp20x.o
 obj-$(CONFIG_MFD_AXP20X_I2C)	+= axp20x-i2c.o
@@ -210,4 +213,9 @@ intel-soc-pmic-$(CONFIG_INTEL_PMC_IPC)	+
 obj-$(CONFIG_INTEL_SOC_PMIC)	+= intel-soc-pmic.o
 obj-$(CONFIG_MFD_MT6397)	+= mt6397-core.o
 
+# Sunxi
+obj-$(CONFIG_MFD_ACX00)		+= acx00-core.o
+obj-$(CONFIG_MFD_AC100)		+= ac100-core.o
+#obj-$(CONFIG_XP_MFD_AC100)	+= ac100.o
+
 obj-$(CONFIG_MFD_ALTERA_A10SR)	+= altera-a10sr.o
