diff -drupN a/drivers/rtc/Makefile b/drivers/rtc/Makefile
--- a/drivers/rtc/Makefile	2018-08-06 17:23:04.000000000 +0300
+++ b/drivers/rtc/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -148,7 +148,14 @@ obj-$(CONFIG_RTC_DRV_STMP)	+= rtc-stmp3x
 obj-$(CONFIG_RTC_DRV_ST_LPC)	+= rtc-st-lpc.o
 obj-$(CONFIG_RTC_DRV_SUN4V)	+= rtc-sun4v.o
 obj-$(CONFIG_RTC_DRV_SUN6I)	+= rtc-sun6i.o
+ifeq ($(CONFIG_ARCH_SUN50IW5T),y)
+obj-$(CONFIG_RTC_DRV_SUNXI)	+= rtc-sunxi-v2.o
+else ifeq ($(CONFIG_ARCH_SUN8IW18),y)
+obj-$(CONFIG_RTC_DRV_SUNXI)	+= rtc-sunxi-v3.o
+else
 obj-$(CONFIG_RTC_DRV_SUNXI)	+= rtc-sunxi.o
+endif
+obj-$(CONFIG_SUNXI_TIMER_ALARM)	+= sunxi_timer_alarm.o
 obj-$(CONFIG_RTC_DRV_TEGRA)	+= rtc-tegra.o
 obj-$(CONFIG_RTC_DRV_TEST)	+= rtc-test.o
 obj-$(CONFIG_RTC_DRV_TILE)	+= rtc-tile.o
