diff -drupN a/arch/arm64/boot/dts/sunxi/sun50iw10p1-pinctrl.dtsi b/arch/arm64/boot/dts/sunxi/sun50iw10p1-pinctrl.dtsi
--- a/arch/arm64/boot/dts/sunxi/sun50iw10p1-pinctrl.dtsi	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/arm64/boot/dts/sunxi/sun50iw10p1-pinctrl.dtsi	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,864 @@
+/*
+ * Allwinner sun50iw10p1 pin config info.
+ */
+
+/ {
+	soc@03000000{
+		r_pio: pinctrl@07022000 {
+			compatible = "allwinner,sun50iw10p1-r-pinctrl";
+			reg = <0x0 0x07022000 0x0 0x400>;
+			interrupts = <GIC_SPI 111 4>;
+			clocks = <&clk_cpurpio>;
+			device_type = "r_pio";
+			//resets = <&apb0_rst 0>;
+			gpio-controller;
+			interrupt-controller;
+			#interrupt-cells = <3>;
+			#size-cells = <0>;
+			#gpio-cells = <6>;
+
+			s_rsb0_pins_a: s_rsb0@0 {
+				allwinner,pins = "PL0", "PL1";
+				allwinner,function = "s_rsb0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <2>;
+				allwinner,pull = <1>;
+			};
+
+			s_uart0_pins_a: s_uart0@0 {
+				allwinner,pins = "PL2", "PL3";
+				allwinner,function = "s_uart0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			s_uart0_pins_b: s_uart0@1 {
+				allwinner,pins = "PL2", "PL3";
+				allwinner,function = "io_disabled";  /* io_disabled */
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+		};
+
+		pio: pinctrl@0300b000 {
+			compatible = "allwinner,sun50iw10p1-pinctrl";
+			reg = <0x0 0x0300b000 0x0 0x400>;
+			interrupts = <GIC_SPI 54 4>,  /* AW1855_GIC_Spec: GPIOB: 86-32=54 */
+						 <GIC_SPI 55 4>,
+						 <GIC_SPI 56 4>,
+						 <GIC_SPI 57 4>,
+						 <GIC_SPI 58 4>,
+						 <GIC_SPI 59 4>,
+						 <GIC_SPI 60 4>,
+						 <GIC_SPI 61 4>,
+						 <GIC_SPI 62 4>;
+			device_type = "pio";
+			clocks = <&clk_pio>;
+			gpio-controller;
+			interrupt-controller;
+			#interrupt-cells = <3>;
+			#size-cells = <0>;
+			#gpio-cells = <6>;
+
+			vdevice_pins_a: vdevice@0 {
+				allwinner,pins = "PB0", "PB1";
+				allwinner,function = "Vdevice";
+				allwinner,muxsel = <5>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+			/* FPGA stage: PB8 PB9, IC stage: PB9 PB10*/
+			uart0_pins_a: uart0@0 {
+				allwinner,pins = "PB8", "PB9";
+				allwinner,pname = "uart0_tx", "uart0_rx";
+				allwinner,function = "uart0";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart0_pins_b: uart0@1 {
+				allwinner,pins = "PB8", "PB9";
+				allwinner,function = "uart0";  /* io_disabled */
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			uart1_pins_a: uart1@0 {
+				allwinner,pins = "PG6", "PG7", "PG8", "PG9";
+				allwinner,pname = "uart1_tx", "uart1_rx",
+						  "uart1_rts", "uart1_cts";
+				allwinner,function = "uart1";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart1_pins_b: uart1@1 {
+				allwinner,pins = "PG6", "PG7", "PG8", "PG9";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			uart2_pins_a: uart2@0 {
+				allwinner,pins = "PJ20", "PJ21", "PJ22", "PJ23";
+				allwinner,pname = "uart2_tx", "uart2_rx",
+						  "uart2_rts", "uart2_cts";
+				allwinner,function = "uart2";
+				allwinner,muxsel = <5>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart2_pins_b: uart2@1 {
+				allwinner,pins = "PJ20", "PJ21", "PJ22", "PJ23";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			uart3_pins_a: uart3@0 {
+				allwinner,pins = "PH4", "PH5", "PH6", "PH7";
+				allwinner,pname = "uart3_tx", "uart3_rx",
+						  "uart3_rts", "uart3_cts";
+				allwinner,function = "uart3";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart3_pins_b: uart3@1 {
+				allwinner,pins = "PH4", "PH5", "PH6", "PH7";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			uart4_pins_a: uart4@0 {
+				allwinner,pins = "PD18", "PD19", "PD20", "PD21";
+				allwinner,pname = "uart4_tx", "uart4_rx",
+						  "uart4_rts", "uart4_cts";
+				allwinner,function = "uart4";
+				allwinner,muxsel = <5>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart4_pins_b: uart4@1 {
+				allwinner,pins = "PD18", "PD19", "PD20", "PD21";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			uart5_pins_a: uart5@0 {
+				allwinner,pins = "PI2", "PI3", "PI4", "PI5";
+				allwinner,pname = "uart5_tx", "uart5_rx",
+						  "uart5_rts", "uart5_cts";
+				allwinner,function = "uart5";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart5_pins_b: uart5@1 {
+				allwinner,pins = "PI2", "PI3", "PI4", "PI5";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			uart6_pins_a: uart6@0 {
+				allwinner,pins = "PI6", "PI7", "PI8", "PI12";
+				allwinner,pname = "uart6_tx", "uart6_rx",
+						  "uart6_rts", "uart6_cts";
+				allwinner,function = "uart6";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			uart6_pins_b: uart6@1 {
+				allwinner,pins = "PI6", "PI7", "PI8", "PI12";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			ir0_pins_a: ir0@0 {
+				allwinner,pins = "PH3";
+				allwinner,pname = "it-tx";
+				allwinner,function = "ir0";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			ir0_pins_b: ir0@1 {
+				allwinner,pins = "PH3";
+				allwinner,pname = "io_disabled";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi0_pins_a: twi0@0 {
+				allwinner,pins = "PB9", "PB10";
+				allwinner,pname = "twi0_scl", "twi0_sda";
+				allwinner,function = "twi0";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi0_pins_b: twi0@1 {
+				allwinner,pins = "PB9", "PB10";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi1_pins_a: twi1@0 {
+				allwinner,pins = "PB4", "PB5";
+				allwinner,pname = "twi1_scl", "twi1_sda";
+				allwinner,function = "twi1";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi1_pins_b: twi1@1 {
+				allwinner,pins = "PB4", "PB5";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi2_pins_a: twi2@0 {
+				allwinner,pins = "PE1", "PE2";
+				allwinner,pname = "twi2_scl", "twi2_sda";
+				allwinner,function = "twi2";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi2_pins_b: twi2@1 {
+				allwinner,pins = "PE1", "PE2";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi3_pins_a: twi3@0 {
+				allwinner,pins = "PE3", "PE4";
+				allwinner,pname = "twi3_scl", "twi3_sda";
+				allwinner,function = "twi3";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi3_pins_b: twi3@1 {
+				allwinner,pins = "PE3", "PE4";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi4_pins_a: twi4@0 {
+				allwinner,pins = "PI0", "PI1";
+				allwinner,pname = "twi4_scl", "twi4_sda";
+				allwinner,function = "twi4";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi4_pins_b: twi4@1 {
+				allwinner,pins = "PI0", "PI1";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi5_pins_a: twi5@0 {
+				allwinner,pins = "PI8", "PI9";
+				allwinner,pname = "twi5_scl", "twi5_sda";
+				allwinner,function = "twi5";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			twi5_pins_b: twi5@1 {
+				allwinner,pins = "PI8", "PI9";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			ts0_pins_a: ts0@0 {
+				allwinner,pins = "PD0", "PD1", "PD2", "PD3",
+						 "PD4", "PD5", "PD6", "PD7",
+						 "PD8", "PD9", "PD10", "PD11";
+				allwinner,pname = "ts0_clk", "ts0_err",
+						  "ts0_sync", "ts0_dvld",
+						  "ts0_d0", "ts0_d1",
+						  "ts0_d2", "ts0_d3",
+						  "ts0_d4", "ts0_d5",
+						  "ts0_d6", "ts0_d7";
+				allwinner,function = "ts0";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			ts0_pins_b: ts0_sleep@0 {
+				allwinner,pins = "PD0", "PD1", "PD2", "PD3",
+						 "PD4", "PD5", "PD6", "PD7",
+						 "PD8", "PD9", "PD10", "PD11";
+				allwinner,pname = "ts0_clk", "ts0_err",
+						  "ts0_sync", "ts0_dvld",
+						  "ts0_d0", "ts0_d1",
+						  "ts0_d2", "ts0_d3",
+						  "ts0_d4", "ts0_d5",
+						  "ts0_d6", "ts0_d7";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			spi0_pins_a: spi0@0 {
+				allwinner,pins = "PC2", "PC4", "PC12", "PC15", "PC16";
+				allwinner,pname = "spi0_mosi", "spi0_miso",
+						  "spi0_sclk", "spi0_wp", "spi0_hold";
+				allwinner,function = "spi0";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			spi0_pins_b: spi0@1 {
+				allwinner,pins = "PC3", "PC7";
+				allwinner,pname = "spi0_cs0", "spi0_cs1";
+				allwinner,function = "spi0";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;	/* only CS should be pulled up */
+			};
+
+			spi0_pins_c: spi0@2 {
+				allwinner,pins = "PC2", "PC3", "PC4", "PC7", "PC12", "PC15", "PC16";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			spi1_pins_a: spi1@0 {
+				allwinner,pins = "PD11", "PD12", "PD13";
+				allwinner,pname = "spi1_sclk", "spi1_mosi",
+						  "spi1_miso";
+				allwinner,function = "spi1";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			spi1_pins_b: spi1@1 {
+				allwinner,pins = "PD10";
+				allwinner,pname = "spi1_cs0";
+				allwinner,function = "spi1";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;	/* only CS should be pulled up */
+			};
+
+			spi1_pins_c: spi1@2 {
+				allwinner,pins = "PD10", "PD11", "PD12", "PD13";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			spi2_pins_a: spi2@0 {
+				allwinner,pins = "PB1", "PB2", "PB3";
+				allwinner,pname = "spi2_sclk", "spi2_mosi",
+						  "spi2_miso";
+				allwinner,function = "spi2";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			spi2_pins_b: spi2@1 {
+				allwinner,pins = "PB0";
+				allwinner,pname = "spi2_cs0";
+				allwinner,function = "spi2";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;	/* only CS should be pulled up */
+			};
+
+			spi2_pins_c: spi2@2 {
+				allwinner,pins = "PB0", "PB1", "PB2", "PB3";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			sdc0_pins_a: sdc0@0 {
+				allwinner,pins = "PF0", "PF1", "PF2",
+						 "PF3", "PF4", "PF5";
+				allwinner,function = "sdc0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			sdc0_pins_b: sdc0@1 {
+				allwinner,pins = "PF0", "PF1", "PF2",
+						 "PF3", "PF4", "PF5";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			sdc1_pins_a: sdc1@0 {
+				allwinner,pins = "PG0", "PG1", "PG2",
+						 "PG3", "PG4", "PG5";
+				allwinner,function = "sdc1";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <3>;
+				allwinner,pull = <1>;
+			};
+
+			sdc1_pins_b: sdc1@1 {
+				allwinner,pins = "PG0", "PG1", "PG2",
+						 "PG3", "PG4", "PG5";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			sdc2_pins_a: sdc2@0 {
+				allwinner,pins = "PC1", "PC5", "PC6",
+						 "PC8", "PC9", "PC10", "PC11",
+						 "PC13", "PC14", "PC15", "PC16";
+				allwinner,function = "sdc2";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <2>;
+				allwinner,pull = <1>;
+			};
+
+			sdc2_pins_b: sdc2@1 {
+				allwinner,pins = "PC0", "PC1", "PC5", "PC6",
+						 "PC8", "PC9", "PC10", "PC11",
+						 "PC13", "PC14", "PC15", "PC16";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			sdc2_pins_c: sdc2@2 {
+				allwinner,pins = "PC0";
+				allwinner,function = "sdc2";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <2>;
+				allwinner,pull = <2>;
+			};
+
+			daudio0_pins_a: daudio0@0 {
+				allwinner,pins = "PB4", "PB5", "PB6", "PB7", "PB8";
+				allwinner,function = "pcm0";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			daudio0_pins_b: daudio0_sleep@0 {
+				allwinner,pins = "PB4", "PB5", "PB6", "PB7", "PB8";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			daudio1_pins_a: daudio1@0 {
+				allwinner,pins = "PG9", "PG10", "PG11", "PG12", "PG13";
+				allwinner,function = "pcm1";
+				allwinner,muxsel = <3>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			daudio1_pins_b: daudio1_sleep@0 {
+				allwinner,pins = "PG9", "PG10", "PG11", "PG12", "PG13";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			daudio2_pins_a: daudio2@0 {
+				allwinner,pins = "PH8", "PH9", "PH10", "PH11", "PH12";
+				allwinner,function = "pcm2";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			daudio2_pins_b: daudio2_sleep@0 {
+				allwinner,pins = "PH8", "PH9", "PH10", "PH11", "PH12";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			daudio3_pins_a: daudio3@0 {
+					allwinner,pins = "PH13", "PH14", "PH15", "PH16", "PH17", "PH18", "PH19";
+					allwinner,function = "pcm3";
+					allwinner,muxsel = <4>;
+					allwinner,drive = <1>;
+					allwinner,pull = <0>;
+			};
+
+			daudio3_pins_b: daudio3_sleep@0 {
+					allwinner,pins = "PH13", "PH14", "PH15", "PH16", "PH17", "PH18", "PH19";
+					allwinner,function = "io_disabled";
+					allwinner,muxsel = <7>;
+					allwinner,drive = <1>;
+					allwinner,pull = <0>;
+			};
+
+			spdif_pins_a: spdif@0 {
+				allwinner,pins = "PH6", "PH7";
+				allwinner,function = "spdif";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			spdif_pins_b: spdif_sleep@0 {
+				allwinner,pins = "PH6", "PH7";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			dmic_pins_a: dmic@0 {
+				allwinner,pins = "PH8", "PH9", "PH10", "PH11", "PH12";
+				allwinner,function = "dmic";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			dmic_pins_b: dmic_sleep@0 {
+				allwinner,pins = "PH8", "PH9", "PH10", "PH11", "PH12";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			csi_cci0_pins_a: csi_cci0@0 {
+				allwinner,pins = "PB0", "PB1";
+				allwinner,pname = "csi_cci0_sck", "csi_cci0_sda";
+				allwinner,function = "csi_cci0";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			csi_cci0_pins_b: csi_cci0@1 {
+				allwinner,pins = "PB0", "PB1";
+				allwinner,pname = "csi_cci0_sck", "csi_cci0_sda";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			csi_mclk0_pins_a: csi_mclk0@0 {
+				allwinner,pins = "PE0";
+				allwinner,pname = "csi_mclk0";
+				allwinner,function = "csi_mclk0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <2>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+			csi_mclk0_pins_b: csi_mclk0@1 {
+				allwinner,pins = "PE0";
+				allwinner,pname = "csi_mclk0";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <2>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+			csi_mclk1_pins_a: csi_mclk1@0 {
+				allwinner,pins = "PE5";
+				allwinner,pname = "csi_mclk1";
+				allwinner,function = "csi_mclk1";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <2>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+			csi_mclk1_pins_b: csi_mclk1@1 {
+				allwinner,pins = "PE5";
+				allwinner,pname = "csi_mclk1";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <2>;
+				allwinner,pull = <0>;
+				allwinner,data = <0>;
+			};
+
+			scr0_pins_a: scr0@0 {
+				allwinner,pins = "PG13", "PG14", "PG10", "PG11", "PG12";
+				allwinner,pname = "scr0_rst", "scr0_det", "scr0_vccen", "scr0_sck", "scr0_sda";
+				allwinner,function = "sim0";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <0>;
+				allwinner,pull = <1>;
+			};
+
+			scr0_pins_b: scr0@1 {
+				allwinner,pins = "PG8", "PG9";
+				allwinner,pname = "scr0_vppen", "scr0_vppp";
+				allwinner,function = "sim0";
+				allwinner,muxsel = <4>;
+				allwinner,drive = <0>;
+				allwinner,pull = <1>;
+			};
+
+			scr0_pins_c: scr0@2 {
+				allwinner,pins = "PG8", "PG9", "PG10", "PG11", "PG12", "PG13", "PG14";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <0>;
+				allwinner,pull = <0>;
+			};
+
+			scr1_pins_a: scr1@0 {
+				allwinner,pins = "PH5", "PH6", "PH2",
+						 "PH3", "PH4";
+				allwinner,pname = "scr1_rst", "scr1_det",
+						  "scr1_vccen", "scr1_sck",
+						  "scr1_sda";
+				allwinner,function = "sim1";
+				allwinner,muxsel = <5>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			scr1_pins_b: scr1@1 {
+				allwinner,pins = "PH0", "PH1";
+				allwinner,pname = "scr1_vppen", "scr1_vppp";
+				allwinner,function = "sim1";
+				allwinner,muxsel = <5>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			scr1_pins_c: scr1@2 {
+				allwinner,pins = "PH0", "PH1", "PH2",
+						 "PH3", "PH4", "PH5",
+						 "PH6";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			nand0_pins_a: nand0@0 {
+				allwinner,pins = "PC0", "PC1", "PC2", "PC4",
+						 "PC6", "PC7", "PC8", "PC9",
+						 "PC10", "PC11", "PC12", "PC13",
+						 "PC14";
+				allwinner,pname= "nand0_we", "nand0_ale","nand0_cle", "nand0_nre",
+						 "nand0_d0", "nand0_d1", "nand0_d2", "nand0_d3",
+						 "nand0_d4", "nand0_d5", "nand0_d6", "nand0_d7",
+						 "nand0_ndqs";
+				allwinner,function = "nand0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			nand0_pins_b: nand0@1 {
+				allwinner,pins = "PC3", "PC5", "PC15", "PC16";
+				allwinner,pname= "nand0_ce0", "nand0_rb0", "nand0_ce1", "nand0_rb1";
+				allwinner,function = "nand0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;	/* only RB&CE should be pulled up */
+			};
+
+			nand0_pins_c: nand0@2 {
+				allwinner,pins = "PC0", "PC1", "PC2", "PC3",
+						 "PC4", "PC5", "PC6", "PC7",
+						 "PC8", "PC9", "PC10", "PC11",
+						 "PC12", "PC13", "PC14", "PC15",
+						 "PC16";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			hdmi_ddc_pin_a: hdmi@0 {
+				allwinner,pins = "PI0","PI1";
+				allwinner,function = "ddc";
+				allwinner,muxsel = <5>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			hdmi_ddc_pin_b: hdmi@1 {
+				allwinner,pins = "PI0","PI1";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			hdmi_cec_pin_a: hdmi@2 {
+				allwinner,pins = "PI2";
+				allwinner,function = "hcec0";
+				allwinner,muxsel = <5>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			hdmi_cec_pin_b: hdmi@3 {
+				allwinner,pins = "PI2";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			ccir_clk_pin_a: ac200@2 {
+				allwinner,pins = "PB0";
+				allwinner,function = "ac200";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			ccir_clk_pin_b: ac200@3 {
+				allwinner,pins = "PB0";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+
+			gmac_pins_a: gmac@0 {
+				allwinner,pins = "PA0", "PA1", "PA2", "PA3",
+						 "PA4", "PA5", "PA6", "PA7",
+						 "PA8", "PA9", "PA10", "PA11",
+						 "PA12", "PA13", "PA14", "PA15",
+						 "PA16", "PA17", "PA18", "PA19",
+						 "PA20", "PA21", "PA22", "PA23",
+						 "PA24", "PA25", "PA26", "PA27",
+						 "PA28", "PA29", "PA30", "PA31";
+				allwinner,function = "gmac0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <3>;
+				allwinner,pull = <0>;
+			};
+
+			gmac_pins_b: gmac@1 {
+				allwinner,pins = "PA0", "PA1", "PA2", "PA3",
+						 "PA4", "PA5", "PA6", "PA7",
+						 "PA8", "PA9", "PA10", "PA11",
+						 "PA12", "PA13", "PA14", "PA15",
+						 "PA16", "PA17", "PA18", "PA19",
+						 "PA20", "PA21", "PA22", "PA23",
+						 "PA24", "PA25", "PA26", "PA27",
+						 "PA28", "PA29", "PA30", "PA31";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <3>;
+				allwinner,pull = <0>;
+			};
+
+			gmac1_pins_a: gmac1@0 {
+				allwinner,pins = "PA0", "PA1", "PA2", "PA3",
+						 "PA4", "PA5", "PA6", "PA7",
+						 "PA8", "PA9", "PA10", "PA11",
+						 "PA12", "PA13", "PA14", "PA15",
+						 "PA16", "PA17", "PA18", "PA19",
+						 "PA20", "PA21", "PA22", "PA23",
+						 "PA24", "PA25", "PA26", "PA27",
+						 "PA28", "PA29", "PA30", "PA31";
+				allwinner,function = "gmac0";
+				allwinner,muxsel = <2>;
+				allwinner,drive = <3>;
+				allwinner,pull = <0>;
+			};
+
+			gmac1_pins_b: gmac1@1 {
+				allwinner,pins = "PA0", "PA1", "PA2", "PA3",
+						 "PA4", "PA5", "PA6", "PA7",
+						 "PA8", "PA9", "PA10", "PA11",
+						 "PA12", "PA13", "PA14", "PA15",
+						 "PA16", "PA17", "PA18", "PA19",
+						 "PA20", "PA21", "PA22", "PA23",
+						 "PA24", "PA25", "PA26", "PA27",
+						 "PA28", "PA29", "PA30", "PA31";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <3>;
+				allwinner,pull = <0>;
+			};
+
+			ledc_pins_a: ledc@0 {
+				allwinner,pins = "PE5";
+				allwinner,function = "ledc";
+				allwinner,muxsel = <5>;
+				allwinner,drive = <1>;
+				allwinner,pull = <1>;
+			};
+
+			ledc_pins_b: ledc@1 {
+				allwinner,pins = "PE5";
+				allwinner,function = "io_disabled";
+				allwinner,muxsel = <7>;
+				allwinner,drive = <1>;
+				allwinner,pull = <0>;
+			};
+		};
+	};
+};
