diff -drupN a/sound/soc/sunxi/sunxi-aio/Kconfig b/sound/soc/sunxi/sunxi-aio/Kconfig
--- a/sound/soc/sunxi/sunxi-aio/Kconfig	1970-01-01 03:00:00.000000000 +0300
+++ b/sound/soc/sunxi/sunxi-aio/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,10 @@
+#
+# embedded vision engine device configuration
+#
+config SUNXI_MPP_AIO
+	tristate "AIO debugfs driver for mpp"
+	default y
+	depends on (ARCH_SUN8IW16 || ARCH_SUN8IW19) && SUNXI_MPP
+	---help---
+	This is the driver for allwinner platform audio input/output debugfs module.
+	To compile this driver as a built-in module.
