--- a/drivers/net/ethernet/hisilicon/hisi-femac/hisi_femac.c
+++ b/drivers/net/ethernet/hisilicon/hisi-femac/hisi_femac.c
@@ -1873,11 +1873,7 @@
                        (unsigned long)phy->phy_id,
                        phy_modes(phy->interface));
 
-    mac_addr = of_get_mac_address(node);
-    if (mac_addr) {
-        ether_addr_copy(ndev->dev_addr, mac_addr);
-    }
-    if (!is_valid_ether_addr(ndev->dev_addr)) {
+    if (eth_platform_get_mac_address(dev, ndev->dev_addr)) {
         eth_hw_addr_random(ndev);
         dev_warn(dev, "using random MAC address %pM\n",
                  ndev->dev_addr);
