--- linux-4.9.37/arch/arm/boot/dts/Makefile	2017-07-12 16:42:41.000000000 +0300
+++ linux-4.9.y/arch/arm/boot/dts/Makefile	2021-06-07 13:01:32.000000000 +0300
@@ -174,6 +174,14 @@
 	hi3519-demb.dtb
 dtb-$(CONFIG_ARCH_HIX5HD2) += \
 	hisi-x5hd2-dkb.dtb
+dtb-$(CONFIG_ARCH_GK7205V200) += \
+	gk7205v200-demb.dtb
+dtb-$(CONFIG_ARCH_GK7205V300) += \
+	gk7205v300-demb.dtb
+dtb-$(CONFIG_ARCH_GK7202V300) += \
+	gk7202v300-demb.dtb
+dtb-$(CONFIG_ARCH_GK7605V100) += \
+	gk7605v100-demb.dtb
 dtb-$(CONFIG_ARCH_INTEGRATOR) += \
 	integratorap.dtb \
 	integratorcp.dtb
