--- linux-4.9.37/drivers/mfd/Kconfig	2017-07-12 16:42:41.000000000 +0300
+++ linux-4.9.y/drivers/mfd/Kconfig	2021-06-07 13:01:33.000000000 +0300
@@ -358,6 +358,16 @@
 	help
 	  Select this option to enable Hisilicon hi655x series pmic driver.
 
+config MFD_GOKE_FMC
+	tristate "Goke Flash Memory Controller"
+	depends on OF
+	depends on ARCH_GOKE
+	select MFD_CORE
+	select REGMAP_MMIO
+	help
+	  Select this option to enable the Goke Flash Memory
+	  Controller(FMC) driver.
+
 config HTC_PASIC3
 	tristate "HTC PASIC3 LED/DS1WM chip support"
 	select MFD_CORE
