--- linux-4.9.37/drivers/net/phy/Kconfig	2017-07-12 16:42:41.000000000 +0300
+++ linux-4.9.y/drivers/net/phy/Kconfig	2021-06-07 13:01:33.000000000 +0300
@@ -89,6 +89,13 @@
 config MDIO_CAVIUM
 	tristate
 
+config MDIO_GOKE_FEMAC
+	tristate "Goke femac MDIO buses"
+	depends on HAS_IOMEM && OF_MDIO
+	help
+	  This module provides a driver for the MDIO busses found in the
+	  Goke SoC that have an Fast Ethernet.
+
 config MDIO_GPIO
 	tristate "GPIO lib-based bitbanged MDIO buses"
 	depends on MDIO_BITBANG && GPIOLIB
