--- linux-4.9.37/drivers/rtc/Kconfig	2017-07-12 16:42:41.000000000 +0300
+++ linux-4.9.y/drivers/rtc/Kconfig	2021-06-07 13:01:34.000000000 +0300
@@ -3,7 +3,7 @@
 #
 
 config RTC_LIB
-	bool
+	bool "Real Time Clock LIB"
 
 config RTC_MC146818_LIB
 	bool
@@ -824,6 +824,14 @@
 # requires <asm/mc146818rtc.h> defining CMOS_READ/CMOS_WRITE, and a
 # global rtc_lock ... it's not yet just another platform_device.
 
+config RTC_DRV_GOKE
+	tristate "Goke RTC support"
+	help
+	  Generic RTC framework driver for Goke SoCs.
+
+	  To compile this driver as a module, choose M here: the module
+	  will be called rtc.
+
 config RTC_DRV_CMOS
 	tristate "PC-style 'CMOS'"
 	depends on X86 || ARM || M32R || PPC || MIPS || SPARC64 || MN10300
