--- linux-4.9.37/drivers/usb/gadget/function/uvc_v4l2.c	2017-07-12 16:42:41.000000000 +0300
+++ linux-4.9.y/drivers/usb/gadget/function/uvc_v4l2.c	2021-06-07 13:01:34.000000000 +0300
@@ -60,8 +60,13 @@
 };
 
 static struct uvc_format uvc_formats[] = {
+#ifndef CONFIG_GOKE_MC
 	{ 16, V4L2_PIX_FMT_YUYV  },
+#else
+	{ 12, V4L2_PIX_FMT_NV21 },
+#endif
 	{ 0,  V4L2_PIX_FMT_MJPEG },
+	{ 0,  V4L2_PIX_FMT_H264 },
 };
 
 static int
