--- linux-4.9.37/drivers/clk/goke/reset.h	1970-01-01 03:00:00.000000000 +0300
+++ linux-4.9.y/drivers/clk/goke/reset.h	2021-06-07 13:01:33.000000000 +0300
@@ -0,0 +1,15 @@
+/*
+ * Copyright (c) Hunan Goke,Chengdu Goke,Shandong Goke. 2021. All rights reserved.
+ */
+
+#ifndef	__GOKE_RESET_H
+#define	__GOKE_RESET_H
+
+struct device_node;
+struct gk_reset_controller;
+
+#ifdef CONFIG_RESET_CONTROLLER
+int __init gk_reset_init(struct device_node *np, int nr_rsts);
+#endif
+
+#endif	/* __GOKE_RESET_H */
