--- linux-4.9.37/drivers/mmc/host/Kconfig	2017-07-12 16:42:41.000000000 +0300
+++ linux-4.9.y/drivers/mmc/host/Kconfig	2021-06-07 13:01:33.000000000 +0300
@@ -176,6 +176,15 @@
 
 	  If unsure, say N.
 
+config MMC_SDHCI_GOKE
+	tristate "SDHCI support on the Goke SoC"
+	depends on ARCH_GK7205V200 || ARCH_GK7205V300 || ARCH_GK7202V300 || ARCH_GK7605V100
+	depends on MMC_SDHCI_PLTFM
+	help
+	  This selects the SDHCI support for Goke System-on-Chip devices.
+	  If you have a controller with this interface, say Y or M here.
+	  If unsure, say N.
+
 config MMC_SDHCI_ESDHC_IMX
 	tristate "SDHCI support for the Freescale eSDHC/uSDHC i.MX controller"
 	depends on ARCH_MXC
@@ -798,3 +807,17 @@
 	  Broadcom STB SoCs.
 
 	  If unsure, say Y.
+
+config MMC_CQ_HCI
+	tristate "Command Queue Support"
+	depends on HAS_DMA
+	help
+	  This selects the Command Queue Host Controller Interface (CQHCI)
+	  support present in host controllers of Qualcomm Technologies, Inc
+	  amongst others.
+	  This controller supports eMMC devices with command queue support.
+
+	  If you have a controller with this interface, say Y or M here.
+
+	  If unsure, say N.
+
