--- linux-4.9.37/drivers/usb/dwc3/Makefile	2017-07-12 16:42:41.000000000 +0300
+++ linux-4.9.y/drivers/usb/dwc3/Makefile	2021-06-07 13:01:34.000000000 +0300
@@ -1,9 +1,12 @@
 # define_trace.h needs to know how to find our header
 CFLAGS_trace.o				:= -I$(src)
 
-obj-$(CONFIG_USB_DWC3)			+= dwc3.o
+obj-$(CONFIG_USB_DWC3)			+= dwc3.o dwc3-goke.o
 
-dwc3-y					:= core.o debug.o trace.o
+dwc3-y					:= core.o proc.o
+ifneq ($(CONFIG_TRACING),)
+dwc3-y					+= trace.o
+endif
 
 ifneq ($(filter y,$(CONFIG_USB_DWC3_HOST) $(CONFIG_USB_DWC3_DUAL_ROLE)),)
 	dwc3-y				+= host.o
