--- linux-4.9.37/drivers/usb/host/xhci.c	2017-07-12 16:42:41.000000000 +0300
+++ linux-4.9.y/drivers/usb/host/xhci.c	2021-06-07 13:01:34.000000000 +0300
@@ -4131,7 +4131,7 @@
 	xhci_dbg(xhci, "%s port %d USB2 hardware LPM\n",
 			enable ? "enable" : "disable", port_num + 1);
 
-	if (enable) {
+	if (enable && !(xhci->quirks & XHCI_HW_LPM_DISABLE)) {
 		/* Host supports BESL timeout instead of HIRD */
 		if (udev->usb2_hw_lpm_besl_capable) {
 			/* if device doesn't have a preferred BESL value use a
