--- linux-4.9.37/drivers/i2c/busses/Makefile	2017-07-12 16:42:41.000000000 +0300
+++ linux-4.9.y/drivers/i2c/busses/Makefile	2021-06-07 13:01:33.000000000 +0300
@@ -51,6 +51,7 @@
 obj-$(CONFIG_I2C_EMEV2)		+= i2c-emev2.o
 obj-$(CONFIG_I2C_EXYNOS5)	+= i2c-exynos5.o
 obj-$(CONFIG_I2C_GPIO)		+= i2c-gpio.o
+obj-$(CONFIG_I2C_GOKE)		+= i2c-goke.o
 obj-$(CONFIG_I2C_HIGHLANDER)	+= i2c-highlander.o
 obj-$(CONFIG_I2C_HIX5HD2)	+= i2c-hix5hd2.o
 obj-$(CONFIG_I2C_IBM_IIC)	+= i2c-ibm_iic.o
