--- linux-4.9.37/drivers/net/ethernet/goke/femac/phy_fix.h	1970-01-01 03:00:00.000000000 +0300
+++ linux-4.9.y/drivers/net/ethernet/goke/femac/phy_fix.h	2021-06-07 13:01:33.000000000 +0300
@@ -0,0 +1,18 @@
+/*
+ * Copyright (c) Hunan Goke,Chengdu Goke,Shandong Goke. 2021. All rights reserved.
+ */
+
+#ifndef __ETH_PHY_FIX_H__
+#define __ETH_PHY_FIX_H__
+
+#define GOKE_PHY_ID_V272     0x20669901
+#define GOKE_PHY_ID_V115     0x20669903
+#define GOKE_PHY_MASK        0xffffffff
+
+#define MII_EXPMD 0x1d
+#define MII_EXPMA 0x1e
+
+void phy_register_fixups(void);
+void phy_unregister_fixups(void);
+
+#endif
