--- linux-4.9.37/drivers/net/ethernet/Kconfig	2017-07-12 16:42:41.000000000 +0300
+++ linux-4.9.y/drivers/net/ethernet/Kconfig	2021-06-07 13:01:33.000000000 +0300
@@ -75,6 +75,7 @@
 source "drivers/net/ethernet/freescale/Kconfig"
 source "drivers/net/ethernet/fujitsu/Kconfig"
 source "drivers/net/ethernet/hisilicon/Kconfig"
+source "drivers/net/ethernet/goke/Kconfig"
 source "drivers/net/ethernet/hp/Kconfig"
 source "drivers/net/ethernet/ibm/Kconfig"
 source "drivers/net/ethernet/intel/Kconfig"
