--- linux-4.9.37/drivers/power/reset/Kconfig	2017-07-12 16:42:41.000000000 +0300
+++ linux-4.9.y/drivers/power/reset/Kconfig	2021-06-07 13:01:34.000000000 +0300
@@ -67,6 +67,12 @@
 	  Say Y here if you have a Broadcom STB board and you wish
 	  to have restart support.
 
+config POWER_RESET_GOKE
+	bool "Goke power-off driver"
+	depends on ARCH_GOKE
+	help
+	  Reboot support for Goke boards.
+
 config POWER_RESET_GPIO
 	bool "GPIO power-off driver"
 	depends on OF_GPIO
