--- linux-4.9.37/drivers/mmc/host/sdhci-pltfm.h	2017-07-12 16:42:41.000000000 +0300
+++ linux-4.9.y/drivers/mmc/host/sdhci-pltfm.h	2021-06-07 13:01:33.000000000 +0300
@@ -109,6 +109,8 @@
 	return (void *)host->private;
 }
 
+int sdhci_pltfm_suspend(struct device *dev);
+int sdhci_pltfm_resume(struct device *dev);
 extern const struct dev_pm_ops sdhci_pltfm_pmops;
 
 #endif /* _DRIVERS_MMC_SDHCI_PLTFM_H */
