--- linux-4.9.37/drivers/mtd/nand/gkfmc100/Kconfig	1970-01-01 03:00:00.000000000 +0300
+++ linux-4.9.y/drivers/mtd/nand/gkfmc100/Kconfig	2021-06-07 13:01:33.000000000 +0300
@@ -0,0 +1,17 @@
+#
+# goke flash memory controller SPI nand device driver version 100
+# drivers/mtd/nand/gkfmc100/Kconfig
+# add by goke 2017.8.7
+#
+
+config MTD_SPI_NAND_FMC100
+	bool "Goke Flash Memory Controller v100 SPI Nand devices support"
+	depends on MFD_GOKE_FMC && MTD_SPI_NAND_GOKE
+	select MISC_FILESYSTEMS
+	select MTD_BLOCK
+	select YAFFS_FS
+	select YAFFS_YAFFS2
+	help
+	  Goke Flash Memory Controller version 100 is called fmc100 for
+	  short. The controller driver support registers and DMA transfers
+	  while reading or writing the SPI nand flash.
