--- linux-4.9.37/fs/fat/fat.h	2017-07-12 16:42:41.000000000 +0300
+++ linux-4.9.y/fs/fat/fat.h	2021-06-07 13:01:34.000000000 +0300
@@ -409,6 +409,10 @@
 			      __le16 __time, __le16 __date, u8 time_cs);
 extern void fat_time_unix2fat(struct msdos_sb_info *sbi, struct timespec *ts,
 			      __le16 *time, __le16 *date, u8 *time_cs);
+#ifdef CONFIG_GOKE_MC
+extern void fat_time_fat2str(struct msdos_sb_info *sbi, char *d_createtime,
+		__le16 __time, __le16 __date, u8 time_cs);
+#endif
 extern int fat_sync_bhs(struct buffer_head **bhs, int nr_bhs);
 
 int fat_cache_init(void);
