--- linux-4.9.37/fs/compat_ioctl.c	2017-07-12 16:42:41.000000000 +0300
+++ linux-4.9.y/fs/compat_ioctl.c	2021-06-07 13:01:34.000000000 +0300
@@ -750,9 +750,9 @@
 	if (!access_ok(VERIFY_READ, udata, sizeof(*udata)))
 		return -EFAULT;
 
-	if (__copy_in_user(&tdata->read_write, &udata->read_write, 2 * sizeof(u8)))
+	if (copy_in_user(&tdata->read_write, &udata->read_write, 2 * sizeof(u8)))
 		return -EFAULT;
-	if (__copy_in_user(&tdata->size, &udata->size, 2 * sizeof(u32)))
+	if (copy_in_user(&tdata->size, &udata->size, 2 * sizeof(u32)))
 		return -EFAULT;
 	if (__get_user(datap, &udata->data) ||
 	    __put_user(compat_ptr(datap), &tdata->data))
