--- linux-4.9.37/include/uapi/linux/i2c-dev.h	2017-07-12 16:42:41.000000000 +0300
+++ linux-4.9.y/include/uapi/linux/i2c-dev.h	2021-06-07 13:01:34.000000000 +0300
@@ -50,6 +50,9 @@
 
 #define I2C_PEC		0x0708	/* != 0 to use PEC with SMBus */
 #define I2C_SMBUS	0x0720	/* SMBus transfer */
+#define I2C_16BIT_REG	0x0709	/* 16BIT REG WIDTH */
+#define I2C_16BIT_DATA	0x070a	/* 16BIT DATA WIDTH */
+#define I2C_DMA		0x070b	/* DMA mode */
 
 
 /* This is the structure as used in the I2C_SMBUS ioctl call */
