--- linux-4.9.37/include/uapi/linux/jffs2.h	2017-07-12 16:42:41.000000000 +0300
+++ linux-4.9.y/include/uapi/linux/jffs2.h	2021-06-07 13:01:34.000000000 +0300
@@ -46,6 +46,7 @@
 #define JFFS2_COMPR_DYNRUBIN	0x05
 #define JFFS2_COMPR_ZLIB	0x06
 #define JFFS2_COMPR_LZO		0x07
+#define JFFS2_COMPR_LZMA	0x08
 /* Compatibility flags. */
 #define JFFS2_COMPAT_MASK 0xc000      /* What do to if an unknown nodetype is found */
 #define JFFS2_NODE_ACCURATE 0x2000
