--- linux-4.9.37/lib/Kconfig	2017-07-12 16:42:41.000000000 +0300
+++ linux-4.9.y/lib/Kconfig	2021-06-07 13:01:34.000000000 +0300
@@ -241,6 +241,12 @@
 
 source "lib/xz/Kconfig"
 
+config LZMA_COMPRESS
+	tristate
+
+config LZMA_DECOMPRESS
+	tristate
+
 #
 # These all provide a common interface (hence the apparent duplication with
 # ZLIB_INFLATE; DECOMPRESS_GZIP is just a wrapper.)
@@ -278,7 +284,7 @@
 #
 config REED_SOLOMON
 	tristate
-	
+
 config REED_SOLOMON_ENC8
 	bool
 
