--- linux-4.9.37/net/ipv4/tcp_timer.c	2017-07-12 16:42:41.000000000 +0300
+++ linux-4.9.y/net/ipv4/tcp_timer.c	2021-06-07 13:01:34.000000000 +0300
@@ -125,6 +125,7 @@
 			mss = tcp_mtu_to_mss(sk, icsk->icsk_mtup.search_low) >> 1;
 			mss = min(net->ipv4.sysctl_tcp_base_mss, mss);
 			mss = max(mss, 68 - tp->tcp_header_len);
+			mss = max(mss, net->ipv4.sysctl_tcp_min_snd_mss);
 			icsk->icsk_mtup.search_low = tcp_mss_to_mtu(sk, mss);
 			tcp_sync_mss(sk, icsk->icsk_pmtu_cookie);
 		}
