diff --git a/home/goody/of_mdio.c b/output/build/linux-4.9.37/drivers/of/of_mdio.c
index 262281bd..8095c00e 100644
--- a/drivers/of/of_mdio.c
+++ b/drivers/of/of_mdio.c
@@ -25,6 +25,11 @@
 MODULE_AUTHOR("Grant Likely <grant.likely@secretlab.ca>");
 MODULE_LICENSE("GPL");
 
+int higmacphy = 0;
+EXPORT_SYMBOL(higmacphy);
+MODULE_PARM_DESC(higmacphy, "Select phyaddr for higmac driver.\n");
+module_param_named(higmacphy, higmacphy, int, 0600);
+
 /* Extract the clause 22 phy ID from the compatible string of the form
  * ethernet-phy-idAAAA.BBBB */
 static int of_get_phy_id(struct device_node *device, u32 *phy_id)
@@ -233,6 +238,11 @@ int of_mdiobus_register(struct mii_bus *mdio, struct device_node *np)
 			continue;
 		}
 
+		if (higmacphy==1) {
+			dev_info(&mdio->dev, "higmacphy addr is forced to be 0 for phy %s\n", child->name);
+			addr = 0;
+		}
+
 		if (of_mdiobus_child_is_phy(child))
 			of_mdiobus_register_phy(mdio, child, addr);
 		else
