diff -urN a/drivers/sstar/cambricon/Kconfig b/drivers/sstar/cambricon/Kconfig
--- a/drivers/sstar/cambricon/Kconfig	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/sstar/cambricon/Kconfig	2021-04-25 16:59:01.364842246 +0300
@@ -0,0 +1 @@
+#
\ В конце файла нет новой строки
diff -urN a/drivers/sstar/cambricon/Makefile b/drivers/sstar/cambricon/Makefile
--- a/drivers/sstar/cambricon/Makefile	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/sstar/cambricon/Makefile	2021-04-25 16:58:56.208925987 +0300
@@ -0,0 +1 @@
+#
\ В конце файла нет новой строки
diff -urN a/drivers/sstar/camdriver/Kconfig b/drivers/sstar/camdriver/Kconfig
--- a/drivers/sstar/camdriver/Kconfig	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/sstar/camdriver/Kconfig	2021-04-25 16:59:23.012490621 +0300
@@ -0,0 +1 @@
+#
\ В конце файла нет новой строки
diff -urN a/drivers/sstar/camdriver/Makefile b/drivers/sstar/camdriver/Makefile
--- a/drivers/sstar/camdriver/Makefile	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/sstar/camdriver/Makefile	2021-04-25 16:59:26.848428308 +0300
@@ -0,0 +1 @@
+#
\ В конце файла нет новой строки
diff -urN a/drivers/sstar/irqchip/Kconfig b/drivers/sstar/irqchip/Kconfig
--- a/drivers/sstar/irqchip/Kconfig	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/sstar/irqchip/Kconfig	2021-04-25 16:59:50.288047511 +0300
@@ -0,0 +1 @@
+#
\ В конце файла нет новой строки
diff --git a/ms_pack_modules.sh b/ms_pack_modules.sh
old mode 100644
new mode 100755
diff --git a/release_kernel_to_alkaid.sh b/release_kernel_to_alkaid.sh
old mode 100644
new mode 100755
diff --git a/list_config.sh b/list_config.sh
old mode 100644
new mode 100755
