diff -drupN a/drivers/rtc/Kconfig b/drivers/rtc/Kconfig
--- a/drivers/rtc/Kconfig	2017-10-21 18:09:07.000000000 +0300
+++ b/drivers/rtc/Kconfig	2022-06-09 05:02:33.000000000 +0300
@@ -139,6 +139,15 @@ config RTC_DRV_TEST
 	  This driver can also be built as a module. If so, the module
 	  will be called rtc-test.
 
+config RTC_DRV_INGENIC
+    tristate "INGENIC RTC"
+        help
+        If you say Y here you will get access to the real time clock
+          built into your INGENIC CPU.
+
+          To compile this driver as a module, choose M here: the
+          module will be called rtc-ingenic.
+
 comment "I2C RTC drivers"
 	depends on I2C
 
@@ -419,6 +428,15 @@ config RTC_DRV_PCF8523
 	  This driver can also be built as a module. If so, the module
 	  will be called rtc-pcf8523.
 
+config RTC_DRV_BM8563
+	tristate "BM8563 "
+	help
+	  If you say yes here you get support for the NXP BM8563 RTC
+	  chips.
+
+	  This driver can also be built as a module. If so, the module
+	  will be called rtc-bm8563.
+
 config RTC_DRV_PCF8563
 	tristate "Philips PCF8563/Epson RTC8564"
 	help
