diff -drupN a/lib/Kconfig b/lib/Kconfig
--- a/lib/Kconfig	2017-10-21 18:09:07.000000000 +0300
+++ b/lib/Kconfig	2022-06-09 05:02:36.000000000 +0300
@@ -254,6 +254,7 @@ config DECOMPRESS_BZIP2
 
 config DECOMPRESS_LZMA
 	tristate
+    default y
 
 config DECOMPRESS_XZ
 	select XZ_DEC
@@ -267,6 +268,10 @@ config DECOMPRESS_LZ4
 	select LZ4_DECOMPRESS
 	tristate
 
+config DECOMPRESS_LZMA_NEEDED
+	 boolean
+     default y
+
 #
 # Generic allocator support is selected if needed
 #
