diff -drupN a/drivers/media/platform/ingenic-vcodec/felix/Makefile b/drivers/media/platform/ingenic-vcodec/felix/Makefile
--- a/drivers/media/platform/ingenic-vcodec/felix/Makefile	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/media/platform/ingenic-vcodec/felix/Makefile	2022-06-09 05:02:29.000000000 +0300
@@ -0,0 +1,12 @@
+
+ccflags-y += -Idrivers/media/platform/ingenic-vcodec/felix/libh264/
+ccflags-y += -Idrivers/media/platform/ingenic-vcodec/felix/libh264/libavcodec/
+ccflags-y += -Idrivers/media/platform/ingenic-vcodec/felix/libh264/libavutil/
+
+obj-y += felix_ops.o
+obj-y += felix_drv.o
+
+obj-y += libh264/
+
+
+
