diff -drupN a/drivers/crypto/Kconfig b/drivers/crypto/Kconfig
--- a/drivers/crypto/Kconfig	2017-10-21 18:09:07.000000000 +0300
+++ b/drivers/crypto/Kconfig	2022-06-09 05:02:28.000000000 +0300
@@ -308,6 +308,15 @@ config CRYPTO_DEV_OMAP_DES
 	  the ECB and CBC modes of operation supported by the driver. Also
 	  accesses made on unaligned boundaries are also supported.
 
+config CRYPTO_DEV_INGENIC_AES
+	tristate "Support for INGENIC AES hw engine"
+	depends on MARCH_XBURST1 || MACH_XBURST2
+	select CRYPTO_AES
+	select CRYPTO_BLKCIPHER2
+	help
+	  INGENIC processors have AES module accelerator. Select this if you
+	  want to use the INGENIC module for AES algorithms.
+
 config CRYPTO_DEV_PICOXCELL
 	tristate "Support for picoXcell IPSEC and Layer2 crypto engines"
 	depends on ARCH_PICOXCELL && HAVE_CLK
@@ -481,6 +490,19 @@ config CRYPTO_DEV_IMGTEC_HASH
 	  hardware hash accelerator. Supporting MD5/SHA1/SHA224/SHA256
 	  hashing algorithms.
 
+config CRYPTO_DEV_INGENIC_SHA
+	tristate "Support for Ingenic SHA hw accelerator"
+	depends on MARCH_XBURST1 || MACH_XBURST2
+	select CRYPTO_ALGAPI
+	help
+	  Some Ingenic processors have MD5/SHA1/SHA224/SHA256/SHA384/SHA512
+	  hw accelerator.
+	  Select this if you want to use the Ingenic module for
+	  MD5/SHA1/SHA224/SHA256/SHA384/SHA512 algorithms.
+
+	  To compile this driver as a module, choose M here: the module
+	  will be called ingenic-sha.
+
 config CRYPTO_DEV_SUN4I_SS
 	tristate "Support for Allwinner Security System cryptographic accelerator"
 	depends on ARCH_SUNXI
