diff -drupN a/sound/soc/soc-io.c b/sound/soc/soc-io.c
--- a/sound/soc/soc-io.c	2017-10-21 18:09:07.000000000 +0300
+++ b/sound/soc/soc-io.c	2022-06-09 05:02:37.000000000 +0300
@@ -41,6 +41,19 @@ int snd_soc_component_read(struct snd_so
 }
 EXPORT_SYMBOL_GPL(snd_soc_component_read);
 
+unsigned int snd_soc_component_read32(struct snd_soc_component *component,
+	unsigned int reg)
+{
+	unsigned int val;
+	int ret;
+
+	ret = snd_soc_component_read(component, reg, &val);
+	if (ret < 0)
+		return -1;
+	return val;
+}
+EXPORT_SYMBOL_GPL(snd_soc_component_read32);
+
 /**
  * snd_soc_component_write() - Write register value
  * @component: Component to write to
