diff -drupN a/drivers/video/fbdev/ingenic/fb_v12/jz_mipi_dsi/Makefile b/drivers/video/fbdev/ingenic/fb_v12/jz_mipi_dsi/Makefile
--- a/drivers/video/fbdev/ingenic/fb_v12/jz_mipi_dsi/Makefile	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/video/fbdev/ingenic/fb_v12/jz_mipi_dsi/Makefile	2022-06-09 05:02:34.000000000 +0300
@@ -0,0 +1 @@
+obj-y += jz_mipi_dsi.o jz_mipi_dsi_lowlevel.o jz_mipi_dsih_hal.o
