diff -drupN a/drivers/misc/Kconfig b/drivers/misc/Kconfig
--- a/drivers/misc/Kconfig	2017-10-21 18:09:07.000000000 +0300
+++ b/drivers/misc/Kconfig	2022-06-09 05:02:30.000000000 +0300
@@ -496,6 +496,14 @@ config USB_SWITCH_FSA9480
 	  stereo and mono audio, video, microphone and UART data to use
 	  a common connector port.
 
+config BCM_43438_RFKILL
+	tristate "Bluetooth power control driver for BCM-43438 module"
+	depends on RFKILL
+	default n
+	---help---
+	Creates an rfkill entry in sysfs for power control of Bluetooth
+	bcm-xxxx chips.
+
 config LATTICE_ECP3_CONFIG
 	tristate "Lattice ECP3 FPGA bitstream configuration via SPI"
 	depends on SPI && SYSFS
@@ -525,6 +533,22 @@ config VEXPRESS_SYSCFG
 	  bus. System Configuration interface is one of the possible means
 	  of generating transactions on this bus.
 
+config INGENIC_RSA
+	bool "JZ RSA Driver"
+	depends on MACH_XBURST2
+	help
+		this driver is used to Encrypt/Decrypt by RSA.
+	default n
+
+config RMEM
+	bool "Reserved memory driver(rmem)"
+	default y
+
+config LOGGER
+	tristate "Log driver porting from Android"
+	default y
+
+
 source "drivers/misc/c2port/Kconfig"
 source "drivers/misc/eeprom/Kconfig"
 source "drivers/misc/cb710/Kconfig"
