diff -drupN a/scripts/setlocalversion b/scripts/setlocalversion
--- a/scripts/setlocalversion	2017-10-21 18:09:07.000000000 +0300
+++ b/scripts/setlocalversion	2022-06-09 05:02:37.000000000 +0300
@@ -53,7 +53,6 @@ scm_version()
 			# If only the short version is requested, don't bother
 			# running further git commands
 			if $short; then
-				echo "+"
 				return
 			fi
 			# If we are past a tagged commit (like
@@ -167,7 +166,6 @@ else
 	# LOCALVERSION= is not specified
 	if test "${LOCALVERSION+set}" != "set"; then
 		scm=$(scm_version --short)
-		res="$res${scm:++}"
 	fi
 fi
 
