diff -drupN a/init/main.c b/init/main.c
--- a/init/main.c	2017-10-21 18:09:07.000000000 +0300
+++ b/init/main.c	2022-06-09 05:02:36.000000000 +0300
@@ -498,6 +498,7 @@ asmlinkage __visible void __init start_k
 {
 	char *command_line;
 	char *after_dashes;
+    unsigned  long flags = 0;
 
 	/*
 	 * Need to run as early as possible, to initialize the
@@ -594,6 +595,8 @@ asmlinkage __visible void __init start_k
 	perf_event_init();
 	profile_init();
 	call_function_init();
+    flags = read_cp0_30_flags();
+    printk("ERROR epc 0x%lx\n", flags);
 	WARN(!irqs_disabled(), "Interrupts were enabled early\n");
 	early_boot_irqs_disabled = false;
 	local_irq_enable();
