diff -drupN a/drivers/media/platform/soc_camera/ingenic/x2000/mipi_csi.h b/drivers/media/platform/soc_camera/ingenic/x2000/mipi_csi.h
--- a/drivers/media/platform/soc_camera/ingenic/x2000/mipi_csi.h	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/media/platform/soc_camera/ingenic/x2000/mipi_csi.h	2022-06-09 05:02:29.000000000 +0300
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) 2012 Ingenic Semiconductor Co., Ltd.
+ *              http://www.ingenic.com/
+ *
+ * Core file for Ingenic Display Controller driver
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ */
+
+#ifndef __MIPI_CSI_H__
+#define __MIPI_CSI_H__
+
+/* csi host regs, base addr should be defined in board cfg */
+#define DWC_CSI_CTRL_BASE		0xB3015000
+#define INNO_CSI_PHY_BASE		0xB3016000
+
+/* csi controller */
+#define VERSION				0x00
+#define N_LANES				0x04
+#define PHY_SHUTDOWNZ			0x08
+#define DPHY_RSTZ			0x0C
+#define CSI2_RESETN			0x10
+#define PHY_STATE			0x14
+#define DATA_IDS_1			0x18
+#define DATA_IDS_2			0x1C
+#define ERR1				0x20
+#define ERR2				0x24
+#define MASK1				0x28
+#define MASK2				0x2C
+#define PHY_TST_CTRL0			0x30
+#define PHY_TST_CTRL1			0x34
+
+/* csi d-phy */
+#define LANE_EN				0x000
+#define CLK_CNT_TIME			0x100
+#define L0_CNT_TIME			0x180
+#define L1_CNT_TIME			0x200
+#define L2_CNT_TIME			0x280
+#define L3_CNT_TIME			0x300
+#define CLK_CON_MODE			0x128
+#define SWITCH_LVDS_BANK		0x080
+#define MODEL_EN			0x2cc
+#define LVDS_LOGICAL_EN			0x300
+
+typedef enum
+{
+	ERR_NOT_INIT = 0xFE,
+	ERR_ALREADY_INIT = 0xFD,
+	ERR_NOT_COMPATIBLE = 0xFC,
+	ERR_UNDEFINED = 0xFB,
+	ERR_OUT_OF_BOUND = 0xFA,
+	SUCCESS = 0
+} csi_error_t;
+
+
+#define dwc_csi_readl(reg)              \
+	readl((unsigned int *)(DWC_CSI_CTRL_BASE + reg))
+#define dwc_csi_writel(reg, value)          \
+	writel((value), (unsigned int *)(DWC_CSI_CTRL_BASE + reg))
+
+#define csi_core_write(addr, value) dwc_csi_writel(addr, value)
+#define csi_core_read(addr) dwc_csi_readl(addr)
+
+#define csi_phy_read(reg)              \
+	readl((unsigned int *)(INNO_CSI_PHY_BASE + reg))
+#define csi_phy_write(reg, value)          \
+	writel((value), (unsigned int *)(INNO_CSI_PHY_BASE + reg))
+
+/* function */
+extern int csi_phy_init(void);
+extern int csi_phy_release(void);
+extern int csi_phy_start(unsigned int id, unsigned int freq, unsigned int lans);
+extern int csi_phy_stop(unsigned int id);
+
+extern void dump_csi_reg(void);
+extern void check_csi_error(void);
+extern unsigned char csi_set_on_lanes(unsigned char lanes);
+
+#endif/*__MIPI_CSI_H__*/
