diff -drupN a/drivers/mtd/devices/ingenic_sfc_v1/sfc.h b/drivers/mtd/devices/ingenic_sfc_v1/sfc.h
--- a/drivers/mtd/devices/ingenic_sfc_v1/sfc.h	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/mtd/devices/ingenic_sfc_v1/sfc.h	2022-06-09 05:02:30.000000000 +0300
@@ -0,0 +1,58 @@
+#ifndef __SFC_H
+#define __SFC_H
+
+#include <linux/list.h>
+#include <linux/mtd/mtd.h>
+#include <linux/irqreturn.h>
+#include <soc/sfc.h>
+
+struct cmd_info{
+	uint8_t cmd;
+	uint8_t dataen;
+	uint8_t pollen;
+	uint8_t sta_exp;
+	uint8_t sta_msk;
+};
+
+struct sfc_transfer {
+	uint8_t direction;
+
+	struct cmd_info cmd_info;
+
+	uint8_t addr_len;
+	uint32_t addr;
+	uint32_t addr_plus;
+	uint8_t *data;
+	uint8_t data_dummy_bits;/*addr + data_dummy_bits + data*/
+
+	uint32_t len;
+	uint32_t cur_len;
+
+	uint8_t sfc_mode;
+	uint8_t ops_mode;
+	uint8_t phase_format;/*we just use default value;phase1:cmd+dummy+addr... phase0:cmd+addr+dummy...*/
+
+	struct list_head list;
+};
+
+struct sfc{
+
+	void __iomem	*iomem;
+	struct resource *ioarea;
+	int			irq;
+	struct clk		*clk;
+	struct clk		*clk_gate;
+	unsigned long src_clk;
+	struct completion	done;
+	uint32_t		threshold;
+	irqreturn_t (*irq_callback)(struct sfc *);
+	unsigned long		phys;
+
+	struct sfc_transfer *transfer;
+};
+
+
+
+#endif
+
+
