diff -drupN a/arch/mips/xburst2/common/cpu_ddr_test/Makefile b/arch/mips/xburst2/common/cpu_ddr_test/Makefile
--- a/arch/mips/xburst2/common/cpu_ddr_test/Makefile	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/mips/xburst2/common/cpu_ddr_test/Makefile	2022-06-09 05:02:27.000000000 +0300
@@ -0,0 +1,11 @@
+
+obj-y += multithread_test.o
+obj-y += ddr_bandwidth_monitor.o
+obj-y += ddr_change_freq.o
+obj-y += l2c_test.o
+obj-y += cpu_spinlock_test.o
+obj-y += cpu_ccu_test.o
+obj-y += raw_dma_test.o
+obj-y += cpu_dma_test.o
+obj-y += cpu_watch_test.o
+ccflags-y += -Wno-unused-function
