diff -drupN a/arch/mips/include/asm/processor.h b/arch/mips/include/asm/processor.h
--- a/arch/mips/include/asm/processor.h	2017-10-21 18:09:07.000000000 +0300
+++ b/arch/mips/include/asm/processor.h	2022-06-09 05:02:27.000000000 +0300
@@ -19,6 +19,7 @@
 #include <asm/cpu-info.h>
 #include <asm/mipsregs.h>
 #include <asm/prefetch.h>
+#include <asm/cpu-features.h>
 
 /*
  * Return current * instruction pointer ("program counter").
@@ -82,9 +83,19 @@ extern unsigned int vced_count, vcei_cou
  */
 #define TASK_UNMAPPED_BASE PAGE_ALIGN(TASK_SIZE / 3)
 
-
 #define NUM_FPU_REGS	32
 
+#if cpu_has_mxu
+#define NUM_MXU_VPR_REGS  16      //vpr0~15
+#define NUM_MXU_VSR_REGS  0
+#endif
+
+#if cpu_has_mxuv3
+#define NUM_MXU_VPR_REGS  32      //vpr0~31
+#define NUM_MXU_VSR_REGS  4       //vsr0~3
+#endif
+
+
 #ifdef CONFIG_CPU_HAS_MSA
 # define FPU_REG_WIDTH	128
 #else
@@ -96,10 +107,30 @@ union fpureg {
 	__u64	val64[FPU_REG_WIDTH / 64];
 };
 
+typedef union mxuvec {
+#if cpu_has_mxu
+    uint8_t    val128[16];     //simd128
+#endif
+#if cpu_has_mxuv3
+    uint8_t    val512[64];     //simd512
+#endif
+} mxuvec_t;
+
+
+//#ifdef CONFIG_MACH_XBURST
+struct xburst_mxu_struct {
+        mxuvec_t   vpr[NUM_MXU_VPR_REGS];
+        mxuvec_t   vsr[NUM_MXU_VSR_REGS];
+		uint32_t	csr;
+};
+//#endif
+
+
+
 #ifdef CONFIG_CPU_LITTLE_ENDIAN
 # define FPR_IDX(width, idx)	(idx)
 #else
-# define FPR_IDX(width, idx)	((idx) ^ ((64 / (width)) - 1))
+# define FPR_IDX(width, idx)   ((FPU_REG_WIDTH / (width)) - 1 - (idx))
 #endif
 
 #define BUILD_FPR_ACCESS(width) \
@@ -154,8 +185,20 @@ struct mips3264_watch_reg_state {
 union mips_watch_reg_state {
 	struct mips3264_watch_reg_state mips3264;
 };
+#if defined(CONFIG_XBURST_MXUV2)
+typedef union {
+	u64 val64[2];
+} vpr_t;
 
-#if defined(CONFIG_CPU_CAVIUM_OCTEON)
+struct xburst_cop2_state {
+	u32 mxu_csr;
+	vpr_t vr[32];
+};
+
+#define COP2_INIT						\
+	.cp2			= {0,},
+
+#elif defined(CONFIG_CPU_CAVIUM_OCTEON)
 
 struct octeon_cop2_state {
 	/* DMFC2 rt, 0x0201 */
@@ -224,6 +267,13 @@ struct nlm_cop2_state {
 #define COP2_INIT
 #endif
 
+#ifdef CONFIG_PMON_DEBUG
+struct xburst_perf_cnt {
+       u32 perfctrl;
+       u64 perfcnt;
+};
+#endif
+
 typedef struct {
 	unsigned long seg;
 } mm_segment_t;
@@ -262,6 +312,11 @@ struct thread_struct {
 	/* Saved state of the DSP ASE, if available. */
 	struct mips_dsp_state dsp;
 
+//#ifdef CONFIG_MACH_XBURST
+	/* Saved registers of the MXU, if available. */
+	struct xburst_mxu_struct mxu;
+//#endif
+
 	/* Saved watch register state, if available. */
 	union mips_watch_reg_state watch;
 
@@ -270,13 +325,18 @@ struct thread_struct {
 	unsigned long cp0_baduaddr;	/* Last kernel fault accessing USEG */
 	unsigned long error_code;
 	unsigned long trap_nr;
-#ifdef CONFIG_CPU_CAVIUM_OCTEON
+#if defined(CONFIG_XBURST_MXUV2)
+	struct xburst_cop2_state cp2;
+#elif defined(CONFIG_CPU_CAVIUM_OCTEON)
 	struct octeon_cop2_state cp2 __attribute__ ((__aligned__(128)));
 	struct octeon_cvmseg_state cvmseg __attribute__ ((__aligned__(128)));
 #endif
 #ifdef CONFIG_CPU_XLP
 	struct nlm_cop2_state cp2;
 #endif
+#ifdef CONFIG_PMON_DEBUG
+        struct xburst_perf_cnt pfc[2];
+#endif
 	struct mips_abi *abi;
 };
 
@@ -288,6 +348,14 @@ struct thread_struct {
 #define FPAFF_INIT
 #endif /* CONFIG_MIPS_MT_FPAFF */
 
+#ifdef CONFIG_PMON_DEBUG
+#define PFC_INIT                                               \
+       .pfc                    = {{0, 0}, {0, 0}},
+#else
+#define PFC_INIT
+#endif /* CONFIG_MACH_XBURST */
+
+
 #define INIT_THREAD  {						\
 	/*							\
 	 * Saved main processor registers			\
@@ -341,6 +409,7 @@ struct thread_struct {
 	 * Platform specific cop2 registers(null if no COP2)	\
 	 */							\
 	COP2_INIT						\
+        PFC_INIT                                                \
 }
 
 struct task_struct;
