diff -drupN a/arch/mips/xburst2/soc-x2000-v12/include/libdmmu.h b/arch/mips/xburst2/soc-x2000-v12/include/libdmmu.h
--- a/arch/mips/xburst2/soc-x2000-v12/include/libdmmu.h	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/mips/xburst2/soc-x2000-v12/include/libdmmu.h	2022-06-09 05:02:27.000000000 +0300
@@ -0,0 +1,14 @@
+#ifndef _LIBDMMU_
+#define _LIBDMMU_
+
+#include <linux/device.h>
+
+unsigned long dmmu_map(struct device *dev,unsigned long vaddr,unsigned long len);
+int dmmu_flush_cache(unsigned long vaddr, unsigned long len);
+int dmmu_unmap(struct device *dev,unsigned long vaddr, int len);
+int dmmu_unmap_all(struct device *dev);
+void dmmu_dump_vaddr(unsigned long vaddr);
+int dmmu_memory_smash(unsigned long vaddr, int smash_mode);
+int dmmu_dump_map(unsigned long vaddr);
+
+#endif
