diff -drupN a/sound/soc/ingenic/README b/sound/soc/ingenic/README
--- a/sound/soc/ingenic/README	1970-01-01 03:00:00.000000000 +0300
+++ b/sound/soc/ingenic/README	2022-06-09 05:02:37.000000000 +0300
@@ -0,0 +1,31 @@
+DIRs:
+.
+├── as-v1/
+├── as-v2/
+├── boards/
+├── icodec/
+├── ecodec/
+├── Kconfig
+├── Makefile
+└── README
+
+---
+as-v1:
+	Audio System Version 1.
+	Support SOCs
+		-X1000 Series.
+---
+as-v2:	Audio System Version 2.
+	Support SOCs
+		-X2000 Series.
+---
+boards: Boards config.
+icodec: Internal Codecs.
+ecodec: External Codecs.
+
+
+## Tested Usable External codecs
+
+ak4458
+ak5558
+wm8594
