diff -drupN a/drivers/media/platform/soc_camera/Kconfig b/drivers/media/platform/soc_camera/Kconfig
--- a/drivers/media/platform/soc_camera/Kconfig	2017-10-21 18:09:07.000000000 +0300
+++ b/drivers/media/platform/soc_camera/Kconfig	2022-06-09 05:02:29.000000000 +0300
@@ -89,3 +89,4 @@ config VIDEO_ATMEL_ISI
 	  This module makes the ATMEL Image Sensor Interface available
 	  as a v4l2 device.
 
+source "drivers/media/platform/soc_camera/ingenic/Kconfig"
