diff -drupN a/fs/squashfs/Kconfig b/fs/squashfs/Kconfig
--- a/fs/squashfs/Kconfig	2017-10-21 18:09:07.000000000 +0300
+++ b/fs/squashfs/Kconfig	2022-06-09 05:02:35.000000000 +0300
@@ -164,6 +164,52 @@ config SQUASHFS_XZ
 	  file systems will be readable without selecting this option.
 
 	  If unsure, say N.
+#config SQUASHFS_LZMA
+#    bool "Include support for LZMA compressed file systems"
+#    depends on SQUASHFS
+#    select LZMADEVICE
+#    select DECOMPRESS_LZMA
+#    select DECOMPRESS_LZMA_NEEDED
+#    default y
+#    help
+#      Saying Y here includes support for reading Squashfs file systems
+#      compressed with LZMA compression.  LZMA gives better compression than
+#      the default zlib compression, at the expense of greater CPU and
+#      memory overhead.
+#
+#      LZMA is not the standard compression used in Squashfs and so most
+#      file systems will be readable without selecting this option.
+#     If unsure, say N.
+
+menuconfig SQUASHFS_LZMA
+    bool "  Include support for LZMA compressed file systems"
+
+    depends on SQUASHFS
+
+    select LZMADEVICE
+    select DECOMPRESS_LZMA
+    select DECOMPRESS_LZMA_NEEDED
+    default y
+
+if SQUASHFS_LZMA
+choice
+	prompt "Lzma Decompressor parallelisation options"
+	depends on SQUASHFS
+config S_DECOMPRESS_LZMA
+bool "SOFTWARE	support for LZMA  file systems"
+    depends on SQUASHFS
+    select DECOMPRESS_LZMA
+    select DECOMPRESS_LZMA_NEEDED
+
+config H_DECOMPRESS_LZMA
+bool "JZ_HARDWARE	support for LZMA  file systems"
+    depends on SQUASHFS
+    select DECOMPRESS_LZMA
+    select DECOMPRESS_LZMA_NEEDED
+    select LZMADEVICE
+endchoice
+endif
+
 
 config SQUASHFS_4K_DEVBLK_SIZE
 	bool "Use 4K device block size?"
