diff -drupN a/arch/mips/xburst2/core/include/jz_proc.h b/arch/mips/xburst2/core/include/jz_proc.h
--- a/arch/mips/xburst2/core/include/jz_proc.h	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/mips/xburst2/core/include/jz_proc.h	2022-06-09 05:02:27.000000000 +0300
@@ -0,0 +1,18 @@
+#ifndef _JZ_PROC_H_
+#define _JZ_PROC_H_
+
+struct jz_single_file_ops {
+	ssize_t (*read) (struct seq_file *, void *);
+	ssize_t (*write) (struct file *, const char __user *, size_t, loff_t *);
+	void *data;
+};
+
+struct proc_dir_entry * jz_proc_mkdir(char *s);
+struct proc_dir_entry *jz_proc_create_data(
+	const char *name, umode_t mode, struct proc_dir_entry *parent,
+	struct jz_single_file_ops *proc_fops, void *data);
+struct proc_dir_entry *jz_proc_create(
+	const char *name, umode_t mode, struct proc_dir_entry *parent,
+	struct jz_single_file_ops *proc_fops);
+struct proc_dir_entry * get_jz_proc_root(void);
+#endif /* _JZ_PROC_H_ */
