diff -drupN a/arch/mips/xburst/soc-x1000/include/soc/tcsm_layout.h b/arch/mips/xburst/soc-x1000/include/soc/tcsm_layout.h
--- a/arch/mips/xburst/soc-x1000/include/soc/tcsm_layout.h	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/mips/xburst/soc-x1000/include/soc/tcsm_layout.h	2022-06-09 05:02:27.000000000 +0300
@@ -0,0 +1,53 @@
+#ifndef __TCSM_LAYOUT_H__
+#define __TCSM_LAYOUT_H__
+
+/**
+ *      |-------------|
+ *      |    BANK0    |
+ *      |-------------|     <--- SLEEP_TCSM_BOOTCODE_TEXT
+ *      |             |
+ *      |             |
+ *      |             |
+ *      |             |
+ *      |             |
+ *      |             |
+ *      |             |
+ *      |-------------|
+ *      | VO DMA DESC |
+ *      |_____________|     <--- VOICE_TCSM_DMA_DESC
+ *
+ *
+ *      |-------------|
+ *      |    BANK1    |
+ *      |-------------|     <--- VOICE_TCSM_DATA_BUF
+ *      |             |
+ *      |             |
+ *      |    ...      |
+ *      |  VOICE DATA |
+ *      |    ...      |
+ *      |             |
+ *      |             |
+ *      |_____________|
+ *
+ */
+
+/*
+ * If you use the VOICE TRIGGER,
+ * and you change the VOICE_* and TCSM_BANK_LEN,
+ * while compiling wakeup module
+ */
+#define SLEEP_TCSM_SPACE           0xb3422000
+#define VOICE_TCSM_DATA_BUF	       0xb3423000
+#define TCSM_BANK_LEN              4096
+
+#define SLEEP_TCSM_BOOT_TEXT       (SLEEP_TCSM_SPACE)
+
+#define VOICE_TCSM_DMA_DESC_LEN    512
+#define SLEEP_TCSM_SPACE_END       (SLEEP_TCSM_SPACE + TCSM_BANK_LEN - VOICE_TCSM_DMA_DESC_LEN)
+#define VOICE_TCSM_DMA_DESC_ADDR   SLEEP_TCSM_SPACE_END
+
+#define SLEEP_TCSM_CPU_RESMUE_SP   SLEEP_TCSM_SPACE_END - 4
+
+#define VOICE_TCSM_DATA_BUF_SIZE   TCSM_BANK_LEN
+
+#endif	/*  __TCSM_LAYOUT_H__ */
