diff -drupN a/arch/mips/xburst/soc-x1800/include/soc/extal.h b/arch/mips/xburst/soc-x1800/include/soc/extal.h
--- a/arch/mips/xburst/soc-x1800/include/soc/extal.h	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/mips/xburst/soc-x1800/include/soc/extal.h	2022-06-09 05:02:27.000000000 +0300
@@ -0,0 +1,8 @@
+
+#ifndef __JZSOC_EXTAL_H__
+#define __JZSOC_EXTAL_H__
+
+#define JZ_EXTAL_RTC	32768     /* RTC extal freq: 32.768 KHz */
+#define JZ_EXTAL	(CONFIG_EXTAL_CLOCK * 1000000)
+
+#endif
