diff -drupN a/arch/mips/xburst/soc-x1800/include/soc/irq.h b/arch/mips/xburst/soc-x1800/include/soc/irq.h
--- a/arch/mips/xburst/soc-x1800/include/soc/irq.h	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/mips/xburst/soc-x1800/include/soc/irq.h	2022-06-09 05:02:27.000000000 +0300
@@ -0,0 +1,84 @@
+/*
+ * IRQ number in JZ47xx INTC definition.
+ *   Only support 4770 now. 2011-9-23
+ *
+ * Copyright (C) 2010 Ingenic Semiconductor Co., Ltd.
+ */
+
+#ifndef __INTC_IRQ_H__
+#define __INTC_IRQ_H__
+
+#include <irq.h>
+
+enum {
+// interrupt controller interrupts
+	IRQ_DMIC=XBURST_SOC_IRQ_BASE,
+	IRQ_AIC0,
+	IRQ_BCH,
+	IRQ_SYS_OST,
+	IRQ_CSI,
+	IRQ_OHCI,
+	IRQ_IPU,
+	IRQ_SFC,
+	IRQ_SSI1,
+	IRQ_SSI0,
+	IRQ_PDMA,
+	IRQ_RESERVED11,
+	IRQ_GPIO5,
+	IRQ_GPIO4,
+	IRQ_GPIO3,
+	IRQ_GPIO2,
+	IRQ_GPIO1,
+	IRQ_GPIO0,
+#define IRQ_GPIO_PORT(N) (IRQ_GPIO0 - (N))
+	IRQ_SADC,
+	IRQ_VO,
+	IRQ_EHCI,
+	IRQ_OTG,
+	IRQ_RESERVED22,
+	IRQ_AES,
+	IRQ_DES,
+	IRQ_TCU2,
+	IRQ_TCU1,
+	IRQ_TCU0,
+	IRQ_RESERVED28,
+	IRQ_ISP,
+	IRQ_RESERVED30,
+	IRQ_LCD,
+
+	IRQ_RTC,
+	IRQ_RESERVED33,
+	IRQ_RESERVED34,
+	IRQ_MSC2,
+	IRQ_MSC1,
+	IRQ_MSC0,
+	IRQ_RESERVED38,
+	IRQ_NFI,
+	IRQ_PCM0,
+	IRQ_RESERVED41,
+	IRQ_RESERVED42,
+	IRQ_RESERVED43,
+	IRQ_HARB2,
+	IRQ_HARB1,
+	IRQ_HARB0,
+	IRQ_CPM,
+	IRQ_UART3,
+	IRQ_UART2,
+	IRQ_UART1,
+	IRQ_UART0,
+	IRQ_DDR,
+	IRQ_RESERVED53,
+	IRQ_EFUSE,
+	IRQ_GMAC,
+	IRQ_I2C4,
+	IRQ_I2C3,
+	IRQ_I2C2,
+	IRQ_I2C1,
+	IRQ_I2C0,
+	IRQ_PDMAM,
+#define IRQ_VPU(ID)	(IRQ_VPU0 + ID)
+	IRQ_VPU0,
+	IRQ_VPU1,
+};
+
+#endif
