diff -drupN a/drivers/video/ingenic_i2d/Kconfig b/drivers/video/ingenic_i2d/Kconfig
--- a/drivers/video/ingenic_i2d/Kconfig	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/video/ingenic_i2d/Kconfig	2022-06-09 05:02:34.000000000 +0300
@@ -0,0 +1,13 @@
+menuconfig JZ_I2D
+    bool "JZ I2D Driver"
+    default y
+    help
+        Support for Ingenic i2d Driver
+
+config INGENIC_I2D
+	bool "Ingenic I2D Driver"
+    depends on JZ_I2D
+    default y
+	help
+	  Support for Ingenic i2d  operations.
+
