diff -drupN a/include/linux/decompress/mm.h b/include/linux/decompress/mm.h
--- a/include/linux/decompress/mm.h	2017-10-21 18:09:07.000000000 +0300
+++ b/include/linux/decompress/mm.h	2022-06-09 05:02:35.000000000 +0300
@@ -61,7 +61,7 @@ static void free(void *where)
 #define large_malloc(a) malloc(a)
 #define large_free(a) free(a)
 
-#define INIT
+//#define INIT
 
 #else /* STATIC */
 
@@ -83,7 +83,7 @@ static void free(void *where)
 #define large_malloc(a) vmalloc(a)
 #define large_free(a) vfree(a)
 
-#define INIT __init
+//#define INIT __init
 #define STATIC
 
 #include <linux/init.h>
