diff -drupN a/drivers/mmc/core/core.c b/drivers/mmc/core/core.c
--- a/drivers/mmc/core/core.c	2017-10-21 18:09:07.000000000 +0300
+++ b/drivers/mmc/core/core.c	2022-06-09 05:02:30.000000000 +0300
@@ -1485,7 +1485,7 @@ int mmc_regulator_get_supply(struct mmc_
 	if (IS_ERR(mmc->supply.vmmc)) {
 		if (PTR_ERR(mmc->supply.vmmc) == -EPROBE_DEFER)
 			return -EPROBE_DEFER;
-		dev_info(dev, "No vmmc regulator found\n");
+		//dev_info(dev, "No vmmc regulator found\n");
 	} else {
 		ret = mmc_regulator_get_ocrmask(mmc->supply.vmmc);
 		if (ret > 0)
@@ -1497,7 +1497,7 @@ int mmc_regulator_get_supply(struct mmc_
 	if (IS_ERR(mmc->supply.vqmmc)) {
 		if (PTR_ERR(mmc->supply.vqmmc) == -EPROBE_DEFER)
 			return -EPROBE_DEFER;
-		dev_info(dev, "No vqmmc regulator found\n");
+		//dev_info(dev, "No vqmmc regulator found\n");
 	}
 
 	return 0;
@@ -2578,7 +2578,7 @@ void mmc_rescan(struct work_struct *work
 	/* If there is a non-removable card registered, only scan once */
 	if ((host->caps & MMC_CAP_NONREMOVABLE) && host->rescan_entered)
 		return;
-	host->rescan_entered = 1;
+//	host->rescan_entered = 1;
 
 	mmc_bus_get(host);
 
