diff -drupN a/arch/mips/boot/zcompressed/tools/filesize b/arch/mips/boot/zcompressed/tools/filesize
--- a/arch/mips/boot/zcompressed/tools/filesize	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/mips/boot/zcompressed/tools/filesize	2022-06-09 05:02:27.000000000 +0300
@@ -0,0 +1,7 @@
+#!/bin/sh
+HOSTNAME=`uname`
+if [ "$HOSTNAME" = "Linux" ]; then
+echo `ls -l $1 | awk '{print $5}'`
+else
+echo `ls -l $1 | awk '{print $6}'`
+fi
