diff -drupN a/drivers/video/fbdev/ingenic/Kconfig b/drivers/video/fbdev/ingenic/Kconfig
--- a/drivers/video/fbdev/ingenic/Kconfig	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/video/fbdev/ingenic/Kconfig	2022-06-09 05:02:34.000000000 +0300
@@ -0,0 +1,80 @@
+menuconfig FB_INGENIC
+	tristate "Ingenic Framebuffer Driver"
+	depends on FB
+	select FB_CFB_FILLRECT
+	select FB_CFB_COPYAREA
+	select FB_CFB_IMAGEBLIT
+    default n
+	help
+		LCD Driver for INGENIC
+
+config FB_VSYNC_SKIP_DISABLE
+        bool "Disable Vsync skip"
+	depends on FB_INGENIC
+        help
+          Disable Vsync skip
+
+config FB_VSYNC_SKIP
+        int "Vsync skip ratio[0..9]"
+	depends on FB_INGENIC
+        default 9
+        help
+          Vsync skip ratio
+
+config FB_INGENIC_NR_FRAMES
+	int "how many frames support(max=3)"
+	depends on FB_INGENIC
+	default 2
+
+config FB_INGENIC_NR_LAYERS
+	int "how many layers support(max=4)"
+	depends on FB_INGENIC
+	default 1
+
+config FB_INGENIC_DEBUG
+        bool "fb test for displaying color bar"
+        depends on FB_INGENIC
+        default n
+        help
+           fb test for displaying color bar in your board.
+
+config SLCDC_CONTINUA
+    tristate "SLCDC CONTINUA TRANFER"
+    depends on FB_INGENIC
+    default n
+
+config SLCDC_USE_TE
+	tristate "SLCDC USE TE SIGNAL"
+	depends on FB_INGENIC
+	default n
+
+config FB_INGENIC_MIPI_DSI
+	tristate "Ingenic MIPI DSI Interface"
+	depends on FB_INGENIC
+	default n
+
+config MIPI_TX_LANE
+	tristate "MIPI_TX_LANES"
+	depends on FB_INGENIC_MIPI_DSI
+	choice
+		prompt "2 LANE or 4 LANE"
+		depends on MIPI_TX_LANE
+		config MIPI_4LANE
+			bool "4LANE"
+		config MIPI_2LANE
+			bool "2LANE"
+	endchoice
+
+config INGENIC_FB_BOOT_PATTERN
+	tristate "Ingenic Janus boot image"
+	depends on FB_INGENIC
+	default n
+	help
+	    display the boot image of Janus in your board
+
+if FB_INGENIC
+source "drivers/video/fbdev/ingenic/fb_v10/Kconfig"
+source "drivers/video/fbdev/ingenic/fb_v11/Kconfig"
+source "drivers/video/fbdev/ingenic/fb_v12/Kconfig"
+endif
+
