diff -drupN a/fs/squashfs/decompressor_single.c b/fs/squashfs/decompressor_single.c
--- a/fs/squashfs/decompressor_single.c	2017-10-21 18:09:07.000000000 +0300
+++ b/fs/squashfs/decompressor_single.c	2022-06-09 05:02:35.000000000 +0300
@@ -68,8 +68,11 @@ int squashfs_decompress(struct squashfs_
 	struct squashfs_stream *stream = msblk->stream;
 
 	mutex_lock(&stream->mutex);
+//	printk("squashfs single decompress start: %s : %d time is %d\n",__func__,__LINE__,1000 * *(volatile u32 *)0xb0002078 * 16 / ((24 * 1000 * 1000) / 512) );
+//	printk("offset = 0x%x   length = 0x%x\n",offset,length);
 	res = msblk->decompressor->decompress(msblk, stream->stream, bh, b,
 		offset, length, output);
+//	printk("squashfs single decompress end: %s : %d time is %d\n",__func__,__LINE__,1000 * *(volatile u32 *)0xb0002078 * 16 / ((24 * 1000 * 1000) / 512) );
 	mutex_unlock(&stream->mutex);
 
 	if (res < 0)
