diff -drupN a/arch/mips/Kconfig b/arch/mips/Kconfig
--- a/arch/mips/Kconfig	2017-10-21 18:09:07.000000000 +0300
+++ b/arch/mips/Kconfig	2022-06-09 05:02:26.000000000 +0300
@@ -299,7 +299,7 @@ config MACH_JAZZ
 	 Olivetti M700-10 workstations.
 
 config MACH_INGENIC
-	bool "Ingenic SoC based machines"
+	bool "Ingenic SoC based machines, only for jz4740,4780"
 	select SYS_SUPPORTS_32BIT_KERNEL
 	select SYS_SUPPORTS_LITTLE_ENDIAN
 	select SYS_SUPPORTS_ZBOOT_UART16550
@@ -312,6 +312,53 @@ config MACH_INGENIC
 	select USE_OF
 	select LIBFDT
 
+config MACH_XBURST
+	bool "Ingenic Xburst based machines"
+	select SYS_SUPPORTS_32BIT_KERNEL
+	select SYS_SUPPORTS_LITTLE_ENDIAN
+	select SYS_HAS_CPU_MIPS32_R1
+	select SYS_HAS_CPU_MIPS32_R2
+	select SYS_HAS_EARLY_PRINTK
+	select MIPS_EXTERNAL_TIMER
+	select DMA_NONCOHERENT
+	select HAVE_CLK
+	select ARCH_REQUIRE_GPIOLIB
+	select SYS_SUPPORTS_HIGHMEM
+	select XBURST_CPU_SCACHE
+	select DMA_INGENIC_HIGHMEM_FLUSH
+	select HAVE_KERNEL_LZMA
+	select IRQ_MIPS_CPU
+	select USE_OF
+	select LIBFDT
+	select COMMON_CLK
+
+config MACH_XBURST2
+	bool "Ingenic Xburst2 based machines"
+	select SYS_SUPPORTS_32BIT_KERNEL
+	select SYS_SUPPORTS_LITTLE_ENDIAN
+	select SYS_HAS_CPU_MIPS32_R1
+	select SYS_HAS_CPU_MIPS32_R2
+	select SYS_HAS_EARLY_PRINTK
+	select SYS_SUPPORTS_SMP
+	select SYS_SUPPORTS_HOTPLUG_CPU
+	select MIPS_EXTERNAL_TIMER
+	select DMA_NONCOHERENT
+	select HAVE_CLK
+	select ARCH_REQUIRE_GPIOLIB
+	select SYS_SUPPORTS_HIGHMEM
+	select XBURST2_CPU_SCACHE
+	select DMA_INGENIC_HIGHMEM_FLUSH
+    select SYS_SUPPORTS_ZBOOT
+	select IRQ_MIPS_CPU
+	select USE_OF
+	select LIBFDT
+	select MIPS_O32_FP64_SUPPORT
+	select WEAK_ORDERING
+	select WEAK_REORDERING_BEYOND_LLSC
+	select COMMON_CLK
+	select OF_RESERVED_MEM
+
+
 config LANTIQ
 	bool "Lantiq based platforms"
 	select DMA_NONCOHERENT
@@ -994,6 +1041,8 @@ source "arch/mips/loongson64/Kconfig"
 source "arch/mips/netlogic/Kconfig"
 source "arch/mips/paravirt/Kconfig"
 source "arch/mips/xilfpga/Kconfig"
+source "arch/mips/xburst/Kconfig"
+source "arch/mips/xburst2/Kconfig"
 
 endmenu
 
@@ -1086,6 +1135,9 @@ config DMA_MAYBE_COHERENT
 	select DMA_NONCOHERENT
 	bool
 
+config DMA_INGENIC_HIGHMEM_FLUSH
+	bool
+
 config DMA_COHERENT
 	bool
 
@@ -2119,6 +2171,15 @@ config MIPS_CPU_SCACHE
 	bool
 	select BOARD_SCACHE
 
+config XBURST_CPU_SCACHE
+	bool
+	select BOARD_SCACHE
+
+config XBURST2_CPU_SCACHE
+	bool
+	select BOARD_SCACHE
+	select CPU_SUPPORTS_CPUFREQ
+
 config R5000_CPU_SCACHE
 	bool
 	select BOARD_SCACHE
