diff -drupN a/arch/mips/include/asm/fpu.h b/arch/mips/include/asm/fpu.h
--- a/arch/mips/include/asm/fpu.h	2017-10-21 18:09:07.000000000 +0300
+++ b/arch/mips/include/asm/fpu.h	2022-06-09 05:02:27.000000000 +0300
@@ -41,7 +41,7 @@ extern void _restore_fp(struct task_stru
  */
 enum fpu_mode {
 	FPU_32BIT = 0,		/* FR = 0 */
-	FPU_64BIT,		/* FR = 1, FRE = 0 */
+	FPU_64BIT,		/* FR = 1 */
 	FPU_AS_IS,
 	FPU_HYBRID,		/* FR = 1, FRE = 1 */
 
@@ -74,11 +74,17 @@ static inline int __enable_fpu(enum fpu_
 		goto fr_common;
 
 	case FPU_64BIT:
-#if !(defined(CONFIG_CPU_MIPSR2) || defined(CONFIG_CPU_MIPSR6) \
-      || defined(CONFIG_64BIT))
-		/* we only have a 32-bit FPU */
+               change_c0_status(ST0_CU1 | ST0_FR, ST0_CU1 | ST0_FR);
+               enable_fpu_hazard();
+               /* check FR has the desired value */
+               if (read_c0_status() & ST0_FR)
+                       return 0;
+
+               /* unsupported FR value */
+               __disable_fpu();
+               /* we only have a 32-bit FPU. not used for xburst2. qiao */
 		return SIGFPE;
-#endif
+
 		/* fall through */
 	case FPU_32BIT:
 		if (cpu_has_fre) {
