diff -drupN a/include/dt-bindings/dma/ingenic-pdma.h b/include/dt-bindings/dma/ingenic-pdma.h
--- a/include/dt-bindings/dma/ingenic-pdma.h	1970-01-01 03:00:00.000000000 +0300
+++ b/include/dt-bindings/dma/ingenic-pdma.h	2022-06-09 05:02:35.000000000 +0300
@@ -0,0 +1,47 @@
+#ifndef __INGENIC_PDMA_H__
+#define __INGENIC_PDMA_H__
+
+#define INGENIC_DMA_REQ_AIC_LOOP_RX	0x5
+#define INGENIC_DMA_REQ_AIC_TX	    0x6
+#define INGENIC_DMA_REQ_AIC_F_RX	0x7
+#define INGENIC_DMA_REQ_AUTO_TX	    0x8
+#define INGENIC_DMA_REQ_SADC_RX     0x9
+#define INGENIC_DMA_REQ_DMIC_RX	    0xd
+#define INGENIC_DMA_REQ_UART3_TX	0xe
+#define INGENIC_DMA_REQ_UART3_RX	0xf
+#define INGENIC_DMA_REQ_UART2_TX	0x10
+#define INGENIC_DMA_REQ_UART2_RX	0x11
+#define INGENIC_DMA_REQ_UART1_TX	0x12
+#define INGENIC_DMA_REQ_UART1_RX	0x13
+#define INGENIC_DMA_REQ_UART0_TX	0x14
+#define INGENIC_DMA_REQ_UART0_RX	0x15
+#define INGENIC_DMA_REQ_SSI0_TX	    0x16
+#define INGENIC_DMA_REQ_SSI0_RX	    0x17
+#define INGENIC_DMA_REQ_SSI1_TX	    0x18
+#define INGENIC_DMA_REQ_SSI1_RX	    0x19
+#define INGENIC_DMA_REQ_SLV_TX	    0x1a
+#define INGENIC_DMA_REQ_SLV_RX	    0x1b
+#define INGENIC_DMA_REQ_I2C0_TX	    0x24
+#define INGENIC_DMA_REQ_I2C0_RX	    0x25
+#define INGENIC_DMA_REQ_I2C1_TX	    0x26
+#define INGENIC_DMA_REQ_I2C1_RX	    0x27
+#define INGENIC_DMA_REQ_I2C2_TX	    0x28
+#define INGENIC_DMA_REQ_I2C2_RX	    0x29
+#define INGENIC_DMA_REQ_I2C3_TX	    0x2a
+#define INGENIC_DMA_REQ_I2C3_RX	    0x2b
+#define INGENIC_DMA_REQ_DES_TX	    0x2e
+#define INGENIC_DMA_REQ_DES_RX	    0x2f
+
+#define INGENIC_DMA_TYPE_REQ_MSK	0xff
+#define INGENIC_DMA_TYPE_CH_SFT	    8
+#define INGENIC_DMA_TYPE_CH_MSK	    (0xff << INGENIC_DMA_TYPE_CH_SFT)
+#define INGENIC_DMA_TYPE_CH_EN	    (1 << 16)
+#define INGENIC_DMA_TYPE_PROG		(1 << 17)
+#define INGENIC_DMA_TYPE_SPEC		(1 << 18)
+
+#define INGENIC_DMA_CH(ch)		((((ch) << INGENIC_DMA_TYPE_CH_SFT) & INGENIC_DMA_TYPE_CH_MSK) | INGENIC_DMA_TYPE_CH_EN)
+#define INGENIC_DMA_TYPE(type)	((type) & INGENIC_DMA_TYPE_REQ_MSK)
+#define INGENIC_DMA_TYPE_CH(type, ch)	(INGENIC_DMA_TYPE((type)) | INGENIC_DMA_CH((ch)))
+#define INGENIC_DMA_PG_CH(type, ch)	(INGENIC_DMA_TYPE_CH((type), (ch)) | INGENIC_DMA_TYPE_PROG_MSK)
+#define INGENIC_DMA_SP_CH(type, ch)	(INGENIC_DMA_PG_CH(type, id) | INGENIC_DMA_TYPE_SPEC_MSK)
+#endif	/* __INGENIC_PDMA_H__ */
