diff -drupN a/drivers/misc/Makefile b/drivers/misc/Makefile
--- a/drivers/misc/Makefile	2017-10-21 18:09:07.000000000 +0300
+++ b/drivers/misc/Makefile	2022-06-09 05:02:30.000000000 +0300
@@ -45,6 +45,7 @@ obj-$(CONFIG_ARM_CHARLCD)	+= arm-charlcd
 obj-$(CONFIG_PCH_PHUB)		+= pch_phub.o
 obj-y				+= ti-st/
 obj-y				+= lis3lv02d/
+obj-$(CONFIG_BCM_43438_RFKILL)   += bt_power_bluesleep.o
 obj-$(CONFIG_USB_SWITCH_FSA9480) += fsa9480.o
 obj-$(CONFIG_ALTERA_STAPL)	+=altera-stapl/
 obj-$(CONFIG_INTEL_MEI)		+= mei/
@@ -56,3 +57,6 @@ obj-$(CONFIG_GENWQE)		+= genwqe/
 obj-$(CONFIG_ECHO)		+= echo/
 obj-$(CONFIG_VEXPRESS_SYSCFG)	+= vexpress-syscfg.o
 obj-$(CONFIG_CXL_BASE)		+= cxl/
+obj-$(CONFIG_RMEM)		+= rmem.o
+obj-$(CONFIG_INGENIC_RSA) 	+= ingenic_rsa.o
+obj-$(CONFIG_LOGGER)		+= logger.o
