diff -drupN a/arch/mips/xburst2/core/include/irq_cpu.h b/arch/mips/xburst2/core/include/irq_cpu.h
--- a/arch/mips/xburst2/core/include/irq_cpu.h	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/mips/xburst2/core/include/irq_cpu.h	2022-06-09 05:02:27.000000000 +0300
@@ -0,0 +1,6 @@
+#ifndef IRQ_CPU_H
+#define IRQ_CPU_H
+
+void ingenic_irq_migration(int lock);
+void ingenic_irq_cpumask_idle(int idle);
+#endif /* IRQ_CPU_H */
