diff -drupN a/drivers/media/platform/ingenic-vcodec/helix/h264enc/common.c b/drivers/media/platform/ingenic-vcodec/helix/h264enc/common.c
--- a/drivers/media/platform/ingenic-vcodec/helix/h264enc/common.c	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/media/platform/ingenic-vcodec/helix/h264enc/common.c	2022-06-09 05:02:29.000000000 +0300
@@ -0,0 +1,11 @@
+#include "common.h"
+
+
+
+int c_clip3(int v, int i_min, int i_max)
+{
+    return ((v < i_min) ? i_min : ((v > i_max) ? i_max : v));
+}
+
+
+
