diff -drupN a/arch/mips/include/asm/cop2.h b/arch/mips/include/asm/cop2.h
--- a/arch/mips/include/asm/cop2.h	2017-10-21 18:09:07.000000000 +0300
+++ b/arch/mips/include/asm/cop2.h	2022-06-09 05:02:27.000000000 +0300
@@ -11,7 +11,15 @@
 
 #include <linux/notifier.h>
 
-#if defined(CONFIG_CPU_CAVIUM_OCTEON)
+#if defined(CONFIG_XBURST_MXUV2)
+#define cop2_present		1
+extern void xburst_cop2_save(struct xburst_cop2_state *);
+extern void xburst_cop2_restore(struct xburst_cop2_state *);
+#define cop2_save(r)            xburst_cop2_save(&(r)->thread.cp2)
+#define cop2_restore(r)         xburst_cop2_restore(&(r)->thread.cp2)
+#define cop2_lazy_restore       0
+
+#elif defined(CONFIG_CPU_CAVIUM_OCTEON)
 
 extern void octeon_cop2_save(struct octeon_cop2_state *);
 extern void octeon_cop2_restore(struct octeon_cop2_state *);
