diff -drupN a/drivers/mmc/host/Kconfig b/drivers/mmc/host/Kconfig
--- a/drivers/mmc/host/Kconfig	2017-10-21 18:09:07.000000000 +0300
+++ b/drivers/mmc/host/Kconfig	2022-06-09 05:02:30.000000000 +0300
@@ -4,6 +4,17 @@
 
 comment "MMC/SD/SDIO Host Controller Drivers"
 
+config MMC_SDHCI_INGENIC
+        tristate "Ingenic(XBurst2)  MMC/SD Card Controller(MSC) support"
+        depends on SOC_T40
+        select MMC_SDHCI
+        help
+          This selects the Ingenic XBurst2 SD/MMC Card Controller MSC.
+          If you have platform with a SD/Multimedia Card slot and compact
+          with this version, say Y or M here.
+          If unsure, say N.
+
+
 config MMC_ARMMMCI
 	tristate "ARM AMBA Multimedia Card Interface support"
 	depends on ARM_AMBA
@@ -689,6 +700,23 @@ config MMC_JZ4740
 	  If you have a board based on such a SoC and with a SD/MMC slot,
 	  say Y or M here.
 
+config INGENIC_MMC
+	tristate "Ingenic(XBurst)  MMC/SD Card Controller(MSC) support"
+	depends on MACH_XBURST
+	help
+	  This selects the Ingenic XBurst SD/MMC Card Controller.
+	  If you have platform with a SD/Multimedia Card slot and compact
+	  with this version, say Y or M here.
+	  If unsure, say N.
+
+config INGENIC_MMC_MMC0
+	bool "INGENIC_MMC MMC0"
+	depends on INGENIC_MMC
+
+config INGENIC_MMC_MMC1
+	bool "INGENIC_MMC MMC1"
+	depends on INGENIC_MMC
+
 config MMC_VUB300
 	tristate "VUB300 USB to SDIO/SD/MMC Host Controller support"
 	depends on USB
