diff -Naur a/arch/mips/boot/dts/ingenic/shark.dts b/arch/mips/boot/dts/ingenic/shark.dts
--- a/arch/mips/boot/dts/ingenic/shark.dts	2023-04-21 19:13:07.904606139 +0300
+++ b/arch/mips/boot/dts/ingenic/shark.dts	2023-04-21 21:14:03.645013643 +0300
@@ -118,7 +118,7 @@
 	ingenic,rst-ms = <10>;
 	ingenic,mac-mode = <RMII>;
 	ingenic,mode-reg = <0xb00000e4>;
-	ingenic,phy-clk-freq = <50000000>;
+	ingenic,phy-clk-freq = <25000000>;
 };
 
 &sfc {
