diff -drupN a/drivers/clk/ingenic/Makefile b/drivers/clk/ingenic/Makefile
--- a/drivers/clk/ingenic/Makefile	2017-10-21 18:09:07.000000000 +0300
+++ b/drivers/clk/ingenic/Makefile	2022-06-09 05:02:28.000000000 +0300
@@ -1,3 +1,7 @@
-obj-y				+= cgu.o
-obj-$(CONFIG_MACH_JZ4740)	+= jz4740-cgu.o
-obj-$(CONFIG_MACH_JZ4780)	+= jz4780-cgu.o
+
+
+obj-$(CONFIG_COMMON_CLK) += clk.o clk-pll.o clk-div.o clk-bus.o power-gate.o
+obj-$(CONFIG_SOC_X2000_V12)	+= clk-x2000-v12.o
+obj-$(CONFIG_SOC_T40)	+= clk-t40.o
+
+
