diff -drupN a/drivers/usb/core/hub.c b/drivers/usb/core/hub.c
--- a/drivers/usb/core/hub.c	2017-10-21 18:09:07.000000000 +0300
+++ b/drivers/usb/core/hub.c	2022-06-09 05:02:34.000000000 +0300
@@ -2414,7 +2414,7 @@ int usb_new_device(struct usb_device *ud
 			(((udev->bus->busnum-1) * 128) + (udev->devnum-1)));
 
 	/* Tell the world! */
-	announce_device(udev);
+	/*announce_device(udev);*/
 
 	if (udev->serial)
 		add_device_randomness(udev->serial, strlen(udev->serial));
