diff -drupN a/drivers/block/zram/zram_drv.c b/drivers/block/zram/zram_drv.c
--- a/drivers/block/zram/zram_drv.c	2017-10-21 18:09:07.000000000 +0300
+++ b/drivers/block/zram/zram_drv.c	2022-06-09 05:02:28.000000000 +0300
@@ -41,7 +41,7 @@ static int zram_major;
 static const char *default_compressor = "lzo";
 
 /* Module params (documentation at end) */
-static unsigned int num_devices = 1;
+static unsigned int num_devices = 2;
 
 static inline void deprecated_attr_warn(const char *name)
 {
