diff -drupN a/drivers/video/fbdev/ingenic/fb_v11/displays/Makefile b/drivers/video/fbdev/ingenic/fb_v11/displays/Makefile
--- a/drivers/video/fbdev/ingenic/fb_v11/displays/Makefile	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/video/fbdev/ingenic/fb_v11/displays/Makefile	2022-06-09 05:02:34.000000000 +0300
@@ -0,0 +1,3 @@
+obj-$(CONFIG_PANEL_X2000_BYD9158B)	+= panel-byd9158B.o
+obj-$(CONFIG_PANEL_X2000_KD035HVFMD057)	+= panel-kd035hvfmd057.o
+obj-$(CONFIG_PANEL_X2000_KD035HVFBD037)	+= panel-kd035hvfbd037.o
