diff -drupN a/fs/squashfs/decompressor.h b/fs/squashfs/decompressor.h
--- a/fs/squashfs/decompressor.h	2017-10-21 18:09:07.000000000 +0300
+++ b/fs/squashfs/decompressor.h	2022-06-09 05:02:35.000000000 +0300
@@ -58,4 +58,8 @@ extern const struct squashfs_decompresso
 extern const struct squashfs_decompressor squashfs_zlib_comp_ops;
 #endif
 
+#ifdef CONFIG_SQUASHFS_LZMA
+extern const struct squashfs_decompressor squashfs_lzma_comp_ops;
+#endif
+
 #endif
