diff -drupN a/arch/mips/boot/dts/ingenic/halley2v20.dts b/arch/mips/boot/dts/ingenic/halley2v20.dts
--- a/arch/mips/boot/dts/ingenic/halley2v20.dts	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/mips/boot/dts/ingenic/halley2v20.dts	2022-06-09 05:02:27.000000000 +0300
@@ -0,0 +1,366 @@
+/dts-v1/;
+
+/memreserve/ 0x01f00000 0x00100000;     /* dmic reserved */
+
+#include <dt-bindings/input/input.h>
+#include <dt-bindings/interrupt-controller/irq.h>
+#include "x1000.dtsi"
+
+&msc0 {
+	status = "okay";
+	pinctrl-names ="default";
+	pinctrl-0 = <&msc0_4bit_pa>;
+	cap-sd-highspeed;
+	cap-mmc-highspeed;
+	cd-inverted;
+	max-frequency = <50000000>;
+	bus-width = <4>;
+	voltage-ranges = <1800 3300>;
+	ingenic,sdio_clk = <0>;
+	/* special property */
+	ingenic,wp-gpios = <0>;
+	ingneic,cd-gpios = <0>;
+	ingenic,rst-gpios = <0>;
+	ingenic,removal-dontcare; /*removal-dontcare, removal-nonremovable, removal-removable, removal-manual*/
+};
+
+&msc1 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+
+	status = "okay";
+	pinctrl-names ="default","enable", "disable";
+	pinctrl-0 = <&msc1_4bit_pc>;
+	pinctrl-1 = <&rtc32k_enable>;
+	pinctrl-2 = <&rtc32k_disable>;
+	cap-sd-highspeed;
+	cap-mmc-highspeed;
+	max-frequency = <48000000>;
+	bus-width = <4>;
+	voltage-ranges = <1800 3300>;
+	non-removable;
+
+	ingenic,sdio_clk = <1>;
+	keep-power-in-suspend;
+	/* special property */
+	ingenic,wp-gpios = <0>;
+	ingneic,cd-gpios = <0>;
+	ingenic,rst-gpios = <0>;
+	ingenic,removal-manual; /*removal-dontcare, removal-nonremovable, removal-removable, removal-manual*/
+
+	#address-cells = <1>;
+	#size-cells = <0>;
+	bcmdhd_wlan: bcmdhd_wlan {
+		 compatible = "android,bcmdhd_wlan";
+		 ingenic,sdio-irq = <&gpc 16 IRQ_TYPE_LEVEL_HIGH INGENIC_GPIO_NOBIAS>;
+		 ingenic,sdio-reset = <&gpc 17 GPIO_ACTIVE_LOW INGENIC_GPIO_NOBIAS>;
+	};
+};
+
+&i2c0 {
+	status = "okay";
+	clock-frequency = <100000>;
+	timeout = <1000>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&i2c0_pb>;
+
+	ov5640: ov5640@0x3c {
+		status = "disabled";
+		compatible = "ovti,ov5640";
+		reg = <0x3c>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&cim_pa>;
+
+		resetb-gpios 	= <&gpd 5 GPIO_ACTIVE_LOW INGENIC_GPIO_NOBIAS>;
+		pwdn-gpios 	= <&gpd 4 GPIO_ACTIVE_HIGH INGENIC_GPIO_NOBIAS>;
+		vcc-en-gpios	= <&gpd 3 GPIO_ACTIVE_HIGH INGENIC_GPIO_NOBIAS>;
+
+		/*gsensor-supply = <&LDO3>;*/
+		/*lcd-supply = <&LDO5>;*/
+		/*supplies-name = "vcc_gsensor", "lcd_3v3";*/
+
+		port {
+			ov5640_0: endpoint {
+				remote-endpoint = <&cim_0>;
+			};
+		};
+	};
+	gc2155: gc2155@0x3c {
+		status = "okay";
+		compatible = "ovti,gc2155";
+		reg = <0x3c>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&cim_pa>;
+
+		resetb-gpios 	= <&gpd 5 GPIO_ACTIVE_LOW INGENIC_GPIO_NOBIAS>;
+		pwdn-gpios 	= <&gpd 4 GPIO_ACTIVE_HIGH INGENIC_GPIO_NOBIAS>;
+		vcc-en-gpios	= <&gpd 3 GPIO_ACTIVE_HIGH INGENIC_GPIO_NOBIAS>;
+
+		/*gsensor-supply = <&LDO3>;*/
+		/*lcd-supply = <&LDO5>;*/
+		/*supplies-name = "vcc_gsensor", "lcd_3v3";*/
+
+		port {
+			gc2155_0: endpoint {
+				remote-endpoint = <&cim_0>;
+			};
+		};
+	};
+};
+
+&spi0 {
+	status = "disabled";
+	pinctrl-names = "default";
+	pinctrl-0 = <&spi_pd>;
+
+	spi-max-frequency = <54000000>;
+	num-cs = <2>;
+	cs-gpios = <0>, <0>;
+/*	cs-gpios = <&gpa 27 GPIO_ACTIVE_HIGH INGENIC_GPIO_NOBIAS>, <&gpa 27 GPIO_ACTIVE_HIGH INGENIC_GPIO_NOBIAS>;*/
+	ingenic,chnl = <0>;
+	ingenic,allow_cs_same = <1>;
+	ingenic,bus_num = <0>;
+	ingenic,has_dma_support = <0>;
+	ingenic,spi-src-clk = <0>;/*0, ext; 1, sfc*/
+
+	dac0: dh2228@0 {
+		compatible = "spidev";
+		reg = <2>;
+		chip_select= <0>;
+		spi-max-frequency = <100000>;
+	};
+};
+
+&cim {
+	status = "okay";
+	port {
+		cim_0: endpoint@0 {
+			/*remote-endpoint = <&ov5640_0>;*/
+			remote-endpoint = <&gc2155_0>;
+			bus-width = <8>;
+			/*vsync-active = 1;*/
+			/*hsync-active = 1;*/
+		};
+	};
+};
+
+&lcd {
+	status = "okay";
+	pinctrl-names = "default";
+	pinctrl-0 = <&lcd_pa>, <&lcd_pb>;
+};
+
+&i2c1 {
+	clock-frequency = <400000>;
+	timeout = <1000>;
+	status = "disabled";
+	pinctrl-names = "default";
+	pinctrl-0 = <&i2c1_pc>;
+};
+
+&i2c2 {
+	status = "disabled";
+};
+
+&uart0 {
+	pinctrl-names = "default";
+	status = "okay";
+	pinctrl-0 = <&uart0_pc_fc>;
+};
+
+&uart2 {
+	pinctrl-names = "default";
+	status = "okay";
+	pinctrl-0 = <&uart2_pc>;
+};
+
+&pinctrl {
+	mac_rmii_p0: mac-rmii-p0 {
+		mac_rmii_p0_normal: mac-rmii-p0-normal {
+			ingenic,pinmux = <&gpb 7 9>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+		};
+		mac_rmii_p0_rst: mac-rmii-p0-rst {
+			ingenic,pinmux = <&gpb 7 9>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCLOLVL>;
+		};
+	};
+	mac_rmii_p1: mac-rmii-p1 {
+		ingenic,pinmux = <&gpb 10 15>;
+		ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+	};
+	mac_rmii_nomdio_p1: mac-rmii-nomdio-p1 {
+		ingenic,pinmux = <&gpb 10 12>, <&gpb 15 15>;
+		ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+	};
+	gpiokeys: gpiokeys {
+		gpio_power_key:gpio_power_key {
+			ingenic,pinmux = <&gpb 31 31>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCLOLVL>;
+		};
+	};
+};
+
+&mac {
+	pinctrl-names = "default", "reset";
+	/*	pinctrl-0 = <&mac_rmii_p0_normal>, <&mac_rmii_p1>;
+	pinctrl-1 = <&mac_rmii_p0_rst>, <&mac_rmii_p1>;*/
+	pinctrl-0 = <&mac_rmii_p0_normal>, <&mac_rmii_nomdio_p1>;
+	pinctrl-1 = <&mac_rmii_p0_rst>, <&mac_rmii_nomdio_p1>;
+	status = "okay";
+	ingenic,rst-gpio = <&gpb 3 GPIO_ACTIVE_LOW INGENIC_GPIO_NOBIAS>;
+	ingenic,rst-ms = <10>;
+};
+
+&sfc {
+	status = "okay";
+	ingenic,sfc-max-frequency = <150000000>;
+	ingenic,use_board_info      = /bits/ 8 <0>;
+};
+
+&otg {
+	g-use-dma;
+	status = "okay";
+};
+
+&otg_phy {
+	dr_mode = "otg";
+	ingenic,drvvbus-gpio = <&gpb 25 GPIO_ACTIVE_HIGH INGENIC_GPIO_NOBIAS>;
+};
+
+/ {
+	model = "halley2-v20";
+
+	memory {
+		device_type = "memory";
+		linux,usable-memory = <0x00000000 0x2000000>;
+	};
+
+	aliases: aliases {
+	};
+	dump_dmic_codec:dump_dmic_codec{
+		compatible = "ingenic,dmic-dump-codec";
+		status = "ok";
+	};
+	dump_spdif_codec:dump_spdif_codec{
+		compatible = "ingenic,spdif-dump-codec";
+		status = "ok";
+	};
+	dump_pcm_codec:dump_pcm_codec{
+		compatible = "ingenic,pcm-dump-codec";
+		status = "ok";
+	};
+	sound_halley2_icdc {
+		status = "ok";
+		compatible = "ingenic,x1000-sound";
+		ingenic,model = "halley2";
+		ingenic,dai-link = "i2s-icdc", "dmic", "pcm";
+		ingenic,stream = "main", "dmic", "pcm";
+		ingenic,cpu-dai = <&i2s>, <&dmic>, <&pcm>;
+		ingenic,platform = <&aic>, <&dmic>, <&pcm>;
+		ingenic,codec = <&codec>, <&dump_dmic_codec>, <&dump_pcm_codec>;
+		ingenic,codec-dai = "icdc-d3-hifi", "dmic-dump", "pcm-dump";
+		ingenic,audio-routing = "Speaker", "DO_LO_PWM",
+			"MICBIAS", "Mic Buildin",
+			"DMIC", "DMic";
+		/*ingenic,spken-gpio = <&gpb 0 GPIO_ACTIVE_LOW INGENIC_GPIO_NOBIAS>;*/
+	};
+	sound_halley2_spdif {
+		status = "disabled";
+		compatible = "ingenic,x1000-sound";
+		ingenic,model = "halley2";
+		ingenic,dai-link = "spdif", "dmic", "pcm";
+		ingenic,stream = "main", "dmic", "pcm";
+		ingenic,cpu-dai = <&spdif>, <&dmic>, <&pcm>;
+		ingenic,platform = <&aic>, <&dmic>, <&pcm>;
+		ingenic,codec = <&dump_spdif_codec>, <&dump_dmic_codec>, <&dump_pcm_codec>;
+		ingenic,codec-dai = "spdif-dump", "dmic-dump", "pcm-dump";
+		/*ingenic,spken-gpio = <&gpb 0 GPIO_ACTIVE_LOW INGENIC_GPIO_NOBIAS>;*/
+	};
+
+	panel_truly@0 {
+			compatible = "ingenic,truly_tft240240_2_e";
+			status = "okay";
+			ingenic,cs-gpio = <&gpb 18 GPIO_ACTIVE_LOW INGENIC_GPIO_NOBIAS>;
+			ingenic,rst-gpio = <&gpd 0 GPIO_ACTIVE_LOW INGENIC_GPIO_NOBIAS>;
+			ingenic,blken-gpio = <&gpd 1 GPIO_ACTIVE_HIGH INGENIC_GPIO_NOBIAS>;
+			/*ingenic,pwmen-gpio = <&gpc 25 GPIO_ACTIVE_HIGH INGENIC_GPIO_NOBIAS>;*/
+	};
+
+	backlight {
+		compatible = "pwm-backlight";
+		pwms = <&pwmz 0 500000>;
+		brightness-levels = <0 4 8 16 24 32 40 64 128 255>;
+		default-brightness-level = <5>;
+	};
+
+pwmz:pwm {
+		compatible = "ingenic,pwm";
+		#pwm-cells = <2>;
+		pinctrl-0 = <&pwm0_pc>;
+		pinctrl-names = "default";
+		pwm0:pwm@0 {
+			ingenic,timer-parent = <&channel0>;
+			status = "okay";
+		};
+		pwm1:pwm@1 {
+			ingenic,timer-parent = <&channel1>;
+			status = "okay";
+		};
+		pwm2:pwm@2 {
+			ingenic,timer-parent = <&channel2>;
+			status = "disabled";
+		};
+		pwm3:pwm@3 {
+			ingenic,timer-parent = <&channel3>;
+			status = "disabled";
+		};
+		pwm4:pwm@4 {
+			ingenic,timer-parent = <&channel4>;
+			status = "disabled";
+		};
+	};
+
+	extclk: extclk {
+		clock-frequency = <24000000>;
+	};
+	gpio_keys: gpio_keys {
+		compatible = "gpio-keys";
+
+		power {
+		   label = "Power";
+		   linux,code = <KEY_POWER>;
+           gpios = <&gpb 31 GPIO_ACTIVE_LOW INGENIC_GPIO_NOBIAS>;
+		   gpio-key,wakeup;
+		};
+	};
+	bt_power{
+		compatible = "ingenic,bt_power";
+		ingenic,reg-on-gpio = <&gpc 18 GPIO_ACTIVE_LOW INGENIC_GPIO_NOBIAS>;
+		ingenic,wake-gpio = <&gpc 20 GPIO_ACTIVE_LOW INGENIC_GPIO_NOBIAS>;
+	};
+};
+
+&gpa {
+	ingenic,gpio-sleep-pull = <0 1 2 3 4 5 6 7 8 9 10 11
+		12 13 14 15 16 17 18 19 20 21 22 23 24 25 26
+		28 29 30 31>;
+	ingenic,gpio-sleep-high = <27>;
+};
+
+&gpb {
+	ingenic,gpio-sleep-pull = <0 1 2 3 4 5 6 7 8 9 10 11 12 13 14
+		15 16 17 18 19 20 21 22 23 24 26>;
+	ingenic,gpio-sleep-npul = <28 29 30>;
+	ingenic,gpio-sleep-low = <25 27>;
+};
+
+&gpc {
+	ingenic,gpio-sleep-pull = <21 23 24 25 26 27>;
+	ingenic,gpio-sleep-npul = <0 1 2 3 4 5 6 7 8 9 10 11 12 13 22>;
+};
+
+&gpd {
+	ingenic,gpio-sleep-pull = <0 1 2 3 4 5>;
+};
+
