diff -drupN a/drivers/irqchip/Makefile b/drivers/irqchip/Makefile
--- a/drivers/irqchip/Makefile	2017-10-21 18:09:07.000000000 +0300
+++ b/drivers/irqchip/Makefile	2022-06-09 05:02:29.000000000 +0300
@@ -55,3 +55,7 @@ obj-$(CONFIG_RENESAS_H8S_INTC)		+= irq-r
 obj-$(CONFIG_ARCH_SA1100)		+= irq-sa11x0.o
 obj-$(CONFIG_INGENIC_IRQ)		+= irq-ingenic.o
 obj-$(CONFIG_IMX_GPCV2)			+= irq-imx-gpcv2.o
+obj-$(CONFIG_INGENIC_INTC)		+= irq-ingenic.o
+obj-$(CONFIG_IRQ_INGENIC_CPU)		+= irq-ingenic-cpu.o
+obj-$(CONFIG_INGENIC_INTC_CHIP)		+= irq-ingenic-chip.o
+
