diff -drupN a/drivers/spi/Kconfig b/drivers/spi/Kconfig
--- a/drivers/spi/Kconfig	2017-10-21 18:09:07.000000000 +0300
+++ b/drivers/spi/Kconfig	2022-06-09 05:02:33.000000000 +0300
@@ -53,6 +53,17 @@ if SPI_MASTER
 
 comment "SPI Master Controller Drivers"
 
+config INGENIC_SPI
+	tristate "Ingenic SPI Controller"
+	depends on MACH_XBURST2
+	select SPI_BITBANG
+	help
+		SPI driver for Ingenic series SoCs
+
+config INGENIC_SPI0
+    bool "Ingenic SoC SSI controller 0 for SPI Host driver"
+	depends on INGENIC_SPI
+
 config SPI_ALTERA
 	tristate "Altera SPI Controller"
 	select SPI_BITBANG
