diff -drupN a/sound/soc/Kconfig b/sound/soc/Kconfig
--- a/sound/soc/Kconfig	2017-10-21 18:09:07.000000000 +0300
+++ b/sound/soc/Kconfig	2022-06-09 05:02:37.000000000 +0300
@@ -67,6 +67,7 @@ source "sound/soc/txx9/Kconfig"
 source "sound/soc/ux500/Kconfig"
 source "sound/soc/xtensa/Kconfig"
 source "sound/soc/zte/Kconfig"
+source "sound/soc/ingenic/Kconfig"
 
 # Supported codecs
 source "sound/soc/codecs/Kconfig"
