diff -drupN a/drivers/net/Kconfig b/drivers/net/Kconfig
--- a/drivers/net/Kconfig	2017-10-21 18:09:07.000000000 +0300
+++ b/drivers/net/Kconfig	2022-06-09 05:02:30.000000000 +0300
@@ -25,9 +25,6 @@ menuconfig NETDEVICES
 # that for each of the symbols.
 if NETDEVICES
 
-config MII
-	tristate
-
 config NET_CORE
 	default y
 	bool "Network core driver support"
@@ -103,6 +100,13 @@ config NET_FC
 	  adaptor below. You also should have said Y to "SCSI support" and
 	  "SCSI generic support".
 
+config MII
+	tristate "Generic Media Independent Interface device support"
+	help
+	  Most ethernet controllers have MII transceiver either as an external
+	  or internal device.  It is safe to say Y or M here even if your
+	  ethernet card lacks MII.
+
 config IFB
 	tristate "Intermediate Functional Block support"
 	depends on NET_CLS_ACT
@@ -423,6 +427,6 @@ config FUJITSU_ES
 	  This driver provides support for Extended Socket network device
           on Extended Partitioning of FUJITSU PRIMEQUEST 2000 E2 series.
 
-source "drivers/net/hyperv/Kconfig"
+source "drivers/net/ethernet/Kconfig"
 
 endif # NETDEVICES
