diff -drupN a/arch/mips/xburst2/common/cpu_ddr_test/multithread_test.h b/arch/mips/xburst2/common/cpu_ddr_test/multithread_test.h
--- a/arch/mips/xburst2/common/cpu_ddr_test/multithread_test.h	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/mips/xburst2/common/cpu_ddr_test/multithread_test.h	2022-06-09 05:02:27.000000000 +0300
@@ -0,0 +1,9 @@
+#ifndef _MULTITHREAD_TEST_H_
+#define _MULTITHREAD_TEST_H_
+
+typedef int (*TEST_FUNC)(void *);
+int multithread_test_add_func(void *handle,const char *name,TEST_FUNC testfunc,void *param);
+void* multithread_test_init(const char *name,int threadcount);
+void multithread_test_deinit(void* handle);
+
+#endif /* _MULTITHREAD_TEST_H_ */
