diff -drupN a/drivers/usb/dwc2/hw.h b/drivers/usb/dwc2/hw.h
--- a/drivers/usb/dwc2/hw.h	2017-10-21 18:09:07.000000000 +0300
+++ b/drivers/usb/dwc2/hw.h	2022-06-09 05:02:34.000000000 +0300
@@ -52,6 +52,8 @@
 #define GOTGCTL_HSTSETHNPEN		(1 << 10)
 #define GOTGCTL_HNPREQ			(1 << 9)
 #define GOTGCTL_HSTNEGSCS		(1 << 8)
+#define GOTGCTL_VBVALIDOVVAL		(1 << 3)
+#define GOTGCTL_VBVALIDOVEN		(1 << 2)
 #define GOTGCTL_SESREQ			(1 << 1)
 #define GOTGCTL_SESREQSCS		(1 << 0)
 
