diff -drupN a/arch/mips/include/asm/r4kcache.h b/arch/mips/include/asm/r4kcache.h
--- a/arch/mips/include/asm/r4kcache.h	2017-10-21 18:09:07.000000000 +0300
+++ b/arch/mips/include/asm/r4kcache.h	2022-06-09 05:02:27.000000000 +0300
@@ -48,6 +48,41 @@ extern void (*r4k_blast_icache)(void);
 	:								\
 	: "i" (op), "R" (*(unsigned char *)(addr)))
 
+#ifdef CONFIG_MACH_XBURST
+#define __inv_btb()	\
+	do {			\
+		unsigned long tmp;		\
+		__asm__ __volatile__(		\
+				".set push\n\t"	\
+				".set noreorder\n\t"	\
+				".set mips32\n\t"	\
+				"mfc0 %0, $16, 7\n\t"	\
+				"nop\n\t"		\
+				"ori %0, 2\n\t"		\
+				"mtc0 %0, $16, 7\n\t"	\
+				"nop\n\t"		\
+				".set pop\n\t"		\
+				: "=&r" (tmp));		\
+	} while(0)
+#define __sync_wb()	__sync()
+#define __bridge_sync_war()	\
+	do {		\
+		if (MIPS_BRIDGE_SYNC_WAR)	\
+			__fast_iob();	\
+	} while (0)
+static inline void blast_inclusive_scache(void)
+{
+	__bridge_sync_war();
+}
+#else
+#define __inv_btb()	do {} while(0)
+#define __sync_wb()	do {} while(0)
+#define __bridge_sync_war() do {} while(0)
+static inline void blast_inclusive_scache(void)
+{
+}
+#endif
+
 #ifdef CONFIG_MIPS_MT
 
 #define __iflush_prologue						\
@@ -76,13 +111,13 @@ extern void (*r4k_blast_icache)(void);
 #else /* CONFIG_MIPS_MT */
 
 #define __iflush_prologue {
-#define __iflush_epilogue }
+#define __iflush_epilogue __inv_btb(); }	/*CONFIG_MACH_XBURST*/
 #define __dflush_prologue {
-#define __dflush_epilogue }
+#define __dflush_epilogue __sync_wb(); }	/*CONFIG_MACH_XBURST*/
 #define __inv_dflush_prologue {
 #define __inv_dflush_epilogue }
 #define __sflush_prologue {
-#define __sflush_epilogue }
+#define __sflush_epilogue __bridge_sync_war(); }	/*CONFIG_MACH_XBURST*/
 #define __inv_sflush_prologue {
 #define __inv_sflush_epilogue }
 
@@ -105,6 +140,7 @@ static inline void flush_dcache_line_ind
 static inline void flush_scache_line_indexed(unsigned long addr)
 {
 	cache_op(Index_Writeback_Inv_SD, addr);
+	__bridge_sync_war();	/*CONFIG_MACH_XBURST*/
 }
 
 static inline void flush_icache_line(unsigned long addr)
@@ -144,6 +180,7 @@ static inline void invalidate_scache_lin
 static inline void flush_scache_line(unsigned long addr)
 {
 	cache_op(Hit_Writeback_Inv_SD, addr);
+	__bridge_sync_war();	/*CONFIG_MACH_XBURST*/
 }
 
 #define protected_cache_op(op,addr)				\
@@ -189,6 +226,7 @@ static inline void protected_flush_icach
 #else
 		protected_cache_op(Hit_Invalidate_I, addr);
 #endif
+		__inv_btb();	/*CONFIG_MACH_XBURST*/
 		break;
 	}
 }
@@ -206,11 +244,14 @@ static inline void protected_writeback_d
 #else
 	protected_cache_op(Hit_Writeback_Inv_D, addr);
 #endif
+	__sync_wb();	/*CONFIG_MACH_XBURST*/
 }
 
 static inline void protected_writeback_scache_line(unsigned long addr)
 {
 	protected_cache_op(Hit_Writeback_Inv_SD, addr);
+
+	__bridge_sync_war();	/*CONFIG_MACH_XBURST*/
 }
 
 /*
