diff -drupN a/arch/mips/include/asm/pgtable-bits.h b/arch/mips/include/asm/pgtable-bits.h
--- a/arch/mips/include/asm/pgtable-bits.h	2017-10-21 18:09:07.000000000 +0300
+++ b/arch/mips/include/asm/pgtable-bits.h	2022-06-09 05:02:27.000000000 +0300
@@ -249,6 +249,17 @@ static inline uint64_t pte_to_entrylo(un
 /* Ingenic uses the WA bit to achieve write-combine memory writes */
 #define _CACHE_UNCACHED_ACCELERATED (1<<_CACHE_SHIFT)
 
+#elif defined(CONFIG_MACH_XBURST) || defined(CONFIG_MACH_XBURST2)
+
+#define _CACHE_CACHABLE_WT_WA      (0<<_CACHE_SHIFT)  /* R4600 only      */
+#define _CACHE_UNCACHED_ACCELERATED    (1<<_CACHE_SHIFT)  /* R4600 only          */
+#define _CACHE_UNCACHED                        (2<<_CACHE_SHIFT)  /* R4[0246]00          */
+#define _CACHE_CACHABLE_WB_WA       (3<<_CACHE_SHIFT)  /* R4[0246]00     */
+#define _CACHE_CACHABLE_WT_WA_S            (4<<_CACHE_SHIFT)  /* R4[04]00MC only */
+#define _CACHE_CACHABLE_WB_WA_S            (5<<_CACHE_SHIFT)  /* R4[04]00MC only */
+#define _CACHE_CACHABLE_WA             _CACHE_UNCACHED_ACCELERATED
+#define _CACHE_CACHABLE_NONCOHERENT _CACHE_CACHABLE_WB_WA
+
 #endif
 
 #ifndef _CACHE_CACHABLE_NO_WA
