diff -drupN a/init/Kconfig b/init/Kconfig
--- a/init/Kconfig	2017-10-21 18:09:07.000000000 +0300
+++ b/init/Kconfig	2022-06-09 05:02:36.000000000 +0300
@@ -44,6 +44,9 @@ config INIT_ENV_ARG_LIMIT
 	  Maximum of each of the number of arguments and environment
 	  variables passed to init from the kernel command line.
 
+config FAST_BOOT
+	bool "Kernel fast start up"
+	default n
 
 config CROSS_COMPILE
 	string "Cross-compiler tool prefix"
@@ -1422,15 +1425,16 @@ config SYSCTL_SYSCALL
 
 config KALLSYMS
 	 bool "Load all symbols for debugging/ksymoops" if EXPERT
+	 depends on !FAST_BOOT
 	 default y
 	 help
 	   Say Y here to let the kernel print out symbolic crash information and
-	   symbolic stack backtraces. This increases the size of the kernel
+	   symbolic stack backtraces. This increases the size of the /kernel
 	   somewhat, as all symbols have to be loaded into the kernel image.
 
 config KALLSYMS_ALL
 	bool "Include all symbols in kallsyms"
-	depends on DEBUG_KERNEL && KALLSYMS
+	depends on DEBUG_KERNEL && KALLSYMS && !FAST_BOOT
 	help
 	   Normally kallsyms only contains the symbols of functions for nicer
 	   OOPS messages and backtraces (i.e., symbols from the text and inittext
