diff -drupN a/drivers/media/platform/soc_camera/Makefile b/drivers/media/platform/soc_camera/Makefile
--- a/drivers/media/platform/soc_camera/Makefile	2017-10-21 18:09:07.000000000 +0300
+++ b/drivers/media/platform/soc_camera/Makefile	2022-06-09 05:02:29.000000000 +0300
@@ -14,3 +14,4 @@ obj-$(CONFIG_VIDEO_PXA27x)		+= pxa_camer
 obj-$(CONFIG_VIDEO_SH_MOBILE_CEU)	+= sh_mobile_ceu_camera.o
 obj-$(CONFIG_VIDEO_SH_MOBILE_CSI2)	+= sh_mobile_csi2.o
 obj-$(CONFIG_VIDEO_RCAR_VIN)		+= rcar_vin.o
+obj-$(CONFIG_VIDEO_INGENIC)		+= ingenic/
