diff -drupN a/drivers/media/v4l2-core/Kconfig b/drivers/media/v4l2-core/Kconfig
--- a/drivers/media/v4l2-core/Kconfig	2017-10-21 18:09:07.000000000 +0300
+++ b/drivers/media/v4l2-core/Kconfig	2022-06-09 05:02:30.000000000 +0300
@@ -99,6 +99,20 @@ config VIDEOBUF2_VMALLOC
 	select VIDEOBUF2_MEMOPS
 	select DMA_SHARED_BUFFER
 
+config VIDEOBUF2_DMA_CONTIG_INGENIC
+        tristate
+        depends on HAS_DMA
+        select VIDEOBUF2_CORE
+        select VIDEOBUF2_MEMOPS
+        select DMA_SHARED_BUFFER
+
+config CAMERA_RESERVE_KB_SIZE
+	int "Reserve memory for camera noncoherent memory. Unit:KByte"
+	default 0
+	depends on VIDEOBUF2_DMA_CONTIG_INGENIC
+
+
+
 config VIDEOBUF2_DMA_SG
 	tristate
 	depends on HAS_DMA
