diff -drupN a/arch/mips/boot/dts/ingenic/x2000-v12-pinctrl.dtsi b/arch/mips/boot/dts/ingenic/x2000-v12-pinctrl.dtsi
--- a/arch/mips/boot/dts/ingenic/x2000-v12-pinctrl.dtsi	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/mips/boot/dts/ingenic/x2000-v12-pinctrl.dtsi	2022-06-09 05:02:27.000000000 +0300
@@ -0,0 +1,508 @@
+#include <dt-bindings/pinctrl/ingenic-pinctrl.h>
+
+&pinctrl {
+	uart0_pin: uart0-pin {
+		uart0_pd: uart0-pd {
+			ingenic,pinmux = <&gpd 23 26>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION2>;
+		};
+	};
+	uart1_pin: uart1-pin {
+		uart1_pc: uart1-pc {
+			ingenic,pinmux = <&gpc 21 24>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+		};
+	};
+
+	uart2_pin: uart2-pin {
+		uart2_pd: uart2-pd {
+			ingenic,pinmux = <&gpd 30 31>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+	};
+
+	uart3_pin: uart3-pin {
+		uart3_pd: uart3-pd {
+			ingenic,pinmux = <&gpd 0 3>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+		};
+	};
+
+	uart4_pin: uart4-pin {
+		uart4_pa: uart4-pa {
+			ingenic,pinmux = <&gpa 0 3>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+		};
+		uart4_pc: uart4-pc {
+			ingenic,pinmux = <&gpc 9 12>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION3>;
+		};
+	};
+
+	uart5_pin: uart5-pin {
+		uart5_pa: uart5-pa {
+			ingenic,pinmux = <&gpa 4 5>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+		};
+		uart5_pc: uart5-pc {
+			ingenic,pinmux = <&gpc 5 6>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION3>;
+		};
+	};
+
+	uart6_pin: uart6-pin {
+		uart6_pa: uart6-pa {
+			ingenic,pinmux = <&gpa 6 7>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+		};
+		uart6_pc: uart6-pc {
+			ingenic,pinmux = <&gpc 7 8>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION3>;
+		};
+	};
+
+	uart7_pin: uart7-pin {
+		uart7_pa: uart7-pa {
+			ingenic,pinmux = <&gpa 8 9>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+		};
+		uart7_pc: uart7-pc {
+			ingenic,pinmux = <&gpc 1 2>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION3>;
+		};
+	};
+
+	uart8_pin: uart8-pin {
+		uart8_pb: uart8-pb {
+			ingenic,pinmux = <&gpb 28 29>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION3>;
+		};
+	};
+
+	uart9_pin: uart9-pin {
+		uart9_pb: uart9-pb {
+			ingenic,pinmux = <&gpb 30 31>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION3>;
+		};
+	};
+
+	i2c0_pin: i2c0-pin {
+		i2c0_pc: i2c0-pc {
+			ingenic,pinmux = <&gpc 13 14>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION3>;
+		};
+	};
+
+	i2c1_pin: i2c1-pin {
+		i2c1_pc: i2c1-pc {
+			ingenic,pinmux = <&gpc 23 24>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION2>;
+		};
+	};
+
+	i2c2_pin: i2c2-pin {
+		i2c2_pb: i2c2-pb {
+			ingenic,pinmux = <&gpb 22 23>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION2>;
+		};
+	};
+
+	i2c3_pin: i2c3-pin {
+		i2c3_pd: i2c3-pd {
+			ingenic,pinmux = <&gpd 30 31>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+		};
+	};
+
+	i2c4_pin: i2c4-pin {
+		i2c4_pd: i2c4-pd {
+			ingenic,pinmux = <&gpd 0 1>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION2>;
+		};
+	};
+
+	i2c5_pin: i2c5-pin {
+		i2c5_pd: i2c5-pd {
+			ingenic,pinmux = <&gpd 4 5>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+		};
+	};
+
+	msc0_pin: msc0-pin {
+		msc0_4bit: msc0-4bit {
+			ingenic,pinmux = <&gpd 17 22>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+		msc0_8bit: msc0-8bit {
+			ingenic,pinmux = <&gpd 17 26>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+	};
+
+	msc1_pin: msc1-pin {
+		msc1_4bit: msc1-4bit {
+			ingenic,pinmux = <&gpd 8 13>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+	};
+
+	msc2_pin: msc2-pin {
+		msc2_4bit: msc2-4bit {
+			ingenic,pinmux = <&gpe 0 5>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+	};
+
+	mac0_rmii_p0: mac0-rmii-p0 {
+		mac0_rmii_p0_normal: mac0-rmii-normal {
+			ingenic,pinmux = <&gpc 6 7>, <&gpc 11 11>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+		};
+		mac0_rmii_p0_rst: mac0-rmii-p0-rst {
+			ingenic,pinmux = <&gpc 6 7>, <&gpc 11 11>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCHILVL>;
+		};
+	};
+
+	mac0_rmii_p1: mac0-rmii-p1 {
+		mac0_rmii_p1_normal: mac0-rmii-p1-normal {
+			ingenic,pinmux = <&gpc 1 3>, <&gpc 10 10>, <&gpc 12 13>, <&gpc 15 15>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+		};
+
+		mac0_rmii_p1_nomdio: mac0-rmii-p1-nomdio {
+			ingenic,pinmux = <&gpc 1 3>, <&gpc 10 10>, <&gpc 15 15>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+		};
+	};
+
+
+	mac0_rgmii_p0: mac0-rgmii-p0 {
+		mac0_rgmii_p0_normal: mac0-rgmii-normal {
+			ingenic,pinmux = <&gpc 6 9>, <&gpc 11 11>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+		};
+		mac0_rgmii_p0_rst: mac0-rmii-p0-rst {
+			ingenic,pinmux = <&gpc 6 9>, <&gpc 11 11>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCHILVL>;
+		};
+	};
+
+	mac0_rgmii_p1: mac0-rgmii-p1 {
+		mac0_rgmii_p1_normal: mac0-rgmii-p1-normal {
+			ingenic,pinmux = <&gpc 1 5>, <&gpc 10 10>, <&gpc 12 15>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+		};
+
+		mac0_rgmii_p1_nomdio: mac0-rmii-p1-nomdio {
+			ingenic,pinmux = <&gpc 1 5>, <&gpc 10 10>, <&gpc 14 15>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+		};
+	};
+
+	mac1_rmii_p0: mac1-rmii-p0 {
+		mac1_rmii_p0_normal: mac1-rmii-normal {
+			ingenic,pinmux = <&gpb 12 13>, <&gpb 18 18>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION3>;
+		};
+		mac1_rmii_p0_rst: mac1-rmii-p0-rst {
+			ingenic,pinmux = <&gpb 12 13>, <&gpb 18 18>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCHILVL>;
+		};
+	};
+
+	mac1_rmii_p1: mac1-rmii-p1 {
+		mac1_rmii_p1_normal: mac1-rmii-p1-normal {
+		ingenic,pinmux = <&gpb 8 9>, <&gpb 17 17>, <&gpb 19 20>, <&gpb 22 23>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION3>;
+		};
+
+		mac1_rmii_p1_nomdio: mac1-rmii-p1-nomdio {
+		ingenic,pinmux = <&gpb 8 9>, <&gpb 17 17>, <&gpb 22 23>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION3>;
+		};
+	};
+
+	mac1_rgmii_p0: mac1-rgmii-p0 {
+		mac1_rgmii_p0_normal: mac1-rgmii-normal {
+			ingenic,pinmux = <&gpb 12 15>, <&gpb 18 18>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION3>;
+		};
+		mac1_rgmii_p0_rst: mac1-rmii-p0-rst {
+			ingenic,pinmux = <&gpb 12 15>, <&gpb 18 18>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCHILVL>;
+		};
+	};
+
+	mac1_rgmii_p1: mac1-rgmii-p1 {
+		mac1_rgmii_p1_normal: mac1-rgmii-p1-normal {
+			ingenic,pinmux = <&gpb 8 11>, <&gpb 17 17>, <&gpb 19 23>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION3>;
+		};
+
+		mac1_rgmii_p1_nomdio: mac1-rmii-p1-nomdio {
+			ingenic,pinmux = <&gpb 8 11>, <&gpb 17 17>, <&gpb 21 23>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION3>;
+		};
+	};
+
+	pwm_pin: pwm-pin{
+		pwm_pc: pwm-pc{
+			ingenic,pinmux = <&gpc 0 15>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+	 };
+
+	sfc_pin: sfc-pin {
+		sfc_pd: sfc-pd {
+			ingenic,pinmux = <&gpd 17 26>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+		};
+
+		sfc_pe: sfc-pe {
+			ingenic,pinmux = <&gpe 16 21>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+	};
+
+	cim_pin: cim-pin {
+		cim_pa: cim-pa {
+			ingenic,pinmux = <&gpa 0 7>, <&gpa 12 16>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+	 };
+	vic_pin: vic-pin {
+		vic_pa: vic-pa {
+			ingenic,pinmux = <&gpa 0 15>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+	 };
+
+	lcd_pins: lcd-pins {
+		tft_lcd_pb: tft-lcd-pb {
+			ingenic,pinmux = <&gpb 0 27>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+		};
+
+		smart_lcd_pb: smart-lcd-pb {
+			ingenic,pinmux = <&gpb 0 15>, <&gpb 25 27>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION2>;
+		};
+	};
+
+	baic1_tmclk_pc: baic1-tmclk-pc {
+		ingenic,pinmux = <&gpc 8 8>;
+		ingenic,pinmux-funcsel = <PINCTL_FUNCTION2>;
+	};
+
+	baic1_rmclk_pc: baic1-rmclk-pc {
+		ingenic,pinmux = <&gpc 1 1>;
+		ingenic,pinmux-funcsel = <PINCTL_FUNCTION2>;
+	};
+
+	baic1_pin: baic1-pin {
+		baic1_pc: baic1-pin-pc {
+			ingenic,pinmux = <&gpc 2 7>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION2>;
+		};
+		baic1_combine_pc: baic1-cpin-pc {
+			ingenic,pinmux = <&gpc 4 7>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION2>;
+		};
+	};
+
+	baic2_mclk_pa: baic2-mclk-pa {
+		ingenic,pinmux = <&gpa 7 7>;
+		ingenic,pinmux-funcsel = <PINCTL_FUNCTION2>;
+	};
+
+	baic2_pa: baic2-pa {
+		ingenic,pinmux = <&gpa 8 13>;
+		ingenic,pinmux-funcsel = <PINCTL_FUNCTION2>;
+	};
+
+	baic3_mclk_pa: baic3-mclk-pa {
+		ingenic,pinmux = <&gpa 0 0>;
+		ingenic,pinmux-funcsel = <PINCTL_FUNCTION2>;
+	};
+
+	baic3_pa: baic3-pa {
+		ingenic,pinmux = <&gpa 1 6>;
+		ingenic,pinmux-funcsel = <PINCTL_FUNCTION2>;
+	};
+
+	baic4_pin: baic4-pin {
+		baic4_pe: baic4-pe {
+			ingenic,pinmux = <&gpe 0 3>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+		};
+
+		baic4_pd: baic4-pd {
+			ingenic,pinmux = <&gpd 2 5>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION2>;
+		};
+	};
+
+	spdif_pc: spdif-pc {
+		ingenic,pinmux = <&gpc 13 14>;
+		ingenic,pinmux-funcsel = <PINCTL_FUNCTION2>;
+	};
+
+	dmic_pc: dmic-pc {
+		ingenic,pinmux = <&gpc 20 24>;
+		ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+	};
+
+	spi0_pb: spi0-pb {
+		ingenic,pinmux = <&gpb 28 31>;
+		ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+	};
+
+	spi1_pc: spi0-pc {
+		ingenic,pinmux = <&gpc 9 12>;
+		ingenic,pinmux-funcsel = <PINCTL_FUNCTION2>;
+	};
+
+	pwm_pins: pwm-pins {
+		pwm0_pc: pwm0_pc {
+			ingenic,pinmux = <&gpc 0 0>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+		pwm1_pc: pwm1_pc {
+			ingenic,pinmux = <&gpc 1 1>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+		pwm2_pc: pwm2_pc {
+			ingenic,pinmux = <&gpc 2 2>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+		pwm3_pc: pwm3_pc {
+			ingenic,pinmux = <&gpc 3 3>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+		pwm4_pc: pwm4_pc {
+			ingenic,pinmux = <&gpc 4 4>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+		pwm5_pc: pwm5_pc {
+			ingenic,pinmux = <&gpc 5 5>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+		pwm6_pc: pwm6_pc {
+			ingenic,pinmux = <&gpc 6 6>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+		pwm7_pc: pwm7_pc {
+			ingenic,pinmux = <&gpc 7 7>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+		pwm8_pc: pwm8_pc {
+			ingenic,pinmux = <&gpc 8 8>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+		pwm9_pc: pwm9_pc {
+			ingenic,pinmux = <&gpc 9 9>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+		pwm10_pc: pwm10_pc {
+			ingenic,pinmux = <&gpc 10 10>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+		pwm10_pe: pwm10_pe {
+			ingenic,pinmux = <&gpe 16 16>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+		};
+		pwm11_pc: pwm11_pc {
+			ingenic,pinmux = <&gpc 11 11>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+		pwm11_pe: pwm11_pe {
+			ingenic,pinmux = <&gpe 17 17>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+		};
+		pwm12_pc: pwm12_pc {
+			ingenic,pinmux = <&gpc 12 12>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+		pwm12_pe: pwm12_pe {
+			ingenic,pinmux = <&gpe 18 18>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+		};
+		pwm13_pc: pwm13_pc {
+			ingenic,pinmux = <&gpc 13 13>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+		pwm13_pe: pwm13_pe {
+			ingenic,pinmux = <&gpe 19 19>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+		};
+		pwm14_pc: pwm14_pc {
+			ingenic,pinmux = <&gpc 14 14>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+		pwm14_pe: pwm14_pe {
+			ingenic,pinmux = <&gpe 20 20>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+		};
+		pwm15_pc: pwm15_pc {
+			ingenic,pinmux = <&gpc 15 15>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+		pwm15_pe: pwm15_pe {
+			ingenic,pinmux = <&gpe 21 21>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+		};
+	};
+	tcu_pin: tcu-pin {
+		tcu0_pc: tcu0_pc {
+			ingenic,pinmux = <&gpc 0 1>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+		tcu1_pc: tcu1_pc {
+			ingenic,pinmux = <&gpc 2 3>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+		tcu2_pc: tcu2_pc {
+			ingenic,pinmux = <&gpc 4 5>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+		tcu3_pc: tcu3_pc {
+			ingenic,pinmux = <&gpc 6 7>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+		tcu4_pc: tcu4_pc {
+			ingenic,pinmux = <&gpc 8 9>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+		tcu5_pc: tcu5_pc {
+			ingenic,pinmux = <&gpc 10 11>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+		tcu5_pe: tcu5_pe {
+			ingenic,pinmux = <&gpe 16 17>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+		};
+		tcu6_pc: tcu6_pc {
+			ingenic,pinmux = <&gpc 12 13>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+		tcu6_pe: tcu6_pe {
+			ingenic,pinmux = <&gpe 18 19>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+		};
+		tcu7_pc: tcu7_pc {
+			ingenic,pinmux = <&gpc 14 15>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION0>;
+		};
+		tcu7_pe: tcu7_pe {
+			ingenic,pinmux = <&gpe 20 21>;
+			ingenic,pinmux-funcsel = <PINCTL_FUNCTION1>;
+		};
+	};
+
+	scc_pc: scc-pc {
+		ingenic,pinmux = <&gpc 3 4>;
+		ingenic,pinmux-funcsel = <PINCTL_FUNCTION3>;
+	};
+};
+
