diff -drupN a/drivers/Makefile b/drivers/Makefile
--- a/drivers/Makefile	2017-10-21 18:09:07.000000000 +0300
+++ b/drivers/Makefile	2022-06-09 05:02:28.000000000 +0300
@@ -20,6 +20,9 @@ obj-$(CONFIG_RAPIDIO)		+= rapidio/
 obj-y				+= video/
 obj-y				+= idle/
 
+# LZMA devices
+obj-y                           += lzma/
+
 # IPMI must come before ACPI in order to provide IPMI opregion support
 obj-$(CONFIG_IPMI_HANDLER)	+= char/ipmi/
 
@@ -66,7 +69,7 @@ obj-$(CONFIG_PARPORT)		+= parport/
 obj-$(CONFIG_NVM)		+= lightnvm/
 obj-y				+= base/ block/ misc/ mfd/ nfc/
 obj-$(CONFIG_LIBNVDIMM)		+= nvdimm/
-obj-$(CONFIG_DMA_SHARED_BUFFER) += dma-buf/
+obj-y				+= dma-buf/
 obj-$(CONFIG_NUBUS)		+= nubus/
 obj-y				+= macintosh/
 obj-$(CONFIG_IDE)		+= ide/
