diff -drupN a/drivers/clk/clk-fractional-divider.c b/drivers/clk/clk-fractional-divider.c
--- a/drivers/clk/clk-fractional-divider.c	2017-10-21 18:09:07.000000000 +0300
+++ b/drivers/clk/clk-fractional-divider.c	2022-06-09 05:02:28.000000000 +0300
@@ -149,8 +149,11 @@ struct clk *clk_register_fractional_divi
 	fd->hw.init = &init;
 
 	clk = clk_register(dev, &fd->hw);
-	if (IS_ERR(clk))
+	if (IS_ERR(clk)) {
 		kfree(fd);
+    } else {
+        __clk_set_flags(clk, 1);
+    }
 
 	return clk;
 }
