diff -drupN a/arch/mips/boot/dts/ingenic/x2000-v12.dtsi b/arch/mips/boot/dts/ingenic/x2000-v12.dtsi
--- a/arch/mips/boot/dts/ingenic/x2000-v12.dtsi	1970-01-01 03:00:00.000000000 +0300
+++ b/arch/mips/boot/dts/ingenic/x2000-v12.dtsi	2022-06-09 05:02:27.000000000 +0300
@@ -0,0 +1,738 @@
+#include <dt-bindings/interrupt-controller/x2000-v12-irq.h>
+#include <dt-bindings/clock/ingenic-tcu.h>
+#include <dt-bindings/sound/ingenic-baic.h>
+#include <dt-bindings/net/ingenic_gmac.h>
+#include <dt-bindings/dma/ingenic-pdma.h>
+
+/ {
+	#address-cells = <1>;
+	#size-cells = <1>;
+	compatible = "ingenic,x2000";
+
+	aliases: aliases {
+		i2c0 = &i2c0;
+		i2c1 = &i2c1;
+		i2c2 = &i2c2;
+		i2c3 = &i2c3;
+		i2c4 = &i2c4;
+		i2c5 = &i2c5;
+		uart0 = &uart0;
+		uart1 = &uart1;
+		uart2 = &uart2;
+		uart3 = &uart3;
+		uart4 = &uart4;
+		uart5 = &uart5;
+		uart6 = &uart6;
+		uart7 = &uart7;
+		uart8 = &uart8;
+		msc0 = &msc0;
+		msc2 = &msc2;
+		mac0 = &mac0;
+		mac1 = &mac1;
+		spi0 = &spi0;
+		spi1 = &spi1;
+	};
+
+	cpus: cpus {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		cpu0: cpu@0 {
+			device_type = "cpu";
+			compatible = "ingenic,xburst2";
+			reg = <0x000>;
+			clock-frequency = <800000000>;
+		};
+
+		cpu1: cpu@1 {
+			device_type = "cpu";
+			compatible = "ingenic,xburst2";
+			reg = <0x001>;
+			clock-frequency = <800000000>;
+		};
+
+		cpu2: cpu@2 {
+			device_type = "cpu";
+			compatible = "ingenic,xburst2";
+			reg = <0x002>;
+			clock-frequency = <800000000>;
+		};
+
+		cpu3: cpu@3 {
+			device_type = "cpu";
+			compatible = "ingenic,xburst2";
+			reg = <0x003>;
+			clock-frequency = <800000000>;
+		};
+	};
+
+	cpuintc: interrupt-controller {
+		#address-cells = <0>;
+		#interrupt-cells = <1>;
+		interrupt-controller;
+		compatible = "ingenic,cpu-interrupt-controller";
+
+	};
+
+	core_intc: core-intc@0x12300000 {
+		compatible = "ingenic,core-intc";
+		reg = <0x12300000 0x1000>;
+		interrupt-controller;
+		#interrupt-cells = <1>;
+		cpu-intc-map = <0 0x000>,
+			           <1 0x100>,
+			           <2 0x200>,
+			           <3 0x300>;
+
+		interrupt-parent = <&cpuintc>;
+		interrupts = <CORE_INTC_IRQ>;
+		interrupt-names ="intc";
+	};
+
+	core_ost: core-ost@0x12000000 {
+		compatible = "ingenic,core-ost";
+		reg = <0x12000000 0x10000>, /*Global ost*/
+		      <0x12100000 0x10000>; /*Core ost*/
+		interrupt-parent = <&cpuintc>;
+		interrupt-names = "sys_ost";
+		interrupts = <CORE_SYS_OST_IRQ>;
+		cpu-ost-map = 	<0 0x000>,
+			<1 0x100>,
+			<2 0x200>,
+			<3 0x300>;
+	};
+
+	extclk: extclk {
+		compatible = "fixed-clock";
+		clock-output-names ="ext";
+		#clock-cells = <0>;
+		clock-frequency  = <24000000>;
+	};
+
+	rtcclk: rtcclk {
+		compatible = "fixed-clock";
+		clock-output-names ="rtc_ext";
+		#clock-cells = <0>;
+		clock-frequency = <32768>;
+	};
+
+	clock: clock-controller@0x10000000 {
+		compatible = "ingenic,x2000-v12-fpga-clocks";
+		reg = <0x10000000 0x100>;
+		clocks = <&extclk>, <&rtcclk>;
+		clock-names = "ext", "rtc_ext";
+		#clock-cells = <1>;
+		little-endian;
+	};
+
+	apb {
+		compatible = "simple-bus";
+		#address-cells = <1>;
+		#size-cells = <1>;
+		ranges = <>;
+		tcu: tcu@0x10002000 {
+				compatible = "ingenic,tcu";
+				reg = <0x10002000 0x140>;
+				interrupt-parent = <&core_intc>;
+				interrupt-names = "tcu_int0", "tcu_int1", "tcu_int2";
+				interrupts = <IRQ_TCU0 IRQ_TCU1 IRQ_TCU2>;
+				interrupt-controller;
+				status = "ok";
+
+				channel0: channel0 {
+					compatible = "ingenic,tcu_chn0";
+					ingenic,channel-info = <CHANNEL_INFO(0, TCU_MODE1, PWM_FUNC, \
+					PWM_IN, NOWORK_SLEEP)>;
+				};
+				channel1: channel1 {
+					compatible = "ingenic,tcu_chn1";
+					ingenic,channel-info = <CHANNEL_INFO(1, TCU_MODE2, PWM_FUNC, \
+					NO_PWM_IN, WORK_SLEEP)>;
+				};
+				channel2: channel2 {
+					compatible = "ingenic,tcu_chn2";
+					ingenic,channel-info = <CHANNEL_INFO(2, TCU_MODE2, PWM_FUNC, \
+					NO_PWM_IN, WORK_SLEEP)>;
+				};
+				channel3: channel3 {
+					compatible = "ingenic,tcu_chn3";
+					ingenic,channel-info = <CHANNEL_INFO(3, TCU_MODE1, \
+					PWM_AND_TRACKBALL_FUNC, PWM_IN, NOWORK_SLEEP)>;
+				};
+				channel4: channel4 {
+					compatible = "ingenic,tcu_chn4";
+					ingenic,channel-info = <CHANNEL_INFO(4, TCU_MODE1, \
+					PWM_AND_TRACKBALL_FUNC, PWM_IN, NOWORK_SLEEP)>;
+				};
+				channel5: channel5 {
+					compatible = "ingenic,tcu_chn5";
+					ingenic,channel-info = <CHANNEL_INFO(5, TCU_MODE1, \
+					TRACKBALL_FUNC, NO_PWM_IN, NOWORK_SLEEP)>;
+				};
+				channel6: channel6 {
+					compatible = "ingenic,tcu_chn6";
+					ingenic,channel-info = <CHANNEL_INFO(6, TCU_MODE1, \
+					TRACKBALL_FUNC, NO_PWM_IN, NOWORK_SLEEP)>;
+				};
+				channel7: channel7 {
+					compatible = "ingenic,tcu_chn7";
+					ingenic,channel-info = <CHANNEL_INFO(7, TCU_MODE1, \
+					TRACKBALL_FUNC, NO_PWM_IN, NOWORK_SLEEP)>;
+				};
+				channel15: channel15 {
+					compatible = "ingenic,tcu_chn15";
+					ingenic,channel-info = <CHANNEL_INFO(15, 0,0,0,0)>;
+				};
+				channel16: channel16 {
+					compatible = "ingenic,watchdog";
+					ingenic,channel-info = <CHANNEL_INFO(16, 0,0,0,0)>;
+				};
+		};
+
+
+		rtc: rtc@0x10003000 {
+			compatible = "ingenic,rtc";
+			reg = <0x10003000 0x4c>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_RTC>;
+			system-power-controller;
+			power-on-press-ms = <1000>;
+		};
+
+		pinctrl: pinctrl@0x10010000 {
+			compatible = "ingenic,x2000-v12-pinctrl";
+			reg = <0x10010000 0x1000>;
+			ingenic,num-chips = <6>;
+			ingenic,regs-offset = <0x100>;
+
+			gpa: gpa {
+				gpio-controller;
+				#gpio-cells = <3>;
+				#ingenic,pincfg-cells = <3>;
+				#ingenic,pinmux-cells = <2>;
+				interrupts = <IRQ_GPIO0>;
+				interrupt-parent = <&core_intc>;
+				interrupt-controller;
+				#interrupt-cells = <3>;
+				ingenic,num-gpios = <17>;
+				ingenic,pull-gpios-low = <0x55555555>;
+				ingenic,pull-gpios-high = <0x55555555>;
+			};
+
+			gpb: gpb {
+				gpio-controller;
+				#gpio-cells = <3>;
+				#ingenic,pincfg-cells = <3>;
+				#ingenic,pinmux-cells = <2>;
+				interrupts = <IRQ_GPIO1>;
+				interrupt-parent = <&core_intc>;
+				interrupt-controller;
+				#interrupt-cells = <3>;
+				ingenic,num-gpios = <32>;
+				ingenic,pull-gpios-low = <0x55565555>;
+				ingenic,pull-gpios-high = <0x55555555>;
+			};
+
+			gpc: gpc {
+				gpio-controller;
+				#gpio-cells = <3>;
+				#ingenic,pincfg-cells = <3>;
+				#ingenic,pinmux-cells = <2>;
+				interrupts = <IRQ_GPIO2>;
+				interrupt-parent = <&core_intc>;
+				interrupt-controller;
+				#interrupt-cells = <3>;
+				ingenic,num-gpios = <25>;
+				ingenic,pull-gpios-low = <0x55555555>;
+				ingenic,pull-gpios-high = <0x55555555>;
+			};
+
+			gpd: gpd {
+				gpio-controller;
+				#gpio-cells = <3>;
+				#ingenic,pincfg-cells = <3>;
+				#ingenic,pinmux-cells = <2>;
+				interrupts = <IRQ_GPIO3>;
+				interrupt-parent = <&core_intc>;
+				interrupt-controller;
+				#interrupt-cells = <3>;
+				ingenic,num-gpios = <32>;
+				ingenic,pull-gpios-low = <0x55555555>;
+				ingenic,pull-gpios-high = <0x55555555>;
+			};
+
+			gpe: gpe {
+				gpio-controller;
+				#gpio-cells = <3>;
+				#ingenic,pincfg-cells = <3>;
+				#ingenic,pinmux-cells = <2>;
+				interrupts = <IRQ_GPIO4>;
+				interrupt-parent = <&core_intc>;
+				interrupt-controller;
+				#interrupt-cells = <3>;
+				ingenic,num-gpios = <32>;
+				ingenic,pull-gpios-low = <0x55555555>;
+				ingenic,pull-gpios-high = <0x55555555>;
+			};
+
+		};
+
+		uart0: serial@0x10030000 {
+			compatible = "ingenic,8250-uart";
+			reg = <0x10030000 0x100>;
+			reg-shift = <2>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_UART0>;
+			pinctrl-names = "default";
+			pinctrl-0 = <&uart0_pd>;
+			status = "disabled";
+		};
+		uart1: serial@0x10031000 {
+			compatible = "ingenic,8250-uart";
+			reg = <0x10031000 0x100>;
+			reg-shift = <2>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_UART1>;
+			pinctrl-names = "default";
+			pinctrl-0 = <&uart1_pc>;
+			status = "disabled";
+		};
+		uart2: serial@0x10032000 {
+			compatible = "ingenic,8250-uart";
+			reg = <0x10032000 0x100>;
+			reg-shift = <2>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_UART2>;
+			pinctrl-names = "default";
+			pinctrl-0 = <&uart2_pd>;
+			status = "disabled";
+		};
+		uart3: serial@0x10033000 {
+			compatible = "ingenic,8250-uart";
+			reg = <0x10033000 0x100>;
+			reg-shift = <2>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_UART3>;
+			pinctrl-names = "default";
+			pinctrl-0 = <&uart3_pd>;
+			status = "disabled";
+		};
+
+		uart4: serial@0x10034000 {
+			compatible = "ingenic,8250-uart";
+			reg = <0x10034000 0x100>;
+			reg-shift = <2>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_UART4>;
+			status = "disabled";
+		};
+		uart5: serial@0x10035000 {
+			compatible = "ingenic,8250-uart";
+			reg = <0x10035000 0x100>;
+			reg-shift = <2>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_UART5>;
+			status = "disabled";
+		};
+		uart6: serial@0x10036000 {
+			compatible = "ingenic,8250-uart";
+			reg = <0x10036000 0x100>;
+			reg-shift = <2>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_UART6>;
+			status = "disabled";
+		};
+		uart7: serial@0x10037000 {
+			compatible = "ingenic,8250-uart";
+			reg = <0x10037000 0x100>;
+			reg-shift = <2>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_UART7>;
+			status = "disabled";
+		};
+
+		uart8: serial@0x10038000 {
+			compatible = "ingenic,8250-uart";
+			reg = <0x10038000 0x100>;
+			reg-shift = <2>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_UART8>;
+			pinctrl-names = "default";
+			pinctrl-0 = <&uart8_pb>;
+			status = "disabled";
+		};
+		uart9: serial@0x10039000 {
+			compatible = "ingenic,8250-uart";
+			reg = <0x10039000 0x100>;
+			reg-shift = <2>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_UART9>;
+			pinctrl-names = "default";
+			pinctrl-0 = <&uart9_pb>;
+			status = "disabled";
+		};
+
+		i2c0: i2c@0x10050000 {
+			compatible = "ingenic,x2000-i2c";
+			reg = <0x10050000 0x1000>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_I2C0>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "disabled";
+		};
+
+		i2c1: i2c@0x10051000 {
+			compatible = "ingenic,x2000-i2c";
+			reg = <0x10051000 0x1000>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_I2C1>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "disabled";
+		};
+
+		i2c2: i2c@0x10052000 {
+			compatible = "ingenic,x2000-i2c";
+			reg = <0x10052000 0x1000>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_I2C2>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "disabled";
+		};
+
+		i2c3: i2c@0x10053000 {
+			compatible = "ingenic,x2000-i2c";
+			reg = <0x10053000 0x1000>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_I2C3>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "disabled";
+		};
+
+		i2c4: i2c@0x10054000 {
+			compatible = "ingenic,x2000-i2c";
+			reg = <0x10054000 0x1000>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_I2C4>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "disabled";
+		};
+
+		i2c5: i2c@0x10055000 {
+			compatible = "ingenic,x2000-i2c";
+			reg = <0x10055000 0x1000>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_I2C5>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "disabled";
+		};
+
+		spi0: spi0@0x10043000 {
+			compatible = "ingenic,spi";
+			reg = <0x10043000 0x1000>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_SSI0>;
+			dmas = <&pdma INGENIC_DMA_TYPE(INGENIC_DMA_REQ_SSI0_TX)>,
+				   <&pdma INGENIC_DMA_TYPE(INGENIC_DMA_REQ_SSI0_RX)>;
+			dma-names = "tx", "rx";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "disabled";
+		};
+
+		spi1: spi1@0x10044000 {
+			compatible = "ingenic,spi";
+			reg = <0x10044000 0x1000>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_SSI1>;
+			dmas = <&pdma INGENIC_DMA_TYPE(INGENIC_DMA_REQ_SSI1_TX)>,
+				   <&pdma INGENIC_DMA_TYPE(INGENIC_DMA_REQ_SSI1_RX)>;
+			dma-names = "tx", "rx";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "disabled";
+		};
+
+		scc: scc@0x10040000 {
+			compatible = "ingenic,scc";
+			reg = <0x10040000 0x100>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_SCC>;
+			status = "disabled";
+		};
+
+		dtrng: dtrng@0x10072000 {
+			compatible = "ingenic,dtrng";
+			reg = <0x10072000 0x100>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_DTRNG>;
+			status = "disabled";
+		};
+
+		otg_phy: otg_phy {
+			compatible = "ingenic,innophy";
+		};
+	};
+
+	ahb2 {
+		compatible = "simple-bus";
+		#address-cells = <1>;
+		#size-cells = <1>;
+		ranges = <>;
+
+		as:as {
+			compatible = "simple-bus";
+			#address-cells = <1>;
+			#size-cells = <1>;
+			ranges = <>;
+
+			as_platform: as-platform {
+				compatible = "ingenic,as-platform";
+				reg = <0x134d0000 0x114>, <0x134d1000 0x100>;
+				reg-names = "dma", "fifo";
+				ingenic,fifo-size = <4096>;
+				interrupt-parent = <&core_intc>;
+				interrupts = <IRQ_AUDIO>;
+				ingenic,fth_quirk;
+			};
+			as_virtual_fe: as-virtual-fe {
+				compatible = "ingenic,as-vir-fe";
+				reg = <0x00000000 0x0>;
+				ingenic,cap-dai-bm = <0xc>;
+				ingenic,num-dais = <4>;
+			};
+			as_fmtcov: as-fmtcov {
+				compatible = "ingenic,as-fmtcov";
+				reg = <0x134d2000 0x28>;
+			};
+			as_fe_dsp: as-dsp {
+				compatible = "ingenic,as-dsp";
+				reg = <0x134d4000 0x30>;
+				ingenic,li-port = <0 1 2 3 4 6 7 8 9 10 11 12>;
+				ingenic,lo-port = <0 1 2 3 4 5 6 7 8 9 10 11>;
+				ingenic,cap-dai-bm = <0x3e0>;
+				ingenic,num-dais = <10>;
+			};
+			as_be_baic: as-baic {
+				compatible = "ingenic,as-baic";
+				reg = <0x134d5000 0x5000>;
+				ingenic,num-dais = <5>;
+				ingenic,dai-mode = <BAIC_3AND(BAIC_PCM_MODE, BAIC_DSP_MODE, BAIC_I2S_MODE)>,
+					<BAIC_3AND(BAIC_PCM_MODE, BAIC_DSP_MODE, BAIC_I2S_MODE)>,
+					<BAIC_4AND(BAIC_I2S_MODE, BAIC_TDM1_MODE, BAIC_TDM2_MODE, BAIC_NO_REPLAY)>,
+					<BAIC_4AND(BAIC_I2S_MODE, BAIC_TDM1_MODE, BAIC_TDM2_MODE, BAIC_NO_RECORD)>,
+					<BAIC_PCM_MODE>;
+				ingenic,dai-pin-num = <1>, <1>, <1>, <1>, <1>;
+				ingenic,dai-pin-split = <0>, <1>, <2>, <3>;
+			};
+			as_dmic: as-dmic {
+				compatible = "ingenic,as-dmic";
+				reg = <0x134da000 0x10>;
+			};
+			as_aux_mixer: as-mixer {
+				compatible = "ingenic,as-mixer";
+				reg = <0x134dc000 0x8>;
+				ingenic,num-mixers = <1>;
+			};
+			as_spdif: as-spdif {
+				compatible = "ingenic,as-spdif";
+				reg = <0x134db000 0x14>, <0x134db100 0x14>;
+				reg-names = "out", "in";
+			};
+		};
+
+		msc0: msc@0x13450000 {
+			compatible = "ingenic,sdhci";
+			reg = <0x13450000 0x10000>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_MSC0>;
+			pinctrl-names ="default";
+			pinctrl-0 = <&msc0_8bit>;
+		};
+
+		msc1: msc@0x13460000 {
+			compatible = "ingenic,sdhci";
+			reg = <0x13460000 0x10000>;
+			status = "disabled";
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_MSC1>;
+			pinctrl-names ="default";
+			pinctrl-0 = <&msc1_4bit>;
+		};
+
+		msc2: msc@0x13490000 {
+			compatible = "ingenic,sdhci";
+			reg = <0x13490000 0x10000>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_MSC2>;
+			pinctrl-names ="default";
+			pinctrl-0 = <&msc2_4bit>;
+		};
+
+		aes: aes@0x13430000 {
+			compatible = "ingenic,aes";
+			reg = <0x13430000 0x10000>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_AES>;
+			status = "ok";
+		};
+
+		hash: hash@0x13470000 {
+			compatible = "ingenic,hash";
+			reg = <0x13470000 0x10000>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_HASH>;
+			status = "ok";
+		};
+
+		rsa: rsa@0x13480000 {
+			compatible = "ingenic,rsa";
+			reg = <0x13480000 0x10000>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_RSA>;
+			status = "ok";
+		};
+
+		mac0: mac@0x134b0000 {
+			compatible = "ingenic,dwc-mac";
+			reg = <0x134b0000 0x2000>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_GMAC0>;
+			status = "disabled";
+			ingenic,rst-ms = <10>;
+		};
+
+		mac1: mac@0x134a0000 {
+			compatible = "ingenic,dwc-mac";
+			reg = <0x134a0000 0x2000>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_GMAC1>;
+			status = "disabled";
+			ingenic,rst-ms = <10>;
+		};
+
+		sfc: sfc@0x13440000 {
+			compatible = "ingenic,sfc";
+			reg = <0x13440000 0x10000>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_SFC>;
+			pinctrl-names = "default";
+			pinctrl-0 = <&sfc_pe>;
+			status = "disabled";
+		};
+
+		pdma: dma@13420000 {
+			compatible = "ingenic,x1000-pdma";
+			reg = <0x13420000 0x10000>;
+			interrupt-parent = <&core_intc>;
+			interrupt-names = "pdma", "pdmad";
+			interrupts = <IRQ_PDMA>, <IRQ_PDMAD>;
+			#dma-channels = <8>;
+			#dma-cells = <1>;
+			ingenic,reserved-chs = <0x3>;
+		};
+
+		pwm: pwm@0x134c0000 {
+			compatible = "ingenic,x2000-pwm";
+			reg = <0x134c0000 0x10000>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_PWM>;
+			pinctrl-names = "default";
+			pinctrl-0 = <&pwm_pc>;
+			status = "ok";
+		};
+
+		otg: otg@0x13500000 {
+			compatible = "ingenic,dwc2-hsotg";
+			reg = <0x13500000 0x40000>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_OTG>;
+			ingenic,usbphy=<&otg_phy>;
+			status = "disabled";
+		};
+
+		efuse: efuse@0x13540000 {
+			compatible = "ingenic,x2000-efuse";
+			reg = <0x13540000 0x10000>;
+			status = "okay";
+		};
+	};
+
+	ahb1 {
+		compatible = "simple-bus";
+		#address-cells = <1>;
+		#size-cells = <1>;
+		ranges = <>;
+
+	};
+
+	ahb0 {
+		compatible = "simple-bus";
+		#address-cells = <1>;
+		#size-cells = <1>;
+		ranges = <>;
+
+		cim: cim@0x13060000 {
+			compatible = "ingenic,cim";
+			reg = <0x13060000 0x10000>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_CIM>;
+			status = "disable";
+		};
+
+		dpu: dpu@0x13050000 {
+			compatible = "ingenic,x2000-dpu";
+			reg = <0x13050000 0x10000>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_LCD>;
+			status = "disabled";
+		};
+
+		rotate: rotate@0x13070000 {
+			compatible = "ingenic,x2000-rotate";
+			reg = <0x13070000 0x10000>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_ROTATE>;
+			status = "okay";
+		};
+
+		helix: helix@0x13200000 {
+			compatible = "ingenic,x2000-helix";
+			reg = <0x13200000 0x100000>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_HELIX>;
+			status = "disabled";
+		};
+
+		felix: felix@0x13300000 {
+			compatible = "ingenic,x2000-felix";
+			reg = <0x13300000 0x100000>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_FELIX>;
+			status = "disabled";
+		};
+
+		isp0: isp0@0x13700000 {
+			compatible = "ingenic,x2000-isp0";
+			reg = <0x13700000 0x100000>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_ISP0>;
+			status = "disabled";
+		};
+
+		isp1: isp1@0x13800000 {
+			compatible = "ingenic,x2000-isp1";
+			reg = <0x13800000 0x100000>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <IRQ_ISP1>;
+			status = "disabled";
+		};
+	};
+
+};
+#include "x2000-v12-pinctrl.dtsi"
