diff -drupN a/drivers/media/platform/ingenic-vcodec/helix/default_sliceinfo.c b/drivers/media/platform/ingenic-vcodec/helix/default_sliceinfo.c
--- a/drivers/media/platform/ingenic-vcodec/helix/default_sliceinfo.c	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/media/platform/ingenic-vcodec/helix/default_sliceinfo.c	2022-06-09 05:02:29.000000000 +0300
@@ -0,0 +1,362 @@
+#include "api/helix_x264_enc.h"
+
+_H264E_SliceInfo default_H264E_Sliceinfo = {
+	.frame_type = 1,
+	.mb_width = 22,
+	.mb_height = 18,
+	.first_mby = 0,
+	.last_mby = 17,
+	.frame_width = 352,
+	.frame_height = 288,
+
+	.frm_re = {0x00, 0x00, 0x00, 0x00},
+	.pskip_en = 1,
+	.mref_en = 0,
+	.dct8x8_en = 0,
+	.scl = 0x03,
+	.hpel_en = 0x01,
+	.qpel_en = 0x01,
+	.ref_mode = 0x01,
+	.max_sech_step_i = 0x3f,
+	.max_mvrx_i = 0x1f4,
+	.max_mvry_i = 0x1f4,
+	.lambda_scale_parameter = 0x08,
+	.fs_en = 0x00,
+	.fs_md = 0x00,
+	.fs_px = 0x00,
+	.fs_py = 0x00,
+	.fs_rx = -1,
+	.fs_ry = -1,
+	.frm_mv_en = 0x00,
+	.frm_mv_size = 0x00,
+	.glb_mv_en = 0x00,
+	.glb_mvx = 0,
+	.glb_mvy = 0,
+	.me_step_en = 0x0,
+	.me_step_0 = 0x0,
+	.me_step_1 = 0x0,
+	.scaling_list = {
+		{0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, },
+		{0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, },
+		{0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, },
+		{0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, },
+	},
+	.scaling_list8 = {
+		{0x06, 0x0a, 0x0d, 0x10, 0x12, 0x16, 0x18, 0x1a, 0x0a, 0x0b, 0x10, 0x12, 0x16, 0x18, 0x1a, 0x1c, 0x0d, 0x10, 0x12, 0x16, 0x18, 0x1a, 0x1c, 0x1e, 0x10, 0x12, 0x16, 0x18, 0x1a, 0x1c, 0x1e, 0x20, 0x12, 0x16, 0x18, 0x1a, 0x1c, 0x1e, 0x20, 0x23, 0x16, 0x18, 0x1a, 0x1c, 0x1e, 0x20, 0x23, 0x25, 0x18, 0x1a, 0x1c, 0x1e, 0x20, 0x23, 0x25, 0x27, 0x1a, 0x1c, 0x1e, 0x20, 0x23, 0x25, 0x27, 0x29, },
+		{0x09, 0x0d, 0x0f, 0x11, 0x13, 0x15, 0x16, 0x18, 0x0d, 0x0d, 0x11, 0x13, 0x15, 0x16, 0x18, 0x19, 0x0f, 0x11, 0x13, 0x15, 0x16, 0x18, 0x19, 0x1b, 0x11, 0x13, 0x15, 0x16, 0x18, 0x19, 0x1b, 0x1c, 0x13, 0x15, 0x16, 0x18, 0x19, 0x1b, 0x1c, 0x1e, 0x15, 0x16, 0x18, 0x19, 0x1b, 0x1c, 0x1e, 0x20, 0x16, 0x18, 0x19, 0x1b, 0x1c, 0x1e, 0x20, 0x21, 0x18, 0x19, 0x1b, 0x1c, 0x1e, 0x20, 0x21, 0x23, },
+	},
+
+	.deadzone =  { 0x00000015, 0x0000000b, 0x00000015, 0x0000000b, 0x00000015, 0x0000000b, 0x00000015, 0x00000015, 0x0000000b},
+	.acmask_mode = 0x000095ea,
+	.intra_mode_msk = 0x0,
+	.i_4x4_dis = 0x0,
+	.i_8x8_dis = 0x0,
+	.i_16x16_dis = 0x0,
+	.p_l0_dis = 0x0,
+	.p_t8_dis = 0x0,
+	.p_skip_dis = 0x0,
+	.p_skip_pl0f_dis = 0x0,
+	.p_skip_pt8f_dis = 0x0,
+	.cost_bias_en = 0x0,
+	.cost_bias_i_4x4 = 0x0,
+	.cost_bias_i_8x8 = 0x0,
+	.cost_bias_i_16x16 = 0x0,
+	.cost_bias_p_l0 = 0x0,
+	.cost_bias_p_t8 = 0x0,
+	.cost_bias_p_skip = 0x0,
+	.intra_lambda_y_bias_en = 0x0,
+	.intra_lambda_c_bias_en = 0x0,
+	.intra_lambda_bias_qp0 = 0x0,
+	.intra_lambda_bias_qp1 = 0x0,
+	.intra_lambda_bias_0 = 0x0,
+	.intra_lambda_bias_1 = 0x0,
+	.intra_lambda_bias_2 = 0x0,
+	.chroma_sse_bias_en = 0x0,
+	.chroma_sse_bias_qp0 = 0x0,
+	.chroma_sse_bias_qp1 = 0x0,
+	.chroma_sse_bias_0 = 0x0,
+	.chroma_sse_bias_1 = 0x0,
+	.chroma_sse_bias_2 = 0x0,
+	.sse_lambda_bias_en = 0x0,
+	.sse_lambda_bias = 0x0,
+	.fbc_ep = 0xcc,
+	.jm_lambda2_en = 0x0,
+	.inter_nei_en = 0x0,
+	.skip_bias_en = 0x0,
+	.info_en = 0x1,
+	.mvd_sum_all = 0x0,
+	.mvd_sum_abs = 0x0,
+	.mv_sum_all = 0x0,
+	.mv_sum_abs = 0x0,
+	.ysse_thr = 0x0,
+	.csse_thr = 0x0,
+	.cfg_size_x = 0x4,
+	.cfg_size_y = 0x4,
+	.cfg_iw_thr = 0x0,
+	.cfg_mvr_thr1 = 0x0,
+	.cfg_mvr_thr2 = 0x0,
+	.cfg_mvr_thr3 = 0x0,
+	.dcm_en = 0x0,
+	.dcm_param = 0x4304,
+	.sde_prior = 0x5,
+	.db_prior = 0x5,
+	.mb_mode_val = 0x1,
+	.bit_16_en = 0x0,
+	.bit_8_en = 0x1,
+	.bit_4_en = 0x1,
+	.bit_uv_en = 0x0,
+	.lamb_16_en = 0x0,
+	.lamb_8_en = 0x0,
+	.lamb_4_en = 0x0,
+	.lamb_uv_en = 0x0,
+	.c_16_en = 0x0,
+	.c_8_en = 0x0,
+	.c_4_en = 0x0,
+	.c_uv_en = 0x0,
+	.pri_16 = 0x0,
+	.pri_8 = 0x0,
+	.pri_4 = 0x0,
+	.pri_uv = 0x0,
+	.ref_neb_4 = 0x1,
+	.ref_neb_8 = 0x1,
+	.bit_16 = {0,0,0,0},
+	.bit_uv = {0,0,0,0},
+	.bit_4 = {4,4,4,4},
+	.bit_8 = {4,4,4,4},
+	.lambda_info16 = 0x0,
+	.lambda_info8 = 0x0,
+	.lambda_info4 = 0x0,
+	.lambda_infouv = 0x0,
+	.ref_4 = 0x4,
+	.ref_8 = 0x3,
+	.const_16 = {0,0,0,0},
+	.const_uv = {0,0,0,0},
+	.const_4 = {0,0,0,0},
+	.const_8 = {0,0,0,0},
+	.deblock = 1,
+	.rotate = 0,
+	.alpha_c0_offset = 0,
+	.beta_offset = 0,
+	.state = 0xf7edfc94,
+	.bs = 0x00000000,
+	.qp = 0x26,
+	.bs_size_en = 0x1,
+	.bs_size = 0x800,
+	.skip_en = 1,
+	.cqp_offset = 0,
+	.use_intra_in_pframe = 0x1,
+	.use_fast_mvp = 0x0,
+	.mode_ctrl = 0x0,
+	.mode_ctrl_param = {0},
+	.fb = {{0,0},{0,0},{0,0}},
+	.raw = {0,0,0},
+	.stride = {0,0},
+	.raw_format = 8,
+	.size_mode = 0,
+	.step_mode = 0,
+	.des_va = 0,
+	.des_pa = 0,
+	.emc_bs_pa = 0,
+	.emc_dblk_pa = 0,
+//	.emc_bs_va = 0,
+	.emc_dblk_va = 0,
+	.emc_recon_va = 0,
+	.emc_mv_va = 0,
+	.emc_se_va = 0,
+	.emc_recon_va = 0,
+	.emc_recon_pa = 0,
+	.emc_mv_pa = 0,
+	.emc_se_pa = 0,
+	.emc_qpt_va = 0,
+	.emc_qpt_pa = 0,
+	.emc_rc_va = 0,
+	.emc_rc_pa = 0,
+	.emc_cpx_va = 0,
+	.emc_cpx_pa = 0,
+	.emc_mod_va = 0,
+	.emc_mod_pa = 0,
+	.emc_ncu_va = 0,
+	.emc_ncu_pa = 0,
+	.emc_sad_va = 0,
+	.emc_sad_pa = 0,
+	.tlba = 0,
+	.roi_info = {
+		{0,0,0,0,0,0,0},
+		{0,0,0,0,0,0,0},
+		{0,0,0,0,0,0,0},
+		{0,0,0,0,0,0,0},
+		{0,0,0,0,0,0,0},
+		{0,0,0,0,0,0,0},
+		{0,0,0,0,0,0,0},
+		{0,0,0,0,0,0,0},
+	},
+	.base_qp = 38,
+	.max_qp = 51,
+	.min_qp = 26,
+	.qp_tab_mode = 0,
+	.qp_tab_en = 0,
+	.qp_tab_len = 0,
+	.qp_tab = NULL,
+	.sas_en = 0,
+	.crp_en = 0,
+	.sas_mthd = 0,
+	.qpg_mb_thd = {0,0,0,0,0,0,0},
+	.qpg_mbqp_ofst = {0,0,0,0,0,0,0},
+	.qpg_flt_thd = {0,0,0,0,0},
+	.mbrc_qpg_sel = 0,
+	.rc_mb_en = 0,
+	.rc_bu_wait_en = 0,
+	.rc_mb_wait_en = 0,
+	.rc_bu_num = 0x00000012,
+	.rc_bu_size = 0x00000016,
+	.rc_bu_level = 1,
+	.rc_mb_level = 2,
+	.mb_ref_info = NULL,
+	.bu_ref_info = NULL,
+	.rc_frm_tbs = 1000,
+	.avg_bu_bs = 55,
+	.tar_bs_thd = {500, 1000, 1500, 2000, 2500, 3000},
+	.bu_alg0_qpo = {1,2,3,4,5,6},
+	.bu_alg1_qpo = {1,1,1,1,1,1},
+	.mb_cs_qpo = {-3,-2,-1,0,1,2,3},
+	.mb_top_bs_qpo = {-1, 1},
+	.mb_rinfo_qpo = {-1, 1},
+	.mb_target_avg_bs = {2,2},
+	.mb_gp_num = 198,
+	.last_bu_size = 22,
+	.rc_bcfg_mode = 0,
+	.mosaic_en = 0,
+	.mos_gthd = {1,4},
+	.mos_sthd = {8,6},
+	.daisy_chain_en = 0,
+	.curr_thread_id = 5,
+	.jrfcd_flag = 0,
+	.jrfc_enable = 0,
+	.jrfd_enable = 0,
+	.lm_head_total = 0,
+	.cm_head_total = 0,
+	.jh = {{0,0}, {0,0}, {0,0}},
+	.spe_y_addr = 0,
+	.spe_c_addr = 0,
+	.mb_mode_use = 0,
+	.mb_mode_info = NULL,
+	.force_i16dc = 1,
+	.force_i16 = 1,
+	.refresh_en = 0,
+	.refresh_mode = 0,
+	.refresh_bias = 10,
+	.refresh_cplx_thd = 4,
+	.cplx_thd_sel = 0,
+	.diff_cplx_sel = 0,
+	.diff_thd_sel = 0,
+	.i16dc_cplx_thd = 3,
+	.i16dc_qp_base = 33,
+	.i16dc_qp_sel = 0,
+	.i16_qp_base = 28,
+	.i16_qp_sel = 0,
+	.diff_cplx_thd = 10,
+	.diff_qp_base = {33,33},
+	.diff_qp_sel = {0, 0},
+	.cplx_thd_idx = {
+		0, /*cplx_thd_idx[0]*/
+		0, /*cplx_thd_idx[1]*/
+		1, /*cplx_thd_idx[2]*/
+		1, /*cplx_thd_idx[3]*/
+		1, /*cplx_thd_idx[4]*/
+		1, /*cplx_thd_idx[5]*/
+		1, /*cplx_thd_idx[6]*/
+		2, /*cplx_thd_idx[7]*/
+		2, /*cplx_thd_idx[8]*/
+		2, /*cplx_thd_idx[9]*/
+		2, /*cplx_thd_idx[10]*/
+		2, /*cplx_thd_idx[11]*/
+		2, /*cplx_thd_idx[12]*/
+		2, /*cplx_thd_idx[13]*/
+		2, /*cplx_thd_idx[14]*/
+		2, /*cplx_thd_idx[15]*/
+		2, /*cplx_thd_idx[16]*/
+		2, /*cplx_thd_idx[17]*/
+		2, /*cplx_thd_idx[18]*/
+		2, /*cplx_thd_idx[19]*/
+		3, /*cplx_thd_idx[20]*/
+		3, /*cplx_thd_idx[21]*/
+		3, /*cplx_thd_idx[22]*/
+		3, /*cplx_thd_idx[23]*/
+	},
+
+	.cplx_thd = {
+		0, /*cplx_thd[0]*/
+		5, /*cplx_thd[1]*/
+		10, /*cplx_thd[2]*/
+		15, /*cplx_thd[3]*/
+		0, /*cplx_thd[4]*/
+		0, /*cplx_thd[5]*/
+		0, /*cplx_thd[6]*/
+		0, /*cplx_thd[7]*/
+	},
+
+	.diff_thd_base = {
+		384, /*diff_thd_base[0]*/
+		96, /*diff_thd_base[1]*/
+		96, /*diff_thd_base[2]*/
+	},
+
+	.diff_thd_ofst = {
+		48, /*diff_thd_ofst[0]*/
+		12, /*diff_thd_ofst[1]*/
+		12, /*diff_thd_ofst[2]*/
+	},
+	.sas_eigen_en = 0,
+	.crp_eigen_en = 0,
+	.sas_eigen_dump = 0,
+	.crp_eigen_dump = 0,
+	.rrs_en = 0,
+	.rrs_dump_en = 1,
+	.rrs_uv_en = 0,
+	.rrs_size_y = 16,
+	.rrs_size_c = 8,
+	.rrs_thrd_y = 624,
+	.rrs_thrd_u = 156,
+	.rrs_thrd_v = 156,
+	.skin_dt_en = 0,
+	.skin_lvl = 0,
+	.skin_cnt_thd = 20,
+	.skin_pxlu_thd = {
+		{100, 130, },
+		{40, 110, },
+		{120, 160, },
+	},
+	.skin_pxlv_thd = {
+		{140, 175, },
+		{120, 155, },
+		{165, 225, },
+	},
+	.skin_qp_ofst = {
+		-8, /*skin_qp_ofst[0]*/
+		-4, /*skin_qp_ofst[1]*/
+		-2, /*skin_qp_ofst[2]*/
+		1, /*skin_qp_ofst[3]*/
+	},
+	.mult_factor = {
+		3, /*mult_factor[0]*/
+		9, /*mult_factor[1]*/
+		5, /*mult_factor[2]*/
+	},
+	.shift_factor = {
+		{3, 5, 8, },
+		{4, 5, 6, },
+		{4, 5, 6, },
+	},
+	.skin_ofst = {
+		97, /*skin_ofst[0]*/
+		183, /*skin_ofst[1]*/
+		271, /*skin_ofst[2]*/
+		884, /*skin_ofst[3]*/
+	},
+	.ncu_mov_en = 0,
+	.ncu_move_len = 0,
+	.ncu_move_info = NULL,
+	.buf_share_en = 0,
+	.buf_share_size = 0,
+	.frame_idx = 2,
+};
