diff -drupN a/arch/mips/Kconfig.debug b/arch/mips/Kconfig.debug
--- a/arch/mips/Kconfig.debug	2017-10-21 18:09:07.000000000 +0300
+++ b/arch/mips/Kconfig.debug	2022-06-09 05:02:26.000000000 +0300
@@ -8,7 +8,7 @@ source "lib/Kconfig.debug"
 
 config EARLY_PRINTK
 	bool "Early printk" if EXPERT
-	depends on SYS_HAS_EARLY_PRINTK
+	depends on SYS_HAS_EARLY_PRINTK && !FAST_BOOT
 	default y
 	help
 	  This option enables special console drivers which allow the kernel
