diff -drupN a/drivers/char/random.c b/drivers/char/random.c
--- a/drivers/char/random.c	2017-10-21 18:09:07.000000000 +0300
+++ b/drivers/char/random.c	2022-06-09 05:02:28.000000000 +0300
@@ -1467,9 +1467,9 @@ urandom_read(struct file *file, char __u
 	if (unlikely(nonblocking_pool.initialized == 0) &&
 	    maxwarn > 0) {
 		maxwarn--;
-		printk(KERN_NOTICE "random: %s: uninitialized urandom read "
+	/*	printk(KERN_NOTICE "random: %s: uninitialized urandom read "
 		       "(%zd bytes read, %d bits of entropy available)\n",
-		       current->comm, nbytes, nonblocking_pool.entropy_total);
+		       current->comm, nbytes, nonblocking_pool.entropy_total);*/
 	}
 
 	nbytes = min_t(size_t, nbytes, INT_MAX >> (ENTROPY_SHIFT + 3));
