diff -drupN a/drivers/bluetooth/Kconfig b/drivers/bluetooth/Kconfig
--- a/drivers/bluetooth/Kconfig	2017-10-21 18:09:07.000000000 +0300
+++ b/drivers/bluetooth/Kconfig	2022-06-09 05:02:28.000000000 +0300
@@ -86,6 +86,17 @@ config BT_HCIUART_H4
 
 	  Say Y here to compile support for HCI UART (H4) protocol.
 
+config BT_HCIUART_RTKH5
+        bool "Realtek H5 protocol support"
+    depends on BT_HCIUART
+    help
+          Realtek H5 is serial protocol for communication
+          between Realtek Bluetooth device and host. This protocol is required for
+          Realtek uart h5 bluetooth controller
+
+          Say Y here to compile support for Realtek HCI H5 protocol.
+
+
 config BT_HCIUART_BCSP
 	bool "BCSP protocol support"
 	depends on BT_HCIUART
