diff -drupN a/drivers/mtd/devices/ingenic_sfc_v1/spinor.h b/drivers/mtd/devices/ingenic_sfc_v1/spinor.h
--- a/drivers/mtd/devices/ingenic_sfc_v1/spinor.h	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/mtd/devices/ingenic_sfc_v1/spinor.h	2022-06-09 05:02:30.000000000 +0300
@@ -0,0 +1,117 @@
+#ifndef __SPINOR_H
+#define __SPINOR_H
+#include "sfc_flash.h"
+#include "spinor_cmd.h"
+
+#define SPIFLASH_PARAMER_OFFSET 0x3c00
+#define NOR_MAJOR_VERSION_NUMBER	2
+#define NOR_MINOR_VERSION_NUMBER	0
+#define NOR_REVERSION_NUMBER	0
+#define NOR_VERSION		(NOR_MAJOR_VERSION_NUMBER | (NOR_MINOR_VERSION_NUMBER << 8) | (NOR_REVERSION_NUMBER << 16))
+
+
+#define SIZEOF_NAME			32
+
+#define NOR_MAGIC	0x726f6e	//ascii "nor"
+#define NOR_PART_NUM	10
+#define NORFLASH_PART_RW	0
+#define NORFLASH_PART_WO	1
+#define NORFLASH_PART_RO	2
+
+
+struct spi_nor_cmd_info {
+	unsigned short cmd;
+	unsigned char dummy_byte;
+	unsigned char addr_nbyte;
+	unsigned char transfer_mode;
+
+};
+
+struct spi_nor_st_info {
+	unsigned short cmd;
+	unsigned char bit_shift;
+	unsigned char mask;
+	unsigned char val;
+	unsigned char len; //length of byte to operate from register
+	unsigned char dummy;
+};
+
+struct spi_nor_info {
+	unsigned char name[32];
+	unsigned int id;
+
+	struct spi_nor_cmd_info read_standard;
+	struct spi_nor_cmd_info read_quad;
+
+	struct spi_nor_cmd_info write_standard;
+	struct spi_nor_cmd_info write_quad;
+
+	struct spi_nor_cmd_info sector_erase;
+
+	struct spi_nor_cmd_info wr_en;
+	struct spi_nor_cmd_info en4byte;
+	struct spi_nor_st_info	quad_set;
+	struct spi_nor_st_info	quad_get;
+	struct spi_nor_st_info	busy;
+
+	unsigned short quad_ops_mode;
+	unsigned short addr_ops_mode;
+
+	unsigned int tCHSH;      //hold
+	unsigned int tSLCH;      //setup
+	unsigned int tSHSL_RD;   //interval
+	unsigned int tSHSL_WR;
+
+	unsigned int chip_size;
+	unsigned int page_size;
+	unsigned int erase_size;
+
+	unsigned char chip_erase_cmd;
+};
+
+struct nor_partition {
+	char name[32];
+	uint32_t size;
+	uint32_t offset;
+	uint32_t mask_flags;//bit 0-1 mask the partiton RW mode, 0:RW  1:W  2:R
+	uint32_t manager_mode;
+};
+
+struct norflash_partitions {
+	struct nor_partition nor_partition[NOR_PART_NUM];
+	uint32_t num_partition_info;
+};
+
+struct nor_private_data {
+	unsigned int fs_erase_size;
+	unsigned char uk_quad;
+};
+
+
+struct burner_params {
+	uint32_t magic;
+	uint32_t version;
+	struct spi_nor_info spi_nor_info;
+	struct norflash_partitions norflash_partitions;
+	struct nor_private_data nor_pri_data;
+};
+
+struct spi_nor_flash_ops {
+	int (*set_4byte_mode)(struct sfc_flash *flash);
+	int (*set_quad_mode)(struct sfc_flash *flash);
+};
+
+struct spinor_flashinfo {
+
+	int quad_succeed;
+	struct spi_nor_flash_ops *nor_flash_ops;
+	struct spi_nor_info *nor_flash_info;
+	struct spi_nor_cmd_info *cur_r_cmd;
+	struct spi_nor_cmd_info *cur_w_cmd;
+	struct norflash_partitions *norflash_partitions;
+	struct nor_private_data *nor_pri_data;
+
+};
+
+#endif
+
