diff -drupN a/drivers/net/ethernet/ingenic/Makefile b/drivers/net/ethernet/ingenic/Makefile
--- a/drivers/net/ethernet/ingenic/Makefile	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/net/ethernet/ingenic/Makefile	2022-06-09 05:02:30.000000000 +0300
@@ -0,0 +1,31 @@
+
+####################################################
+# Explanation for EXTRA_FLAGS
+# -DDEBUG		     => Enable the debug trace for driver
+# -DIPC_OFFLOAD		     => Enables the IP and TCP checksum offloading feature in HW (IPV4 only)
+# -DENH_DESC		     => Enables Enhanced Descriptors
+# -DENH_DESC_8W              => Enable Enhanced Descriptors of 8words or else Descriptor will be of 4words.
+####################################################
+
+
+#with the below flags enables Debug messages
+
+#EXTRA_CFLAGS += -DDEBUG
+#EXTRA_CFLAGS +=-DDEBUG -DIPC_OFFLOAD
+#EXTRA_CFLAGS += -DDEBUG -DIPC_OFFLOAD
+#EXTRA_CFLAGS += -DIPC_OFFLOAD
+
+
+#Enhanced Descriptor
+#EXTRA_CFLAGS += -DENH_DESC
+#EXTRA_CFLAGS += -DENH_DESC -DIPC_OFFLOAD
+EXTRA_CFLAGS += -DENH_DESC -DENH_DESC_8W
+#EXTRA_CFLAGS += -DDEBUG -DENH_DESC -DENH_DESC_8W
+#EXTRA_CFLAGS += -DDEBUG -DENH_DESC -DENH_DESC_8W -DIPC_OFFLOAD
+#EXTRA_CFLAGS += -DENH_DESC -DENH_DESC_8W -DIPC_OFFLOAD
+#EXTRA_CFLAGS += -DENH_DESC -DENH_DESC_8W -DAVB_SUPPORT  -DDEBUG
+#EXTRA_CFLAGS += -DENH_DESC -DENH_DESC_8W -DAVB_SUPPORT
+#EXTRA_CFLAGS += -DENH_DESC -DENH_DESC_8W
+
+obj-$(CONFIG_INGENIC_MAC) += synopGMAC_plat.o synopGMAC_Dev.o
+obj-$(CONFIG_INGENIC_MAC) += ingenic_mac.o
