diff -drupN a/drivers/video/fbdev/ingenic/fb_v11/Kconfig b/drivers/video/fbdev/ingenic/fb_v11/Kconfig
--- a/drivers/video/fbdev/ingenic/fb_v11/Kconfig	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/video/fbdev/ingenic/fb_v11/Kconfig	2022-06-09 05:02:34.000000000 +0300
@@ -0,0 +1,8 @@
+menuconfig FB_INGENIC_X2000
+	tristate "Ingenic Framebuffer Driver for Version 11"
+	depends on FB_INGENIC
+	select FB_INGENIC_DISPLAYS_X2000
+	help
+		LCD Driver for X2000
+
+source "drivers/video/fbdev/ingenic/fb_v11/displays/Kconfig"
