diff -drupN a/drivers/crypto/Makefile b/drivers/crypto/Makefile
--- a/drivers/crypto/Makefile	2017-10-21 18:09:07.000000000 +0300
+++ b/drivers/crypto/Makefile	2022-06-09 05:02:28.000000000 +0300
@@ -29,3 +29,6 @@ obj-$(CONFIG_CRYPTO_DEV_QAT) += qat/
 obj-$(CONFIG_CRYPTO_DEV_QCE) += qce/
 obj-$(CONFIG_CRYPTO_DEV_VMX) += vmx/
 obj-$(CONFIG_CRYPTO_DEV_SUN4I_SS) += sunxi-ss/
+
+obj-$(CONFIG_CRYPTO_DEV_INGENIC_AES) += ingenic-aes.o
+obj-$(CONFIG_CRYPTO_DEV_INGENIC_SHA) += ingenic-hash.o
\ В конце файла нет новой строки
