diff -drupN a/drivers/i2c/busses/Kconfig b/drivers/i2c/busses/Kconfig
--- a/drivers/i2c/busses/Kconfig	2017-10-21 18:09:07.000000000 +0300
+++ b/drivers/i2c/busses/Kconfig	2022-06-09 05:02:29.000000000 +0300
@@ -993,6 +993,32 @@ config I2C_RCAR
 	  This driver can also be built as a module.  If so, the module
 	  will be called i2c-rcar.
 
+config I2C_INGENIC
+        bool "Ingenic SoC based on Xburst2 arch's I2C controler Driver support"
+	help
+	  Say Y here in order to support Ingenic SoC's I2C Controller Drvier
+
+config I2C_NON_RESTART_MODE
+	bool "controler i2c no restart mode"
+	depends on I2C_INGENIC
+	default n
+
+config I2C_FIFO_LEN
+        int "INGENIC I2C Controller FIFO length"
+	default 64
+	depends on I2C_INGENIC
+	help
+	  M200 and JZ4775 have 64 entries FIFO.
+
+config I2C_DEBUG_INFO
+	bool "enable or disable Ingenic Soc's I2C driver debug info"
+	default n
+	depends on I2C_INGENIC
+	help
+	  I2C debug print info, you can get the debug information from
+	  /sys/devices/platform/i2c-ingenic.x/debug_info,
+	  for example, echo 3 > /sys/devices/platform/i2c-ingenic.x/debug
+
 comment "External I2C/SMBus adapter drivers"
 
 config I2C_DIOLAN_U2C
