diff -drupN a/drivers/net/ethernet/ingenic/readme b/drivers/net/ethernet/ingenic/readme
--- a/drivers/net/ethernet/ingenic/readme	1970-01-01 03:00:00.000000000 +0300
+++ b/drivers/net/ethernet/ingenic/readme	2022-06-09 05:02:30.000000000 +0300
@@ -0,0 +1,26 @@
+*************************************README***************************************
+
+
+**********************************************************************************
+
+This is a README file for the sample device driver software for Synopsys Universal
+GMAC IP. The files listed below are available in this directory.
+
+./README (this file)
+./synopGMAC_banner.h
+./synopGMAC_plat.c
+./synopGMAC_plat.h
+./synopGMAC_Dev.c
+./synopGMAC_Dev.h
+./synopGMAC_network_interface.c
+./synopGMAC_network_interface.h
+./synopGMAC_Host.c
+./synopGMAC_Host.h
+./Makefile
+./synopGMAC_Debug.c
+./synopGMAC_driver_ug.pdf
+
+The User Guide (synopGMAC_driver_ug.pdf) provides details about driver architecture, flow
+diagrams and driver APIs. Please refer to compilation subsection of chapter 1 to compile
+the driver and the debug utility. The same section explains how to use the kernel module
+in Linux platform.
