diff -drupN a/fs/squashfs/super.c b/fs/squashfs/super.c
--- a/fs/squashfs/super.c	2017-10-21 18:09:07.000000000 +0300
+++ b/fs/squashfs/super.c	2022-06-09 05:02:35.000000000 +0300
@@ -449,8 +449,8 @@ static int __init init_squashfs_fs(void)
 		destroy_inodecache();
 		return err;
 	}
-
-	pr_info("version 4.0 (2009/01/31) Phillip Lougher\n");
+printk(KERN_INFO "squashfs: version 4.0 (2021/14/49) "
+		"Phillip Lougher \n");
 
 	return 0;
 }
