diff -drupN a/drivers/net/ethernet/Makefile b/drivers/net/ethernet/Makefile
--- a/drivers/net/ethernet/Makefile	2017-10-21 18:09:07.000000000 +0300
+++ b/drivers/net/ethernet/Makefile	2022-06-09 05:02:30.000000000 +0300
@@ -87,3 +87,4 @@ obj-$(CONFIG_NET_VENDOR_VIA) += via/
 obj-$(CONFIG_NET_VENDOR_WIZNET) += wiznet/
 obj-$(CONFIG_NET_VENDOR_XILINX) += xilinx/
 obj-$(CONFIG_NET_VENDOR_XIRCOM) += xircom/
+obj-$(CONFIG_INGENIC_MAC) += ingenic/
