diff -drupN a/drivers/clocksource/Makefile b/drivers/clocksource/Makefile
--- a/drivers/clocksource/Makefile	2017-10-21 18:09:07.000000000 +0300
+++ b/drivers/clocksource/Makefile	2022-06-09 05:02:28.000000000 +0300
@@ -65,3 +65,5 @@ obj-$(CONFIG_H8300_TMR16)		+= h8300_time
 obj-$(CONFIG_H8300_TPU)			+= h8300_tpu.o
 obj-$(CONFIG_CLKSRC_ST_LPC)		+= clksrc_st_lpc.o
 obj-$(CONFIG_X86_NUMACHIP)		+= numachip.o
+obj-$(CONFIG_CLKSRC_INGENIC_SYS_OST)		+= ingenic_sysost.o
+obj-$(CONFIG_CLKSRC_INGENIC_CORE_OST)		+= ingenic_core_ost.o
