diff -drupN a/arch/mips/include/asm/cpu.h b/arch/mips/include/asm/cpu.h
--- a/arch/mips/include/asm/cpu.h	2017-10-21 18:09:07.000000000 +0300
+++ b/arch/mips/include/asm/cpu.h	2022-06-09 05:02:27.000000000 +0300
@@ -45,6 +45,7 @@
 #define PRID_COMP_INGENIC_D0	0xd00000	/* JZ4740, JZ4750 */
 #define PRID_COMP_INGENIC_D1	0xd10000	/* JZ4770, JZ4775 */
 #define PRID_COMP_INGENIC_E1	0xe10000	/* JZ4780 */
+#define PRID_COMP_INGENIC_13	0x130000	/* X2000 */
 
 /*
  * Assigned Processor ID (implementation) values for bits 15:8 of the PRId
@@ -170,11 +171,42 @@
 #define PRID_IMP_CAVIUM_CN78XX 0x9500
 #define PRID_IMP_CAVIUM_CN70XX 0x9600
 
+
+
 /*
- * These are the PRID's for when 23:16 == PRID_COMP_INGENIC_*
+ * These are the PRID's for when 23:16 == PRID_COMP_INGENIC
+ * 31:24, Company Options is reserved;
+ * 23:16, Company ID is 0xd0;
+ * 15:8,  Processor ID;
+ *              -- 15:13, is the Generation of INGENIC's core.
+ *                              0x0 is XBurst1,
+ *                              0x1 is XBurst2,
+ *                              others are reserved.
+ *              -- 12:8, the encode of micro-architecture(uA).
+ *      7:0, Revision;
+ *              -- 7:4, the encode of process.
+ *                              0x0 is SMIC40LP,
+ *                              0x1 is TSMC40LP,
+ *                              others are reserved.
+ *              -- 3:0, the version of a chip's package or upgrades.
  */
+#define PRID_COMPANY_ID_MASK   0x00FF0000
+#define PRID_CPU_ARCH_MASK     0x00001F00
+#define PRID_CPU_FEATURE_MASK       0xFFFF0000
+#define PRID_CPU_CHIPS_MASK         0x000FF000
+#define PRID_IMP_PROCESSOR_ID_MSK 0x0000E000
+#define PRID_IMP_XBURST                         (0x0 << 13)
+#define PRID_IMP_XBURST2                        (0x1 << 13)
+#define PRID_CPU_ISA_MASK           0x00000FFF
 
-#define PRID_IMP_JZRISC	       0x0200
+#define PRID_IMP_JZRISC             0x00010000
+#define PRID_CPU_JZ4775S            0x2ed00000
+#define PRID_CPU_JZ4780             0x3ee00000
+#define PRID_CPU_M200               0x7ae00000
+#define PRID_CPU_X1000		    0x2ed10000
+#define PRID_CPU_X1800	            0x00d00000
+#define PRID_CPU_X2000              0x00130000
+#define PRID_IMP_ISA_R2             0x0000024f
 
 /*
  * These are the PRID's for when 23:16 == PRID_COMP_NETLOGIC
@@ -398,6 +430,6 @@ enum cpu_type_enum {
 #define MIPS_ASE_MIPSMT		0x00000020 /* CPU supports MIPS MT */
 #define MIPS_ASE_DSP2P		0x00000040 /* Signal Processing ASE Rev 2 */
 #define MIPS_ASE_VZ		0x00000080 /* Virtualization ASE */
-#define MIPS_ASE_MSA		0x00000100 /* MIPS SIMD Architecture */
-
+#define MIPS_ASE_CU2           0x00000100 /* Coprocessor 2 implemented */
+#define MIPS_ASE_MSA           0x00080000 /* MIPS SIMD Architecture */
 #endif /* _ASM_CPU_H */
