diff -drupN a/arch/mips/include/asm/bootinfo.h b/arch/mips/include/asm/bootinfo.h
--- a/arch/mips/include/asm/bootinfo.h	2017-10-21 18:09:07.000000000 +0300
+++ b/arch/mips/include/asm/bootinfo.h	2022-06-09 05:02:27.000000000 +0300
@@ -79,6 +79,8 @@ enum loongson_machine_type {
  */
 #define  MACH_INGENIC_JZ4730	0	/* JZ4730 SOC		*/
 #define  MACH_INGENIC_JZ4740	1	/* JZ4740 SOC		*/
+#define  MACH_XBURST		2	/* Xburst based SOC	*/
+#define  MACH_XBURST2		3	/* Xburst2 based SOC	*/
 
 extern char *system_type;
 const char *get_system_type(void);
