diff -drupN a/drivers/clk/clk.c b/drivers/clk/clk.c
--- a/drivers/clk/clk.c	2017-10-21 18:09:07.000000000 +0300
+++ b/drivers/clk/clk.c	2022-06-09 05:02:28.000000000 +0300
@@ -410,6 +410,16 @@ static unsigned long __clk_get_accuracy(
 	return core->accuracy;
 }
 
+void __clk_set_flags(struct clk *clk, unsigned long enable)
+{
+    if (1 == enable) {
+        clk->core->flags |= BIT(1);
+    } else {
+        clk->core->flags &= ~BIT(1);
+    }
+}
+EXPORT_SYMBOL_GPL(__clk_set_flags);
+
 unsigned long __clk_get_flags(struct clk *clk)
 {
 	return !clk ? 0 : clk->core->flags;
