diff -drupN a/arch/mips/boot/Makefile b/arch/mips/boot/Makefile
--- a/arch/mips/boot/Makefile	2017-10-21 18:09:07.000000000 +0300
+++ b/arch/mips/boot/Makefile	2022-06-09 05:02:27.000000000 +0300
@@ -75,7 +75,7 @@ $(obj)/vmlinux.bin.lzo: $(obj)/vmlinux.b
 # Compressed u-boot images
 #
 
-targets += uImage
+#targets += uImage
 targets += uImage.bin
 targets += uImage.bz2
 targets += uImage.gz
@@ -97,6 +97,9 @@ $(obj)/uImage.lzma: $(obj)/vmlinux.bin.l
 $(obj)/uImage.lzo: $(obj)/vmlinux.bin.lzo FORCE
 	$(call if_changed,uimage,lzo)
 
-$(obj)/uImage: $(obj)/uImage.$(suffix-y)
+
+
+targets += uImage
+$(obj)/uImage: $(obj)/uImage.lzma  $(obj)/uImage.lzo FORCE
 	@ln -sf $(notdir $<) $@
 	@echo '  Image $@ is ready'
