diff -drupN a/drivers/usb/dwc2/hcd_intr.c b/drivers/usb/dwc2/hcd_intr.c
--- a/drivers/usb/dwc2/hcd_intr.c	2017-10-21 18:09:07.000000000 +0300
+++ b/drivers/usb/dwc2/hcd_intr.c	2022-06-09 05:02:34.000000000 +0300
@@ -102,6 +102,11 @@ static void dwc2_hc_handle_tt_clear(stru
 	if (!usb_urb || !usb_urb->dev || !usb_urb->dev->tt)
 		return;
 
+	WARN_ON(!usb_urb->dev->tt->hub);
+	if (usb_urb->dev->tt->hub ==
+			dwc2_hsotg_to_hcd(hsotg)->self.root_hub)
+		return;
+
 	if (qtd->urb->status != -EPIPE && qtd->urb->status != -EREMOTEIO) {
 		chan->qh->tt_buffer_dirty = 1;
 		if (usb_hub_clear_tt_buffer(usb_urb))
