diff --git a/drivers/mtd/spi-nor/spi-nor.c b/drivers/mtd/spi-nor/spi-nor.c
index 36a79999..43718cac 100644
--- a/drivers/mtd/spi-nor/spi-nor.c
+++ b/drivers/mtd/spi-nor/spi-nor.c
@@ -2191,7 +2191,6 @@ static int spi_nor_setup(struct spi_nor *nor, const struct flash_info *info,
 
 	enable_quad_io = (SNOR_PROTO_DATA_FROM_PROTO(nor->read_proto) == 4 ||
 			  SNOR_PROTO_DATA_FROM_PROTO(nor->write_proto) == 4);
-	enable_quad_io = 0;
 
 	/* Enable Quad I/O if needed. */
 	if (enable_quad_io && params->enable_quad_io) {
@@ -2458,10 +2457,8 @@ int spi_nor_scan(struct spi_nor *nor, const char *name,
 	if (info->flags & SPI_NOR_NO_FR)
 		modes->rd_modes &= ~SNOR_MODE_1_1_1;
 
-	nor->erase_proto = SNOR_PROTO_1_1_1;
-	nor->read_proto = SNOR_PROTO_1_1_1;
-	nor->write_proto = SNOR_PROTO_1_1_1;
-	modes->rd_modes = SNOR_MODE_1_1_1;
+	modes->rd_modes = SNOR_EON_RD_MODES;
+	modes->wr_modes = SNOR_EON_WR_MODES;
 	printk("@spi_nor_scan(), modes->rd_modes:0x%x.\n", modes->rd_modes);
 
 	nor->program_opcode = SPINOR_OP_PP;

