--- a/arch/arm/boot/dts/infinity6b0-ssc009a-s01a.dts	2021-09-20 15:52:35.181208520 +0300
+++ b/arch/arm/boot/dts/infinity6b0-ssc009a-s01a.dts	2021-09-20 15:07:16.949450278 +0300
@@ -62,8 +62,8 @@
             vif_sr0_parallel_pdn_mode  = <2>;
             vif_sr0_mipi_rst_mode  = <1>;
             vif_sr0_mipi_pdn_mode  = <1>;
-            vif_sr0_hvsync_mode  = <0>;
-            vif_sr0_pck_mode  = <0>;
+            vif_sr0_hvsync_mode  = <1>;
+            vif_sr0_pck_mode  = <1>;
             /* Config sensor 1 pad mux */
             vif_sr1_par_mode = <2>;
             vif_sr1_mipi_mode = <2>;
