diff -drupN a/modules/gpu/mali-midgard/kernel_mode/driver/sconscript b/modules/gpu/mali-midgard/kernel_mode/driver/sconscript
--- a/modules/gpu/mali-midgard/kernel_mode/driver/sconscript	1970-01-01 03:00:00.000000000 +0300
+++ b/modules/gpu/mali-midgard/kernel_mode/driver/sconscript	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,23 @@
+#
+# (C) COPYRIGHT 2010-2014, 2016 ARM Limited. All rights reserved.
+#
+# This program is free software and is provided to you under the terms of the
+# GNU General Public License version 2 as published by the Free Software
+# Foundation, and any use by you of this program is subject to the terms
+# of such GNU licence.
+#
+# A copy of the licence is included with the program, and can also be obtained
+# from Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+# Boston, MA  02110-1301, USA.
+#
+#
+
+
+
+Import('env')
+
+if env['backend'] == 'kernel' and int(env['kernel_modules']) == 1:
+	SConscript('drivers/sconscript')
+
+	if Glob('tests/sconscript'):
+		SConscript('tests/sconscript')
