diff -drupN a/modules/gpu/mali-utgard/driver/src/devicedrv/mali/include/linux/aw/sun8i/sun8iw5p1.h b/modules/gpu/mali-utgard/driver/src/devicedrv/mali/include/linux/aw/sun8i/sun8iw5p1.h
--- a/modules/gpu/mali-utgard/driver/src/devicedrv/mali/include/linux/aw/sun8i/sun8iw5p1.h	1970-01-01 03:00:00.000000000 +0300
+++ b/modules/gpu/mali-utgard/driver/src/devicedrv/mali/include/linux/aw/sun8i/sun8iw5p1.h	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,81 @@
+/**
+ * Copyright (C) 2015-2016 Allwinner Technology Limited. All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * Author: Albert Yu <yuxyun@allwinnertech.com>
+ */
+
+#ifndef _MALI_SUN8I_W5P1_H_
+#define _MALI_SUN8I_W5P1_H_
+
+#define MALI_PP_CORES_NUM 2
+#define AW_MALI_GPU_RESOURCES_MALI400_MPX MALI_GPU_RESOURCES_MALI400_MP2
+
+#define GPU_PBASE           SUNXI_GPU_PBASE
+#define IRQ_GPU_GP          SUNXI_IRQ_GPUGP
+#define IRQ_GPU_GPMMU       SUNXI_IRQ_GPUGPMMU
+#define IRQ_GPU_PP0         SUNXI_IRQ_GPUPP0
+#define IRQ_GPU_PPMMU0      SUNXI_IRQ_GPUPPMMU0
+#define IRQ_GPU_PP1         SUNXI_IRQ_GPUPP1
+#define IRQ_GPU_PPMMU1      SUNXI_IRQ_GPUPPMMU1
+
+aw_private_data aw_private = {
+	.tempctrl      = {
+		.temp_ctrl_status = 1,
+	},
+	.pm            = {
+		.regulator      = NULL,
+		.regulator_id   = "vdd-gpu",
+		.clk[0]         = {
+			.clk_name   = "pll",
+		#ifndef CONFIG_MALI_DT
+			.clk_id     = PLL_GPU_CLK,
+		#endif
+			.clk_handle = NULL,
+		},
+		.clk[1]         = {
+			.clk_name   = "mali",
+		#ifndef CONFIG_MALI_DT
+			.clk_id     = GPU_CLK,
+		#endif
+			.clk_handle = NULL,
+		},
+		.vf_table[0]   = {
+			.vol  = 1100,
+			.freq = 144,
+		},
+		.vf_table[1]   = {
+			.vol  = 1100,
+			.freq = 264,
+		},
+		.vf_table[2]   = {
+			.vol  = 1100,
+			.freq = 384,
+		},
+		.dvfs_status       = 0,
+		.begin_level       = 2,
+		.max_level         = 2,
+		.scene_ctrl_cmd    = 0,
+		.scene_ctrl_status = 1,
+		.independent_pow   = 0,
+		.dvm               = 0,
+	},
+	.debug           = {
+		.enable      = 0,
+		.frequency   = 0,
+		.voltage     = 0,
+		.tempctrl    = 0,
+		.dvfs        = 0,
+		.level       = 0,
+	}
+};
+
+#endif /* _MALI_SUN8I_W5P1_H_ */
