diff -drupN a/modules/gpu/mali-utgard/driver/src/devicedrv/ump/Kconfig b/modules/gpu/mali-utgard/driver/src/devicedrv/ump/Kconfig
--- a/modules/gpu/mali-utgard/driver/src/devicedrv/ump/Kconfig	1970-01-01 03:00:00.000000000 +0300
+++ b/modules/gpu/mali-utgard/driver/src/devicedrv/ump/Kconfig	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,16 @@
+config UMP
+	tristate "UMP support"
+	depends on ARM
+	---help---
+	  This enables support for the UMP memory allocation and sharing API.
+
+	  To compile this driver as a module, choose M here: the module will be
+	  called ump.
+
+config UMP_DEBUG
+	bool "Enable extra debug in UMP"
+	depends on UMP
+	default y
+	---help---
+	  This enabled extra debug checks and messages in UMP.
+
