diff -drupN a/modules/gpu/mali-midgard/kernel_mode/driver/Makefile b/modules/gpu/mali-midgard/kernel_mode/driver/Makefile
--- a/modules/gpu/mali-midgard/kernel_mode/driver/Makefile	1970-01-01 03:00:00.000000000 +0300
+++ b/modules/gpu/mali-midgard/kernel_mode/driver/Makefile	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,16 @@
+.PHONY: all kds mali_base clean
+
+all: kds mali_base dma_buf_test_exporter
+
+kds:
+	$(MAKE) -C drivers/base/kds -j
+
+mali_base:
+	$(MAKE) -C  drivers/gpu/arm/midgard -j
+
+dma_buf_test_exporter:
+	$(MAKE) -C drivers/base/dma_buf_test_exporter -j
+clean:
+	$(MAKE) -C drivers/base/kds clean
+	$(MAKE) -C drivers/gpu/arm/midgard clean
+	$(MAKE) -C drivers/base/dma_buf_test_exporter clean
