diff -drupN a/modules/gpu/mali-midgard/kernel_mode/driver/drivers/gpu/arm/midgard/platform/devicetree/Kbuild b/modules/gpu/mali-midgard/kernel_mode/driver/drivers/gpu/arm/midgard/platform/devicetree/Kbuild
--- a/modules/gpu/mali-midgard/kernel_mode/driver/drivers/gpu/arm/midgard/platform/devicetree/Kbuild	1970-01-01 03:00:00.000000000 +0300
+++ b/modules/gpu/mali-midgard/kernel_mode/driver/drivers/gpu/arm/midgard/platform/devicetree/Kbuild	2022-06-12 05:28:14.000000000 +0300
@@ -0,0 +1,18 @@
+#
+# (C) COPYRIGHT 2012-2016 ARM Limited. All rights reserved.
+#
+# This program is free software and is provided to you under the terms of the
+# GNU General Public License version 2 as published by the Free Software
+# Foundation, and any use by you of this program is subject to the terms
+# of such GNU licence.
+#
+# A copy of the licence is included with the program, and can also be obtained
+# from Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+# Boston, MA  02110-1301, USA.
+#
+#
+
+
+mali_kbase-y += \
+	$(MALI_PLATFORM_THIRDPARTY_DIR)/mali_kbase_config_devicetree.o \
+	$(MALI_PLATFORM_THIRDPARTY_DIR)/mali_kbase_runtime_pm.o
