diff --git a/buffer.c b/buffer.c
index 822d97a5..646763b4 100644
--- a/buffer.c
+++ b/buffer.c
@@ -3070,7 +3070,7 @@ evbuffer_file_segment_materialize(struct evbuffer_file_segment *seg)
 			offset_leftover = offset % page_size;
 			offset_rounded = offset - offset_leftover;
 		}
-		mapped = mmap(NULL, length + offset_leftover,
+		mapped = mmap64(NULL, length + offset_leftover,
 		    PROT_READ,
 #ifdef MAP_NOCACHE
 		    MAP_NOCACHE | /* ??? */
